{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "heading_collapsed": true
   },
   "source": [
    "# Common"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-11-18T16:23:13.843923Z",
     "start_time": "2021-11-18T16:23:13.356704Z"
    },
    "hidden": true
   },
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "from tqdm.contrib.concurrent import process_map\n",
    "#from tqdm import tqdm  # !pip install tqdm\n",
    "from tqdm.notebook import tqdm\n",
    "tqdm.pandas()\n",
    "#from tqdm import trange\n",
    "import multiprocessing as mp\n",
    "import shutil\n",
    "import glob\n",
    "import os\n",
    "import sys\n",
    "if sys.version_info[0] < 3: \n",
    "    from StringIO import StringIO\n",
    "else:\n",
    "    from io import StringIO"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-11-18T16:23:13.968497Z",
     "start_time": "2021-11-18T16:23:13.847331Z"
    },
    "hidden": true
   },
   "outputs": [],
   "source": [
    "!mkdir -p Temp"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-11-18T16:23:13.977535Z",
     "start_time": "2021-11-18T16:23:13.972387Z"
    },
    "hidden": true
   },
   "outputs": [],
   "source": [
    "def bracket_row(row):    \n",
    "    s = row['data']\n",
    "    index = min(s.find('.'), s.find('('))\n",
    "    data = row['data']\n",
    "    row['data'] = data[0:index]\n",
    "    row['bracket'] = data[index:]\n",
    "    return row"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-11-18T16:23:13.984539Z",
     "start_time": "2021-11-18T16:23:13.980698Z"
    },
    "hidden": true
   },
   "outputs": [],
   "source": [
    "def adjust(text,n=7):\n",
    "    text = str(text)    \n",
    "    return \" \" * (n - len(text)) + text"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-11-18T16:23:14.171613Z",
     "start_time": "2021-11-18T16:23:14.160674Z"
    },
    "hidden": true
   },
   "outputs": [],
   "source": [
    "def bracket_to_ct(tag, data, bracket, deltaG, negative_deltaG=True):    \n",
    "    deltaG = deltaG.replace('(','').replace(')','')\n",
    "    deltaG = float(deltaG)\n",
    "    if(deltaG > 0 and negative_deltaG ): # negetive?!\n",
    "        deltaG = -1 * deltaG\n",
    "    stack = []\n",
    "    index = np.zeros((len(bracket)), dtype = int)\n",
    "    values = np.zeros((len(bracket)), dtype = int)\n",
    "    for i in range(len(bracket)):\n",
    "        index[i] = i + 1\n",
    "        if(bracket[i] == '.'):\n",
    "            values[i] = 0\n",
    "        elif(bracket[i] == '('):\n",
    "            stack.append(i)\n",
    "        elif(bracket[i] == ')'):\n",
    "            if(len(stack) == 0 ):\n",
    "                print('structure error!')\n",
    "            values[stack[-1]] = i + 1\n",
    "            values[i]  = stack[-1] + 1\n",
    "            stack.pop()\n",
    "        else:\n",
    "            print('structure error!')\n",
    "    if(len(stack) != 0 ):\n",
    "        print('structure error!')\n",
    "    # body    \n",
    "    ct = f\"{adjust(len(data),6)} dG ={adjust(deltaG,10)} {tag}\\n\"   \n",
    "    for i in range(len(bracket)):\n",
    "        ct += f\"{adjust(index[i],6)} {data[i]} {adjust(i,6)} {adjust((i+2)%(len(data)+1),6)} {adjust(values[i],6)} {adjust(index[i],7)}\\n\"\n",
    "    return ct"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-11-18T16:23:14.487825Z",
     "start_time": "2021-11-18T16:23:14.480533Z"
    },
    "hidden": true
   },
   "outputs": [],
   "source": [
    "def fasta_to_df(path):\n",
    "    with open(path, 'r') as file:\n",
    "        text = file.read()\n",
    "    lines = [line for line in text.split('\\n') if len(line) > 0]\n",
    "    s = ''\n",
    "    tags = []\n",
    "    data = []\n",
    "    for l in lines:\n",
    "        if(l[0]=='>'):\n",
    "            tags.append(l)        \n",
    "            data.append(s)\n",
    "            s = ''\n",
    "        else:\n",
    "            s += l    \n",
    "    data.append(s)\n",
    "    df = pd.DataFrame(\n",
    "            {\n",
    "                'tag': tags,\n",
    "                'data': data[1:]\n",
    "            })\n",
    "    df['tag'] = df['tag'].apply(lambda x: x[1:])    \n",
    "    return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-11-18T16:23:14.830646Z",
     "start_time": "2021-11-18T16:23:14.825745Z"
    },
    "hidden": true
   },
   "outputs": [],
   "source": [
    "def df_to_fasta(df, path):\n",
    "    lines = []\n",
    "    df.apply(lambda row: lines.append(f\">{row['tag']}\\n{row['data']}\\n\"),axis=1)\n",
    "    with open(path,'w') as file:\n",
    "        file.write(''.join(lines))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-11-18T16:23:15.149524Z",
     "start_time": "2021-11-18T16:23:15.145355Z"
    },
    "hidden": true
   },
   "outputs": [],
   "source": [
    "def reformat(path):\n",
    "    return path.replace('(','_').replace(')','_').replace('.','').replace(':','_')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-11-18T16:23:15.634454Z",
     "start_time": "2021-11-18T16:23:15.626507Z"
    },
    "hidden": true
   },
   "outputs": [],
   "source": [
    "def reformatCT(path):\n",
    "    with open(path, 'r') as file:\n",
    "        text = file.read()\n",
    "    text = [l for l in text.split('\\n') if len(l) > 0 ] # remove blank lines\n",
    "    text = '\\n'.join(text)\n",
    "    text = text.replace(\"\\t\",\" \")\n",
    "    while(\"  \" in text):\n",
    "        text = text.replace(\"  \", \" \")\n",
    "    lines = [l for l in text.split('\\n')]\n",
    "    for i in range(len(lines)):\n",
    "        if(lines[i][0] == \" \"):\n",
    "            lines[i] = lines[i][1:]\n",
    "        if(lines[i][-1] == \" \"):\n",
    "            lines[i] = lines[i][:-1]\n",
    "    text = '\\n'.join(lines)\n",
    "    return text"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-11-18T16:23:16.001951Z",
     "start_time": "2021-11-18T16:23:15.996513Z"
    },
    "hidden": true
   },
   "outputs": [],
   "source": [
    "def get_ct_data(ct):\n",
    "    ct = \"\\n\".join(ct.split('\\n')[1:])\n",
    "    df = pd.read_csv(StringIO(ct), sep=\" \", header=None)               \n",
    "    nucleotide = df.iloc[:,1]\n",
    "    index = df.iloc[:,5]\n",
    "    values = df.iloc[:,4]\n",
    "    return [nucleotide, index, values]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-11-18T16:23:16.602536Z",
     "start_time": "2021-11-18T16:23:16.596390Z"
    },
    "hidden": true
   },
   "outputs": [],
   "source": [
    "def ct2dot_bracket(path):\n",
    "    [nucleotide, index, values] = get_ct_data(reformatCT(path))\n",
    "    text = ''.join(nucleotide) + \"\\n\"\n",
    "    watch = []\n",
    "    for i, v in zip(index,values):\n",
    "        if(v == 0):\n",
    "            text += '.'\n",
    "        else:\n",
    "            if( v not in watch):\n",
    "                text += '('\n",
    "                watch.append(i)\n",
    "            if( v in watch):\n",
    "                text += ')'\n",
    "    return text"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-11-18T16:23:16.935444Z",
     "start_time": "2021-11-18T16:23:16.929537Z"
    },
    "hidden": true
   },
   "outputs": [],
   "source": [
    "def is_nested(index, values):\n",
    "    max_value = max(index) + 10 # inf\n",
    "    for i, v in zip(index, values):\n",
    "        if(v < max_value and v != 0):\n",
    "            max_value  = v\n",
    "        if(i >= max_value):\n",
    "            max_value = max(index) + 10 # inf\n",
    "        if(v > max_value):\n",
    "            return False               \n",
    "    return True"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-11-18T16:23:17.425214Z",
     "start_time": "2021-11-18T16:23:17.421277Z"
    },
    "hidden": true
   },
   "outputs": [],
   "source": [
    "'''ct = reformatCT('./secondary_structure/spot_rna/AMWY020598281_2832-3256_+_/AMWY020598281_2832-3256_+_.ct')\n",
    "[nucleotide, index, values] = get_ct_data(ct)\n",
    "print(is_nested( index,  values))\n",
    "''';"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "heading_collapsed": true,
    "hidden": true
   },
   "source": [
    "### rename tag of input genome to new tag id"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "heading_collapsed": true
   },
   "source": [
    "# Download dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-11-14T15:26:00.765914Z",
     "start_time": "2021-11-14T15:26:00.762196Z"
    },
    "hidden": true
   },
   "outputs": [],
   "source": [
    "'''\n",
    "from Bio import Entrez\n",
    "Entrez.email = \"abolhasani.eliya@gmail.com\"     \n",
    "with Entrez.esearch(db='nucleotide', term=\"Arabidopsis thaliana\") as handle:\n",
    "    result = Entrez.read(handle)\n",
    "\n",
    "print(result)\n",
    "genome_ids = result['IdList']\n",
    "\n",
    "for genome_id in genome_ids:\n",
    "    print(genome_id)\n",
    "    record = Entrez.efetch(db=\"nucleotide\", id=genome_id, rettype=\"fasta\", retmode=\"text\")        \n",
    "    with open(f'{genome_id}.fasta', 'w') as f:\n",
    "        f.write(record.read())\n",
    "    break\n",
    "''';\n",
    "'''\n",
    "from Bio import Entrez\n",
    "Entrez.email = \"abolhasani.eliya@gmail.com\"     \n",
    "record = Entrez.efetch(db=\"nucleotide\", id=\"NC_054143.4\", rettype=\"fasta\", retmode=\"text\")        \n",
    "with open(f'data.fasta', 'w') as f:\n",
    "    f.write(record.read())\n",
    "''';"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-11-14T14:40:09.505684Z",
     "start_time": "2021-11-14T14:40:04.087813Z"
    },
    "hidden": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "--2021-11-14 18:10:04--  https://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/000/439/995/GCA_000439995.3_AzaInd2.1/GCA_000439995.3_AzaInd2.1_genomic.fna.gz\n",
      "Resolving ftp.ncbi.nlm.nih.gov (ftp.ncbi.nlm.nih.gov)... 130.14.250.10, 130.14.250.7, 2607:f220:41e:250::11, ...\n",
      "Connecting to ftp.ncbi.nlm.nih.gov (ftp.ncbi.nlm.nih.gov)|130.14.250.10|:443... connected.\n",
      "HTTP request sent, awaiting response... 200 OK\n",
      "Length: 85647577 (82M) [application/x-gzip]\n",
      "Saving to: ‘GCA_000439995.3_AzaInd2.1_genomic.fna.gz’\n",
      "\n",
      "CA_000439995.3_AzaI  27%[====>               ]  22.82M  7.26MB/s    eta 8s     ^C\n"
     ]
    }
   ],
   "source": [
    "!wget https://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/000/439/995/GCA_000439995.3_AzaInd2.1/GCA_000439995.3_AzaInd2.1_genomic.fna.gz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 263,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-10-24T15:35:44.724088Z",
     "start_time": "2021-10-24T15:35:41.652464Z"
    },
    "hidden": true
   },
   "outputs": [],
   "source": [
    "!gzip -d ./GCA_000439995.3_AzaInd2.1_genomic.fna.gz"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "heading_collapsed": true
   },
   "source": [
    "# Download data from Mirbase"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-11-14T15:43:46.403811Z",
     "start_time": "2021-11-14T15:43:46.398998Z"
    },
    "hidden": true
   },
   "outputs": [],
   "source": [
    "directory = 'miRBase_driven_data'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-11-14T15:38:44.732483Z",
     "start_time": "2021-11-14T15:38:31.128444Z"
    },
    "hidden": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "--2021-11-14 19:08:31--  https://www.mirbase.org/ftp/CURRENT/aliases.txt.gz\n",
      "Resolving www.mirbase.org (www.mirbase.org)... 130.88.97.249\n",
      "Connecting to www.mirbase.org (www.mirbase.org)|130.88.97.249|:443... connected.\n",
      "HTTP request sent, awaiting response... 200 OK\n",
      "Length: 480536 (469K) [application/x-gzip]\n",
      "Saving to: ‘./miRBase_driven_data/aliases.txt.gz’\n",
      "\n",
      "aliases.txt.gz      100%[===================>] 469.27K   469KB/s    in 1.0s    \n",
      "\n",
      "2021-11-14 19:08:32 (469 KB/s) - ‘./miRBase_driven_data/aliases.txt.gz’ saved [480536/480536]\n",
      "\n",
      "--2021-11-14 19:08:33--  https://www.mirbase.org/ftp/CURRENT/hairpin.fa.gz\n",
      "Resolving www.mirbase.org (www.mirbase.org)... 130.88.97.249\n",
      "Connecting to www.mirbase.org (www.mirbase.org)|130.88.97.249|:443... connected.\n",
      "HTTP request sent, awaiting response... 200 OK\n",
      "Length: 1547350 (1.5M) [application/x-gzip]\n",
      "Saving to: ‘./miRBase_driven_data/hairpin.fa.gz’\n",
      "\n",
      "hairpin.fa.gz       100%[===================>]   1.48M   995KB/s    in 1.5s    \n",
      "\n",
      "2021-11-14 19:08:34 (995 KB/s) - ‘./miRBase_driven_data/hairpin.fa.gz’ saved [1547350/1547350]\n",
      "\n",
      "--2021-11-14 19:08:35--  https://www.mirbase.org/ftp/CURRENT/hairpin_high_conf.fa.gz\n",
      "Resolving www.mirbase.org (www.mirbase.org)... 130.88.97.249\n",
      "Connecting to www.mirbase.org (www.mirbase.org)|130.88.97.249|:443... connected.\n",
      "HTTP request sent, awaiting response... 200 OK\n",
      "Length: 129596 (127K) [application/x-gzip]\n",
      "Saving to: ‘./miRBase_driven_data/hairpin_high_conf.fa.gz’\n",
      "\n",
      "hairpin_high_conf.f 100%[===================>] 126.56K   208KB/s    in 0.6s    \n",
      "\n",
      "2021-11-14 19:08:36 (208 KB/s) - ‘./miRBase_driven_data/hairpin_high_conf.fa.gz’ saved [129596/129596]\n",
      "\n",
      "--2021-11-14 19:08:36--  https://www.mirbase.org/ftp/CURRENT/mature.fa.gz\n",
      "Resolving www.mirbase.org (www.mirbase.org)... 130.88.97.249\n",
      "Connecting to www.mirbase.org (www.mirbase.org)|130.88.97.249|:443... connected.\n",
      "HTTP request sent, awaiting response... 200 OK\n",
      "Length: 803746 (785K) [application/x-gzip]\n",
      "Saving to: ‘./miRBase_driven_data/mature.fa.gz’\n",
      "\n",
      "mature.fa.gz        100%[===================>] 784.91K   595KB/s    in 1.3s    \n",
      "\n",
      "2021-11-14 19:08:38 (595 KB/s) - ‘./miRBase_driven_data/mature.fa.gz’ saved [803746/803746]\n",
      "\n",
      "--2021-11-14 19:08:38--  https://www.mirbase.org/ftp/CURRENT/mature_high_conf.fa.gz\n",
      "Resolving www.mirbase.org (www.mirbase.org)... 130.88.97.249\n",
      "Connecting to www.mirbase.org (www.mirbase.org)|130.88.97.249|:443... connected.\n",
      "HTTP request sent, awaiting response... 200 OK\n",
      "Length: 69381 (68K) [application/x-gzip]\n",
      "Saving to: ‘./miRBase_driven_data/mature_high_conf.fa.gz’\n",
      "\n",
      "mature_high_conf.fa 100%[===================>]  67.75K   134KB/s    in 0.5s    \n",
      "\n",
      "2021-11-14 19:08:39 (134 KB/s) - ‘./miRBase_driven_data/mature_high_conf.fa.gz’ saved [69381/69381]\n",
      "\n",
      "--2021-11-14 19:08:39--  https://www.mirbase.org/ftp/CURRENT/miRNA.str.gz\n",
      "Resolving www.mirbase.org (www.mirbase.org)... 130.88.97.249\n",
      "Connecting to www.mirbase.org (www.mirbase.org)|130.88.97.249|:443... connected.\n",
      "HTTP request sent, awaiting response... 200 OK\n",
      "Length: 3373763 (3.2M) [application/x-gzip]\n",
      "Saving to: ‘./miRBase_driven_data/miRNA.str.gz’\n",
      "\n",
      "miRNA.str.gz        100%[===================>]   3.22M  1.77MB/s    in 1.8s    \n",
      "\n",
      "2021-11-14 19:08:41 (1.77 MB/s) - ‘./miRBase_driven_data/miRNA.str.gz’ saved [3373763/3373763]\n",
      "\n",
      "--2021-11-14 19:08:41--  https://www.mirbase.org/ftp/CURRENT/miRNA.xls.gz\n",
      "Resolving www.mirbase.org (www.mirbase.org)... 130.88.97.249\n",
      "Connecting to www.mirbase.org (www.mirbase.org)|130.88.97.249|:443... connected.\n",
      "HTTP request sent, awaiting response... 200 OK\n",
      "Length: 2963404 (2.8M) [application/x-gzip]\n",
      "Saving to: ‘./miRBase_driven_data/miRNA.xls.gz’\n",
      "\n",
      "miRNA.xls.gz        100%[===================>]   2.83M  1.65MB/s    in 1.7s    \n",
      "\n",
      "2021-11-14 19:08:43 (1.65 MB/s) - ‘./miRBase_driven_data/miRNA.xls.gz’ saved [2963404/2963404]\n",
      "\n",
      "--2021-11-14 19:08:44--  https://www.mirbase.org/ftp/CURRENT/organisms.txt.gz\n",
      "Resolving www.mirbase.org (www.mirbase.org)... 130.88.97.249\n",
      "Connecting to www.mirbase.org (www.mirbase.org)|130.88.97.249|:443... connected.\n",
      "HTTP request sent, awaiting response... 200 OK\n",
      "Length: 6842 (6.7K) [application/x-gzip]\n",
      "Saving to: ‘./miRBase_driven_data/organisms.txt.gz’\n",
      "\n",
      "organisms.txt.gz    100%[===================>]   6.68K  --.-KB/s    in 0s      \n",
      "\n",
      "2021-11-14 19:08:44 (132 MB/s) - ‘./miRBase_driven_data/organisms.txt.gz’ saved [6842/6842]\n",
      "\n"
     ]
    }
   ],
   "source": [
    "base = \"https://www.mirbase.org/ftp/CURRENT\"        \n",
    "!rm -r {directory}\n",
    "!mkdir -p {directory}\n",
    "!wget {base}/aliases.txt.gz -P ./{directory}/       ; gzip -d ./{directory}/aliases.txt.gz \n",
    "!wget {base}/hairpin.fa.gz -P ./{directory}/           ; gzip -d ./{directory}/hairpin.fa.gz \n",
    "!wget {base}/hairpin_high_conf.fa.gz -P ./{directory}/ ; gzip -d ./{directory}/hairpin_high_conf.fa.gz \n",
    "!wget {base}/mature.fa.gz -P ./{directory}/            ; gzip -d ./{directory}/mature.fa.gz \n",
    "!wget {base}/mature_high_conf.fa.gz -P ./{directory}/  ; gzip -d ./{directory}/mature_high_conf.fa.gz\n",
    "!wget {base}/miRNA.str.gz -P ./{directory}/            ; gzip -d ./{directory}/miRNA.str.gz \n",
    "!wget {base}/miRNA.xls.gz -P ./{directory}/            ; gzip -d ./{directory}/miRNA.xls.gz \n",
    "!wget {base}/organisms.txt.gz -P ./{directory}/        ; gzip -d ./{directory}/organisms.txt.gz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-11-14T15:44:38.104775Z",
     "start_time": "2021-11-14T15:44:37.963817Z"
    },
    "hidden": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(48885, 3)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>tag</th>\n",
       "      <th>data</th>\n",
       "      <th>organism</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>cel-let-7-5p MIMAT0000001 Caenorhabditis elega...</td>\n",
       "      <td>UGAGGUAGUAGGUUGUAUAGUU</td>\n",
       "      <td>cel</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>cel-let-7-3p MIMAT0015091 Caenorhabditis elega...</td>\n",
       "      <td>CUAUGCAAUUUUCUACCUUACC</td>\n",
       "      <td>cel</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                 tag                    data  \\\n",
       "0  cel-let-7-5p MIMAT0000001 Caenorhabditis elega...  UGAGGUAGUAGGUUGUAUAGUU   \n",
       "1  cel-let-7-3p MIMAT0015091 Caenorhabditis elega...  CUAUGCAAUUUUCUACCUUACC   \n",
       "\n",
       "  organism  \n",
       "0      cel  \n",
       "1      cel  "
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = fasta_to_df(f'./{directory}/mature.fa')\n",
    "#df = fasta_to_df('./Data/mature_high_conf.fa')\n",
    "df['organism'] = df['tag'].apply(lambda x: x[:3])\n",
    "print(df.shape)\n",
    "df.head(2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-11-14T15:44:38.486126Z",
     "start_time": "2021-11-14T15:44:38.465726Z"
    },
    "hidden": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(285, 5)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>organism</th>\n",
       "      <th>division</th>\n",
       "      <th>name</th>\n",
       "      <th>tree</th>\n",
       "      <th>NCBI-taxid</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>aqu</td>\n",
       "      <td>AQU</td>\n",
       "      <td>Amphimedon queenslandica</td>\n",
       "      <td>Metazoa;Porifera;</td>\n",
       "      <td>400682</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>nve</td>\n",
       "      <td>NVE</td>\n",
       "      <td>Nematostella vectensis</td>\n",
       "      <td>Metazoa;Cnidaria;</td>\n",
       "      <td>45351</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  organism division                      name               tree  NCBI-taxid\n",
       "0      aqu      AQU  Amphimedon queenslandica  Metazoa;Porifera;      400682\n",
       "1      nve      NVE    Nematostella vectensis  Metazoa;Cnidaria;       45351"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "organism = pd.read_csv(f'./{directory}/organisms.txt',sep='\\t')\n",
    "organism.columns = [c.replace('#','') for c in organism.columns] # remove sharp from columns\n",
    "print(organism.shape)\n",
    "organism.head(2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-11-14T15:44:39.097367Z",
     "start_time": "2021-11-14T15:44:39.085246Z"
    },
    "hidden": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['Viruses;',\n",
       " 'Mycetozoa;',\n",
       " 'Alveolata;',\n",
       " 'Metazoa;Porifera;',\n",
       " 'Metazoa;Cnidaria;',\n",
       " 'Viridiplantae;Chlorophyta;',\n",
       " 'Viridiplantae;Embryophyta;',\n",
       " 'Viridiplantae;Coniferophyta;',\n",
       " 'Viridiplantae;Magnoliophyta;',\n",
       " 'Metazoa;Bilateria;Deuterostoma;',\n",
       " 'Chromalveolata;Heterokontophyta;',\n",
       " 'Metazoa;Bilateria;Ecdysozoa;Nematoda;',\n",
       " 'Metazoa;Bilateria;Lophotrochozoa;Annelida;',\n",
       " 'Metazoa;Bilateria;Lophotrochozoa;Nemertea;',\n",
       " 'Metazoa;Bilateria;Lophotrochozoa;Mollusca;',\n",
       " 'Viridiplantae;Magnoliophyta;monocotyledons;',\n",
       " 'Metazoa;Bilateria;Deuterostoma;Hemichordata;',\n",
       " 'Metazoa;Bilateria;Deuterostoma;Echinodermata;',\n",
       " 'Metazoa;Bilateria;Lophotrochozoa;Brachiopoda;',\n",
       " 'Metazoa;Bilateria;Ecdysozoa;Arthropoda;Hexapoda;',\n",
       " 'Metazoa;Bilateria;Ecdysozoa;Arthropoda;Crustacea;',\n",
       " 'Metazoa;Bilateria;Lophotrochozoa;Platyhelminthes;',\n",
       " 'Metazoa;Bilateria;Ecdysozoa;Arthropoda;Chelicerata;',\n",
       " 'Metazoa;Bilateria;Ecdysozoa;Arthropoda;Mandibulata;',\n",
       " 'Viridiplantae;Magnoliophyta;eudicotyledons;Poaceae;',\n",
       " 'Metazoa;Bilateria;Deuterostoma;Chordata;Urochordata;',\n",
       " 'Viridiplantae;Magnoliophyta;eudicotyledons;Fabaceae;',\n",
       " 'Viridiplantae;Magnoliophyta;eudicotyledons;Vitaceae;',\n",
       " 'Viridiplantae;Magnoliophyta;eudicotyledons;Rosaceae;',\n",
       " 'Viridiplantae;Magnoliophyta;eudicotyledons;Rutaceae;',\n",
       " 'Viridiplantae;Magnoliophyta;eudicotyledons;Lamiales;',\n",
       " 'Viridiplantae;Magnoliophyta;eudicotyledons;Linaceae;',\n",
       " 'Viridiplantae;Magnoliophyta;eudicotyledons;Malvaceae;',\n",
       " 'Viridiplantae;Magnoliophyta;eudicotyledons;Myrtaceae;',\n",
       " 'Viridiplantae;Magnoliophyta;eudicotyledons;Caricaceae;',\n",
       " 'Viridiplantae;Magnoliophyta;eudicotyledons;Salicaceae;',\n",
       " 'Viridiplantae;Magnoliophyta;eudicotyledons;Solanaceae;',\n",
       " 'Viridiplantae;Magnoliophyta;eudicotyledons;Asteraceae;',\n",
       " 'Viridiplantae;Magnoliophyta;eudicotyledons;Araliaceae;',\n",
       " 'Viridiplantae;Magnoliophyta;eudicotyledons;Paeoniaceae;',\n",
       " 'Metazoa;Bilateria;Deuterostoma;Chordata;Cephalochordata;',\n",
       " 'Metazoa;Bilateria;Deuterostoma;Chordata;Vertebrata;Aves;',\n",
       " 'Viridiplantae;Magnoliophyta;eudicotyledons;Brassicaceae;',\n",
       " 'Viridiplantae;Magnoliophyta;eudicotyledons;Papaveraceae;',\n",
       " 'Viridiplantae;Magnoliophyta;eudicotyledons;Ranunculaceae;',\n",
       " 'Viridiplantae;Magnoliophyta;eudicotyledons;Euphorbiaceae;',\n",
       " 'Viridiplantae;Magnoliophyta;eudicotyledons;Cucurbitaceae;',\n",
       " 'Viridiplantae;Magnoliophyta;eudicotyledons;Amaranthaceae;',\n",
       " 'Viridiplantae;Magnoliophyta;eudicotyledons;Rhizophoraceae;',\n",
       " 'Metazoa;Bilateria;Deuterostoma;Chordata;Vertebrata;Sauria;',\n",
       " 'Metazoa;Bilateria;Deuterostoma;Chordata;Vertebrata;Amphibia;',\n",
       " 'Metazoa;Bilateria;Deuterostoma;Chordata;Vertebrata;Teleostei;',\n",
       " 'Metazoa;Bilateria;Deuterostoma;Chordata;Vertebrata;Gnathostomata;',\n",
       " 'Metazoa;Bilateria;Deuterostoma;Chordata;Vertebrata;Mammalia;Suina;',\n",
       " 'Metazoa;Bilateria;Deuterostoma;Chordata;Vertebrata;Agnathostomata;',\n",
       " 'Metazoa;Bilateria;Deuterostoma;Chordata;Vertebrata;Mammalia;Rodentia;',\n",
       " 'Metazoa;Bilateria;Deuterostoma;Chordata;Vertebrata;Mammalia;Carnivora;',\n",
       " 'Metazoa;Bilateria;Deuterostoma;Chordata;Vertebrata;Mammalia;Cingulata;',\n",
       " 'Metazoa;Bilateria;Deuterostoma;Chordata;Vertebrata;Mammalia;Metatheria;',\n",
       " 'Metazoa;Bilateria;Deuterostoma;Chordata;Vertebrata;Mammalia;Ruminantia;',\n",
       " 'Metazoa;Bilateria;Deuterostoma;Chordata;Vertebrata;Mammalia;Lagomorpha;',\n",
       " 'Metazoa;Bilateria;Deuterostoma;Chordata;Vertebrata;Mammalia;Scandentia;',\n",
       " 'Metazoa;Bilateria;Deuterostoma;Chordata;Vertebrata;Mammalia;Afrotheria;',\n",
       " 'Metazoa;Bilateria;Deuterostoma;Chordata;Vertebrata;Mammalia;Prototheria;',\n",
       " 'Metazoa;Bilateria;Deuterostoma;Chordata;Vertebrata;Mammalia;Laurasiatheria;',\n",
       " 'Metazoa;Bilateria;Deuterostoma;Chordata;Vertebrata;Mammalia;Primates;Cebidae;',\n",
       " 'Metazoa;Bilateria;Deuterostoma;Chordata;Vertebrata;Mammalia;Primates;Atelidae;',\n",
       " 'Metazoa;Bilateria;Deuterostoma;Chordata;Vertebrata;Mammalia;Primates;Hominidae;',\n",
       " 'Metazoa;Bilateria;Deuterostoma;Chordata;Vertebrata;Mammalia;Primates;Lemuridae;',\n",
       " 'Metazoa;Bilateria;Deuterostoma;Chordata;Vertebrata;Mammalia;Primates;Galagidae;',\n",
       " 'Metazoa;Bilateria;Deuterostoma;Chordata;Vertebrata;Mammalia;Primates;Hylobatidae;',\n",
       " 'Metazoa;Bilateria;Deuterostoma;Chordata;Vertebrata;Mammalia;Primates;Cheirogaleidae;',\n",
       " 'Metazoa;Bilateria;Deuterostoma;Chordata;Vertebrata;Mammalia;Primates;Daubentoniidae;',\n",
       " 'Metazoa;Bilateria;Deuterostoma;Chordata;Vertebrata;Mammalia;Primates;Cercopithecidae;']"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "items = list(organism['tree'].unique())\n",
    "items.sort(key=len)\n",
    "items"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-11-14T15:44:42.522205Z",
     "start_time": "2021-11-14T15:44:42.505136Z"
    },
    "hidden": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(86, 5)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>organism</th>\n",
       "      <th>division</th>\n",
       "      <th>name</th>\n",
       "      <th>tree</th>\n",
       "      <th>NCBI-taxid</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>66</th>\n",
       "      <td>cre</td>\n",
       "      <td>CRE</td>\n",
       "      <td>Chlamydomonas reinhardtii</td>\n",
       "      <td>Viridiplantae;Chlorophyta;</td>\n",
       "      <td>3055</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>67</th>\n",
       "      <td>pta</td>\n",
       "      <td>PTA</td>\n",
       "      <td>Pinus taeda</td>\n",
       "      <td>Viridiplantae;Coniferophyta;</td>\n",
       "      <td>3352</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>68</th>\n",
       "      <td>ppt</td>\n",
       "      <td>PPT</td>\n",
       "      <td>Physcomitrella patens</td>\n",
       "      <td>Viridiplantae;Embryophyta;</td>\n",
       "      <td>3218</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>69</th>\n",
       "      <td>smo</td>\n",
       "      <td>SMO</td>\n",
       "      <td>Selaginella moellendorffii</td>\n",
       "      <td>Viridiplantae;Embryophyta;</td>\n",
       "      <td>88036</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>70</th>\n",
       "      <td>ath</td>\n",
       "      <td>ATH</td>\n",
       "      <td>Arabidopsis thaliana</td>\n",
       "      <td>Viridiplantae;Magnoliophyta;eudicotyledons;Bra...</td>\n",
       "      <td>3702</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   organism division                        name  \\\n",
       "66      cre      CRE   Chlamydomonas reinhardtii   \n",
       "67      pta      PTA                 Pinus taeda   \n",
       "68      ppt      PPT       Physcomitrella patens   \n",
       "69      smo      SMO  Selaginella moellendorffii   \n",
       "70      ath      ATH        Arabidopsis thaliana   \n",
       "\n",
       "                                                 tree  NCBI-taxid  \n",
       "66                         Viridiplantae;Chlorophyta;        3055  \n",
       "67                       Viridiplantae;Coniferophyta;        3352  \n",
       "68                         Viridiplantae;Embryophyta;        3218  \n",
       "69                         Viridiplantae;Embryophyta;       88036  \n",
       "70  Viridiplantae;Magnoliophyta;eudicotyledons;Bra...        3702  "
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "selectedTree = organism[organism['tree'].apply(lambda x: \"Viridiplantae;\" in x)]\n",
    "print(selectedTree.shape)\n",
    "selectedTree.head(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-11-14T15:44:45.986622Z",
     "start_time": "2021-11-14T15:44:45.960318Z"
    },
    "hidden": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(10414, 3)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>tag</th>\n",
       "      <th>data</th>\n",
       "      <th>organism</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>316</th>\n",
       "      <td>ath-miR156a-5p MIMAT0000166 Arabidopsis thalia...</td>\n",
       "      <td>UGACAGAAGAGAGUGAGCAC</td>\n",
       "      <td>ath</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>317</th>\n",
       "      <td>ath-miR156a-3p MIMAT0031865 Arabidopsis thalia...</td>\n",
       "      <td>GCUCACUGCUCUUUCUGUCAGA</td>\n",
       "      <td>ath</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>318</th>\n",
       "      <td>ath-miR156b-5p MIMAT0000167 Arabidopsis thalia...</td>\n",
       "      <td>UGACAGAAGAGAGUGAGCAC</td>\n",
       "      <td>ath</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>319</th>\n",
       "      <td>ath-miR156b-3p MIMAT0031866 Arabidopsis thalia...</td>\n",
       "      <td>UGCUCACCUCUCUUUCUGUCAGU</td>\n",
       "      <td>ath</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>320</th>\n",
       "      <td>ath-miR156c-5p MIMAT0000168 Arabidopsis thalia...</td>\n",
       "      <td>UGACAGAAGAGAGUGAGCAC</td>\n",
       "      <td>ath</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                   tag  \\\n",
       "316  ath-miR156a-5p MIMAT0000166 Arabidopsis thalia...   \n",
       "317  ath-miR156a-3p MIMAT0031865 Arabidopsis thalia...   \n",
       "318  ath-miR156b-5p MIMAT0000167 Arabidopsis thalia...   \n",
       "319  ath-miR156b-3p MIMAT0031866 Arabidopsis thalia...   \n",
       "320  ath-miR156c-5p MIMAT0000168 Arabidopsis thalia...   \n",
       "\n",
       "                        data organism  \n",
       "316     UGACAGAAGAGAGUGAGCAC      ath  \n",
       "317   GCUCACUGCUCUUUCUGUCAGA      ath  \n",
       "318     UGACAGAAGAGAGUGAGCAC      ath  \n",
       "319  UGCUCACCUCUCUUUCUGUCAGU      ath  \n",
       "320     UGACAGAAGAGAGUGAGCAC      ath  "
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "selected = df[df['organism'].isin(selectedTree['organism'])]\n",
    "print(selected.shape)\n",
    "selected.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-11-14T15:45:02.006479Z",
     "start_time": "2021-11-14T15:45:01.685273Z"
    },
    "hidden": true
   },
   "outputs": [],
   "source": [
    "df_to_fasta(selected,'./Temp/mature_microRNA_queries.fasta')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "heading_collapsed": true
   },
   "source": [
    "# Remove redundant"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "heading_collapsed": true,
    "hidden": true
   },
   "source": [
    "## cdhit-est"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-11-14T15:53:14.610830Z",
     "start_time": "2021-11-14T15:53:14.359020Z"
    },
    "hidden": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "================================================================\n",
      "Program: CD-HIT, V4.8.1 (+OpenMP), Oct 23 2021, 21:45:39\n",
      "Command: cdhit/cd-hit-est -i\n",
      "         ./Temp/mature_microRNA_queries.fasta -o\n",
      "         ./Temp/NR_mature_microRNA_queries.fasta -c 1 -r 0 -G 1\n",
      "         -g 1 -b 30 -l 10 -aL 0 -AL 99999999 -aS 0 -AS 99999999\n",
      "         -s 0 -S 0\n",
      "\n",
      "Started: Sun Nov 14 19:23:14 2021\n",
      "================================================================\n",
      "                            Output                              \n",
      "----------------------------------------------------------------\n",
      "total seq: 10414\n",
      "longest and shortest : 28 and 17\n",
      "Total letters: 222978\n",
      "Sequences have been sorted\n",
      "\n",
      "Approximated minimal memory consumption:\n",
      "Sequence        : 1M\n",
      "Buffer          : 1 X 12M = 12M\n",
      "Table           : 1 X 16M = 16M\n",
      "Miscellaneous   : 0M\n",
      "Total           : 30M\n",
      "\n",
      "Table limit with the given memory limit:\n",
      "Max number of representatives: 4000000\n",
      "Max number of word counting entries: 96149440\n",
      "\n",
      "comparing sequences from          0  to      10414\n",
      "..........    10000  finished       5817  clusters\n",
      "\n",
      "    10414  finished       6028  clusters\n",
      "\n",
      "Approximated maximum memory consumption: 31M\n",
      "writing new database\n",
      "writing clustering information\n",
      "program completed !\n",
      "\n",
      "Total CPU time 0.12\n"
     ]
    }
   ],
   "source": [
    "!cdhit/cd-hit-est -i ./Temp/mature_microRNA_queries.fasta  -o ./Temp/NR_mature_microRNA_queries.fasta \\\n",
    "    -c 1 -r 0 -G 1 -g 1 -b 30 -l 10 -aL 0 -AL 99999999 -aS 0 \\\n",
    "    -AS 99999999 -s 0 -S 0 "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "heading_collapsed": true,
    "hidden": true
   },
   "source": [
    "## reformat"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-11-14T15:53:16.096729Z",
     "start_time": "2021-11-14T15:53:16.056661Z"
    },
    "hidden": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(10414, 2)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqid</th>\n",
       "      <th>cluster</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>cst-miR11332</td>\n",
       "      <td>C0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>stu-miR7994b-5p</td>\n",
       "      <td>C1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             seqid cluster\n",
       "0     cst-miR11332      C0\n",
       "1  stu-miR7994b-5p      C1"
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "with open('./Temp/NR_mature_microRNA_queries.fasta.clstr','r') as file:\n",
    "    text = file.read()\n",
    "lines = [line for line in text.split('\\n') if len(line) > 0]\n",
    "cluster = []\n",
    "seqid = []\n",
    "last_cluster = \"\"\n",
    "for l in lines:\n",
    "    if(l[0]=='>'):        \n",
    "        last_cluster = l.replace('>Cluster ',\"C\")\n",
    "    else:        \n",
    "        cluster.append(last_cluster)\n",
    "        seqid.append(l.split(', >')[1].split('...')[0])                \n",
    "seq2cluster = pd.DataFrame({'seqid': seqid,'cluster': cluster})\n",
    "print(seq2cluster.shape)\n",
    "seq2cluster.head(2)    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-11-14T15:53:16.880158Z",
     "start_time": "2021-11-14T15:53:16.769037Z"
    },
    "hidden": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(10414, 3)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cluster</th>\n",
       "      <th>seqid</th>\n",
       "      <th>tag</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>C5495</td>\n",
       "      <td>ath-miR156a-5p</td>\n",
       "      <td>ath-miR156a-5p MIMAT0000166 Arabidopsis thalia...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>C1199</td>\n",
       "      <td>ath-miR156a-3p</td>\n",
       "      <td>ath-miR156a-3p MIMAT0031865 Arabidopsis thalia...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  cluster           seqid                                                tag\n",
       "0   C5495  ath-miR156a-5p  ath-miR156a-5p MIMAT0000166 Arabidopsis thalia...\n",
       "1   C1199  ath-miR156a-3p  ath-miR156a-3p MIMAT0031865 Arabidopsis thalia..."
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "df = fasta_to_df(\"./Temp/mature_microRNA_queries.fasta\")\n",
    "df['accession'] = df['tag'].apply(lambda x : x.split(' ')[0])\n",
    "seq2cluster = pd.merge(df,seq2cluster,how=\"inner\",left_on='accession',right_on=\"seqid\")[['cluster','seqid','tag']]\n",
    "print(seq2cluster.shape)\n",
    "display(seq2cluster.head(2))\n",
    "seq2cluster.to_csv('./Temp/seq2cluster.csv',index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-11-14T15:53:17.423563Z",
     "start_time": "2021-11-14T15:53:17.402420Z"
    },
    "hidden": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cluster</th>\n",
       "      <th>seqid</th>\n",
       "      <th>tag</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>9422</th>\n",
       "      <td>C0</td>\n",
       "      <td>cst-miR11332</td>\n",
       "      <td>cst-miR11332 MIMAT0044622 Cucumis sativus miR1...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7002</th>\n",
       "      <td>C1</td>\n",
       "      <td>stu-miR7994b-5p</td>\n",
       "      <td>stu-miR7994b-5p MIMAT0031188 Solanum tuberosum...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     cluster            seqid  \\\n",
       "9422      C0     cst-miR11332   \n",
       "7002      C1  stu-miR7994b-5p   \n",
       "\n",
       "                                                    tag  \n",
       "9422  cst-miR11332 MIMAT0044622 Cucumis sativus miR1...  \n",
       "7002  stu-miR7994b-5p MIMAT0031188 Solanum tuberosum...  "
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# todo: sorted first by cluster then by seqid\n",
    "seq2cluster.sort_values(\"cluster\").head(2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-11-14T15:53:18.409180Z",
     "start_time": "2021-11-14T15:53:18.205823Z"
    },
    "hidden": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(6028, 2)\n"
     ]
    }
   ],
   "source": [
    "df = fasta_to_df(\"./Temp/NR_mature_microRNA_queries.fasta\")\n",
    "df['tag'] = df['tag'].apply(lambda x : x.split(' ')[0])\n",
    "df = pd.merge(df,seq2cluster,how=\"inner\",left_on='tag',right_on=\"seqid\")[['cluster','data']]\n",
    "\n",
    "lines = []\n",
    "df.apply(lambda row: lines.append(f\">{row['cluster']}\\n{row['data']}\\n\"),axis=1)\n",
    "print(df.shape)\n",
    "with open('./Temp/BLASTn_queries.fasta','w') as file:\n",
    "    file.write(''.join(lines))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "heading_collapsed": true
   },
   "source": [
    "# BlastN"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "hidden": true
   },
   "source": [
    "!sudo apt-get install ncbi-blast+\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-11-14T15:53:56.806426Z",
     "start_time": "2021-11-14T15:53:45.294463Z"
    },
    "hidden": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\n",
      "Building a new DB, current time: 11/14/2021 19:23:46\n",
      "New DB name:   /home/jupyter/plant_microRNA_prediction/Temp/blastn_database\n",
      "New DB title:  input_genome.fna\n",
      "Sequence type: Nucleotide\n",
      "Keep MBits: T\n",
      "Maximum file size: 1000000000B\n",
      "Adding sequences from FASTA; added 126142 sequences in 9.84965 seconds.\n"
     ]
    }
   ],
   "source": [
    "!makeblastdb -in input_genome.fna \\\n",
    "             -dbtype nucl \\\n",
    "             -out ./Temp/blastn_database"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-11-15T17:35:22.926131Z",
     "start_time": "2021-11-15T17:35:22.922603Z"
    },
    "hidden": true
   },
   "outputs": [],
   "source": [
    "header = 'qseqid sseqid qstart qend sstart send qseq sseq evalue bitscore score length pident nident mismatch positive gapopen gaps ppos frames qframe sframe sstrand qcovs qcovhsp qlen slen'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-11-15T17:37:49.644999Z",
     "start_time": "2021-11-15T17:35:24.046004Z"
    },
    "hidden": true
   },
   "outputs": [],
   "source": [
    "!blastn -query ./Temp/BLASTn_queries.fasta \\\n",
    "        -out ./Temp/BLASTn_result \\\n",
    "        -num_threads {mp.cpu_count()} \\\n",
    "        -db ./Temp/blastn_database \\\n",
    "        -word_size 7 \\\n",
    "        -penalty -3 \\\n",
    "        -reward 2 \\\n",
    "        -gapopen 5 \\\n",
    "        -gapextend 2 \\\n",
    "        -outfmt '6 qseqid sseqid qstart qend sstart send qseq sseq evalue bitscore score length pident nident mismatch positive gapopen gaps ppos frames qframe sframe sstrand qcovs qcovhsp qlen slen'       "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-11-15T17:38:24.524868Z",
     "start_time": "2021-11-15T17:38:23.190352Z"
    },
    "hidden": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(326849, 27)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>qseqid</th>\n",
       "      <th>sseqid</th>\n",
       "      <th>qstart</th>\n",
       "      <th>qend</th>\n",
       "      <th>sstart</th>\n",
       "      <th>send</th>\n",
       "      <th>qseq</th>\n",
       "      <th>sseq</th>\n",
       "      <th>evalue</th>\n",
       "      <th>bitscore</th>\n",
       "      <th>...</th>\n",
       "      <th>gaps</th>\n",
       "      <th>ppos</th>\n",
       "      <th>frames</th>\n",
       "      <th>qframe</th>\n",
       "      <th>sframe</th>\n",
       "      <th>sstrand</th>\n",
       "      <th>qcovs</th>\n",
       "      <th>qcovhsp</th>\n",
       "      <th>qlen</th>\n",
       "      <th>slen</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>C5495</td>\n",
       "      <td>AMWY02099822.1</td>\n",
       "      <td>1</td>\n",
       "      <td>20</td>\n",
       "      <td>1769</td>\n",
       "      <td>1750</td>\n",
       "      <td>TGACAGAAGAGAGTGAGCAC</td>\n",
       "      <td>TGACAGAAGAGAGTGAGCAC</td>\n",
       "      <td>0.004</td>\n",
       "      <td>37.4</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>100.0</td>\n",
       "      <td>1/-1</td>\n",
       "      <td>1</td>\n",
       "      <td>-1</td>\n",
       "      <td>minus</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>20</td>\n",
       "      <td>3308</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>C5495</td>\n",
       "      <td>AMWY02082313.1</td>\n",
       "      <td>1</td>\n",
       "      <td>20</td>\n",
       "      <td>5954</td>\n",
       "      <td>5973</td>\n",
       "      <td>TGACAGAAGAGAGTGAGCAC</td>\n",
       "      <td>TGACAGAAGAGAGTGAGCAC</td>\n",
       "      <td>0.004</td>\n",
       "      <td>37.4</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>100.0</td>\n",
       "      <td>1/1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>plus</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>20</td>\n",
       "      <td>8471</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2 rows × 27 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  qseqid          sseqid  qstart  qend  sstart  send                  qseq  \\\n",
       "0  C5495  AMWY02099822.1       1    20    1769  1750  TGACAGAAGAGAGTGAGCAC   \n",
       "1  C5495  AMWY02082313.1       1    20    5954  5973  TGACAGAAGAGAGTGAGCAC   \n",
       "\n",
       "                   sseq  evalue  bitscore  ...  gaps   ppos  frames  qframe  \\\n",
       "0  TGACAGAAGAGAGTGAGCAC   0.004      37.4  ...     0  100.0    1/-1       1   \n",
       "1  TGACAGAAGAGAGTGAGCAC   0.004      37.4  ...     0  100.0     1/1       1   \n",
       "\n",
       "   sframe  sstrand  qcovs  qcovhsp  qlen  slen  \n",
       "0      -1    minus    100      100    20  3308  \n",
       "1       1     plus    100      100    20  8471  \n",
       "\n",
       "[2 rows x 27 columns]"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_blastn = pd.read_csv('./Temp/BLASTn_result', sep='\\t',header=None)\n",
    "df_blastn.columns = header.replace(\"  \",\" \").split(\" \")\n",
    "print(df_blastn.shape)\n",
    "df_blastn.head(2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-11-15T17:38:25.602420Z",
     "start_time": "2021-11-15T17:38:25.482593Z"
    },
    "hidden": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(80217, 28)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>qseqid</th>\n",
       "      <th>sseqid</th>\n",
       "      <th>qstart</th>\n",
       "      <th>qend</th>\n",
       "      <th>sstart</th>\n",
       "      <th>send</th>\n",
       "      <th>qseq</th>\n",
       "      <th>sseq</th>\n",
       "      <th>evalue</th>\n",
       "      <th>bitscore</th>\n",
       "      <th>...</th>\n",
       "      <th>ppos</th>\n",
       "      <th>frames</th>\n",
       "      <th>qframe</th>\n",
       "      <th>sframe</th>\n",
       "      <th>sstrand</th>\n",
       "      <th>qcovs</th>\n",
       "      <th>qcovhsp</th>\n",
       "      <th>qlen</th>\n",
       "      <th>slen</th>\n",
       "      <th>Nonconformity</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>C5495</td>\n",
       "      <td>AMWY02099822.1</td>\n",
       "      <td>1</td>\n",
       "      <td>20</td>\n",
       "      <td>1769</td>\n",
       "      <td>1750</td>\n",
       "      <td>TGACAGAAGAGAGTGAGCAC</td>\n",
       "      <td>TGACAGAAGAGAGTGAGCAC</td>\n",
       "      <td>0.004</td>\n",
       "      <td>37.4</td>\n",
       "      <td>...</td>\n",
       "      <td>100.0</td>\n",
       "      <td>1/-1</td>\n",
       "      <td>1</td>\n",
       "      <td>-1</td>\n",
       "      <td>minus</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>20</td>\n",
       "      <td>3308</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>C5495</td>\n",
       "      <td>AMWY02082313.1</td>\n",
       "      <td>1</td>\n",
       "      <td>20</td>\n",
       "      <td>5954</td>\n",
       "      <td>5973</td>\n",
       "      <td>TGACAGAAGAGAGTGAGCAC</td>\n",
       "      <td>TGACAGAAGAGAGTGAGCAC</td>\n",
       "      <td>0.004</td>\n",
       "      <td>37.4</td>\n",
       "      <td>...</td>\n",
       "      <td>100.0</td>\n",
       "      <td>1/1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>plus</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>20</td>\n",
       "      <td>8471</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2 rows × 28 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  qseqid          sseqid  qstart  qend  sstart  send                  qseq  \\\n",
       "0  C5495  AMWY02099822.1       1    20    1769  1750  TGACAGAAGAGAGTGAGCAC   \n",
       "1  C5495  AMWY02082313.1       1    20    5954  5973  TGACAGAAGAGAGTGAGCAC   \n",
       "\n",
       "                   sseq  evalue  bitscore  ...   ppos  frames  qframe  sframe  \\\n",
       "0  TGACAGAAGAGAGTGAGCAC   0.004      37.4  ...  100.0    1/-1       1      -1   \n",
       "1  TGACAGAAGAGAGTGAGCAC   0.004      37.4  ...  100.0     1/1       1       1   \n",
       "\n",
       "   sstrand  qcovs  qcovhsp  qlen  slen Nonconformity  \n",
       "0    minus    100      100    20  3308             0  \n",
       "1     plus    100      100    20  8471             0  \n",
       "\n",
       "[2 rows x 28 columns]"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "threshold = 4\n",
    "df_blastn['Nonconformity'] = df_blastn['qlen'] - (abs(df_blastn['qend'] - df_blastn['qstart']) + 1) + df_blastn['gaps'] + df_blastn['mismatch']\n",
    "df_blastn = df_blastn[df_blastn['Nonconformity'] <= threshold]\n",
    "print(df_blastn.shape)\n",
    "df_blastn.head(2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-11-15T17:38:27.384345Z",
     "start_time": "2021-11-15T17:38:26.466806Z"
    },
    "hidden": true,
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(66445, 28)\n"
     ]
    }
   ],
   "source": [
    "# remore redundancy and hold best one base of Nonconformity value\n",
    "df_blastn = df_blastn.sort_values([\"Nonconformity\", \"evalue\"], ascending = (True, True))\n",
    "df_blastn = df_blastn.drop_duplicates(subset=['sseqid','sstart', 'send','sstrand'], keep='first')\n",
    "df_blastn.to_csv('./Temp/filtered_out_blastn.csv')\n",
    "print(df_blastn.shape)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "heading_collapsed": true
   },
   "source": [
    "# Result of the blastn to bed file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-11-15T17:38:28.652838Z",
     "start_time": "2021-11-15T17:38:28.636993Z"
    },
    "hidden": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "<ipython-input-22-d52c72306552>:3: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  df['ones'] = 1\n"
     ]
    }
   ],
   "source": [
    "flanking_value = 200\n",
    "df = df_blastn[['sseqid', 'sstart', 'send', 'sstrand','slen']]\n",
    "df['ones'] = 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-11-15T17:38:39.483787Z",
     "start_time": "2021-11-15T17:38:28.981816Z"
    },
    "hidden": true
   },
   "outputs": [],
   "source": [
    "def switch(row):\n",
    "    if(row['sstart'] > row['send']):        \n",
    "        temp = row['sstart']\n",
    "        row['sstart'] = row['send']\n",
    "        row['send'] = temp\n",
    "    return row\n",
    "df = df.apply(lambda row: switch(row), axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-11-15T17:38:39.513215Z",
     "start_time": "2021-11-15T17:38:39.487160Z"
    },
    "hidden": true
   },
   "outputs": [],
   "source": [
    "def convert(inp):\n",
    "    if(inp == \"plus\"):\n",
    "        return \"forward\"\n",
    "    if(inp == \"minus\"):\n",
    "        return \"reverse\"\n",
    "    raise Exception('Error, sstrand contains illegal word! only \"plus\" and \"minus\" are allowed')\n",
    "df['strand'] = df['sstrand'].apply(lambda x: convert(x))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-11-15T17:38:39.541625Z",
     "start_time": "2021-11-15T17:38:39.516766Z"
    },
    "hidden": true
   },
   "outputs": [],
   "source": [
    "def convert2sign(inp):\n",
    "    if(inp == \"plus\"):\n",
    "        return \"+\"\n",
    "    if(inp == \"minus\"):\n",
    "        return \"-\"\n",
    "    raise Exception('Error, sstrand contains illegal word! only \"plus\" and \"minus\" are allowed')\n",
    "df['sign'] = df['sstrand'].apply(lambda x: convert2sign(x))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-11-15T17:38:41.509655Z",
     "start_time": "2021-11-15T17:38:39.544832Z"
    },
    "hidden": true
   },
   "outputs": [],
   "source": [
    "df['hit_length'] = df.apply(lambda row: abs(row['send'] - row['sstart']) + 1 ,axis=1)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "heading_collapsed": true,
    "hidden": true
   },
   "source": [
    "## convert sstart and send from location to index (range)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-11-15T17:38:41.549426Z",
     "start_time": "2021-11-15T17:38:41.512935Z"
    },
    "hidden": true
   },
   "outputs": [],
   "source": [
    "df['sstart'] = df['sstart'].apply(lambda x: x - 1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-11-15T17:38:41.588656Z",
     "start_time": "2021-11-15T17:38:41.552256Z"
    },
    "hidden": true
   },
   "outputs": [],
   "source": [
    "df['downstream_flanking'] = df['sstart'].apply(lambda x:  flanking_value if x > flanking_value else x)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-11-15T17:38:43.613893Z",
     "start_time": "2021-11-15T17:38:41.591319Z"
    },
    "hidden": true
   },
   "outputs": [],
   "source": [
    "df['upstream_flanking'] = df.apply(lambda row:  flanking_value if (row['send']+flanking_value) <= row['slen'] else row['slen'] - row['send'],axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-11-15T17:38:45.487406Z",
     "start_time": "2021-11-15T17:38:43.619391Z"
    },
    "hidden": true
   },
   "outputs": [],
   "source": [
    "df['hit_start'] = df.apply(lambda row: row['downstream_flanking'] if row['sign'] == \"+\" else row['upstream_flanking'],axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-11-15T17:38:47.794296Z",
     "start_time": "2021-11-15T17:38:45.490590Z"
    },
    "hidden": true
   },
   "outputs": [],
   "source": [
    "df['hit_end'] = df.apply(lambda row: row['downstream_flanking'] + row['hit_length'] if row['sign'] == \"+\" else row['upstream_flanking'] + row['hit_length'],axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-11-15T17:38:49.856166Z",
     "start_time": "2021-11-15T17:38:47.797824Z"
    },
    "hidden": true
   },
   "outputs": [],
   "source": [
    "df['sstart'] = df['sstart'].apply(lambda x: max(x - flanking_value, 0))\n",
    "df['send'] = df.apply(lambda row: min(row['send'] + flanking_value , row['slen']),axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-11-15T17:38:53.067352Z",
     "start_time": "2021-11-15T17:38:49.859005Z"
    },
    "hidden": true
   },
   "outputs": [],
   "source": [
    "df['tag'] = df.apply(lambda row: f\">{row['sseqid']}:{row['sstart']}-{row['send']}({row['sign']})\",axis=1)\n",
    "df['reformated_tag'] = df['tag'].apply(lambda t: reformat(t))\n",
    "df[['tag', 'reformated_tag', 'hit_start', 'hit_end']].to_csv('./Temp/hit_index_info.csv')#, index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-11-15T17:38:56.908475Z",
     "start_time": "2021-11-15T17:38:53.071253Z"
    },
    "hidden": true
   },
   "outputs": [],
   "source": [
    "df['location_tag'] = df.apply(lambda row: f\">{row['sseqid']}|{row['sign']}|{row['sstart'] + 1}-{row['send']}|{row['hit_start']+1}-{row['hit_end']}\",axis=1)\n",
    "df[['location_tag']].to_csv('./Temp/pipe_seprated_location_list.csv',index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-11-15T17:38:57.133643Z",
     "start_time": "2021-11-15T17:38:56.911402Z"
    },
    "hidden": true
   },
   "outputs": [],
   "source": [
    "df[['sseqid','sstart','send','strand','ones', 'sign']].to_csv('./Temp/extension_index.bed', \n",
    "        index=False, header=False, sep=\"\\t\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "heading_collapsed": true
   },
   "source": [
    "# Extention\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-11-15T17:38:57.139683Z",
     "start_time": "2021-11-15T17:38:57.136568Z"
    },
    "hidden": true
   },
   "outputs": [],
   "source": [
    "# !sudo apt-get install bedtools"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-11-15T17:39:00.272370Z",
     "start_time": "2021-11-15T17:38:57.142577Z"
    },
    "hidden": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "index file ./input_genome.fna.fai not found, generating...\r\n"
     ]
    }
   ],
   "source": [
    "!bedtools getfasta -fi ./input_genome.fna -fo ./Temp/extended_original.txt -s -bed ./Temp/extension_index.bed\n",
    "!rm input_genome.fna.fai"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-11-15T14:45:18.658125Z",
     "start_time": "2021-11-15T14:45:18.653842Z"
    },
    "code_folding": [],
    "hidden": true
   },
   "outputs": [],
   "source": [
    "# todo: remove duplicated\n",
    "'''\n",
    "df = fasta_to_df(\"./Temp/extended.txt\")\n",
    "df = df.drop_duplicates(subset=['tag'], keep='first')\n",
    "df_to_fasta(df,\"./Temp/extended.txt\")\n",
    "len(df['tag'].unique())\n",
    "''';"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "heading_collapsed": true
   },
   "source": [
    "# Convert hit region to upper case and other region to lower case"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-11-15T17:39:00.649102Z",
     "start_time": "2021-11-15T17:39:00.276411Z"
    },
    "hidden": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(66445, 5)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>tag</th>\n",
       "      <th>reformated_tag</th>\n",
       "      <th>hit_start</th>\n",
       "      <th>hit_end</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>132836</td>\n",
       "      <td>AMWY02059828.1:2832-3256(+)</td>\n",
       "      <td>&gt;AMWY020598281_2832-3256_+_</td>\n",
       "      <td>200</td>\n",
       "      <td>224</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>300170</td>\n",
       "      <td>AMWY02004761.1:1853-2277(+)</td>\n",
       "      <td>&gt;AMWY020047611_1853-2277_+_</td>\n",
       "      <td>200</td>\n",
       "      <td>224</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Unnamed: 0                          tag               reformated_tag  \\\n",
       "0      132836  AMWY02059828.1:2832-3256(+)  >AMWY020598281_2832-3256_+_   \n",
       "1      300170  AMWY02004761.1:1853-2277(+)  >AMWY020047611_1853-2277_+_   \n",
       "\n",
       "   hit_start  hit_end  \n",
       "0        200      224  \n",
       "1        200      224  "
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ext = fasta_to_df('./Temp/extended_original.txt')\n",
    "info = pd.read_csv('./Temp/hit_index_info.csv')\n",
    "info['tag'] = info['tag'].apply(lambda x: x[1:])\n",
    "print(info.shape)\n",
    "info.head(2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-11-15T17:39:12.591397Z",
     "start_time": "2021-11-15T17:39:00.652410Z"
    },
    "hidden": true
   },
   "outputs": [],
   "source": [
    "ext = ext.sort_values(by=['tag']).reset_index()\n",
    "ext['help_tag'] = ext.apply(lambda r: r['tag']+ str(r.name),axis=1)\n",
    "del ext['tag']\n",
    "\n",
    "info = info.sort_values(by=['tag']).reset_index()\n",
    "info['help_tag'] = info.apply(lambda row: row['tag']+ str(row.name),axis=1)\n",
    "def redefined_tag(row):\n",
    "    tag = row['tag']\n",
    "    [sstart, send] = tag.split(':')[-1].split('(')[0].split('-')\n",
    "    sstart = int(sstart) + 1\n",
    "    sign = tag.split('(')[-1].split(')')[0]    \n",
    "    return f\"{tag.split(':')[0]}|{sstart}-{send}|{row['hit_start']+1}-{row['hit_end']}|{sign}\"\n",
    "info['tag'] = info.apply(lambda row: redefined_tag(row),axis=1)\n",
    "ext = pd.merge(ext,info,how='inner', on='help_tag')\n",
    "\n",
    "def emphasis_hit(row):\n",
    "    seq = list(row['data'].lower())            \n",
    "    s = row['hit_start']\n",
    "    e = row['hit_end']\n",
    "    seq[s:e] = list(''.join(seq[s:e]).upper())    \n",
    "    return ''.join(seq)\n",
    "    \n",
    "ext['data'] = ext.apply(lambda row: emphasis_hit(row),axis=1)\n",
    "df_to_fasta(ext[['tag','data']],\"./Temp/extended_modified.txt\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "heading_collapsed": true
   },
   "source": [
    "# Extended validation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "hidden": true
   },
   "outputs": [],
   "source": [
    "df_blastn['hit'] = df_blastn['sseq'].apply(lambda x: x.replace('-', ''))\n",
    "info = pd.read_csv('./Temp/hit_index_info.csv')\n",
    "ext = fasta_to_df('./Temp/extended2.txt')\n",
    "\n",
    "counter = 0\n",
    "for index in df_blastn.index:\n",
    "    hit = df_blastn['hit'][index]\n",
    "    row = info[info['Unnamed: 0']== index].reset_index()\n",
    "    tag = row['tag'][0][1:]\n",
    "    hs = row['hit_start'][0]\n",
    "    he = row['hit_end'][0]        \n",
    "    tag = tag.replace('(+)',f\"|{hs}-{he}(+)\")\n",
    "    tag = tag.replace('(-)',f\"|{hs}-{he}(-)\")            \n",
    "    seq = ext[ext['tag']==tag]['data'].iloc[0]\n",
    "    seq = ext[ext['tag']==tag]['data'].iloc[0]    \n",
    "    if(seq[hs:he] != hit):\n",
    "        print(tag, df_blastn['slen'][index])\n",
    "        print(seq[hs:he])\n",
    "        print(hit)\n",
    "        print('\\n\\n')\n",
    "        counter += 1                "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 178,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-11-12T16:55:04.208851Z",
     "start_time": "2021-11-12T16:55:03.831213Z"
    },
    "hidden": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>tag</th>\n",
       "      <th>data</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AMWY02000003.1:4084-4506(+)</td>\n",
       "      <td>ATACAATTGTCACATagtttacattattaatttccgcttaatttat...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AMWY02000003.1:4391-4810(+)</td>\n",
       "      <td>ttctaattgcaaatttatgttatatttttaaatagaaaggGAGATT...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                           tag  \\\n",
       "0  AMWY02000003.1:4084-4506(+)   \n",
       "1  AMWY02000003.1:4391-4810(+)   \n",
       "\n",
       "                                                data  \n",
       "0  ATACAATTGTCACATagtttacattattaatttccgcttaatttat...  \n",
       "1  ttctaattgcaaatttatgttatatttttaaatagaaaggGAGATT...  "
      ]
     },
     "execution_count": 178,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = pd.read_csv('check1.csv')\n",
    "df.columns = ['tag', 'data']\n",
    "def do(x):\n",
    "    x = x[1:]\n",
    "    x = x.split('|')\n",
    "    s = int(x[2].split('-')[0]) - 1\n",
    "    e = x[2].split('-')[1]\n",
    "    return f\"{x[0]}:{s}-{e}({x[1]})\"\n",
    "df['tag'] = df['tag'].apply(lambda x: do(x))\n",
    "df.head(2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 176,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-11-12T16:54:03.716073Z",
     "start_time": "2021-11-12T16:47:01.709169Z"
    },
    "hidden": true
   },
   "outputs": [],
   "source": [
    "for t in df['tag']:\n",
    "    check_seq = df[df['tag']==t]['data'].iloc[0]\n",
    "    ext_seq = ext[ext['tag']==t]['data'].iloc[0]\n",
    "    if(check_seq != ext_seq):\n",
    "        print('error')        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 167,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-11-12T16:38:20.621448Z",
     "start_time": "2021-11-12T16:38:20.428902Z"
    },
    "hidden": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>tag</th>\n",
       "      <th>data</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AMWY02059828.1:2832-3256(+)</td>\n",
       "      <td>AAAGAATCAGCAATGGAAAAATAACCGGTTCTTAATTCAGcataac...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AMWY02004761.1:1853-2277(+)</td>\n",
       "      <td>actaataatgCATGGCCATATATATCAAATCTACCATATgccattt...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                           tag  \\\n",
       "0  AMWY02059828.1:2832-3256(+)   \n",
       "1  AMWY02004761.1:1853-2277(+)   \n",
       "\n",
       "                                                data  \n",
       "0  AAAGAATCAGCAATGGAAAAATAACCGGTTCTTAATTCAGcataac...  \n",
       "1  actaataatgCATGGCCATATATATCAAATCTACCATATgccattt...  "
      ]
     },
     "execution_count": 167,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ext = fasta_to_df('./Temp/extended.txt')\n",
    "ext.head(2)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "heading_collapsed": true
   },
   "source": [
    "# RNA 2d prediction"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "hidden": true
   },
   "source": [
    "## Mfold"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-11-18T16:23:27.830630Z",
     "start_time": "2021-11-18T16:23:27.826922Z"
    },
    "hidden": true
   },
   "outputs": [],
   "source": [
    "'''\n",
    "# installation\n",
    "!wget http://www.unafold.org/download/mfold-3.6.tar.gz\n",
    "!tar -xvf ./mfold-3.6.tar.gz; rm ./mfold-3.6.tar.gz\n",
    "%cd ./mfold-3.6\n",
    "!./configure\n",
    "!make\n",
    "!make install\n",
    "%cd ..\n",
    "!sudo apt install texlive-font-utils\n",
    "''';"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-11-18T16:23:28.114502Z",
     "start_time": "2021-11-18T16:23:28.111002Z"
    },
    "hidden": true
   },
   "outputs": [],
   "source": [
    "#todo : add all hyperparameter(options) to GUI"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-11-18T16:24:02.726047Z",
     "start_time": "2021-11-18T16:24:02.077975Z"
    },
    "hidden": true
   },
   "outputs": [],
   "source": [
    "counter = 0\n",
    "base = \"./secondary_structure/mfold/\"\n",
    "!rm -r {base}\n",
    "!mkdir -p {base}\n",
    "df = fasta_to_df('./Temp/extended_modified.txt')\n",
    "\n",
    "for index, row in df.iterrows():    \n",
    "    tag = reformat(row['tag'])\n",
    "    if(not os.path.exists(base + tag)):\n",
    "        os.makedirs(base + tag)            \n",
    "    with open(base + f\"{tag}/SEQ.FASTA\",'w') as file:\n",
    "        file.write(f\">{row['tag']}\\n{row['data']}\")\n",
    "    counter += 1    \n",
    "    if(counter >= 200):\n",
    "        break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-11-18T16:25:50.356513Z",
     "start_time": "2021-11-18T16:24:05.915818Z"
    },
    "hidden": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/home/jupyter/plant_microRNA_prediction/secondary_structure/mfold/AMWY020000031|4085-4506|201-222|+/home/jupyter/plant_microRNA_prediction/secondary_structure/mfold/AMWY020000261|484-903|201-220|+/home/jupyter/plant_microRNA_prediction/secondary_structure/mfold/AMWY020000391|321-573|201-218|+/home/jupyter/plant_microRNA_prediction/secondary_structure/mfold/AMWY020000141|194-614|201-221|+/home/jupyter/plant_microRNA_prediction/secondary_structure/mfold/AMWY020000471|1432-1851|201-220|-/home/jupyter/plant_microRNA_prediction/secondary_structure/mfold/AMWY020000511|4616-5029|201-214|-/home/jupyter/plant_microRNA_prediction/secondary_structure/mfold/AMWY020000531|455-874|201-220|+/home/jupyter/plant_microRNA_prediction/secondary_structure/mfold/AMWY020000231|268-683|201-216|-/home/jupyter/plant_microRNA_prediction/secondary_structure/mfold/AMWY020000311|5367-5786|201-220|+/home/jupyter/plant_microRNA_prediction/secondary_structure/mfold/AMWY020000881|2344-2731|201-219|+/home/jupyter/plant_microRNA_prediction/secondary_structure/mfold/AMWY020000631|238-660|201-223|+/home/jupyter/plant_microRNA_prediction/secondary_structure/mfold/AMWY020000811|7360-7778|201-219|-/home/jupyter/plant_microRNA_prediction/secondary_structure/mfold/AMWY020000501|1-282|201-219|-/home/jupyter/plant_microRNA_prediction/secondary_structure/mfold/AMWY020000811|6067-6488|201-222|+/home/jupyter/plant_microRNA_prediction/secondary_structure/mfold/AMWY020000791|1166-1585|201-220|-/home/jupyter/plant_microRNA_prediction/secondary_structure/mfold/AMWY020000571|1714-2132|201-219|-/home/jupyter/plant_microRNA_prediction/secondary_structure/mfold/AMWY020000611|618-947|114-130|-/home/jupyter/plant_microRNA_prediction/secondary_structure/mfold/AMWY020001171|151-568|201-218|-/home/jupyter/plant_microRNA_prediction/secondary_structure/mfold/AMWY020000931|2123-2541|201-219|+\n",
      "/home/jupyter/plant_microRNA_prediction/secondary_structure/mfold/AMWY020001081|1319-1736|201-218|-/home/jupyter/plant_microRNA_prediction/secondary_structure/mfold/AMWY020001171|72-490|201-219|-/home/jupyter/plant_microRNA_prediction/secondary_structure/mfold/AMWY020001361|2266-2684|201-219|-/home/jupyter/plant_microRNA_prediction/secondary_structure/mfold/AMWY020001351|2646-3065|201-220|-\n",
      "\n",
      "/home/jupyter/plant_microRNA_prediction/secondary_structure/mfold/AMWY020001221|102-520|201-219|-\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "mfold version 3.6\n",
      "mfold version 3.6\n",
      "mfold version 3.6\n",
      "mfold version 3.6\n",
      "mfold version 3.6\n",
      "mfold version 3.6\n",
      "mfold version 3.6\n",
      "mfold version 3.6\n",
      "mfold version 3.6\n",
      "mfold version 3.6\n",
      "mfold version 3.6\n",
      "mfold version 3.6\n",
      "mfold version 3.6\n",
      "mfold version 3.6\n",
      "mfold version 3.6\n",
      "mfold version 3.6\n",
      "mfold version 3.6\n",
      "mfold version 3.6\n",
      "mfold version 3.6\n",
      "mfold version 3.6\n",
      "mfold version 3.6\n",
      "mfold version 3.6\n",
      "mfold version 3.6\n",
      "mfold version 3.6\n",
      "REUSE= NO\n",
      "REUSE= NO\n",
      "REUSE= NO\n",
      "REUSE= NO\n",
      "REUSE= NO\n",
      "REUSE= NO\n",
      "REUSE= NO\n",
      "REUSE= NO\n",
      "REUSE= NO\n",
      "REUSE= NO\n",
      "REUSE= NO\n",
      "REUSE= NO\n",
      "REUSE= NO\n",
      "REUSE= NO\n",
      "REUSE= NO\n",
      "REUSE= NO\n",
      "REUSE= NO\n",
      "REUSE= NO\n",
      "REUSE= NO\n",
      "REUSE= NO\n",
      "REUSE= NO\n",
      "REUSE= NO\n",
      "REUSE= NO\n",
      "REUSE= NO\n",
      "SEQ.pnt created.SEQ.pnt created.\n",
      "SEQ.pnt created.\n",
      "SEQ.pnt created.\n",
      "SEQ.pnt created.\n",
      "SEQ.pnt created.\n",
      "SEQ.pnt created.\n",
      "SEQ.pnt created.\n",
      "SEQ.pnt created.\n",
      "SEQ.pnt created.\n",
      "Sequence length is 420\n",
      "SEQ.pnt created.\n",
      "SEQ.pnt created.\n",
      "SEQ.pnt created.\n",
      "SEQ.pnt created.\n",
      "Sequence length is 282\n",
      "SEQ.pnt created.\n",
      "Sequence length is 421\n",
      "SEQ.pnt created.\n",
      "Sequence length is 253\n",
      "SEQ.pnt created.\n",
      "Sequence length is 420\n",
      "SEQ.pnt created.\n",
      "Sequence length is 419\n",
      "SEQ.pnt created.\n",
      "Sequence length is 414\n",
      "SEQ.pnt created.\n",
      "Sequence length is 420\n",
      "SEQ.pnt created.\n",
      "Sequence length is 422\n",
      "SEQ.pnt created.\n",
      "Sequence length is 418\n",
      "SEQ.pnt created.\n",
      "Sequence length is 416\n",
      "SEQ.pnt created.\n",
      "Sequence length is 420\n",
      "\n",
      "Sequence length is 330\n",
      "Sequence length is 422\n",
      "Sequence length is 423\n",
      "Sequence length is 419\n",
      "Sequence length is 418\n",
      "Sequence length is 388\n",
      "Sequence length is 420\n",
      "Sequence length is 419\n",
      "Sequence length is 420\n",
      "Sequence length is 419\n",
      "Sequence length is 419\n",
      "Sequence length is 419\n",
      "RNA free energy files (version 2.3) at 20 degrees created.\n",
      "RNA free energy files (version 2.3) at 20 degrees created.\n",
      "RNA free energy files (version 2.3) at 20 degrees created.\n",
      "RNA free energy files (version 2.3) at 20 degrees created.\n",
      "RNA free energy files (version 2.3) at 20 degrees created.\n",
      "RNA free energy files (version 2.3) at 20 degrees created.\n",
      "RNA free energy files (version 2.3) at 20 degrees created.\n",
      "RNA free energy files (version 2.3) at 20 degrees created.\n",
      "RNA free energy files (version 2.3) at 20 degrees created.\n",
      "RNA free energy files (version 2.3) at 20 degrees created.\n",
      "RNA free energy files (version 2.3) at 20 degrees created.\n",
      "RNA free energy files (version 2.3) at 20 degrees created.\n",
      "RNA free energy files (version 2.3) at 20 degrees created.\n",
      "RNA free energy files (version 2.3) at 20 degrees created.\n",
      "RNA free energy files (version 2.3) at 20 degrees created.\n",
      "RNA free energy files (version 2.3) at 20 degrees created.\n",
      "RNA free energy files (version 2.3) at 20 degrees created.\n",
      "RNA free energy files (version 2.3) at 20 degrees created.\n",
      "RNA free energy files (version 2.3) at 20 degrees created.\n",
      "RNA free energy files (version 2.3) at 20 degrees created.\n",
      "RNA free energy files (version 2.3) at 20 degrees created.\n",
      "RNA free energy files (version 2.3) at 20 degrees created.\n",
      "RNA free energy files (version 2.3) at 20 degrees created.\n",
      "RNA free energy files (version 2.3) at 20 degrees created.\n",
      "10,20,30,40,10,20,30,10,20,30,40,10,20,30,40,50,10,20,30,40,10,20,30,40,50,10,20,30,40,10,20,30,40,50,10,20,30,40,50,10,20,30,40,50,60,10,20,30,40,50,60,70,10,20,30,40,50,10,20,30,40,50,60,70,10,20,30,40,50,60,70,10,20,30,40,50,60,10,20,30,40,50,60,70,10,20,30,40,50,60,10,20,30,40,50,60,70,10,20,30,40,50,60,70,10,20,30,40,50,60,70,80,10,20,30,40,50,60,70,50,60,70,10,20,30,40,50,60,10,20,30,40,50,10,20,30,40,50,60,70,50,60,70,40,50,60,70,70,80,90,60,70,80,50,60,70,60,70,80,90,50,60,70,80,60,70,80,90,80,90,100,60,70,80,80,90,100,80,90,100,70,80,90,60,70,80,90,60,70,80,80,90,100,80,90,100,110,70,80,90,90,100,80,90,70,80,90,80,90,100,80,90,100,110,80,90,80,90,100,100,110,80,90,100,90,100,110,80,90,100,90,100,110,90,100,110,120,130,100,110,120,110,120,100,90,100,110,120,100,110,120,130,140,110,100,110,120,120,130,110,120,130,110,120,130,110,120,110,120,120,100,110,120,110,120,100,110,100,110,100,110,110,140,150,130,140,120,130,130,140,110,120,110,120,130,120,130,130,140,140,120,130,140,150,130,140,130,140,120,130,150,160,170,140,150,130,140,120,130,140,150,130,140,140,150,150,130,140,120,130,120,130,160,170,150,160,170,120,130,140,150,150,130,180,140,140,160,140,150,160,140,150,130,140,160,170,150,160,170,150,160,170,150,160,180,160,150,160,140,150,150,190,200,150,160,180,190,140,150,140,150,160,160,150,160,150,180,170,180,160,170,160,160,170,180,160,170,170,180,170,180,190,180,190,180,190,200,150,160,170,180,210,190,200,170,170,180,160,200,210,220,230,240,160,170,160,170,180,170,160,190,190,190,170,180,210,220,230,180,200,170,170,180,190,200,180,170,190,180,220,230,190,200,210,210,220,210,220,180,190,180,190,170,180,190,200,190,200,180,190,200,210,190,170,180,240,250,260,270,230,240,250,190,250, \n",
      "End of Fill\n",
      "Save file created using nafold.\n",
      "190,200,200,210,200,200,210,200,180,190,200,230,210,220,230,240,220,190,200,200,200,210,190,190,280, \n",
      "End of Fill\n",
      "Save file created using nafold.\n",
      "Minimum folding energy is -101.20 kcal/mol.\n",
      "Energy increment is 5.06 kcal/mol.\n",
      "210,200,210,240,250,210,210,210,220,230,220,230,210,200,220,230,200,210,220,230,240,250,220,230,240,230,240,260,270,210,Minimum folding energy is -71.80 kcal/mol.\n",
      "Energy increment is 3.59 kcal/mol.\n",
      "H-num file created from plot file.\n",
      "200,H-num file created from plot file.\n",
      "210,220,210,220,220,220,220,230,260,200,220,230,250,260,250,260,240,240,240,250,220,230,210,280,290,300,210,250,260,260,270,220,230,240,210,220,270,280,290,230,220,1,2,3,4,5,6,7,240,250,260,230,240,230,270,280,1,2,3,4,5,6,7,8,9, \n",
      "260,250,270,240,220,220,240,240,240,250,250,290,300,270,280,290,270,250,260,280,8,9,10,11,12, \n",
      "310,320,230,230,230,300,310,320,280,250,260,270,240,250,260,270,260,250,250,260,230,240,240,300,310,320, \n",
      "End of Fill\n",
      "310,320,240,280,290,290,300,230,260,270,280,250,290,300,330,340,350,360,280,290,240,280,290,330,340,350,360,270,260,270,270,240,270,280,290,330,340,350,360,260,270,280,290,270,310,320,260,270,250,260,Suboptimal foldings created.\n",
      "250,260,370,380,390,400,410, \n",
      "End of Fill\n",
      "Save file created using nafold.\n",
      "290,300,310,300,310,310,320,330,250,300,310,Save file created using nafold.\n",
      "Minimum folding energy is -115.90 kcal/mol.\n",
      "Energy increment is 5.79 kcal/mol.\n",
      "300,310,280,280,290,250,370,380, \n",
      "End of Fill\n",
      "Save file created using nafold.\n",
      "Minimum folding energy is -114.60 kcal/mol.\n",
      "280,250,300,280,290,300,Energy dot plot created.\n",
      "1\t270,280,290,370,380,390,400,410, \n",
      "End of Fill\n",
      "270,330,340,350,360,320,330,340,350,260,270,340,350,360,320,330,Minimum folding energy is -86.90 kcal/mol.\n",
      "Energy increment is 4.34 kcal/mol.\n",
      "320,330,340,290,H-num file created from plot file.\n",
      "Energy increment is 5.73 kcal/mol.\n",
      "260,270,Suboptimal foldings created.\n",
      "Energy dot plot created.\n",
      "10\t300,320,330,340,300,310,320,260,310,320,330,290,300,310,Save file created using nafold.\n",
      "Minimum folding energy is -120.20 kcal/mol.\n",
      "Energy increment is 6.01 kcal/mol.\n",
      "300,310,280,360,370,380,390,400,410,280,370,380,390,H-num file created from plot file.\n",
      "340,350,350,360,280,290,370,380,390,400,410, \n",
      "End of Fill\n",
      "Save file created using nafold.\n",
      "300,280,1,2,3,4,5,6,7,8,H-num file created from plot file.\n",
      "330,340,350,310,320,330,340,350,310,320,270,350,360,370,380,320,320,330, \n",
      "End of Fill\n",
      "Save file created using nafold.\n",
      "Minimum folding energy is -117.50 kcal/mol.\n",
      "290,290,300,400,410, \n",
      "End of Fill\n",
      "Save file created using nafold.\n",
      "Minimum folding energy is -148.90 kcal/mol.\n",
      "370,380,390,400,410, \n",
      "1,2,3,4,360,370,380,390,Minimum folding energy is -99.20 kcal/mol.\n",
      "Energy increment is 4.96 kcal/mol.\n",
      "310,320,1,2,3,4,290,330,340,340,350,9,10,11,12,13,14,15,280,300,360,370,380,390,330,340,350,360,370,390,400,410,420, \n",
      "End of Fill\n",
      "Save file created using nafold.\n",
      "Energy increment is 5.88 kcal/mol.\n",
      "H-num file created from plot file.\n",
      "310,300,310,Energy increment is 7.45 kcal/mol.\n",
      "340,5,6, \n",
      "400,410, \n",
      "End of Fill\n",
      "H-num file created from plot file.\n",
      "End of Fill\n",
      "Save file created using nafold.\n",
      "Minimum folding energy is -129.90 kcal/mol.\n",
      "Energy increment is 6.50 kcal/mol.\n",
      "330,340,5,6,7,360,370,380,390,16,17, \n",
      "300,290,360,370,350,360,370,Minimum folding energy is -169.00 kcal/mol.\n",
      "Energy increment is 8.45 kcal/mol.\n",
      "1,2,3,4,5,6,380,390,400,410,420,350,360,370,310,320,330,400,410, \n",
      "End of Fill\n",
      "Save file created using nafold.\n",
      "Minimum folding energy is -121.00 kcal/mol.\n",
      "Energy increment is 6.05 kcal/mol.\n",
      "320,320,Save file created using nafold.\n",
      "Minimum folding energy is -153.40 kcal/mol.\n",
      "Energy increment is 7.67 kcal/mol.\n",
      "1,2,3,8,9,10, \n",
      "H-num file created from plot file.\n",
      "350,360,370,380,390,400,410, \n",
      "End of Fill\n",
      "300,310,400,410, \n",
      "End of Fill\n",
      "Save file created using nafold.\n",
      "7,8,9,10,380,390,400,410, \n",
      "End of Fill\n",
      "Save file created using nafold.\n",
      " \n",
      "End of Fill\n",
      "Save file created using nafold.\n",
      "Minimum folding energy is -107.40 kcal/mol.\n",
      "Energy increment is 5.37 kcal/mol.\n",
      "340,350,310,320,H-num file created from plot file.\n",
      "380,390,400,410,330,340,350,330,340,H-num file created from plot file.\n",
      "4,5,6,7,8,320,Minimum folding energy is -171.60 kcal/mol.\n",
      "Energy increment is 8.58 kcal/mol.\n",
      "380,390,400,410, \n",
      "End of Fill\n",
      "Save file created using nafold.\n",
      "Minimum folding energy is -116.70 kcal/mol.\n",
      "Energy increment is 5.83 kcal/mol.\n",
      "1,2,3,4,5,6,11,12,13,H-num file created from plot file.\n",
      "Minimum folding energy is -120.00 kcal/mol.\n",
      "Energy increment is 6.00 kcal/mol.\n",
      "H-num file created from plot file.\n",
      "1,2,3,4,5,330,340,360,370,380,390,420, \n",
      "End of Fill\n",
      "Save file created using nafold.\n",
      "Minimum folding energy is -101.60 kcal/mol.\n",
      "H-num file created from plot file.\n",
      "360,370,380,390,350,360,370,1,2,3,9,10,11,Save file created using nafold.\n",
      "Minimum folding energy is -112.20 kcal/mol.\n",
      "Energy increment is 5.61 kcal/mol.\n",
      "7,8,9,10,11,330,340,Suboptimal foldings created.\n",
      "Energy dot plot created.\n",
      "1\t1,2,3,4,5,14,15, \n",
      "1,2,3,4,6,7,8,9,10,H-num file created from plot file.\n",
      "Energy increment is 5.08 kcal/mol.\n",
      "350,360,400,410, \n",
      "End of Fill\n",
      "Save file created using nafold.\n",
      "1,2,3,4,400,410, \n",
      "End of Fill\n",
      "Save file created using nafold.\n",
      "380,390,400,4,5,6,7,8,9, \n",
      "12,13,H-num file created from plot file.\n",
      "350,360,H-num file created from plot file.\n",
      "H-num file created from plot file.\n",
      "6,7,8,5,6,7,8,11,12,13,14,H-num file created from plot file.\n",
      "370,380,390,Minimum folding energy is -116.50 kcal/mol.\n",
      "Energy increment is 5.83 kcal/mol.\n",
      "410, \n",
      "End of Fill\n",
      "Save file created using nafold.\n",
      "Minimum folding energy is -143.50 kcal/mol.\n",
      "Energy increment is 7.17 kcal/mol.\n",
      "5,6,7,8,9,Minimum folding energy is -143.70 kcal/mol.\n",
      "Energy increment is 7.18 kcal/mol.\n",
      "10,11,12,13,14,1,2,3,9,10,11,12,14,15,16,17,1,2,3,4,1,2,3,4,5,H-num file created from plot file.\n",
      "1,2,3,4,370,380,390,9,10,11,12,1,2,3,15,16,17,Suboptimal foldings created.\n",
      "Energy dot plot created.\n",
      "400,410, \n",
      "End of Fill\n",
      "Save file created using nafold.\n",
      "10,11,12,4,5,6,7,8,15,16,17,18, \n",
      "5,6,7,1,2,3,400,410,420, \n",
      "End of Fill\n",
      "5,6,7,8,6,7,8,9,13,14,Suboptimal foldings created.\n",
      "H-num file created from plot file.\n",
      "4,5,6,7,Minimum folding energy is -142.50 kcal/mol.\n",
      "Energy increment is 7.12 kcal/mol.\n",
      "18,19,20,21,H-num file created from plot file.\n",
      "9,10,11,12,13,H-num file created from plot file.\n",
      " \n",
      "10\t13,14,15,16,19,20,21,22,8,9,4,5,6,7,9,10,11,12,13,Save file created using nafold.\n",
      "Minimum folding energy is -131.50 kcal/mol.\n",
      "Energy increment is 6.58 kcal/mol.\n",
      "10,11,12,13,14,15,15,16, \n",
      "1,2,3,4,5,6,Energy dot plot created.\n",
      "10\t8,9,10,11,12,13, \n",
      "1,2,14, \n",
      "1,2,3,4,5,17, \n",
      "23,24,25,26,10,11,12,13,8,9,10,11,12,13,16,17,18,19, \n",
      "7,8,9,10,11,Suboptimal foldings created.\n",
      "Energy dot plot created.\n",
      "10\tH-num file created from plot file.\n",
      "14,15,16,3,4,5,6,7,8,6,7,8,9,10,11,27, \n",
      "14,15,16,14,15,16,17,18,19,20,21,22, \n",
      "H-num file created from plot file.\n",
      "12,13,14,15,1,2,3,4,5,6,17,18, \n",
      "9,10,11,12,13,14,Suboptimal foldings created.\n",
      "Energy dot plot created.\n",
      "12,13,14,15, \n",
      "20,21,22,23,1,2,3,4,16,17,18,19,Suboptimal foldings created.\n",
      "7,8,9,10,11,15,16,17,10\t16,17,18,19,20,21,24,25, \n",
      "5,6,7,8,20,21, \n",
      "Energy dot plot created.\n",
      "10\t12,13,14,15,16,18,19,20,21,22,22,23,24,9, \n",
      "17,18,19,20,23,24,25,25,26,27,Suboptimal foldings created.\n",
      "Energy dot plot created.\n",
      "10\t21,22,23,24,26, \n",
      " \n",
      "Suboptimal foldings created.\n",
      "Energy dot plot created.\n",
      "25,26,27,28, \n",
      "Suboptimal foldings created.\n",
      "Energy dot plot created.\n",
      "Suboptimal foldings created.\n",
      "Energy dot plot created.\n",
      "10\t10\tSuboptimal foldings created.\n",
      "Energy dot plot created.\n",
      "10\t10\tSuboptimal foldings created.\n",
      "Energy dot plot created.\n",
      "1\tSuboptimal foldings created.\n",
      "Energy dot plot created.\n",
      "10\tSuboptimal foldings created.\n",
      "Energy dot plot created.\n",
      "10\tSuboptimal foldings created.\n",
      "Energy dot plot created.\n",
      "10\tSuboptimal foldings created.\n",
      "Energy dot plot created.\n",
      "10\tSuboptimal foldings created.\n",
      "Energy dot plot created.\n",
      "Suboptimal foldings created.\n",
      "Energy dot plot created.\n",
      "10\t10\tSuboptimal foldings created.\n",
      "Energy dot plot created.\n",
      "10\tSuboptimal foldings created.\n",
      "Energy dot plot created.\n",
      "10\tSuboptimal foldings created.\n",
      "Energy dot plot created.\n",
      "Suboptimal foldings created.\n",
      "Energy dot plot created.\n",
      "10\t10\t11\t11\t11\t11\t11\t11\t11\t11\t11\t11\t11\t11\t11\t1\t11\t11\t11\t2\t2\t11\t2\t11\t11\t11\t3\t2\t12\t12\t12\t12\t3\t1\t12\t12\t12\t12\t12\t12\t12\t12\t12\t12\t12\t12\t12\t3\t12\t12\t1\t4\t3\t13\t13\t13\t2\t13\t13\t13\t4\t13\t13\t13\t13\t13\t13\t13\t1\t13\t13\t13\t13\t4\t5\t2\t4\t14\t14\t14\t14\t3\t14\t5\t14\t1\t14\t14\t2\t14\t14\t14\t14\t5\t14\t14\t14\t14\t3\t6\t15\t15\t15\t15\t15\t15\t6\t15\t15\t5\t2\t6\t3\t15\t4\t15\t15\t15\t15\t1\t15\t15\t4\t7\t16\t16\t16\t1616\t16\t116\t6\t\n",
      "Structure plots generated.\n",
      "All done.\n",
      "3\t16\t\t7\t5\t4\t\t16\t16\t16\t16\t16\t16\t5\t8\t2\t/home/jupyter/plant_microRNA_prediction\n",
      "/home/jupyter/plant_microRNA_prediction/secondary_structure/mfold/AMWY020000341|123-432|201-219|+\n",
      "17\t17\t17\tmfold version 3.6\n",
      "17\t2\t17\t7\t17\t6\t4\t17\t5\t1\t8\tREUSE= NO\n",
      "17\t17\t17\t17\t1\t6\tSEQ.pnt created.\n",
      "Sequence length is 310\n",
      "9\t3\t17\tRNA free energy files (version 2.3) at 20 degrees created.\n",
      "10,20,30,40,50,60,1\t18\t18\t70,80,90,2\t5\t18\t18\t8\t3\t6\t9\t18\t18\t18\t7\t100,110,18\t1\t120,130,\n",
      "Structure plots generated.\n",
      "All done.\n",
      "1\t2\t140,7\t4\t150,18\t160,170,/home/jupyter/plant_microRNA_prediction\n",
      "/home/jupyter/plant_microRNA_prediction/secondary_structure/mfold/AMWY020000411|1710-2128|201-219|-\n",
      "2\t19\t180,8\t19\tmfold version 3.6\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "190,7\t9\t19\t6\t\n",
      "Structure plots generated.\n",
      "All done.\n",
      "3\t19\tREUSE= NO\n",
      "19\t19\t19\t4\t200,210,19\t2\t8\t2\t220,230,SEQ.pnt created.\n",
      "Sequence length is 419\n",
      "3\t240,5\tRNA free energy files (version 2.3) at 20 degrees created.\n",
      "250,260,/home/jupyter/plant_microRNA_prediction\n",
      "/home/jupyter/plant_microRNA_prediction/secondary_structure/mfold/AMWY020000191|1-291|201-219|-10,20,30,40,50,60,\n",
      "270,280,290,300, \n",
      "1\t70,80,90,1\t9\tEnd of Fill\n",
      "Save file created using nafold.\n",
      "Minimum folding energy is -81.90 kcal/mol.\n",
      "4\t\n",
      "Structure plots generated.\n",
      "All done.\n",
      "mfold version 3.6\n",
      "3\t1\t100,110,7\tEnergy increment is 4.09 kcal/mol.\n",
      "1\tREUSE= NO\n",
      "1\t8\t1\t120,130,1\t5\tH-num file created from plot file.\n",
      "SEQ.pnt created.\n",
      "1\t9\t3\t1\t140,150,3\tSequence length is 291\n",
      "1,2,3,4,5,6,7,4\t160,170,/home/jupyter/plant_microRNA_prediction8,9,10,11,12,\n",
      "/home/jupyter/plant_microRNA_prediction/secondary_structure/mfold/AMWY020000911|11460-11868|201-219|+180,RNA free energy files (version 2.3) at 20 degrees created.\n",
      "10,20,\n",
      "6\t13,14, \n",
      "190,30,40,50,60,2\tmfold version 3.6\n",
      "200,\n",
      "Structure plots generated.\n",
      "All done.\n",
      "20\t4\t70,80,REUSE= NO\n",
      "210,9\t8\t20\t5\t90,100,20\t20\t6\tSEQ.pnt created.\n",
      "220,\n",
      "Structure plots generated.\n",
      "All done.\n",
      "20\t20\tSequence length is 409\n",
      "110,120,20\t4\t20\t230,240,4\tSuboptimal foldings created.130,250,260,5\t/home/jupyter/plant_microRNA_prediction\n",
      "/home/jupyter/plant_microRNA_prediction/secondary_structure/mfold/AMWY020000541|1731-2152|201-222|-RNA free energy files (version 2.3) at 20 degrees created.\n",
      "\n",
      "7\t140,\n",
      "Energy dot plot created.\n",
      "10\t270,10,20,30,40,150,160,mfold version 3.6\n",
      "280,290,50,60,70,/home/jupyter/plant_microRNA_prediction\n",
      "170,21\t3\tREUSE= NO\n",
      "/home/jupyter/plant_microRNA_prediction/secondary_structure/mfold/AMWY020000501|377-795|201-219|-\n",
      "300,5\t80,90,180,190,21\t310,320,330,21\t\n",
      "Structure plots generated.\n",
      "All done.\n",
      "100,110,SEQ.pnt created.\n",
      "Sequence length is 422\n",
      "6\t9\t200,mfold version 3.6\n",
      "7\t21\t21\t120,RNA free energy files (version 2.3) at 20 degrees created.\n",
      "340,350,360,210,220,21\tREUSE= NO\n",
      "21\t5\t21\t130,370,380,390,400,410, \n",
      "End of Fill\n",
      "230,240,10,20,30,40,50,5\tSEQ.pnt created.\n",
      "Sequence length is 419\n",
      "6\t140,150,11\tSave file created using nafold.\n",
      "Minimum folding energy is -128.00 kcal/mol.\n",
      "Energy increment is 6.40 kcal/mol.\n",
      "250,260,270,280,60,70,80,RNA free energy files (version 2.3) at 20 degrees created.\n",
      "8\t/home/jupyter/plant_microRNA_prediction290, \n",
      "End of Fill\n",
      "Save file created using nafold.\n",
      "Minimum folding energy is -72.40 kcal/mol.\n",
      "160,170,\n",
      "/home/jupyter/plant_microRNA_prediction/secondary_structure/mfold/AMWY020000511|4796-5213|201-218|+90,100,110,10,20,30,40,50,\n",
      "H-num file created from plot file.\n",
      "6\t180,Energy increment is 3.62 kcal/mol.\n",
      "4\t22\t1,2,120,130,60,70,2\t190,22\t7\tH-num file created from plot file.\n",
      "1,8\t140,150,3,4,5,mfold version 3.6\n",
      "\n",
      "Structure plots generated.\n",
      "All done.\n",
      "80,90,100,200,210,2,3,4,5,6,7,8,9,6,7, \n",
      "22\t22\t110,120,130,REUSE= NO\n",
      "160,22\t2\t22\t6\t7\t10,11,12,13,14,15,16,220,140,170,180,6\tSEQ.pnt created.\n",
      "150,160,17,18,19,20,21,22,23,230,Sequence length is 418\n",
      "190,12\t9\t170,240,/home/jupyter/plant_microRNA_prediction24, \n",
      "200,RNA free energy files (version 2.3) at 20 degrees created.\n",
      "\n",
      "7\t/home/jupyter/plant_microRNA_prediction/secondary_structure/mfold/AMWY020001361|3637-4056|201-220|+Suboptimal foldings created.\n",
      "Energy dot plot created.\n",
      "1\t250,5\t210,180,10,20,30,\n",
      "260,3\t23\t190,220,23\t40,50,60,270,mfold version 3.6\n",
      "8\t2\t200,230,70,80,23\t9\t280,290,300,REUSE= NO\n",
      "210,23\t240,7\t23\t90,100,110,3\t7\t220,230,310,320,13\t250,260,120,130,SEQ.pnt created.\n",
      "Sequence length is 420\n",
      "240,330,340,350,360,8\t8\t270,140,150,RNA free energy files (version 2.3) at 20 degrees created.\n",
      "370,380,390,400, \n",
      "End of Fill\n",
      "250,\n",
      "Structure plots generated.\n",
      "All done.\n",
      "2\tSuboptimal foldings created.\n",
      "Energy dot plot created.\n",
      "10\t160,24\t10,20,260,4\t6\t24\tSave file created using nafold.\n",
      "Minimum folding energy is -126.00 kcal/mol.\n",
      "280,290,170,30,40,50,60,70,9\t270,3\t300,310,\n",
      "Structure plots generated.\n",
      "All done.\n",
      "Energy increment is 6.30 kcal/mol.\n",
      "180,24\t80,90,24\t320,280,290,H-num file created from plot file.\n",
      "8\t190,100,110,120,24\t8\t300,310,1,2,3,/home/jupyter/plant_microRNA_prediction330,340,200,130,\n",
      "320,330,/home/jupyter/plant_microRNA_prediction/secondary_structure/mfold/AMWY020000661|2162-2581|201-220|-4\t4,5,6,7,\n",
      "350,360,370,14\t210,140,150,9\t8,9,340,350,380,390,400,220,9\t160,/home/jupyter/plant_microRNA_predictionmfold version 3.6\n",
      "\n",
      "360,370,380,11\t410,420, \n",
      "End of Fill\n",
      "Save file created using nafold.\n",
      "/home/jupyter/plant_microRNA_prediction/secondary_structure/mfold/AMWY020001141|1-226|73-96|-10,11, \n",
      "170,REUSE= NO\n",
      "\n",
      "25\t5\t230,3\t\n",
      "Structure plots generated.\n",
      "All done.\n",
      "180,390,400,410, \n",
      "End of Fill\n",
      "Save file created using nafold.\n",
      "7\tMinimum folding energy is -156.40 kcal/mol.\n",
      "Energy increment is 7.82 kcal/mol.\n",
      "SEQ.pnt created.\n",
      "Sequence length is 420\n",
      "25\tmfold version 3.6\n",
      "Minimum folding energy is -94.90 kcal/mol.\n",
      "Energy increment is 4.75 kcal/mol.\n",
      "240,4\t190,200,REUSE= NO\n",
      "25\tRNA free energy files (version 2.3) at 20 degrees created.\n",
      "10,25\tH-num file created from plot file.\n",
      "250,25\t210,9\tH-num file created from plot file.\n",
      "20,30,40,50,60,9\t260,270,1,2,3,4,5,SEQ.pnt created.\n",
      "Sequence length is 226\n",
      "220,230,Suboptimal foldings created.\n",
      "Energy dot plot created.\n",
      "10\t5\t1\t1,2,3,4,5,70,80,6,7,8,240,\n",
      "Structure plots generated.\n",
      "All done.\n",
      "RNA free energy files (version 2.3) at 20 degrees created.\n",
      "/home/jupyter/plant_microRNA_prediction280,\n",
      "\n",
      "Structure plots generated.\n",
      "All done.\n",
      "/home/jupyter/plant_microRNA_prediction/secondary_structure/mfold/AMWY020000811|6241-6659|201-219|-\n",
      "6,7,8,9,90,100,110,10,20,30,40,9,10,11,12,250,290,12\t10,11,26\tmfold version 3.6\n",
      "6\t120,130,50,60,70,80,260,4\tREUSE= NO\n",
      "300,310,8\t140,12, \n",
      "90,100,13,26\t270,280,5\t320,330,150,290, \n",
      "110,120,/home/jupyter/plant_microRNA_prediction26\tSEQ.pnt created.\n",
      "Sequence length is 419\n",
      "\n",
      "Structure plots generated.\n",
      "All done.\n",
      "\n",
      "340,350,160,/home/jupyter/plant_microRNA_prediction/home/jupyter/plant_microRNA_prediction/secondary_structure/mfold/AMWY020000621|3340-3758|201-219|-\n",
      "26\t/home/jupyter/plant_microRNA_prediction/secondary_structure/mfold/AMWY020000471|1435-1851|201-217|+\n",
      "\n",
      "\n",
      "Structure plots generated.\n",
      "All done.\n",
      "130,140,150,300,310,RNA free energy files (version 2.3) at 20 degrees created.\n",
      "2\t360,370,380,170,180,320,330,10,20,30,160,170,11\tmfold version 3.6\n",
      "mfold version 3.6\n",
      "2\t6\t190,390,400,410, \n",
      "End of Fill\n",
      "40,50,60,180,190,200,210,220,340,350,REUSE= NO\n",
      "REUSE= NO\n",
      "200,13\tSuboptimal foldings created.\n",
      "Energy dot plot created.\n",
      "10\tSave file created using nafold.\n",
      "Minimum folding energy is -141.10 kcal/mol.\n",
      "Energy increment is 7.05 kcal/mol.\n",
      "360,370,380,70,80,90, \n",
      "End of Fill\n",
      "Save file created using nafold.\n",
      "Minimum folding energy is -86.10 kcal/mol.\n",
      "Energy increment is 4.30 kcal/mol.\n",
      "27\tSEQ.pnt created.\n",
      "Sequence length is 417\n",
      "Suboptimal foldings created.\n",
      "Energy dot plot created.\n",
      "SEQ.pnt created.\n",
      "Sequence length is 419\n",
      "210,9\t/home/jupyter/plant_microRNA_prediction7\t\n",
      "/home/jupyter/plant_microRNA_prediction/secondary_structure/mfold/AMWY020000271|2935-3353|201-219|-5\t390,400,410, \n",
      "End of Fill\n",
      "Save file created using nafold.\n",
      "100,110,\n",
      "10\tRNA free energy files (version 2.3) at 20 degrees created.\n",
      "RNA free energy files (version 2.3) at 20 degrees created.\n",
      "H-num file created from plot file.\n",
      "Minimum folding energy is -111.30 kcal/mol.\n",
      "Energy increment is 5.57 kcal/mol.\n",
      "2\t/home/jupyter/plant_microRNA_predictionH-num file created from plot file.\n",
      "\n",
      "220,10,20,30,40,10,20,30,40,120,/home/jupyter/plant_microRNA_prediction/secondary_structure/mfold/AMWY020001351|777-1200|201-224|-\n",
      "27\tmfold version 3.6\n",
      "6\t1,2,3,4,5,6,27\t50,60,70,1,2,3,4,5,6,7,8,50,60,70,130,140,H-num file created from plot file.\n",
      "REUSE= NO\n",
      "230,240,mfold version 3.6\n",
      "9, \n",
      "3\t3\t80,90,7,8, \n",
      "80,90,100,1,2,3,4,REUSE= NO\n",
      "150,160,170,100,110,1\t250,260,7\t110,SEQ.pnt created.\n",
      "5,6,7,8,9,180,14\t120,130,SEQ.pnt created.\n",
      "Suboptimal foldings created.\n",
      "Sequence length is 419\n",
      "10,11,12,120,130,Sequence length is 424\n",
      "190,11\t140,\n",
      "Structure plots generated.\n",
      "All done.8\t270,280,Energy dot plot created.\n",
      "1\t\n",
      "11\t13,14,15,16,17,RNA free energy files (version 2.3) at 20 degrees created.\n",
      "10,Suboptimal foldings created.\n",
      "Energy dot plot created.\n",
      "140,150,150,160,200,28\tRNA free energy files (version 2.3) at 20 degrees created.\n",
      "10,20,30,290,300,6\t \n",
      "20,30,40,50,60,170,180,1\t160,210,220,310,320,40,50,60,70,3\t7\t170,70,80,90,2\t4\t190,230,2\t80,90,100,330,340,350,2\t100,110,4\t180,190,240,200,/home/jupyter/plant_microRNA_prediction360,370,380,\n",
      "110,120,8\t/home/jupyter/plant_microRNA_prediction/secondary_structure/mfold/AMWY020000031|4392-4810|201-219|+\n",
      "120,130,250,390,400,410, \n",
      "End of Fill\n",
      "Save file created using nafold.\n",
      "210,220,130,2\t200,15\tmfold version 3.6\n",
      "140,140,150,260,230,Minimum folding energy is -161.10 kcal/mol.\n",
      "Energy increment is 8.05 kcal/mol.\n",
      "12\t9\tREUSE= NO\n",
      "160,170,150,210,240,2\t270,280,12\t7\t220,250,2\t180,160,4\t290,SEQ.pnt created.\n",
      "Sequence length is 419\n",
      "H-num file created from plot file.\n",
      "Suboptimal foldings created.\n",
      "Energy dot plot created.\n",
      "8\t170,230,260,270,3\t5\t190,300,1,2,3,3\tRNA free energy files (version 2.3) at 20 degrees created.\n",
      "10\t240,250,180,190,280,290,200,9\t4,5,6,7,5\t260,310,320,330,10,20,30,40,3\t200,300,310,210,220,50,60,70,340,350,270,8,9,10,11,12,3\t16\t3\t320,330,\n",
      "Structure plots generated.\n",
      "All done.\n",
      "230,1\t210,360,370,380,390,280,290,80,90,100,110,13,14,15,220,340,350,360,240,250,300,13\t3\t400,410, \n",
      "End of Fill\n",
      " \n",
      "120,130,\n",
      "Structure plots generated.\n",
      "All done.\n",
      "5\t230,4\t310,Save file created using nafold.\n",
      "Minimum folding energy is -160.50 kcal/mol.\n",
      "Energy increment is 8.03 kcal/mol.\n",
      "6\t260,270,280,370,380,390,400,410,140,150,160,9\t240,11\t320,330,4\t290,170,\n",
      "Structure plots generated.\n",
      " \n",
      "End of Fill\n",
      "Save file created using nafold.\n",
      "Minimum folding energy is -170.40 kcal/mol.\n",
      "Energy increment is 8.52 kcal/mol.\n",
      "All done.\n",
      "250,6\t340,350,/home/jupyter/plant_microRNA_prediction\n",
      "H-num file created from plot file.\n",
      "/home/jupyter/plant_microRNA_prediction/secondary_structure/mfold/AMWY020000811|2425-2842|201-218|-300,310,\n",
      "180,260,4\t360,370,4\t1,17\tH-num file created from plot file.\n",
      "320,330,270,280,190,380,390,400,410, \n",
      "End of Fill\n",
      "Save file created using nafold.\n",
      "/home/jupyter/plant_microRNA_predictionmfold version 3.6\n",
      "2,3,4,5,\n",
      "1,2,3,340,350,/home/jupyter/plant_microRNA_prediction/secondary_structure/mfold/AMWY020000411|3319-3666|201-222|+290,\n",
      "200,2\tMinimum folding energy is -136.30 kcal/mol.\n",
      "Energy increment is 6.82 kcal/mol.\n",
      "REUSE= NO\n",
      "6,7,8,9,10,Suboptimal foldings created.\n",
      "Energy dot plot created.\n",
      "10\t4\t360,370,380,390,4,5,6,7,300,310,210,SEQ.pnt created.\n",
      "Sequence length is 418\n",
      "11,12,13,14,/home/jupyter/plant_microRNA_prediction7\t8,9,10,mfold version 3.6\n",
      "1\t400,410, \n",
      "End of Fill\n",
      "Save file created using nafold.\n",
      "320,220,H-num file created from plot file.\n",
      "\n",
      "4\t/home/jupyter/plant_microRNA_prediction/secondary_structure/mfold/AMWY020000251|1857-2180|104-124|-\n",
      "12\tREUSE= NO\n",
      "330,340,\n",
      "Structure plots generated.\n",
      "5\t230,6\tAll done.\n",
      "1,2,11,12,13,14,15,15,16,RNA free energy files (version 2.3) at 20 degrees created.\n",
      "Minimum folding energy is -152.60 kcal/mol.\n",
      "Energy increment is 7.63 kcal/mol.\n",
      "mfold version 3.6\n",
      "350,SEQ.pnt created.\n",
      "Sequence length is 348\n",
      "240,250,5\t3,4,5,6,16,17,17,18,19,20,21,10,20,30,40,50,60,7\tREUSE= NO\n",
      "360,370,5\t5\t7,8,9,10,18,19,20,18\tH-num file created from plot file.\n",
      "1,70,80,90,380,390,400,22, \n",
      "SEQ.pnt created.\n",
      "Sequence length is 324\n",
      "RNA free energy files (version 2.3) at 20 degrees created.\n",
      "260,270,11,12, \n",
      "5\t21,22,23, \n",
      "100,110,2,3,4,5,6,7,8,10,20,30,40,50,60,RNA free energy files (version 2.3) at 20 degrees created.\n",
      "410,420, \n",
      "End of Fill\n",
      "Save file created using nafold.\n",
      "Minimum folding energy is -153.30 kcal/mol.\n",
      "280,290,/home/jupyter/plant_microRNA_prediction120,3\t9,10,11,12,\n",
      "10,20,/home/jupyter/plant_microRNA_prediction/secondary_structure/mfold/AMWY020000601|277-497|201-218|+\n",
      "300,11\t2\t70,80,Energy increment is 7.67 kcal/mol.\n",
      "130,140,5\t13,14,8\t30,40,50,60,7\t310,320,330,H-num file created from plot file.\n",
      "6\t13\t150,160,90,100,15,16,17,mfold version 3.6\n",
      "REUSE= NO\n",
      "70,80,340,110,1,2,18,19,20,21,170,6\t90,100,110,350,360,370,380,3,4,5,6,19\t6\tSEQ.pnt created.\n",
      "Sequence length is 221\n",
      "120,130,180,22,120,8\t6\t390,400,410, \n",
      "End of Fill\n",
      "7,8,9,10,11,12,Suboptimal foldings created.\n",
      "Energy dot plot created.\n",
      "10\t23,24, \n",
      "190,200,130,140,140,150,Save file created using nafold.\n",
      "Minimum folding energy is -118.70 kcal/mol.\n",
      "Energy increment is 5.93 kcal/mol.\n",
      "RNA free energy files (version 2.3) at 20 degrees created.\n",
      "10,13,14,15,4\t6\tSuboptimal foldings created.\n",
      "Energy dot plot created.\n",
      "10\t210,220,20,30,40,50,160,16,17,18,19,150,160,12\t230,H-num file created from plot file.\n",
      "9\t3\t170,60,70,80,7\t20,21,22,14\tSuboptimal foldings created.\n",
      "Energy dot plot created.\n",
      "6\t170,1,2,240,250,23,24,10\t8\t180,3,4,5,90,180,190,25, \n",
      "7\t260,190,200,1\t7\t6,7,8,100,110,120,130,200,210,270,11\t210,7\t9\t140,150,160,170,180,190,200,9,10,11,12,13,220,280,290,220,7\t11\t230,240,250,230,240,5\t\n",
      "Structure plots generated.\n",
      "All done.\n",
      "300,310,320,14,15, \n",
      "210,220, \n",
      "End of Fill\n",
      "Save file created using nafold.\n",
      "260,270,280,Suboptimal foldings created.\n",
      "330,340,350,250,8\t13\t7\tMinimum folding energy is -64.90 kcal/mol.\n",
      "Energy increment is 3.25 kcal/mol.\n",
      "4\t260,290,300,310,15\tEnergy dot plot created.\n",
      "10\t360,370,380,9\t11\t390,400,410, \n",
      "End of Fill\n",
      "270,20\t320, \n",
      "End of Fill\n",
      "Save file created using nafold.\n",
      "H-num file created from plot file.\n",
      "8\t280,290,300,Save file created using nafold.\n",
      "Minimum folding energy is -143.10 kcal/mol.\n",
      "Energy increment is 7.16 kcal/mol.\n",
      "8\t/home/jupyter/plant_microRNA_prediction\n",
      "Structure plots generated.\n",
      "All done.Minimum folding energy is -67.10 kcal/mol.\n",
      "Energy increment is 3.35 kcal/mol.\n",
      "1,2,3,4,5,6,7, \n",
      "\n",
      "\n",
      "8\t/home/jupyter/plant_microRNA_prediction/secondary_structure/mfold/AMWY020000381|375-652|57-78|-\n",
      "12\t310,320,H-num file created from plot file.\n",
      "H-num file created from plot file.\n",
      "Suboptimal foldings created.\n",
      "Energy dot plot created.\n",
      "330,340, \n",
      "End of Fill\n",
      "Save file created using nafold.\n",
      "12\t8\tmfold version 3.6\n",
      "1,2,1,2,3,4,5,6\tSuboptimal foldings created.\n",
      "Energy dot plot created.\n",
      "10\t10\tMinimum folding energy is -120.20 kcal/mol.\n",
      "Energy increment is 6.01 kcal/mol.\n",
      "3,4,5,6,6,7,8,9,10,11,Suboptimal foldings created.REUSE= NO\n",
      "12,13,14,15, \n",
      "7,8,9,8\t\n",
      "Energy dot plot created.\n",
      "1\t9\tH-num file created from plot file.\n",
      "14\t21\t/home/jupyter/plant_microRNA_predictionSEQ.pnt created.\n",
      "Sequence length is 278\n",
      "16\t11\t\n",
      "Structure plots generated.\n",
      "All done.\n",
      "10,11,12,1,2,\n",
      "/home/jupyter/plant_microRNA_prediction/secondary_structure/mfold/AMWY020001221|1173-1592|201-220|+\n",
      "\n",
      "Structure plots generated.\n",
      "All done.\n",
      "5\t12\tRNA free energy files (version 2.3) at 20 degrees created.\n",
      "13, \n",
      "3,4,5,6,7,8,9\tmfold version 3.6\n",
      "10,20,30,40,50,1\t9,10,11,12,13,14,REUSE= NO\n",
      "60,70,9\t \n",
      "Suboptimal foldings created.\n",
      "7\t9\t11\tSEQ.pnt created.\n",
      "Sequence length is 420\n",
      "13\t80,90,100,Energy dot plot created.\n",
      "10\t/home/jupyter/plant_microRNA_prediction\n",
      "/home/jupyter/plant_microRNA_prediction/secondary_structure/mfold/AMWY020001171|2724-3142|201-219|-\n",
      "/home/jupyter/plant_microRNA_prediction\n",
      "2\t/home/jupyter/plant_microRNA_prediction/secondary_structure/mfold/AMWY020001151|2666-3084|201-219|+\n",
      "11\t110,120,RNA free energy files (version 2.3) at 20 degrees created.\n",
      "9\t\n",
      "Structure plots generated.\n",
      "All done.\n",
      "mfold version 3.6\n",
      "Suboptimal foldings created.\n",
      "22\t130,140,15\t17\t10,20,30,40,mfold version 3.6\n",
      "12\tEnergy dot plot created.\n",
      "10\tREUSE= NO\n",
      "150,50,60,70,13\tSuboptimal foldings created.\n",
      "REUSE= NO\n",
      "6\t\n",
      "Structure plots generated.\n",
      "All done.\n",
      "160,170,80,90,Energy dot plot created.\n",
      "10\tSEQ.pnt created.\n",
      "Sequence length is 419\n",
      "\n",
      "\n",
      "Structure plots generated.\n",
      "All done.2\tSEQ.pnt created.\n",
      "Sequence length is 419\n",
      "180,190,100,110,8\t11\t/home/jupyter/plant_microRNA_prediction\n",
      "Structure plots generated.\n",
      "All done.\n",
      "RNA free energy files (version 2.3) at 20 degrees created.\n",
      "\n",
      "12\tRNA free energy files (version 2.3) at 20 degrees created.\n",
      "3\t200,210,/home/jupyter/plant_microRNA_prediction/secondary_structure/mfold/AMWY020000831|499-917|201-219|+\n",
      "120,10,20,30,40,50,10,20,30,40,50,14\t220,230,240,130,140,mfold version 3.6\n",
      "12\t60,70,80,60,70,80,/home/jupyter/plant_microRNA_prediction250,260,270, \n",
      "End of Fill\n",
      "1\t150,\n",
      "\n",
      "Structure plots generated.\n",
      "All done.\n",
      "REUSE= NO\n",
      "/home/jupyter/plant_microRNA_prediction/secondary_structure/mfold/AMWY020000931|2528-2946|201-219|+\n",
      "23\t/home/jupyter/plant_microRNA_prediction\n",
      "90,100,/home/jupyter/plant_microRNA_prediction/secondary_structure/mfold/AMWY020000931|1938-2356|201-219|-90,100,1\tSave file created using nafold.\n",
      "Minimum folding energy is -94.00 kcal/mol.\n",
      "Energy increment is 4.70 kcal/mol.\n",
      "160,SEQ.pnt created.\n",
      "Sequence length is 419\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "11\t14\t/home/jupyter/plant_microRNA_prediction110,13\t110,\n",
      "170,180,/home/jupyter/plant_microRNA_prediction/secondary_structure/mfold/AMWY020001191|1555-1973|201-219|-mfold version 3.6\n",
      "\n",
      "mfold version 3.6\n",
      "11\t120,120,RNA free energy files (version 2.3) at 20 degrees created.\n",
      "10,REUSE= NO\n",
      "H-num file created from plot file.\n",
      "3\t7\t190,130,140,mfold version 3.6\n",
      "REUSE= NO\n",
      "130,140,1,2,3,4,5,6,20,30,40,50,4\t200,/home/jupyter/plant_microRNA_prediction\n",
      "13\tREUSE= NO\n",
      "9\t\n",
      "150,/home/jupyter/plant_microRNA_prediction/secondary_structure/mfold/AMWY020000511|6024-6442|201-219|-SEQ.pnt created.\n",
      "Sequence length is 419\n",
      "150,7,8,9,10,11,60,70,80,12\tSEQ.pnt created.\n",
      "Sequence length is 419\n",
      "210,160,SEQ.pnt created.\n",
      "Sequence length is 419\n",
      "160,170,RNA free energy files (version 2.3) at 20 degrees created.\n",
      "12, \n",
      "RNA free energy files (version 2.3) at 20 degrees created.\n",
      "90,100,15\t2\t170,180,mfold version 3.6\n",
      "10,20,30,40,50,13\t180,220,24\tRNA free energy files (version 2.3) at 20 degrees created.\n",
      "10,110,120,130,10,20,30,40,190,190,REUSE= NO\n",
      "230,240,12\t60,70,2\t15\t140,20,30,40,50,60,70,200,14\t200,50,60,70,80,90,80,90,100,80,90,150,160,250,SEQ.pnt created.\n",
      "Sequence length is 419\n",
      "210,8\t4\t100,110,120,Suboptimal foldings created.\n",
      "Energy dot plot created.\n",
      "10\t\n",
      "Structure plots generated.\n",
      "All done.\n",
      "260,210,170,220,12\t100,110,120,130,5\t110,120,14\tRNA free energy files (version 2.3) at 20 degrees created.\n",
      "13\t180,190,230,240,130,140,270,280,220,10,20,30,40,50,60,140,150,130,140,16\t2\t3\t250,70,80,200,150,160,14\t150,160,230,240,260,270,160,170,210,13\t90,100,290,300,310,/home/jupyter/plant_microRNA_prediction170,180,\n",
      "3\t\n",
      "170,180,/home/jupyter/plant_microRNA_prediction/secondary_structure/mfold/AMWY020000501|517-935|201-219|-220,250,280,290,15\t180,190,320,330,110,120,130,190,16\t9\t190,5\t6\t300,310,13\t260,230,140,150,mfold version 3.6\n",
      "11\t200,200,200,210,340,350,14\t270,160,REUSE= NO\n",
      "320,240,250,15\t210,220,360,370,380,390,400,210,220,230,280,290,SEQ.pnt created.\n",
      "Sequence length is 419\n",
      "260,330,340,350,360,170,17\t15\t410, \n",
      "End of Fill\n",
      "Save file created using nafold.\n",
      "3\t300,220,230,370,380,390,400,410,270,180,190,4\tRNA free energy files (version 2.3) at 20 degrees created.\n",
      "240,310,320,230,200,1\t240,Minimum folding energy is -140.20 kcal/mol.\n",
      "Energy increment is 7.01 kcal/mol.\n",
      " \n",
      "End of Fill\n",
      "Save file created using nafold.\n",
      "Minimum folding energy is -156.80 kcal/mol.\n",
      "Energy increment is 7.84 kcal/mol.\n",
      "280,290,250,260,10,20,30,40,50,16\t4\tH-num file created from plot file.\n",
      "240,210,330,340,250,260,300,7\t60,70,80,1,2,3,4,5,250,17\t\n",
      "Structure plots generated.\n",
      "All done.\n",
      "350,360,370,270,280,290,14\t270,220,6\t310,320,330,90,100,110,15\t260,270,1\t6,7,8,9,H-num file created from plot file.\n",
      "1,380,390,400,410, \n",
      "End of Fill\n",
      "12\t280,290,230,240,300,310,120,130,340,350,280,2,3,4,5,6,7,8,9,Save file created using nafold.\n",
      "Minimum folding energy is -129.40 kcal/mol.\n",
      "Energy increment is 6.47 kcal/mol.\n",
      "10,11,12,13,14,15,16,250,300,310,320,330,340,140,150,4\t360,370,380,390,290,300,17,18,19,20,10,11,12,13,14,160,260,270,18\t5\t320,330,350,360,370,400,410, \n",
      "End of Fill\n",
      "Save file created using nafold.\n",
      "16\t/home/jupyter/plant_microRNA_prediction310,320,H-num file created from plot file.\n",
      "\n",
      "21,22,23,15,16,17,170,180,/home/jupyter/plant_microRNA_prediction/secondary_structure/mfold/AMWY020000541|1748-2165|201-218|-Minimum folding energy is -126.80 kcal/mol.\n",
      "Energy increment is 6.34 kcal/mol.\n",
      "280,17\t380,390,400,410, \n",
      "End of Fill\n",
      "Save file created using nafold.\n",
      "340,350,360,\n",
      "1,2,330,340,350,5\t24, \n",
      "\n",
      "Structure plots generated.\n",
      "All done.\n",
      "18,19,290,300,190,2\t370,380,390,400,18\t360,370,Minimum folding energy is -111.90 kcal/mol.\n",
      "Energy increment is 5.59 kcal/mol.\n",
      "3,4,5,1\tmfold version 3.6\n",
      "20,21,22,310,1\t200,210,410, \n",
      "End of Fill\n",
      "Save file created using nafold.\n",
      "Minimum folding energy is -148.90 kcal/mol.\n",
      "Energy increment is 7.45 kcal/mol.\n",
      "380,390,400,410, \n",
      "End of Fill\n",
      "1\tH-num file created from plot file.\n",
      "1,2\t7\tREUSE= NO\n",
      "6,7,8,9,10,220,320,330,23,24,H-num file created from plot file.\n",
      "1,Save file created using nafold.\n",
      "Minimum folding energy is -99.90 kcal/mol.\n",
      "Energy increment is 5.00 kcal/mol.\n",
      "2,3,4,SEQ.pnt created.\n",
      "340,350, \n",
      "5\t230,11,12,13,14,15,360,370,380,2,3,4,5,6,7,5,6,7,8,9,Sequence length is 418\n",
      "240,250,16,17,18,19,H-num file created from plot file.\n",
      "1,H-num file created from plot file.\n",
      "/home/jupyter/plant_microRNA_prediction390,400,410, \n",
      "End of Fill\n",
      "\n",
      "6\t8,9,10,11,19\t10,11,12,13,14,15,RNA free energy files (version 2.3) at 20 degrees created.\n",
      "260,/home/jupyter/plant_microRNA_prediction/secondary_structure/mfold/AMWY020000611|1-346|129-146|+20,21, \n",
      "\n",
      "2,3,4,5,17\tSave file created using nafold.\n",
      "Minimum folding energy is -118.40 kcal/mol.\n",
      "Energy increment is 5.92 kcal/mol.\n",
      "1,2,3,12,13,14,16,17,270,10,20,30,40,50,18\t6,7,8,9,10,3\t19\t6\tmfold version 3.6\n",
      "280,290,300,60,70,80,8\t15,16,17,18,19,20,18,19,20,21,2\tH-num file created from plot file.\n",
      "4,5,6,7, \n",
      "2\t11,12,13,REUSE= NO\n",
      "3\t2\t310,320,1,2,3,22, \n",
      "21,22,23,24,25,90,100,110,14,15,330,340,6Suboptimal foldings created.\n",
      "Energy dot plot created.\n",
      "26,27,28,SEQ.pnt created.\n",
      "Sequence length is 346\n",
      "120,130,4,5,6,7,8,350,360,370,\t10\t16,17,18,19, \n",
      "RNA free energy files (version 2.3) at 20 degrees created.\n",
      "9,10,11,12,13,14,15,Suboptimal foldings created.\n",
      "140,150,380,390,400,410, \n",
      "20,21,22,16,17,10,20,30,40,7\tEnergy dot plot created.\n",
      "1\t1\t18\tEnd of Fill\n",
      "Save file created using nafold.\n",
      "Minimum folding energy is -89.90 kcal/mol.\n",
      "Energy increment is 4.50 kcal/mol.\n",
      "160,170,23, \n",
      "Suboptimal foldings created.\n",
      "Energy dot plot created.\n",
      "18,19,20,50,60,70,19\t1\t7\t21,22,23,24,25,80,90,180,190,4\t10\t3\tSuboptimal foldings created.\n",
      "Energy dot plot created.\n",
      "H-num file created from plot file.\n",
      "9\t3\t26,27,28,29,3\t100,110,200,1,2,3,4,5,6,4\t710\t120, \n",
      "210,\t7,8,11\t220,130,140,9,10, \n",
      "Suboptimal foldings created.\n",
      "Energy dot plot created.\n",
      "2\t8\t150,160,19\t10\t230,240,250,20\t1\t8\tSuboptimal foldings created.\n",
      "Energy dot plot created.\n",
      "170,180,190,20\t11\t260,5\t4\t5\t4\t200,Suboptimal foldings created.\n",
      "270,280,10\t11\t\n",
      "Structure plots generated.\n",
      "4\t8\tAll done.\n",
      "210,220,290,Suboptimal foldings created.\n",
      "Energy dot plot created.Energy dot plot created.\n",
      "10\t230,240,300,310,\n",
      "10\t320,330,12\t250,260,270,280,290,3\t340,9\t11\t/home/jupyter/plant_microRNA_prediction1\t\n",
      "300,310,320,Suboptimal foldings created.\n",
      "Energy dot plot created.\n",
      "6\t21\t350,360,/home/jupyter/plant_microRNA_prediction/secondary_structure/mfold/AMWY020000621|4390-4809|201-220|+20\t\n",
      "21\t9\t12\t330,340, \n",
      "End of Fill\n",
      "Save file created using nafold.\n",
      "11\t5\t5\t10\t370,380,390,400,Minimum folding energy is -81.50 kcal/mol.\n",
      "Energy increment is 4.08 kcal/mol.\n",
      "mfold version 3.6\n",
      "6\t9\t12\t5\t410, \n",
      "End of Fill\n",
      "Save file created using nafold.\n",
      "REUSE= NO\n",
      "Minimum folding energy is -159.70 kcal/mol.\n",
      "Energy increment is 7.99 kcal/mol.\n",
      "1\t11\tH-num file created from plot file.\n",
      "1,13\t2,3,4,5,6,7,4\tSEQ.pnt created.\n",
      "Sequence length is 420\n",
      "8, \n",
      "H-num file created from plot file.\n",
      "\n",
      "Structure plots generated.\n",
      "All done.\n",
      "12\t\n",
      "Structure plots generated.\n",
      "All done.\n",
      "20\t22\t1,2,3,4,21\t22\t6\t11\t7\tRNA free energy files (version 2.3) at 20 degrees created.\n",
      "10,5,6,7,13\t6\t12\t\n",
      "Structure plots generated.\n",
      "All done.\n",
      "20,30,40,50,60,Suboptimal foldings created.\n",
      "8,9,10,13\t70,80,6\t7\t11,12,Energy dot plot created.\n",
      "1\t2\t90,100,12\t \n",
      "/home/jupyter/plant_microRNA_prediction\n",
      "/home/jupyter/plant_microRNA_prediction/home/jupyter/plant_microRNA_prediction/secondary_structure/mfold/AMWY020001361|5149-5567|201-219|+\n",
      "/home/jupyter/plant_microRNA_prediction/secondary_structure/mfold/AMWY020000701|933-1352|201-220|+\n",
      "\n",
      "110,120,14\t/home/jupyter/plant_microRNA_prediction\n",
      "130,5\t/home/jupyter/plant_microRNA_prediction/secondary_structure/mfold/AMWY020000211|1481-1899|201-219|+mfold version 3.6\n",
      "\n",
      "mfold version 3.6\n",
      "140,REUSE= NO\n",
      "7\t13\t21\t22\tmfold version 3.6\n",
      "7\t23\t8\tREUSE= NO\n",
      "150,160,2\t12\t14\t13\tSEQ.pnt created.\n",
      "Sequence length is 420\n",
      "SEQ.pnt created.\n",
      "Sequence length is 419\n",
      "170,7\tREUSE= NO\n",
      "2\tSuboptimal foldings created.\n",
      "Energy dot plot created.\n",
      "8\tRNA free energy files (version 2.3) at 20 degrees created.\n",
      "RNA free energy files (version 2.3) at 20 degrees created.\n",
      "180,190,14\t3\t10\t10,10,20,30,SEQ.pnt created.\n",
      "40,50,13\t20,30,40,50,60,200,210,Sequence length is 419\n",
      "15\t60,6\t70,80,220,70,RNA free energy files (version 2.3) at 20 degrees created.\n",
      "10,20,30,40,90,100,110,8\t8\t3\t23\t2\t13\t80,90,230,240,9\t22\t50,60,14\t15\t14\t120,130,250,260,100,110,3\t9\t70,80,90,8\t15\t140,150,120,130,11\t270,100,110,120,160,170,4\t140,150,280,290,130,140,14\t180,16\t160,170,150,160,7\t300,310,190,9\t24\t180,170,3\t\n",
      "Structure plots generated.\n",
      "200,All done.\n",
      "320,330,340,4\t9\t190,210,23\t350,360,15\t15\t16\t14\t4\t180,220,200,210,370,380,390,400,410,\n",
      "Structure plots generated.\n",
      "All done.\n",
      "190,230,12\t220,9\t \n",
      "End of Fill\n",
      "Save file created using nafold.\n",
      "Minimum folding energy is -124.80 kcal/mol.\n",
      "200,16\t15\t5\t240,230,/home/jupyter/plant_microRNA_predictionEnergy increment is 6.24 kcal/mol.\n",
      "\n",
      "210,220,/home/jupyter/plant_microRNA_prediction/secondary_structure/mfold/AMWY020000811|5090-5507|201-218|+250,\n",
      "\n",
      "Structure plots generated.\n",
      "All done.\n",
      "240,230,260,H-num file created from plot file.\n",
      "\n",
      "Structure plots generated.\n",
      "All done.\n",
      "17\t250,mfold version 3.6\n",
      "/home/jupyter/plant_microRNA_prediction240,\n",
      "1,2,3,270,/home/jupyter/plant_microRNA_prediction/secondary_structure/mfold/AMWY020000111|50-468|201-219|-\n",
      "2\t4\t260,REUSE= NO\n",
      "250,4,5,6,280,290,\n",
      "Structure plots generated.\n",
      "All done.\n",
      "5\t16\t15\tSEQ.pnt created.\n",
      "Sequence length is 418\n",
      "24\t/home/jupyter/plant_microRNA_prediction16\tmfold version 3.6\n",
      "7,8,9,10,11,\n",
      "270,5\t/home/jupyter/plant_microRNA_prediction/secondary_structure/mfold/AMWY020001201|1226-1645|201-220|+300,310,260,17\t\n",
      "12,13, \n",
      "\n",
      "Structure plots generated.\n",
      "320,330,280,290,REUSE= NO\n",
      "All done.\n",
      "270,280,/home/jupyter/plant_microRNA_prediction\n",
      "RNA free energy files (version 2.3) at 20 degrees created.\n",
      "/home/jupyter/plant_microRNA_prediction/secondary_structure/mfold/AMWY020001411|1-290|73-90|+\n",
      "1\tmfold version 3.6\n",
      "17\t300,310,340,350,360,SEQ.pnt created.\n",
      "Sequence length is 419\n",
      "6\t16\t290,10,20,30,40,50,REUSE= NO\n",
      "/home/jupyter/plant_microRNA_predictionRNA free energy files (version 2.3) at 20 degrees created.\n",
      "320,330,\n",
      "370,380,390,mfold version 3.6\n",
      "300,310,/home/jupyter/plant_microRNA_prediction/secondary_structure/mfold/AMWY020001481|2931-3190|201-222|+\n",
      "60,70,340,350,360,10,20,30,18\t400,410, \n",
      "End of Fill\n",
      "Save file created using nafold.\n",
      "REUSE= NO\n",
      "SEQ.pnt created.\n",
      "Sequence length is 420\n",
      "320,330,340,80,90,100,3\t370,380,390,400,410,40,50,60,/home/jupyter/plant_microRNA_prediction\n",
      "mfold version 3.6\n",
      "/home/jupyter/plant_microRNA_prediction/secondary_structure/mfold/AMWY020000261|1715-2132|201-218|+RNA free energy files (version 2.3) at 20 degrees created.\n",
      "\n",
      "SEQ.pnt created.\n",
      "Sequence length is 290\n",
      "350,360,Minimum folding energy is -139.70 kcal/mol.\n",
      "Energy increment is 6.99 kcal/mol.\n",
      "6\t110,120, \n",
      "End of Fill\n",
      "Save file created using nafold.\n",
      "70,80,5\t17\tREUSE= NO\n",
      "10,20,30,40,17\tRNA free energy files (version 2.3) at 20 degrees created.\n",
      "16\tSuboptimal foldings created.\n",
      "90,100,110,Minimum folding energy is -150.70 kcal/mol.\n",
      "Energy increment is 7.54 kcal/mol.\n",
      "370,380,390,130,6\t25\tSEQ.pnt created.\n",
      "Sequence length is 260\n",
      "mfold version 3.6\n",
      "50,60,400,410, \n",
      "End of Fill\n",
      "Save file created using nafold.\n",
      "18\tH-num file created from plot file.\n",
      "2\t140,150,120,Energy dot plot created.\n",
      "10\t10,20,30,40,50,REUSE= NO\n",
      "Minimum folding energy is -162.60 kcal/mol.\n",
      "Energy increment is 8.13 kcal/mol.\n",
      "70,80,90,H-num file created from plot file.\n",
      "RNA free energy files (version 2.3) at 20 degrees created.\n",
      "7\t60,70,80,160,1,2,3,18\t130,140,1,2,3,4,17\t10,20,30,40,50,100,110,SEQ.pnt created.\n",
      "Sequence length is 418\n",
      "90,100,110,4,5,6,7,8,5,6,7,8,60,70,80,170,120,130,150,H-num file created from plot file.\n",
      "90,100,110,RNA free energy files (version 2.3) at 20 degrees created.\n",
      "19\t120,7\t180,190,4\t160,170,9,10,11,12,140,1,2,3,4,9,10,11,18\t120,6\t200,2\t10,20,30,40,150,160,13,14,15,130,140,150,5,6,7,8,9,17\t180,12,13,14,15,18\t7\t19\t130,140, \n",
      "50,60,70,80,11\t16,17,18,19,160,170,210,10,11,12,190,170,3\t8\t150,160,170,20,21, \n",
      "13,14,15,16,180,220,180,190,19\t200,90,100,180,190,200,18\t200,210,220,210,220,230,17,18,19,20,21,190,110,120,210,220,230,240,250, \n",
      "End of Fill\n",
      "230,240,250,230,1\t240,250,130,140,150,200,210,8\t22,23,24, \n",
      "5\t240,Save file created using nafold.\n",
      "Minimum folding energy is -89.10 kcal/mol.\n",
      "Energy increment is 4.46 kcal/mol.\n",
      "260,260,270,160,3\t7\t8\t19\t250,220,18\t270,12\tH-num file created from plot file.\n",
      "1,1\t19\tSuboptimal foldings created.\n",
      "Energy dot plot created.\n",
      "280, \n",
      "End of Fill\n",
      "Save file created using nafold.\n",
      "Minimum folding energy is -97.20 kcal/mol.\n",
      "Energy increment is 4.86 kcal/mol.\n",
      "170,180,230,240,2,3,4,5,6,7,8,9,10,11,12,10\t260,270,H-num file created from plot file.\n",
      "280,290,9\t250,190,1\t4\t13, \n",
      "19\t280,1,2,3,Suboptimal foldings created.\n",
      "Energy dot plot created.\n",
      "300,20\t200,260,4,5,6, \n",
      "10\t290,300,310,9\t270,210,\n",
      "Structure plots generated.\n",
      "All done.\n",
      "1\t6\t19\t220,280,290,4\t310,320,13\t320,330,340,Suboptimal foldings created.\n",
      "Energy dot plot created.\n",
      "1\t8\tSuboptimal foldings created.\n",
      "Energy dot plot created.\n",
      "20\t1\t300,230,350,360,370,330,340,10\t11\tSuboptimal foldings created.\n",
      "Energy dot plot created.\n",
      "240,310,320,380,390,400,410, \n",
      "End of Fill\n",
      "350,360,370,\n",
      "Structure plots generated.\n",
      "All done.\n",
      "10\t250,5\t330,340,Save file created using nafold.\n",
      "Minimum folding energy is -160.00 kcal/mol.\n",
      "Energy increment is 8.00 kcal/mol.\n",
      "380,390,400,410, \n",
      "End of Fill\n",
      "20\t/home/jupyter/plant_microRNA_prediction350,360,260,\n",
      "1\t/home/jupyter/plant_microRNA_prediction/secondary_structure/mfold/AMWY020001481|2932-3190|201-221|+Save file created using nafold.\n",
      "Minimum folding energy is -162.20 kcal/mol.\n",
      "\n",
      "21\t11\t270,280,370,380,\n",
      "Structure plots generated.\n",
      "All done.\n",
      "Energy increment is 8.11 kcal/mol.\n",
      "H-num file created from plot file.\n",
      "mfold version 3.6\n",
      "290,390,400,410, \n",
      "End of Fill\n",
      "Save file created using nafold.\n",
      "2\tH-num file created from plot file.\n",
      "1\t7\t11\t/home/jupyter/plant_microRNA_prediction\n",
      "1,2,3,4,REUSE= NO\n",
      "Minimum folding energy is -129.30 kcal/mol.\n",
      "Energy increment is 6.46 kcal/mol.\n",
      "/home/jupyter/plant_microRNA_prediction/secondary_structure/mfold/AMWY020001481|2934-3190|201-218|+20\t300,20\t1\t\n",
      "9\t5\t1,21\t12\t5,6,7,8,SEQ.pnt created.\n",
      "Sequence length is 259\n",
      "310,320,2,3,4,5,H-num file created from plot file.\n",
      "9,10,11,RNA free energy files (version 2.3) at 20 degrees created.\n",
      "10,mfold version 3.6\n",
      "330,340,11\t/home/jupyter/plant_microRNA_prediction\n",
      "6,7,8,1,2,3,4,12,13,14,15,16,17,/home/jupyter/plant_microRNA_prediction/secondary_structure/mfold/AMWY020000811|7255-7674|201-220|+20,30,40,50,60,6\t21\t350,360,REUSE= NO\n",
      "\n",
      "5,6,7,8,18,19,20,9,10, \n",
      "22\t20\t370,380,390,400,70,80,90,SEQ.pnt created.\n",
      "Sequence length is 257\n",
      "12\t9,10,11,21, \n",
      "100,110,120,mfold version 3.6\n",
      "410, \n",
      "End of Fill\n",
      "Save file created using nafold.\n",
      "Minimum folding energy is -161.10 kcal/mol.\n",
      "Energy increment is 8.05 kcal/mol.\n",
      "8\tRNA free energy files (version 2.3) at 20 degrees created.\n",
      "12,REUSE= NO\n",
      "3\t12\t130,20\t13,14,21\t\n",
      "Structure plots generated.\n",
      "All done.\n",
      "10,20,30,40,6\t21\tH-num file created from plot file.\n",
      "22\t140,150,2\t13\t15, \n",
      "50,60,70,SEQ.pnt created.\n",
      "Sequence length is 420\n",
      "1,2,3,160,170,80,90,100,12\tSuboptimal foldings created.\n",
      "Energy dot plot created.\n",
      "10\t7\t4,5,6,7,RNA free energy files (version 2.3) at 20 degrees created.\n",
      "2\t180,190,200,110,120,8,9,10,21\t10,20,30,40,50,60,13\t210,220,230,23\t/home/jupyter/plant_microRNA_prediction130,140,240,250, \n",
      "End of Fill\n",
      "Save file created using nafold.\n",
      "11,12,13,70,80,\n",
      "13\t4\t/home/jupyter/plant_microRNA_prediction/secondary_structure/mfold/AMWY020000471|9036-9454|201-219|-150,160,Minimum folding energy is -89.10 kcal/mol.\n",
      "Energy increment is 4.46 kcal/mol.\n",
      "\n",
      "90,100,110,Suboptimal foldings created.\n",
      "Energy dot plot created.\n",
      "14,15,22\t9\t7\t170,180,190,14\tSuboptimal foldings created.\n",
      "H-num file created from plot file.\n",
      "21\t10\t120, \n",
      "3\t22\tmfold version 3.6\n",
      "23\t1,2,3,200,210,Energy dot plot created.\n",
      "10\t130,140,220,230,240,250, \n",
      "4,5,6,7,8,9,10,13\t1\t8\t150,160,REUSE= NO\n",
      "3\t22\tEnd of Fill\n",
      "Save file created using nafold.\n",
      "Minimum folding energy is -89.10 kcal/mol.\n",
      "Energy increment is 4.46 kcal/mol.\n",
      "11,12, \n",
      "170,180,SEQ.pnt created.\n",
      "Sequence length is 419\n",
      "24\t1\t14\t190,200,H-num file created from plot file.\n",
      "RNA free energy files (version 2.3) at 20 degrees created.\n",
      "5\t210,Suboptimal foldings created.\n",
      "\n",
      "Structure plots generated.\n",
      "All done.\n",
      "1,2,3,10,20,30,40,23\t220,4\tSuboptimal foldings created.\n",
      "11\tEnergy dot plot created.\n",
      "10\t24\t22\t4,5,6,7,8,9,10,2\t50,60,70,8\t15\t11\t230,Energy dot plot created.\n",
      "10\t11,12, \n",
      "80,90,100,2\t14240,250,9\t4\t\t110,120,130,260,15\t23\t/home/jupyter/plant_microRNA_prediction\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2\t/home/jupyter/plant_microRNA_prediction/secondary_structure/mfold/AMWY020000301|687-1105|201-219|-140,150,270,280,\n",
      "2\tSuboptimal foldings created.\n",
      "6\t160,290,300,mfold version 3.6\n",
      "Energy dot plot created.\n",
      "10\t170,180,24\t11\t12\tREUSE= NO\n",
      "23\t310,320,3\t11\t190,9\t5\tSEQ.pnt created.\n",
      "330,340,350,360,1\t2\t12\t200,3\tSequence length is 419\n",
      "\n",
      "Structure plots generated.\n",
      "All done.\n",
      "370,380,390,400,15\t210,220,5\tRNA free energy files (version 2.3) at 20 degrees created.\n",
      "16\t410, \n",
      "End of Fill\n",
      "Save file created using nafold.\n",
      "230,2\tMinimum folding energy is -145.10 kcal/mol.\n",
      "Energy increment is 7.25 kcal/mol.\n",
      "10,20,30,40,50,3\t\n",
      "Structure plots generated.\n",
      "All done.\n",
      "3\t240,H-num file created from plot file.\n",
      "60,11\t250,1,70,80,90,/home/jupyter/plant_microRNA_prediction\n",
      "/home/jupyter/plant_microRNA_prediction/secondary_structure/mfold/AMWY020001541|671-1094|201-224|+4\t\n",
      "2,3,4,5,6,260,12\t100,24\t13\t25\t12\t\n",
      "Structure plots generated.\n",
      "All done.\n",
      "7,8,9,6\t270,110,3\tmfold version 3.6\n",
      "4\t10,11,12,/home/jupyter/plant_microRNA_prediction280,290,300,13\t2\t\n",
      "REUSE= NO\n",
      "120,130,16\t13,14,15,/home/jupyter/plant_microRNA_prediction/secondary_structure/mfold/AMWY020001481|20-441|201-222|-310,320,\n",
      "140,6\t3\t17\t16,17,18,4\tSEQ.pnt created.\n",
      "Sequence length is 424\n",
      "330,340,150,4\t19, \n",
      "12\tmfold version 3.6\n",
      "/home/jupyter/plant_microRNA_prediction\n",
      "/home/jupyter/plant_microRNA_prediction/secondary_structure/mfold/AMWY020001361|1083-1501|201-219|-160,350,360,\n",
      "REUSE= NO\n",
      "RNA free energy files (version 2.3) at 20 degrees created.\n",
      "370,380,390,400,1\t170,SEQ.pnt created.\n",
      "Sequence length is 422\n",
      "13\t25\t5\t10,20,30,40,50,60,14\t26\tmfold version 3.6\n",
      "410, \n",
      "End of Fill\n",
      "Save file created using nafold.\n",
      "180,70,80,REUSE= NO\n",
      "7\tRNA free energy files (version 2.3) at 20 degrees created.\n",
      "4\t14\tMinimum folding energy is -166.40 kcal/mol.\n",
      "Energy increment is 8.32 kcal/mol.\n",
      "5\t3\t190,90,100,SEQ.pnt created.\n",
      "17\t10,20,30,40,200,18\t110,120,130,5\t4\tSequence length is 419\n",
      "H-num file created from plot file.\n",
      "50,60,70,80,7\tRNA free energy files (version 2.3) at 20 degrees created.\n",
      "10,20,30,1\t210,140,150,5\t90,1,2,3,2\t40,50,60,220,230,100,Suboptimal foldings created.\n",
      "Energy dot plot created.\n",
      "160,170,4,5,6,7,8,26\t27\t70,80,90,100,15\t10\t110,120,14\t180,9,10,11,6\t110,240,5\t130,140,190,200,8\t4\t12,13,14,15,18\t120,130,250,150,6\t210,15\t16,17,18,19,8\t5\t140,260,6\t160,170,19\t220,230,2\t270,280,150,160,20,21,22,23,24,6\t240,3\t290,180,190,11\t25,15\t300,16\t27\t250,260,28\t7\t \n",
      "200,170,180,190,310,320,210,6\t200,270,280,5\t9\t1\t7\t7\t330,340,220,230,19\t210,220,290,300,350,240,6\t310,320,330,3\t1\t4\t250,230,360,370,9\t340,350,240,250,260,380,390,400,7\t12\t360,370,380,1\t17\t270,280,2\t8\t410, \n",
      "End of Fill\n",
      "Save file created using nafold.\n",
      "Minimum folding energy is -126.60 kcal/mol.\n",
      "390,400,410,28\t260,290,Energy increment is 6.33 kcal/mol.\n",
      "\n",
      "Structure plots generated.\n",
      "All done.\n",
      "420, \n",
      "End of Fill\n",
      "Save file created using nafold.\n",
      "300,270,280,8\t7\t6\t290,Minimum folding energy is -136.10 kcal/mol.\n",
      "Energy increment is 6.80 kcal/mol.\n",
      "2\t310,320,1\tSuboptimal foldings created.\n",
      "Energy dot plot created.\n",
      "10\tH-num file created from plot file.\n",
      "8\t330,300,310,1,2,3,7\t4\t\n",
      "Structure plots generated.\n",
      "All done.\n",
      "H-num file created from plot file.\n",
      "320,340,350,360,4,5,6,7,5\t20\t1,2,370,380,/home/jupyter/plant_microRNA_prediction330,340,8,9,10,11,12,8\t18\t\n",
      "3,4,5,6,7,/home/jupyter/plant_microRNA_prediction/secondary_structure/mfold/AMWY020001621|7042-7462|201-221|+390,400,410,420, \n",
      "End of Fill\n",
      "2\t13\t350,360,370,3\t\n",
      "13,14,15,16,29\tSave file created using nafold.\n",
      "Minimum folding energy is -169.60 kcal/mol.\n",
      "8,9,10,9\t380,390,400,410, 17,18,19,20,21,22,/home/jupyter/plant_microRNA_prediction11,12,13,14,15,Energy increment is 8.48 kcal/mol.\n",
      "mfold version 3.6\n",
      "\n",
      "/home/jupyter/plant_microRNA_prediction/secondary_structure/mfold/AMWY020001171|364-784|201-221|+9\t\n",
      "8\t\n",
      "End of Fill\n",
      "Save file created using nafold.\n",
      "Minimum folding energy is -145.80 kcal/mol.\n",
      "Energy increment is 7.29 kcal/mol.\n",
      "7\t23,24,25,26,3\t16,17,18,19,11\t9\t20\tREUSE= NO\n",
      "27,28,H-num file created from plot file.\n",
      "20,mfold version 3.6\n",
      "6\tSEQ.pnt created.\n",
      "Sequence length is 421\n",
      "5\t29, \n",
      "8\t21\tREUSE= NO\n",
      "21,22,23,1,2,3,4,H-num file created from plot file.\n",
      "RNA free energy files (version 2.3) at 20 degrees created.\n",
      " \n",
      "5,6,7,10,1,2,3,4,9\t4\t3\t19\t20,30,40,50,8,9,10,11, \n",
      "SEQ.pnt created.\n",
      "5,6,7,8,9,14\t2\t60,70,\n",
      "Structure plots generated.\n",
      "All done.\n",
      "10,11,Sequence length is 421\n",
      "8\t\n",
      "Structure plots generated.\n",
      "All done.\n",
      "80,90,9\t4\t12,13,14,15,12\tRNA free energy files (version 2.3) at 20 degrees created.\n",
      "10,100,21\t6\t16,17,\n",
      "Structure plots generated.\n",
      "All done.\n",
      "20,30,40,50,110,7\t60,70, \n",
      "120,130,9\t2\t80,90,100,Suboptimal foldings created.\n",
      "Energy dot plot created.\n",
      "\n",
      "Structure plots generated.\n",
      "All done.\n",
      "/home/jupyter/plant_microRNA_prediction\n",
      "140,\n",
      "/home/jupyter/plant_microRNA_prediction/secondary_structure/mfold/AMWY020000881|2343-2731|201-220|+/home/jupyter/plant_microRNA_prediction110,5\t\n",
      "10\t/home/jupyter/plant_microRNA_prediction/secondary_structure/mfold/AMWY020001621|747-1166|201-220|-150,\n",
      "4\t120,130,1\t/home/jupyter/plant_microRNA_predictionmfold version 3.6\n",
      "3\t160,\n",
      "140,150,160,/home/jupyter/plant_microRNA_prediction/secondary_structure/mfold/AMWY020001481|2932-3190|201-220|+\n",
      "mfold version 3.6\n",
      "15\t170,180,9\t5\tREUSE= NO\n",
      "REUSE= NO\n",
      "/home/jupyter/plant_microRNA_predictionSuboptimal foldings created.\n",
      "Energy dot plot created.\n",
      "10\t170,\n",
      "\n",
      "Structure plots generated.\n",
      "7\t/home/jupyter/plant_microRNA_prediction/secondary_structure/mfold/AMWY020001301|1906-2324|201-219|+13\tAll done.\n",
      "190,SEQ.pnt created.\n",
      "Sequence length is 389\n",
      "\n",
      "mfold version 3.6\n",
      "SEQ.pnt created.\n",
      "Sequence length is 420\n",
      "180,Suboptimal foldings created.\n",
      "Energy dot plot created.\n",
      "RNA free energy files (version 2.3) at 20 degrees created.\n",
      "8\t200,Suboptimal foldings created.\n",
      "Energy dot plot created.\n",
      "\n",
      "Structure plots generated.\n",
      "All done.\n",
      "22\tREUSE= NO\n",
      "190,200,10\t3\tmfold version 3.6\n",
      "10,20,30,40,210,210,10\tRNA free energy files (version 2.3) at 20 degrees created.\n",
      "REUSE= NO\n",
      "SEQ.pnt created.\n",
      "Sequence length is 259\n",
      "50,60,70,80,220,230,10,20,30,40,50,11\t220,6\t/home/jupyter/plant_microRNA_prediction\n",
      "5\t20\t60,70,80,240,/home/jupyter/plant_microRNA_prediction/secondary_structure/mfold/AMWY020001631|753-1171|201-219|+90,100,110,\n",
      "SEQ.pnt created.\n",
      "Sequence length is 419230,240,4\tRNA free energy files (version 2.3) at 20 degrees created.\n",
      "250,16\t\n",
      "Structure plots generated.\n",
      "All done.\n",
      "\n",
      "90,100,/home/jupyter/plant_microRNA_prediction\n",
      "6\t250,120,130,140,/home/jupyter/plant_microRNA_prediction/secondary_structure/mfold/AMWY020000981|438-709|201-224|+8\t\n",
      "11\tmfold version 3.6\n",
      "260,10,20,30,40,50,60,110,120,130,RNA free energy files (version 2.3) at 20 degrees created.\n",
      "150,160,260,270,140,150,14\t270,9\tREUSE= NO\n",
      "mfold version 3.6\n",
      "70,80,90,10,20,30,40,50,23\t170,280,290,11\t280,100,11\tSEQ.pnt created.\n",
      "4\t60,70,80,160,REUSE= NO\n",
      "300,290,180,190,110,120,/home/jupyter/plant_microRNA_prediction7\t\n",
      "/home/jupyter/plant_microRNA_prediction/secondary_structure/mfold/AMWY020001651|1553-1972|201-220|-\n",
      "1\t170,180,310,320,Sequence length is 419\n",
      "90,100,21\t130,140,150,200,210,300,310,SEQ.pnt created.\n",
      "RNA free energy files (version 2.3) at 20 degrees created.\n",
      "10,6\t110,120,330,mfold version 3.6\n",
      "160,170,220,230,190,320,330,17\t130,140,Sequence length is 272\n",
      "7\t9\t5\t20,30,40,50,REUSE= NO\n",
      "\n",
      "Structure plots generated.\n",
      "All done.\n",
      "340,350,360,370,240,200,340,350,360,180,190,150,RNA free energy files (version 2.3) at 20 degrees created.\n",
      "10,60,70,80,380,390,400,410,420, \n",
      "End of Fill\n",
      "Save file created using nafold.\n",
      "15\t250,260,370,380,390,400,200,210,220,210,12\tSEQ.pnt created.\n",
      "Sequence length is 420\n",
      "20,30,40,50,12\t160,90,100,Minimum folding energy is -144.70 kcal/mol.\n",
      "Energy increment is 7.24 kcal/mol.\n",
      "220,230,410,420, \n",
      "End of Fill\n",
      "Save file created using nafold.\n",
      "Minimum folding energy is -128.80 kcal/mol.\n",
      "230,240,250, \n",
      "End of Fill\n",
      "RNA free energy files (version 2.3) at 20 degrees created.\n",
      "24\t12\t270,60,70,170,180,110,5\tSave file created using nafold.\n",
      "Minimum folding energy is -89.10 kcal/mol.\n",
      "Energy increment is 4.46 kcal/mol.\n",
      "2\t10,20,30,40,240,Energy increment is 6.44 kcal/mol.\n",
      "H-num file created from plot file.\n",
      "190,280,290,300,310,120,130,8\t2\t80,90,50,60,/home/jupyter/plant_microRNA_predictionH-num file created from plot file.\n",
      "\n",
      "250,140,200,320,330,/home/jupyter/plant_microRNA_prediction/secondary_structure/mfold/AMWY020001481|2933-3190|201-219|+1,2,3,4,5,\n",
      "H-num file created from plot file.\n",
      "100,110,7\t\n",
      "Structure plots generated.\n",
      "All done.\n",
      "260,70,80,150,160,210,18\t1,2,8\t1,2,3,4,5,6,7,8,340,350,360,370,120,6\tmfold version 3.6\n",
      "6,7,8,9,10,220,90,100,110,3,4,5,6,7,8,9,10,11,12, \n",
      "130,140,380, \n",
      "End of Fill\n",
      "Save file created using nafold.\n",
      "270,280,170,REUSE= NO\n",
      "230,11,12,13,14,15,9,10,11,290,300,150,13\t16\tMinimum folding energy is -115.20 kcal/mol.\n",
      "Energy increment is 5.76 kcal/mol.\n",
      "13\t \n",
      "240,12,13,14,180,310,160,120,130,SEQ.pnt created.\n",
      "Sequence length is 258\n",
      "2\t13\t15, \n",
      "250,260,190,200,170,140,150,3\t6\tRNA free energy files (version 2.3) at 20 degrees created.\n",
      "H-num file created from plot file.\n",
      "320,/home/jupyter/plant_microRNA_prediction\n",
      "/home/jupyter/plant_microRNA_prediction/secondary_structure/mfold/AMWY020001481|2934-3190|201-221|+9\t160,270,\n",
      "180,190,Suboptimal foldings created.\n",
      "Energy dot plot created.\n",
      "10\t1,2,3,4,330,340,350,10,20,30,40,50,210,3\t8\t170,19\t280,290,60,70,80,220,200,210,220,230,240,360,370,380,9\t5,6,7,8,mfold version 3.6\n",
      "180,90,100,110,390,400,410, \n",
      "End of Fill\n",
      "250,260,270, \n",
      "End of Fill\n",
      "Save file created using nafold.\n",
      "9,10, \n",
      "REUSE= NO\n",
      "300,310,7\tSuboptimal foldings created.\n",
      "Energy dot plot created.\n",
      "10\t190,120,230,14\t14\tSave file created using nafold.\n",
      "Minimum folding energy is -133.90 kcal/mol.\n",
      "Energy increment is 6.70 kcal/mol.\n",
      "Minimum folding energy is -63.70 kcal/mol.\n",
      "Energy increment is 3.19 kcal/mol.\n",
      "320,330,340,17\t200,240,250,130,140,Suboptimal foldings created.\n",
      "Energy dot plot created.\n",
      "3\tSEQ.pnt created.\n",
      "H-num file created from plot file.\n",
      "350,360,370,380,390,210,14\t7\t4\tSequence length is 257\n",
      "260,11\t10\t150,160,170,1,2,3,4,5,6, \n",
      "400,410, \n",
      "End of Fill\n",
      "Save file created using nafold.\n",
      "\n",
      "Structure plots generated.\n",
      "All done.\n",
      "220,H-num file created from plot file.\n",
      "1\tRNA free energy files (version 2.3) at 20 degrees created.\n",
      "180,190,230,Minimum folding energy is -148.10 kcal/mol.\n",
      "Energy increment is 7.41 kcal/mol.\n",
      "270,280,4\tSuboptimal foldings created.\n",
      "Energy dot plot created.\n",
      "10\t9\t\n",
      "Structure plots generated.\n",
      "All done.\n",
      "1,2,3,10,20,30,40,50,60,200,210,220,230,240,240,290,300,H-num file created from plot file.\n",
      "Suboptimal foldings created.\n",
      "Energy dot plot created.\n",
      "8\t250, \n",
      "End of Fill\n",
      "Save file created using nafold.\n",
      "4,5,6,7,310,1,70,80,1\t/home/jupyter/plant_microRNA_prediction11\t250,2,3,4,5,6,15\t320,90,100,110,8,9,10,11,\n",
      "/home/jupyter/plant_microRNA_prediction/secondary_structure/mfold/AMWY020001671|2603-3023|201-221|+Minimum folding energy is -89.10 kcal/mol.\n",
      "Energy increment is 4.46 kcal/mol.\n",
      "\n",
      "18\t15\t \n",
      "4\t260,120,130,330,340,12,13,14,15\tH-num file created from plot file.\n",
      "270,/home/jupyter/plant_microRNA_prediction\n",
      "mfold version 3.6\n",
      "/home/jupyter/plant_microRNA_prediction/secondary_structure/mfold/AMWY020001731|1512-1929|201-218|+5\t8\t350,360,370,140,150,1,2,3,4,15,16,17,\n",
      "280,11\t12\tREUSE= NO\n",
      "380,390,400,410,Suboptimal foldings created.\n",
      "5,6,7,8,18,19,20,160,170,180,\n",
      "Structure plots generated.\n",
      "All done.\n",
      "2\t5\tSEQ.pnt created.\n",
      " \n",
      "End of Fill\n",
      "Save file created using nafold.\n",
      "290,1\t190,200,9,10,11,12, \n",
      "mfold version 3.6\n",
      "Energy dot plot created.\n",
      "1\t \n",
      "2\tMinimum folding energy is -104.80 kcal/mol.\n",
      "Energy increment is 5.24 kcal/mol.\n",
      "9\tSequence length is 421\n",
      "300,310,210,220,REUSE= NO\n",
      "RNA free energy files (version 2.3) at 20 degrees created.\n",
      "10,320,330,230,240,250, \n",
      "End of Fill\n",
      "Save file created using nafold.\n",
      "12\tH-num file created from plot file.\n",
      "1,SEQ.pnt created.\n",
      "19\t16\t340,350,16\t20,30,40,50,Minimum folding energy is -89.10 kcal/mol.\n",
      "Energy increment is 4.46 kcal/mol.\n",
      "/home/jupyter/plant_microRNA_prediction2,3,4,5,Sequence length is 418\n",
      "5\t\n",
      "Suboptimal foldings created.\n",
      "Energy dot plot created.\n",
      "360,370,/home/jupyter/plant_microRNA_prediction/secondary_structure/mfold/AMWY020001841|252-670|201-219|+\n",
      "9\t60,70,80,16\t6\t6,7,8,9,10,11,10\t380,390,400,RNA free energy files (version 2.3) at 20 degrees created.\n",
      "10,90,mfold version 3.6\n",
      "1\t12\tH-num file created from plot file.\n",
      "1,12,13,2\t20,30,40,6\t100,110,3\t2,3,4,5,6,7,8,REUSE= NO\n",
      "410, \n",
      "End of Fill\n",
      "Save file created using nafold.\n",
      "Minimum folding energy is -116.30 kcal/mol.\n",
      "Energy increment is 5.82 kcal/mol.\n",
      "\n",
      "Structure plots generated.\n",
      "All done.\n",
      "14,15,16,50,60,70,2\t9,10,11,12,Suboptimal foldings created.\n",
      "Energy dot plot created.\n",
      "120,SEQ.pnt created.\n",
      "Sequence length is 419\n",
      "3\t17, \n",
      " \n",
      "10\t130,140,H-num file created from plot file.\n",
      "80,90,1\t17\tRNA free energy files (version 2.3) at 20 degrees created.\n",
      "17\t13\t100,110,150,1,2,3,4,11\t\n",
      "Structure plots generated.\n",
      "All done.\n",
      "2\t10,20,30,40,50,6\t17\t5,6,7,8,9,/home/jupyter/plant_microRNA_prediction\n",
      "120,130,160,13\t/home/jupyter/plant_microRNA_prediction/secondary_structure/mfold/AMWY020001901|276-697|201-222|+7\t60,70,Suboptimal foldings created.\n",
      "Energy dot plot created.\n",
      "10\t\n",
      "170,10,11,12,140,150,3\t4\t80,90,100,3\t160,170,mfold version 3.6\n",
      "180,7\t13,14,15,16,17,110,120,4\t180,REUSE= NO\n",
      "/home/jupyter/plant_microRNA_prediction190,200,\n",
      "130,18,19,20,/home/jupyter/plant_microRNA_prediction/secondary_structure/mfold/AMWY020001941|1048-1466|201-219|-11\tSuboptimal foldings created.\n",
      "Energy dot plot created.\n",
      "10\tSEQ.pnt created.\n",
      "\n",
      "190,140,150,14\t18\t18\t210,3\t21,22,23,24,20\t200,1\t160,Sequence length is 422\n",
      "12\tmfold version 3.6\n",
      "220,14\t7\t25,26,27,28, \n",
      "210,170,8\tRNA free energy files (version 2.3) at 20 degrees created.\n",
      "10,20,30,REUSE= NO\n",
      "11\t230,4\t220,180,SEQ.pnt created.\n",
      "8\t40,50,5\t240,60,70,80,230,5\t4\tSequence length is 419\n",
      "190,200,240,RNA free energy files (version 2.3) at 20 degrees created.\n",
      "10,20,30,90,100,250,12\t210,19\t19\t15\t250,11\t110,40,50,60,260,270,4\t260,220,1\t21\t2\t70,80,90,120,130,9\t15\t12\t8\t270,230,240,280,290,5\t6\t100,110,140,150,6\t250,280,290,300,9\t160,120,130,300,310,310,320,5\t260,270,170,Suboptimal foldings created.\n",
      "Energy dot plot created.\n",
      "320,330,13\t140,150,330,1\t12\t280,1\t2\t1\t10\t340,350,360,1\t340,350,5\t290,3\t180,160,22\t360,370,370,380,390,400,410, \n",
      "End of Fill\n",
      "9\t1\t170,190,300,310,6\t380,390,Save file created using nafold.\n",
      "Minimum folding energy is -100.80 kcal/mol.\n",
      "Energy increment is 5.04 kcal/mol.\n",
      "\n",
      "Structure plots generated.\n",
      "All done.\n",
      "320,330,200,180,400,410,420, \n",
      "End of Fill\n",
      "Save file created using nafold.\n",
      "Minimum folding energy is -108.00 kcal/mol.\n",
      "Energy increment is 5.40 kcal/mol.H-num file created from plot file.\n",
      "\n",
      "Structure plots generated.\n",
      "All done.\n",
      "7\t340,350,\n",
      "Structure plots generated.\n",
      "190,6\t1,2,All done.\n",
      "210,\n",
      "360,370,11\t3,4,220,14\t3\t380,390,400,200,230,5,6,7,8,13\tH-num file created from plot file.\n",
      "1,2\t/home/jupyter/plant_microRNA_prediction210,2\t410, \n",
      "End of Fill\n",
      "Save file created using nafold.\n",
      "\n",
      "9,10,11,12,240,/home/jupyter/plant_microRNA_prediction/secondary_structure/mfold/AMWY020001961|176-594|201-219|+20\t2,3,4,5,\n",
      "6\t/home/jupyter/plant_microRNA_predictionMinimum folding energy is -156.00 kcal/mol.\n",
      "Energy increment is 7.80 kcal/mol.\n",
      "\n",
      "20\t23\t\n",
      "Structure plots generated.\n",
      "All done.\n",
      "/home/jupyter/plant_microRNA_prediction/secondary_structure/mfold/AMWY020001481|2931-3190|201-219|+4\t13,14, \n",
      "\n",
      "Structure plots generated.\n",
      "All done.\n",
      "\n",
      "2\t/home/jupyter/plant_microRNA_prediction6,7,8,9,10,\n",
      "/home/jupyter/plant_microRNA_prediction/secondary_structure/mfold/AMWY020002001|10002-10424|201-223|+\n",
      "220,mfold version 3.6\n",
      "250,11,12,13, \n",
      "230,H-num file created from plot file.\n",
      "mfold version 3.6\n",
      "REUSE= NO\n",
      "mfold version 3.6\n",
      "260,8\t240,1,2,3,4,REUSE= NO\n",
      "SEQ.pnt created.\n",
      "REUSE= NO\n",
      "4\t250,5,6,7,8,9,10,Sequence length is 419\n",
      "7\t270,/home/jupyter/plant_microRNA_prediction15\t/home/jupyter/plant_microRNA_prediction\n",
      "12\tSEQ.pnt created.\n",
      "Sequence length is 423\n",
      "\n",
      "SEQ.pnt created.\n",
      "Sequence length is 260\n",
      "3\t260,11,12,/home/jupyter/plant_microRNA_prediction/secondary_structure/mfold/AMWY020002201|395-814|201-220|+/home/jupyter/plant_microRNA_prediction/secondary_structure/mfold/AMWY020002061|1-215|66-83|-RNA free energy files (version 2.3) at 20 degrees created.\n",
      "280,290,\n",
      "14\t\n",
      "RNA free energy files (version 2.3) at 20 degrees created.\n",
      "RNA free energy files (version 2.3) at 20 degrees created.\n",
      "270,7\t13,14,15,16,21\t300,10,20,30,40,50,3\t280,mfold version 3.6\n",
      "10,20,30,40,mfold version 3.6\n",
      "10,20,30,40,50,17, \n",
      "5\t21\t3\t60,70,80,Suboptimal foldings created.\n",
      "Energy dot plot created.\n",
      "10\t310,320,Suboptimal foldings created.\n",
      "Energy dot plot created.\n",
      "REUSE= NO\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "290,50,60,70,24\t60,70,80,REUSE= NO\n",
      "90,100,330,10\t300,310,80,90,90,100,SEQ.pnt created.\n",
      "340,350,110,9\tSEQ.pnt created.\n",
      "Sequence length is 215\n",
      "110,100,110,320,330,13\t5\t8\tSequence length is 420\n",
      "360,370,380,120,130,140,16\tRNA free energy files (version 2.3) at 20 degrees created.\n",
      "10,20,30,120,130,4\t120,130,140,340,350,15\tRNA free energy files (version 2.3) at 20 degrees created.\n",
      "10,20,30,390,400,410,420, \n",
      "End of Fill\n",
      "22\t40,50,60,70,150,160,140,150,160,8\t360,370,150,160,4\t40,50,60,70,Save file created using nafold.\n",
      "Minimum folding energy is -135.00 kcal/mol.\n",
      "Energy increment is 6.75 kcal/mol.\n",
      "170,180,6\t80,90,100,170,180,380,390,400,170,180,4\t25\t22\t11\t80,90,190,110,120,190,200,Suboptimal foldings created.\n",
      "Energy dot plot created.\n",
      "190,200,H-num file created from plot file.\n",
      "410, \n",
      "End of Fill\n",
      "Save file created using nafold.\n",
      "11\t100,110,200,130,140,210,220,10\t\n",
      "Structure plots generated.\n",
      "All done.\n",
      "210,Minimum folding energy is -103.20 kcal/mol.\n",
      "Energy increment is 5.16 kcal/mol.\n",
      "120,130,230,240,250, \n",
      "End of Fill\n",
      "Save file created using nafold.\n",
      "150,160,170,180,210,14\t1,2,3,4,5,6,9\t6\t220,140,150,Minimum folding energy is -89.10 kcal/mol.\n",
      "Energy increment is 4.46 kcal/mol.\n",
      "190,200,210, \n",
      "End of Fill\n",
      "Save file created using nafold.\n",
      "17\t220,7,8,9,10,11,5\tH-num file created from plot file.\n",
      "9\t230,240,23\t160,230,12,13,14,15,16\tMinimum folding energy is -55.80 kcal/mol.\n",
      "Energy increment is 2.79 kcal/mol.\n",
      "1,2,3,4,5,6,250,H-num file created from plot file.\n",
      "170,240,250,16,17,18,19,12\t5\t/home/jupyter/plant_microRNA_prediction7,8,9,10,11,5\t\n",
      "1,2,3,4,5,6,7,8,/home/jupyter/plant_microRNA_prediction/secondary_structure/mfold/AMWY020002261|95-512|201-218|+180,190,23\t20,21,22,260,270,260,7\t\n",
      "H-num file created from plot file.\n",
      "1,2\t280,200,9,10,11,12,13, \n",
      "12,13,14,15,23,24, \n",
      "270,11\t12\t2,3,4, \n",
      "mfold version 3.6\n",
      "210,290,280,16,17,18,\n",
      "Structure plots generated.\n",
      "All done.\n",
      "\n",
      "Structure plots generated.\n",
      "All done.\n",
      "7\t15\tREUSE= NO\n",
      "6\t300,310,290,300,220,19,20, \n",
      "Suboptimal foldings created.\n",
      "Energy dot plot created.\n",
      "1\t17\tSEQ.pnt created.\n",
      "320,330,340,310,230,18\t350,320,Sequence length is 418\n",
      "240,250,Suboptimal foldings created.\n",
      "Energy dot plot created.\n",
      "10\t24\t360,370,380,390,330,340,RNA free energy files (version 2.3) at 20 degrees created.\n",
      "10,6\t260,8\t13\t2\t/home/jupyter/plant_microRNA_prediction/home/jupyter/plant_microRNA_prediction\n",
      "\n",
      "3400,410, \n",
      "End of Fill\n",
      "Save file created using nafold.\n",
      "6\t350,360,/home/jupyter/plant_microRNA_prediction/secondary_structure/mfold/AMWY020002401|668-1086|201-219|-/home/jupyter/plant_microRNA_prediction/secondary_structure/mfold/AMWY020001481|2932-3190|201-218|+\n",
      "270,20,30,40,50,60,\n",
      "12\t\tMinimum folding energy is -140.80 kcal/mol.\n",
      "Energy increment is 7.04 kcal/mol.\n",
      "370,380,280,13\t70,80,mfold version 3.6\n",
      "390,400,410,420, \n",
      "mfold version 3.6\n",
      "7\t8\t90,100,290,300,REUSE= NO\n",
      "End of Fill\n",
      "Save file created using nafold.\n",
      "H-num file created from plot file.\n",
      "16\tREUSE= NO\n",
      "2\t310,320,110,120,1\tSuboptimal foldings created.\n",
      "Energy dot plot created.\n",
      "19\tMinimum folding energy is -107.50 kcal/mol.\n",
      "Energy increment is 5.38 kcal/mol.\n",
      "1,2,3,4,Suboptimal foldings created.\n",
      "130,SEQ.pnt created.\n",
      "10\tH-num file created from plot file.\n",
      "330,340,140,150,SEQ.pnt created.\n",
      "Sequence length is 419\n",
      "25\tEnergy dot plot created.\n",
      "10\t11\t7\tSequence length is 259\n",
      "5,6,7,8,9,1,350,360,370,380,7\t14\tRNA free energy files (version 2.3) at 20 degrees created.\n",
      "160,9\t3\t10,11,12,13,14,15,16,RNA free energy files (version 2.3) at 20 degrees created.\n",
      "10,20,2,3,4,5,6,390,400,410, \n",
      "End of Fill\n",
      "10,20,30,4\t9\t1\t170,17,18,19,30,40,50,60,70,7,8,8\t13\tSave file created using nafold.\n",
      "Minimum folding energy is -127.20 kcal/mol.\n",
      "Energy increment is 6.36 kcal/mol.\n",
      "40,50,60,180,20,21,3\t80,90,100,70,80,17\t190,9,10,11,22,23,24,110,120,H-num file created from plot file.\n",
      "1\t90,100,2\t11\t \n",
      "25,26,27, \n",
      "200,130,12\t1,2,11\t110,120,826\t140,150,160,\n",
      "Structure plots generated.\n",
      "All done.\n",
      "1\t8\t\t3,4,5,6,7,8,210,130,140,170,180,4\t\n",
      "Structure plots generated.\n",
      "All done.\n",
      "9,10,11,12,2\t5\t220,9\t190,200,210,150,14\t4\t13, \n",
      "230,Suboptimal foldings created.\n",
      "Energy dot plot created.\n",
      "10\t160,170,220,230,240,18\t240,250, \n",
      "End of Fill\n",
      "Save file created using nafold.\n",
      "/home/jupyter/plant_microRNA_prediction\n",
      "/home/jupyter/plant_microRNA_prediction/secondary_structure/mfold/AMWY020002411|5665-6082|201-218|+20\t12\t180,\n",
      "250,260,Minimum folding energy is -89.10 kcal/mol.\n",
      "Energy increment is 4.46 kcal/mol.\n",
      "12\t13\t3\t/home/jupyter/plant_microRNA_prediction190,\n",
      "/home/jupyter/plant_microRNA_prediction/secondary_structure/mfold/AMWY020001481|2934-3190|201-220|+\n",
      "270,9\t27\t9\t200,mfold version 3.6\n",
      "H-num file created from plot file.\n",
      "1,280,5\tSuboptimal foldings created.\n",
      "Energy dot plot created.\n",
      "2\tmfold version 3.6\n",
      "REUSE= NO\n",
      "\n",
      "Structure plots generated.\n",
      "All done.\n",
      "6\t2,3,4,5,6,7,8,290,10\tSuboptimal foldings created.\n",
      "3\t\n",
      "Structure plots generated.\n",
      "15\t210,REUSE= NO\n",
      "All done.\n",
      "300,9,10,11,Energy dot plot created.\n",
      "10\tSEQ.pnt created.\n",
      "Sequence length is 418\n",
      "SEQ.pnt created.\n",
      "Sequence length is 257\n",
      "220,230,11\t12, \n",
      "310,320,19\tRNA free energy files (version 2.3) at 20 degrees created.\n",
      "RNA free energy files (version 2.3) at 20 degrees created.\n",
      "240,330,340,13\t10,20,30,\n",
      "/home/jupyter/plant_microRNA_prediction250,350,2\t/home/jupyter/plant_microRNA_prediction/secondary_structure/mfold/AMWY020001481|2934-3190|201-219|+\n",
      "/home/jupyter/plant_microRNA_prediction13\t1\t4\t\n",
      "40,50,60,10,20,30,40,/home/jupyter/plant_microRNA_prediction/secondary_structure/mfold/AMWY020002101|3109-3527|201-219|+260,360,370,\n",
      "28\tmfold version 3.6\n",
      "\n",
      "Structure plots generated.\n",
      "All done.\n",
      "\n",
      "Structure plots generated.\n",
      "All done.\n",
      "Suboptimal foldings created.\n",
      "Energy dot plot created.\n",
      "10\t270,70,80,90,6\t3\t50,60,70,380,390,400,410, \n",
      "End of Fill\n",
      "REUSE= NO\n",
      "mfold version 3.6\n",
      "7\t280,290,11\t100,110,Save file created using nafold.\n",
      "Minimum folding energy is -104.10 kcal/mol.\n",
      "Energy increment is 5.21 kcal/mol.\n",
      "REUSE= NO\n",
      "4\t80,90,100,110,120,300,SEQ.pnt created.\n",
      "Sequence length is 257\n",
      "11\t130,1\t16\t310,H-num file created from plot file.\n",
      "120,130,SEQ.pnt created.\n",
      "Sequence length is 419\n",
      "140,/home/jupyter/plant_microRNA_prediction/home/jupyter/plant_microRNA_prediction\n",
      "\n",
      "14\t/home/jupyter/plant_microRNA_prediction/secondary_structure/mfold/AMWY020001621|7045-7462|201-218|+320,330,2\t/home/jupyter/plant_microRNA_prediction/secondary_structure/mfold/AMWY020002441|1025-1442|201-218|+140,150,RNA free energy files (version 2.3) at 20 degrees created.\n",
      "\n",
      "3\t1,2,3,150,RNA free energy files (version 2.3) at 20 degrees created.\n",
      "\n",
      "1\t5\t160,170,180,4,5,6,7,8,340,350,160,170,29\t14\t10,20,30,40,mfold version 3.6\n",
      "10,20,30,40,50,60,70,11\t360,370,380,390,9,10,11,12,190,200,mfold version 3.6\n",
      "180,4\t80,90,100,50,60,70,7\t210,220,230,240,REUSE= NO\n",
      "8\t400,410, \n",
      "End of Fill\n",
      "Save file created using nafold.\n",
      "190,110,13,14,15,16,17,5\t12\tREUSE= NO\n",
      "80,90,100,250, \n",
      "End of Fill\n",
      "Save file created using nafold.\n",
      "Minimum folding energy is -95.50 kcal/mol.\n",
      "Energy increment is 4.78 kcal/mol.\n",
      "120,130,200,110,SEQ.pnt created.2\tSEQ.pnt created.\n",
      "12\t140,Minimum folding energy is -89.10 kcal/mol.\n",
      "Energy increment is 4.46 kcal/mol.\n",
      "17\t18,19,20,21,22,23,120,210,\n",
      "Sequence length is 418\n",
      "150,160,15\tSequence length is 418\n",
      "H-num file created from plot file.\n",
      "1,24,25,26, \n",
      "130,220,4\t3\tH-num file created from plot file.\n",
      "RNA free energy files (version 2.3) at 20 degrees created.\n",
      "170,180,2,3,4,2\t20\t6\tRNA free energy files (version 2.3) at 20 degrees created.\n",
      "140,230,12\t15\t1,2,3,4,190,200,5,6,7,8,10,20,30,40,150,160,240,10,20,30,40,50,60,8\t5,6,7,8,9,9,10,11,12, \n",
      "9\t5\t170,50,60,70,80,90,210,220,230,240,250, \n",
      "End of Fill10,11,70,80,250,6\t180,13\t100,110,\n",
      "Save file created using nafold.\n",
      "Minimum folding energy is -89.10 kcal/mol.\n",
      "Energy increment is 4.46 kcal/mol.\n",
      "260,270,90,100,110,12,13,14,15,3\t190,H-num file created from plot file.\n",
      "1\t16,17,18,19,20,280,290,120,120,130,140,200,16\t13\t1,2,3,4,4\tSuboptimal foldings created.\n",
      "Energy dot plot created.21,22,130,140,300,150,160,1\t5\t5,6,7,8,9,10,7\t3\t21\t150,170,23, \n",
      "\n",
      "10\t16\t310,320,210,11,12, \n",
      "Suboptimal foldings created.\n",
      "160,6\t\n",
      "Structure plots generated.\n",
      "All done.\n",
      "220,180,190,9\t330,340,Energy dot plot created.\n",
      "10\t170,230,240,350,360,370,380,200,1\t7\t180,210,220,250,390,190,2\t4\t260,230,240,400,410, \n",
      "End of Fill\n",
      "Save file created using nafold.\n",
      "200,14\t5\t11\t17\t/home/jupyter/plant_microRNA_predictionSuboptimal foldings created.\n",
      "Energy dot plot created.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "10\t\n",
      "/home/jupyter/plant_microRNA_prediction/secondary_structure/mfold/AMWY020002451|340-759|201-220|-\n",
      "210,Minimum folding energy is -139.20 kcal/mol.\n",
      "Energy increment is 6.96 kcal/mol.\n",
      "270,6\t250,22\t2\t8\t4\t220,230,17\tH-num file created from plot file.\n",
      "1,280,mfold version 3.6\n",
      "260,270,7\t2,3,4,\n",
      "Structure plots generated.\n",
      "All done.\n",
      "REUSE= NO\n",
      "240,280,290,290,300,250,5,6,7,11\t300,310,8\tSuboptimal foldings created.\n",
      "Energy dot plot created.\n",
      "2\tSEQ.pnt created.\n",
      "Sequence length is 420\n",
      "310,320,8, \n",
      "260,320,330,340,5\t10\t3\tRNA free energy files (version 2.3) at 20 degrees created.\n",
      "15\t270,280,12\t11\t3\t350,360,370,380,330,340,350,360,6\t10,20,30,18\t/home/jupyter/plant_microRNA_prediction290,7\t\n",
      "9\t23\t390,400,410, \n",
      "End of Fill\n",
      "370,380,390,/home/jupyter/plant_microRNA_prediction/secondary_structure/mfold/AMWY020001611|1642-2061|201-220|-40,50,300,5\t\n",
      "Suboptimal foldings created.\n",
      "18\t400,410, \n",
      "End of Fill\n",
      "Save file created using nafold.\n",
      "Save file created using nafold.\n",
      "Minimum folding energy is -126.10 kcal/mol.\n",
      "Energy increment is 6.30 kcal/mol.\n",
      "8\t310,320,Energy dot plot created.\n",
      "1\t60,70,80,Minimum folding energy is -132.60 kcal/mol.\n",
      "Energy increment is 6.63 kcal/mol.\n",
      "mfold version 3.6\n",
      "12\t330,90,100,110,3\t9\t11\t340,350,H-num file created from plot file.\n",
      "1,REUSE= NO\n",
      "120,130,2,3,4,5,6\t16\tH-num file created from plot file.\n",
      "140,360,370,12\t1\t4\t4\t6,7,8,1,2,3,4,380,390,400,SEQ.pnt created.\n",
      "Sequence length is 420\n",
      "150,160,19\t7\t\n",
      "Structure plots generated.\n",
      "All done.\n",
      "9,10,11,12,13,5,6,7,8,170,180,410, \n",
      "End of Fill\n",
      "Save file created using nafold.\n",
      "Minimum folding energy is -127.30 kcal/mol.\n",
      "8\t24\tRNA free energy files (version 2.3) at 20 degrees created.\n",
      "14,15,16,6\tEnergy increment is 6.37 kcal/mol.\n",
      "9,10,11,12,13,2\t10,20,30,40,50,19\t9\t190,17,18, \n",
      "13\t\n",
      "Structure plots generated.\n",
      "All done.\n",
      "14,15,16,17,18,60,70,200,H-num file created from plot file.\n",
      "12\t19,20,21,22,23,24,80,90,100,4\t2\t210,220,1,2,3,4,/home/jupyter/plant_microRNA_prediction25,26,17\t7\t110,1\t230,\n",
      "5,6,7,/home/jupyter/plant_microRNA_prediction/secondary_structure/mfold/AMWY020001631|9099-9518|201-220|+\n",
      "5\t27,28,29,30,8\t120,130,9\t5\t240,8,9,10,/home/jupyter/plant_microRNA_prediction\n",
      "31,32,33, \n",
      "mfold version 3.6\n",
      "/home/jupyter/plant_microRNA_prediction/secondary_structure/mfold/AMWY020001681|1001-1419|201-219|-\n",
      "11,12,13,14,140,150,250,25\t1\tREUSE= NO\n",
      "\n",
      "Structure plots generated.\n",
      "15,16,160,170,All done.\n",
      "260,1\t7\tmfold version 3.6\n",
      "3\tSEQ.pnt created.\n",
      "Sequence length is 420\n",
      "270,14\t \n",
      "180,Suboptimal foldings created.\n",
      "Energy dot plot created.\n",
      "10\tRNA free energy files (version 2.3) at 20 degrees created.\n",
      "REUSE= NO\n",
      "3\t280,190,200,13\t5\tSEQ.pnt created.10,20,30,40,9\t290,300,8\t6\t2\t210,6\t50,60,70,80,\n",
      "Sequence length is 419\n",
      "18\t310,320,\n",
      "Structure plots generated.\n",
      "/home/jupyter/plant_microRNA_predictionAll done.\n",
      "\n",
      "/home/jupyter/plant_microRNA_prediction/secondary_structure/mfold/AMWY020001731|1512-1929|201-218|-220,90,100,\n",
      "RNA free energy files (version 2.3) at 20 degrees created.\n",
      "330,10,20,30,230,340,110,20\t26\tmfold version 3.6\n",
      "40,50,60,120,240,250,350,360,370,380,390,8\tSuboptimal foldings created.\n",
      "Energy dot plot created.\n",
      "10\t20\t70,80,90,130,140,REUSE= NO\n",
      "/home/jupyter/plant_microRNA_prediction4\t4\t400,410, \n",
      "End of Fill\n",
      "\n",
      "260,270,150,/home/jupyter/plant_microRNA_prediction/secondary_structure/mfold/AMWY020001631|10043-10462|201-220|+\n",
      "100,110,SEQ.pnt created.\n",
      "Sequence length is 418\n",
      "280,290,15\tSave file created using nafold.\n",
      "Minimum folding energy is -125.80 kcal/mol.\n",
      "Energy increment is 6.29 kcal/mol.\n",
      "11\t14\t\n",
      "Structure plots generated.\n",
      "160,7\t6\tAll done.\n",
      "300,120,130,RNA free energy files (version 2.3) at 20 degrees created.\n",
      "9\tmfold version 3.6\n",
      "170,180,19\t7\t3\t140,150,310,320,10,20,30,40,50,H-num file created from plot file.\n",
      "REUSE= NO\n",
      "190,Suboptimal foldings created.\n",
      "330,340,160,60,70,1,2,3,4,SEQ.pnt created.\n",
      "200,210,21\t350,360,370,Energy dot plot created.\n",
      "10\t170,180,27\t80,90,2\t5,6,7,8,9,Sequence length is 420\n",
      "220,/home/jupyter/plant_microRNA_prediction9\t\n",
      "380,390,400,410, \n",
      "End of Fill\n",
      "100,110,/home/jupyter/plant_microRNA_prediction/secondary_structure/mfold/AMWY020002481|3479-3899|201-221|-11\tRNA free energy files (version 2.3) at 20 degrees created.\n",
      "190,5\t\n",
      "10,11,12, \n",
      "230,5\tSave file created using nafold.\n",
      "Minimum folding energy is -124.20 kcal/mol.\n",
      "Energy increment is 6.21 kcal/mol.\n",
      "120,12\t10,20,30,40,50,60,200,210,240,16\t8\t15\t130,7\t4\t\n",
      "mfold version 3.6\n",
      "220,70,80,90,100,Structure plots generated.\n",
      "All done.\n",
      "250,8\t140,150,H-num file created from plot file.\n",
      "1,2,REUSE= NO\n",
      "1\t110,120,260,230,240,160,130,140,3,4,5,6,7,250,170,270,28\tSEQ.pnt created.\n",
      "Sequence length is 421\n",
      "11\t150,3\t8,9,10,11,12,\n",
      "Structure plots generated.\n",
      "All done.\n",
      "6\t260,180,22\t280,290,Suboptimal foldings created.\n",
      "RNA free energy files (version 2.3) at 20 degrees created.\n",
      "13,14,15,16,/home/jupyter/plant_microRNA_prediction160,170,\n",
      "270,280,190,/home/jupyter/plant_microRNA_prediction/secondary_structure/mfold/AMWY020002001|1249-1668|201-220|-\n",
      "13\t300,310,10,20,30,40,180,Energy dot plot created.\n",
      "10\t12\t9\t290,17,18,6\t200,320,330,16\t17\tmfold version 3.6\n",
      "340,300,310,190,50,60,70,80,5\t8\t19,20,21,22,/home/jupyter/plant_microRNA_prediction9\t\n",
      "20\t/home/jupyter/plant_microRNA_prediction/secondary_structure/mfold/AMWY020002481|3481-3899|201-219|+\n",
      "REUSE= NO\n",
      "350,210,320,330,90,100,200,210,23,24, \n",
      "360,370,380,340,220,220,110,mfold version 3.6\n",
      "12\t350,360,230,240,390,400,410, \n",
      "End of Fill\n",
      "7\tREUSE= NO\n",
      "SEQ.pnt created.\n",
      "230,240,120,4\tSave file created using nafold.\n",
      "Minimum folding energy is -141.10 kcal/mol.\n",
      "Energy increment is 7.05 kcal/mol.\n",
      "23\t11\tSEQ.pnt created.\n",
      "Sequence length is 419\n",
      "250,130,140,Sequence length is 420\n",
      "370,380,390,2\t14\t\n",
      "Structure plots generated.\n",
      "All done.\n",
      "13\t250,RNA free energy files (version 2.3) at 20 degrees created.\n",
      "260,150,RNA free energy files (version 2.3) at 20 degrees created.\n",
      "400,410, \n",
      "End of Fill\n",
      "Save file created using nafold.\n",
      "260,270,\n",
      "Structure plots generated.\n",
      "All done.\n",
      "7\t270,H-num file created from plot file.\n",
      "160,10,20,30,40,50,10,20,30,40,Minimum folding energy is -164.50 kcal/mol.\n",
      "Energy increment is 8.22 kcal/mol.\n",
      "21\t280,17\t18\t1,2,3,4,5,280,290,170,6\t60,70,50,60,70,80,9\t290,300,300,6,H-num file created from plot file.\n",
      "180,90,100,80,90,310,320,/home/jupyter/plant_microRNA_prediction\n",
      "13\t/home/jupyter/plant_microRNA_prediction/secondary_structure/mfold/AMWY020002591|394-813|201-220|-\n",
      "7,8, \n",
      "1,2,3,100,110,310,320,110,120,190,330,340,/home/jupyter/plant_microRNA_prediction8\t120,\n",
      "330,340,4,5,6,7,200,210,mfold version 3.6\n",
      "130,140,/home/jupyter/plant_microRNA_prediction/secondary_structure/mfold/AMWY020001851|225-641|201-217|-350,360,370,\n",
      "14\t12\t24\t130,220,Suboptimal foldings created.\n",
      "Energy dot plot created.\n",
      "8,9,10,11,12,150,15\t5\t3\tREUSE= NO\n",
      "380,390,400,410, \n",
      "End of Fill\n",
      "Save file created using nafold.\n",
      "350,360,370,13,14,15,16,10\t230,140,160,mfold version 3.6\n",
      "19\tMinimum folding energy is -140.20 kcal/mol.\n",
      "Energy increment is 7.01 kcal/mol.\n",
      "380,390,400,410, \n",
      "End of Fill\n",
      "Suboptimal foldings created.\n",
      "Energy dot plot created.\n",
      "22\tSEQ.pnt created.\n",
      "7\t18\t240,8\t150,160,170,180,REUSE= NO\n",
      "17,18,19,\n",
      "Structure plots generated.\n",
      "All done.\n",
      "Save file created using nafold.\n",
      "Minimum folding energy is -106.30 kcal/mol.\n",
      "Energy increment is 5.32 kcal/mol.\n",
      "1\t250,190,Sequence length is 420\n",
      "RNA free energy files (version 2.3) at 20 degrees created.\n",
      "20,21, \n",
      "170,H-num file created from plot file.\n",
      "SEQ.pnt created.\n",
      "10,20,30,40,50,260,200,H-num file created from plot file.\n",
      "9\t180,14\t60,70,1,2,3,4,Sequence length is 417\n",
      "210,1,2,3,4,5,6,270,280,1\t80,90,190,5,6,7,8,290,7,8,9,10,RNA free energy files (version 2.3) at 20 degrees created.\n",
      "10,220,/home/jupyter/plant_microRNA_prediction4\t\n",
      "/home/jupyter/plant_microRNA_prediction/secondary_structure/mfold/AMWY020002201|539-957|201-219|-9, \n",
      "\n",
      "2\t11\t230,240,15\t100,110,120,300,310,6\t16\t\n",
      "Structure plots generated.\n",
      "20,30,40,50,200,210,All done.\n",
      "11,12,13, \n",
      "1\tmfold version 3.6\n",
      "19\t23\t320,130,140,220,250,60,70,80,8\tREUSE= NO\n",
      "2\t330,340,230,260,270,90,100,150,\n",
      "Structure plots generated.\n",
      "All done.\n",
      "350,360,370,280,110,240,160,170,380,Suboptimal foldings created.\n",
      "Suboptimal foldings created.\n",
      "Energy dot plot created.\n",
      "SEQ.pnt created.\n",
      "Sequence length is 419\n",
      "15\t/home/jupyter/plant_microRNA_prediction250,180,390,400,410,420, \n",
      "End of Fill\n",
      "290,Suboptimal foldings created.\n",
      "Energy dot plot created.\n",
      "Energy dot plot created.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "10\t\n",
      "2\tRNA free energy files (version 2.3) at 20 degrees created.\n",
      "120,130,1\t/home/jupyter/plant_microRNA_prediction/secondary_structure/mfold/AMWY020002411|8269-8690|201-222|+\n",
      "Save file created using nafold.\n",
      "Minimum folding energy is -146.80 kcal/mol.\n",
      "Energy increment is 7.34 kcal/mol.\n",
      "260,17\t190,200,10\t300,310,320,10,20,30,40,140,/home/jupyter/plant_microRNA_prediction3\t\n",
      "20\t7\t16\t12\tmfold version 3.6\n",
      "5\t/home/jupyter/plant_microRNA_prediction/secondary_structure/mfold/AMWY020002691|1434-1854|201-221|+50,60,270,280,210,\n",
      "24\t1\t220,H-num file created from plot file.\n",
      "330,340,350,360,9\t150,160,70,80,REUSE= NO\n",
      "290,3\t300,310,90,100,1,2,3,4,5,230,240,mfold version 3.6\n",
      "170,370,380,390,400,410, \n",
      "End of Fill\n",
      "Save file created using nafold.\n",
      "SEQ.pnt created.\n",
      "Sequence length is 422\n",
      "Minimum folding energy is -133.00 kcal/mol.\n",
      "Energy increment is 6.65 kcal/mol.\n",
      "REUSE= NO\n",
      "180,6,7,8,9,10,320,330,110,120,250,RNA free energy files (version 2.3) at 20 degrees created.\n",
      "190,16\t11,12,13,14,11\t340,350,2\t3\t10,20,30,40,SEQ.pnt created.\n",
      "Sequence length is 421\n",
      "130,140,260,18\t11\t15, \n",
      "H-num file created from plot file.\n",
      "1,360,370,380,RNA free energy files (version 2.3) at 20 degrees created.200,8\t50,60,70,270,280,1\t13\t150,160,6\t2,3,4,5,4\t\n",
      "10,390,400,410, \n",
      "End of Fill\n",
      "Save file created using nafold.\n",
      "\n",
      "Structure plots generated.\n",
      "All done.\n",
      "21\t25\t210,20\t80,90,100,290,300,310,170,180,20,30,40,50,6,7,8,9,10,4\tMinimum folding energy is -122.00 kcal/mol.\n",
      "Energy increment is 6.10 kcal/mol.\n",
      "320,330,220,110,120,190,11,12,13,60,70,230,340,350,80,90,130,200,17\t \n",
      "360,370,380,240,H-num file created from plot file.\n",
      "100,140,150,3\t210,Suboptimal foldings created.\n",
      "Energy dot plot created.\n",
      "12\t1,2,/home/jupyter/plant_microRNA_prediction390,400,410, \n",
      "End of Fill\n",
      "4\t\n",
      "110,120,160,10\t/home/jupyter/plant_microRNA_prediction/secondary_structure/mfold/AMWY020002691|643-1061|201-219|-220,230,1\t14\t12\t250,\n",
      "9\t3,4,5,6,Save file created using nafold.\n",
      "Minimum folding energy is -119.50 kcal/mol.\n",
      "Energy increment is 5.97 kcal/mol.\n",
      "130,2\t26\t240,5\t260,170,180,7\t7,8,9,140,150,mfold version 3.6\n",
      "22\t21\t270,250,190,Suboptimal foldings created.\n",
      "10,11,12,5\t160,H-num file created from plot file.\n",
      "280,200,260,REUSE= NO\n",
      "Energy dot plot created.\n",
      "10\t170,13,14,210,290,270,180,15,16, \n",
      "18\tSEQ.pnt created.\n",
      "Sequence length is 419\n",
      "220,4\t1,2,3,4,300,310,13\t190,280,290,13\tRNA free energy files (version 2.3) at 20 degrees created.\n",
      "\n",
      "Structure plots generated.\n",
      "All done.\n",
      "5\t230,11\t320,2\t200,5,6,7,8,9,10,11,12,13,14,15,300,310,10,20,30,40,3\t15\t330,320,210,27\t16,17,18,19,20,21,240,50,60,70,6\t220,330,340,340,8\t23\t22\t22, \n",
      "11\t80,90,250,6\t350,360,370,350,360,230,/home/jupyter/plant_microRNA_prediction260,\n",
      "100,110,/home/jupyter/plant_microRNA_prediction/secondary_structure/mfold/AMWY020001951|29-448|201-220|+19\t\n",
      "Suboptimal foldings created.\n",
      "Energy dot plot created.\n",
      "370,380,390,380,390,400,410, \n",
      "End of Fill\n",
      "240,120,130,5\t270,280,10\tmfold version 3.6\n",
      "1\t140,14\tSave file created using nafold.\n",
      "Minimum folding energy is -140.70 kcal/mol.\n",
      "Energy increment is 7.04 kcal/mol.\n",
      "400,410, \n",
      "End of Fill\n",
      "Save file created using nafold.\n",
      "Minimum folding energy is -117.40 kcal/mol.\n",
      "REUSE= NO\n",
      "6\t250,260,12\t290,300,150,160,Energy increment is 5.87 kcal/mol.\n",
      "270,16\tH-num file created from plot file.\n",
      "310,320,4\t24\t170,3\t1,2,3,4,9\t280,2\t330,SEQ.pnt created.\n",
      "Sequence length is 420\n",
      "H-num file created from plot file.\n",
      "180,7\t290,300,12\t5,6,7,8,23\t340,350,1,2,3,RNA free energy files (version 2.3) at 20 degrees created.\n",
      "190,9,10,11,12,13,1\t7\tSuboptimal foldings created.\n",
      "Energy dot plot created.\n",
      "310,4,5,6,7,10,20,30,40,50,360,370,380,390,200,6\t14,15, \n",
      "8,9,10,320,330,340,11\t10\t210,400,410,420, \n",
      "End of Fill\n",
      "Save file created using nafold.\n",
      "Minimum folding energy is -132.10 kcal/mol.\n",
      "Energy increment is 6.61 kcal/mol.\n",
      "60,70,80,90,2\t13\t7\t350,15\t11,12,13,14,220,5\t100,110,4\t25\t360,370,17\t15,16,17,18,19,20,H-num file created from plot file.\n",
      "\n",
      "Structure plots generated.\n",
      "All done.\n",
      "3\t230,120,130,21,22,23,24,25,1,2,3,380,390,400,410,420,8\t2\t240,250,140,150,26,27,28, \n",
      "End of Fill\n",
      "Save file created using nafold.\n",
      "Minimum folding energy is -138.90 kcal/mol.\n",
      "Energy increment is 6.95 kcal/mol.\n",
      "13\t160,4,5,6,7,260,29,20\t8\tSuboptimal foldings created.\n",
      "11\t12\t170,180,270,8,9,10,11,12,13,30,31, \n",
      "7\tEnergy dot plot created.\n",
      "10\t/home/jupyter/plant_microRNA_prediction\n",
      "3\tH-num file created from plot file.14\t/home/jupyter/plant_microRNA_prediction/secondary_structure/mfold/AMWY020001661|153-453|84-101|-14,15,16, \n",
      "190,\n",
      "280,290,18\t16\t8\t6\t\n",
      "1,2,3,4,4\t300,310,200,5\t5,6,7,8,9,26\tmfold version 3.6\n",
      "320,3\t210,1\t10,11,12,9\t330,340,350,220,230,REUSE= NO\n",
      "13,14,15,16,17,21\t360,370,240,\n",
      "Structure plots generated.\n",
      "All done.\n",
      "8\t18,19,20,380,390,400,13\tSEQ.pnt created.\n",
      "Sequence length is 301\n",
      "11\t12\t4\t250,260,21,22,23,24,25,410, \n",
      "End of Fill\n",
      "Save file created using nafold.\n",
      "15\tSuboptimal foldings created.\n",
      "26,27,270,19\tRNA free energy files (version 2.3) at 20 degrees created.\n",
      "Minimum folding energy is -158.20 kcal/mol.\n",
      "Energy increment is 7.91 kcal/mol.\n",
      "9\tEnergy dot plot created.\n",
      "10\t17\t5\t28,29,30,7\t10,20,30,280,290,/home/jupyter/plant_microRNA_prediction\n",
      "/home/jupyter/plant_microRNA_prediction/secondary_structure/mfold/AMWY020001641|1-306|201-221|-31, \n",
      "6\t\n",
      "Structure plots generated.\n",
      "All done.\n",
      "2\t40,50,60,\n",
      "300,310,2\tH-num file created from plot file.\n",
      "4\t70,80,Suboptimal foldings created.\n",
      "Energy dot plot created.\n",
      "320,330,mfold version 3.6\n",
      "22\t90,100,1,2,3,10\tREUSE= NO\n",
      "340,350,12\t4,5,6,7,8,9\t14\t13\t110,120,5\t360,370,380,390,1\t/home/jupyter/plant_microRNA_prediction\n",
      "Structure plots generated.\n",
      "All done.\n",
      "\n",
      "/home/jupyter/plant_microRNA_prediction/secondary_structure/mfold/AMWY020001911|2660-3081|201-222|+9,10,\n",
      "SEQ.pnt created.\n",
      "Sequence length is 306\n",
      "130,140,400,410, \n",
      "End of Fill\n",
      "18\t6\t1\tRNA free energy files (version 2.3) at 20 degrees created.\n",
      "150,Save file created using nafold.\n",
      "Minimum folding energy is -146.90 kcal/mol.\n",
      "Energy increment is 7.34 kcal/mol.\n",
      "11,12,13,14,8\tmfold version 3.6\n",
      "11\t7\t10,20,30,40,3\t5\tREUSE= NO\n",
      "3\t15, \n",
      "160,50,60,/home/jupyter/plant_microRNA_predictionSEQ.pnt created.\n",
      "Sequence length is 422\n",
      "H-num file created from plot file.\n",
      "\n",
      "/home/jupyter/plant_microRNA_prediction/secondary_structure/mfold/AMWY020002461|1-329|185-202|+170,180,\n",
      "11\t70,23\t1,2,3,4,5,6,\n",
      "Structure plots generated.\n",
      "All done.\n",
      "190,RNA free energy files (version 2.3) at 20 degrees created.\n",
      "80,90,100,mfold version 3.6\n",
      "7,8,9,10,11,13\t6\t200,210,14\t10,20,30,40,50,15\t110,REUSE= NO\n",
      "2\t12,13,14,220,230,240,12\t60,70,80,20\t19\t120,130,SEQ.pnt created.\n",
      "Suboptimal foldings created.\n",
      "Energy dot plot created.\n",
      "10\t15,16,17,18,19,250,260,9\t7\t90,100,110,140,Sequence length is 329\n",
      "Suboptimal foldings created.\n",
      "20,21,/home/jupyter/plant_microRNA_prediction\n",
      "270,280,290,300, \n",
      "End of Fill\n",
      "Save file created using nafold.\n",
      "4\t4\t/home/jupyter/plant_microRNA_prediction/secondary_structure/mfold/AMWY020001631|10045-10464|201-220|+\n",
      "8\t150,RNA free energy files (version 2.3) at 20 degrees created.\n",
      "6\tEnergy dot plot created.\n",
      "10\t120,130, \n",
      "12\tmfold version 3.6\n",
      "Minimum folding energy is -85.70 kcal/mol.\n",
      "Energy increment is 4.29 kcal/mol.\n",
      "10,20,30,160,140,150,H-num file created from plot file.\n",
      "40,50,60,70,7\t14\tREUSE= NO\n",
      "24\t16\t160,170,180,15\t1,2,80,90,3\t170,SEQ.pnt created.\n",
      "Sequence length is 420\n",
      "190,100,110,3,4,5,180,13\t1\tRNA free energy files (version 2.3) at 20 degrees created.\n",
      "200,210,21\t11\t120,8\t190,5\t6,7,8, \n",
      "\n",
      "Structure plots generated.\n",
      "All done.\n",
      "10,20,30,220,230,7\t5\t11\t40,50,60,200,130,140,9\t13\t210,150,160,70,80,90,240,250,260,220,100,170,180,15\t25\t270,280,290,Suboptimal foldings created.\n",
      "Energy dot plot created.\n",
      "1\t8\t1\t230,/home/jupyter/plant_microRNA_predictionSuboptimal foldings created.\n",
      "Energy dot plot created.\n",
      "10\t110,120,130,16\t190,\n",
      "300, \n",
      "End of Fill\n",
      "Save file created using nafold.\n",
      "Minimum folding energy is -68.20 kcal/mol.\n",
      "Energy increment is 3.41 kcal/mol.\n",
      "/home/jupyter/plant_microRNA_prediction/secondary_structure/mfold/AMWY020001621|7266-7689|201-224|+4\t\n",
      "140,14\t200,210,240,250,20\t150,mfold version 3.6\n",
      "12\t22\tH-num file created from plot file.\n",
      "1,220,230,260,270,9\t6\t160,REUSE= NO\n",
      "2,3,4,5,6,\n",
      "Structure plots generated.\n",
      "All done.\n",
      "240,250,260,280,12\t8\t170,180,270,280,290,7,8,9,10,11,6\t14\tSEQ.pnt created.\n",
      "Sequence length is 424\n",
      "190,290,300,RNA free energy files (version 2.3) at 20 degrees created.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "26\t12, \n",
      "1\t17\t300,310,320, \n",
      "End of Fill\n",
      "200,210,2\t9\t310,320,10,20,30,40,11\t5\tSave file created using nafold.\n",
      "Minimum folding energy is -77.40 kcal/mol.\n",
      "Energy increment is 3.87 kcal/mol.\n",
      "2\t220,330,340,350,/home/jupyter/plant_microRNA_prediction\n",
      "50,60,/home/jupyter/plant_microRNA_prediction/secondary_structure/mfold/AMWY020002441|1034-1450|201-217|-15\t\n",
      "H-num file created from plot file.\n",
      "\n",
      "Structure plots generated.\n",
      "All done.\n",
      "230,240,70,80,21\t23\t13\t360,370,380,Suboptimal foldings created.\n",
      "Energy dot plot created.\n",
      "1,2,3,4,7\t90,100,mfold version 3.6\n",
      "390,400,410,420, \n",
      "End of Fill\n",
      "5,6,7,8,9,250,9\t10\t15\t13\t7\t110,120,260,270,10,11,12,Save file created using nafold.\n",
      "Minimum folding energy is -145.90 kcal/mol.\n",
      "Energy increment is 7.29 kcal/mol.\n",
      "27\tREUSE= NO\n",
      "\n",
      "Structure plots generated.\n",
      "All done.\n",
      "130,140,280,3\t13,14,15,16, \n",
      "SEQ.pnt created.\n",
      "Sequence length is 417\n",
      "/home/jupyter/plant_microRNA_prediction\n",
      "150,160,2\t/home/jupyter/plant_microRNA_prediction/secondary_structure/mfold/AMWY020002001|10002-10421|201-220|+\n",
      "290,300,18\t6\tH-num file created from plot file.\n",
      "12\t170,180,RNA free energy files (version 2.3) at 20 degrees created.\n",
      "310,320,1,3\tmfold version 3.6\n",
      "16\t190,200,330,340,2,3,4,5,10,20,30,40,REUSE= NO\n",
      "8\t24\t11\t50,60,70,6,7,8,9,210,/home/jupyter/plant_microRNA_prediction2\t\n",
      "14\t350,360,370,/home/jupyter/plant_microRNA_prediction/secondary_structure/mfold/AMWY020001731|3859-4278|201-220|+\n",
      "Structure plots generated.\n",
      "All done.\n",
      "\n",
      "16\t80,90,220,SEQ.pnt created.\n",
      "Sequence length is 420\n",
      "380,390,400,410, \n",
      "End of Fill\n",
      "Suboptimal foldings created.\n",
      "Energy dot plot created.\n",
      "10,11,12,13,14,8\t230,10\tSave file created using nafold.\n",
      "Minimum folding energy is -140.10 kcal/mol.\n",
      "100,110,RNA free energy files (version 2.3) at 20 degrees created.\n",
      "4\t15,16,17,18,28\t14\t240,mfold version 3.6\n",
      "Energy increment is 7.00 kcal/mol.\n",
      "120,130,10,20,30,7\t19\t19,20, \n",
      "250,260,3\t13\tREUSE= NO\n",
      "SEQ.pnt created.\n",
      "140,H-num file created from plot file.\n",
      "40,50,60,270,Sequence length is 420\n",
      "150,160,4\t/home/jupyter/plant_microRNA_prediction1,2,3,\n",
      "70,80,12\t/home/jupyter/plant_microRNA_prediction/secondary_structure/mfold/AMWY020002411|462-880|201-219|-280,\n",
      "RNA free energy files (version 2.3) at 20 degrees created.\n",
      "1\t4,5,15\t90,100,3\t290,300,170,110,120,2\t9\t10,20,30,40,50,6,7,15\t180,17\t310,320,mfold version 3.6\n",
      "130,11\t60,70,80,8,9,10, \n",
      "330,340,350,9\t5\t29\t190,200,REUSE= NO\n",
      "140,1\t8\t90,100,360,370,380,210,4\tSEQ.pnt created.\n",
      "Sequence length is 419\n",
      "150,160,110,120,390,400,410,420, \n",
      "End of Fill\n",
      "RNA free energy files (version 2.3) at 20 degrees created.\n",
      "220,14\t170,Suboptimal foldings created.\n",
      "Energy dot plot created.\n",
      "130,2\t5\t1\tSave file created using nafold.\n",
      "Minimum folding energy is -101.90 kcal/mol.\n",
      "Energy increment is 5.09 kcal/mol.\n",
      "10,20,30,40,230,\n",
      "Structure plots generated.\n",
      "4\tAll done.\n",
      "10\t140,180,190,3\t16\t240,Suboptimal foldings created.\n",
      "Energy dot plot created.\n",
      "50,60,70,150,1\t200,250,10\t80,90,160,18\t12\tH-num file created from plot file.\n",
      "6\t\n",
      "Structure plots generated.\n",
      "All done.\n",
      "210,260,100,170,9\t1,2,3,270,110,120,130,2\t20\t/home/jupyter/plant_microRNA_prediction220,180,\n",
      "4,5,6,140,280,5\t/home/jupyter/plant_microRNA_prediction/secondary_structure/mfold/AMWY020002571|634-1052|201-219|-190,\n",
      "230,15\t7,8,9,150,290,/home/jupyter/plant_microRNA_prediction10,11,12,\n",
      "240,/home/jupyter/plant_microRNA_prediction/secondary_structure/mfold/AMWY020002321|5520-5938|201-219|+11\t\n",
      "2\t300,5\t160,6\t200,13,14,3\tmfold version 3.6\n",
      "4\t250,170,310,17\tmfold version 3.6\n",
      "210,260,180,320,15,7\t19\tREUSE= NO\n",
      "2\tREUSE= NO\n",
      "1\t270,220,330,340,230,16,17,18,190,13\t280,290,\n",
      "Structure plots generated.\n",
      "350,360,All done.\n",
      "21\tSEQ.pnt created.\n",
      "Sequence length is 419\n",
      "300,240,19,20, \n",
      "200,210,SEQ.pnt created.\n",
      "Sequence length is 419\n",
      "6\t370,380,390,220,30\t310,320,RNA free energy files (version 2.3) at 20 degrees created.\n",
      "250,16\t3\tRNA free energy files (version 2.3) at 20 degrees created.\n",
      "230,240,330,340,7\t12\t10,20,30,40,50,400,410, \n",
      "End of Fill\n",
      "Save file created using nafold.\n",
      "260,270,10,20,30,4\t6\t350,360,370,380,Minimum folding energy is -170.50 kcal/mol.\n",
      "Energy increment is 8.53 kcal/mol.\n",
      "/home/jupyter/plant_microRNA_prediction\n",
      "60,70,80,280,290,250,/home/jupyter/plant_microRNA_prediction/secondary_structure/mfold/AMWY020002481|3480-3898|201-219|-40,50,60,18\t1\t14\t390,400,410, \n",
      "End of Fill\n",
      "\n",
      "5\t8\t3\t300,70,80,90,90,100,110,2\tH-num file created from plot file.\n",
      "260,270,Save file created using nafold.\n",
      "Minimum folding energy is -106.80 kcal/mol.\n",
      "Energy increment is 5.34 kcal/mol.\n",
      "310,320,100,22\t120,130,mfold version 3.6\n",
      "1,2,3,4,5,280,290,300,110,120,140,150,330,REUSE= NO\n",
      "310,H-num file created from plot file.\n",
      "6,7,8,9,10,7\t31\tSuboptimal foldings created.\n",
      "Energy dot plot created.\n",
      "130,140,160,170,320,330,4\t17\tSEQ.pnt created.\n",
      "Sequence length is 419\n",
      "11,12,13,14,15,340,350,13\t8\t1,2,180,340,350,RNA free energy files (version 2.3) at 20 degrees created.\n",
      "16,17,18,10\t150,7\t360,370,380,5\t3,4,5,6,19,20,21,10,20,30,190,360,370,380,160,170,6\t19\t\n",
      "Structure plots generated.\n",
      "All done.\n",
      "390,400,7,8, \n",
      "22,23,24,20\t180,200,40,50,60,3\t390,400,410, \n",
      "End of Fill\n",
      "Save file created using nafold.\n",
      "15\t4\t410, \n",
      "End of Fill\n",
      "Save file created using nafold.\n",
      "210,25,26,27,190,Minimum folding energy is -102.80 kcal/mol.\n",
      "Energy increment is 5.14 kcal/mol.\n",
      "70,80,90,200,100,Minimum folding energy is -100.00 kcal/mol.\n",
      "Energy increment is 5.00 kcal/mol.\n",
      "28,29,220,2\tH-num file created from plot file.\n",
      "32\tH-num file created from plot file.\n",
      "110,230,240,30, \n",
      "/home/jupyter/plant_microRNA_prediction\n",
      "18\t5\t/home/jupyter/plant_microRNA_prediction/secondary_structure/mfold/AMWY020001671|2375-2791|201-217|+8\t210,220,14\t\n",
      "1,120,130,9\tSuboptimal foldings created.\n",
      "Energy dot plot created.\n",
      "1\t11\t250,8\t1,2,3,230,2,3,4,5,140,260,270,280,7\t240,mfold version 3.6\n",
      "1\t4,5,6,7,8,9,10,6\t5\t150,160,6,7,8,9,10,16\t290,300,4\t250,260,21\t11,12,13,14,15,16,REUSE= NO\n",
      "11,12,13,14,15,170,310,320,SEQ.pnt created.\n",
      "270,3\t16,17,18,180,17,18,19,20,330,340,350,Sequence length is 417\n",
      "280,360,370,33\t190,200,19,20, \n",
      "21,22,23,\n",
      "Structure plots generated.\n",
      "All done.\n",
      "6\t290,300,2\t15\tRNA free energy files (version 2.3) at 20 degrees created.\n",
      "19\t9\t12\t210,380,390,400,410, \n",
      "End of Fill\n",
      "9\t10,20,30,40,24,25, \n",
      "310,220,Save file created using nafold.\n",
      "Minimum folding energy is -124.20 kcal/mol.\n",
      "Energy increment is 6.21 kcal/mol.\n",
      "20\t7\t8\t6\t320,330,340,230,1\t50,60,70,5\tH-num file created from plot file.\n",
      "22\t350,360,240,80,90,100,/home/jupyter/plant_microRNA_prediction1,2,\n",
      "Suboptimal foldings created.\n",
      "/home/jupyter/plant_microRNA_prediction/secondary_structure/mfold/AMWY020002001|13296-13714|201-219|+370,380,390,400,\n",
      "250,110,120,3,4,Energy dot plot created.\n",
      "10\t4\t410, \n",
      "End of Fill\n",
      "260,7\tSuboptimal foldings created.\n",
      "Energy dot plot created.\n",
      "Save file created using nafold.\n",
      "Minimum folding energy is -128.40 kcal/mol.\n",
      "Energy increment is 6.42 kcal/mol.\n",
      "5,6,7,8,9,3\t270,mfold version 3.6\n",
      "130,140,\n",
      "Structure plots generated.\n",
      "All done.\n",
      "133\t10\t16\t10,11,12,\n",
      "Structure plots generated.\n",
      "All done.\n",
      "150,280,290,REUSE= NO\n",
      "1\t8\t\tH-num file created from plot file.\n",
      "13,14,15, \n",
      "300,SEQ.pnt created.\n",
      "160,9\t1,2,3,21\t2\t310,320,Sequence length is 419\n",
      "7\t170,23\t4,5,6,7,8,6\tSuboptimal foldings created.\n",
      "RNA free energy files (version 2.3) at 20 degrees created.\n",
      "180,330,340,/home/jupyter/plant_microRNA_prediction9,10,\n",
      "/home/jupyter/plant_microRNA_prediction/secondary_structure/mfold/AMWY020002221|924-1340|201-217|-5\t\n",
      "Energy dot plot created.\n",
      "10\t8\t/home/jupyter/plant_microRNA_prediction\n",
      "190,10,20,30,40,50,11\t350,360,/home/jupyter/plant_microRNA_prediction/secondary_structure/mfold/AMWY020001681|1238-1655|201-218|+11,12,13,\n",
      "200,60,70,370,380,390,400,mfold version 3.6\n",
      " \n",
      "mfold version 3.6\n",
      "210,4\t410, \n",
      "End of Fill\n",
      "Save file created using nafold.\n",
      "80,90,100,20\t11\tREUSE= NO\n",
      "4\t17\tMinimum folding energy is -146.80 kcal/mol.\n",
      "Energy increment is 7.34 kcal/mol.\n",
      "110,Suboptimal foldings created.\n",
      "14\tSEQ.pnt created.\n",
      "Sequence length is 417\n",
      "\n",
      "Structure plots generated.\n",
      "All done.\n",
      "REUSE= NO\n",
      "9\t3\t220,22\tEnergy dot plot created.\n",
      "10\t120,130,230,RNA free energy files (version 2.3) at 20 degrees created.\n",
      "24\t7\t8\tH-num file created from plot file.\n",
      "SEQ.pnt created.\n",
      "Sequence length is 418\n",
      "140,150,240,10,20,30,40,9\tRNA free energy files (version 2.3) at 20 degrees created.\n",
      "1,2,3,4,5,6,12\t160,11\tSuboptimal foldings created.\n",
      "Energy dot plot created.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "10\t50,60,70,6\t250,7,8,9,10,20,30,40,50,170,80,260,5\t12\t10,11,12,13,60,70,80,/home/jupyter/plant_microRNA_prediction\n",
      "5\t21\t90,100,18\t180,15\t270,14,15, \n",
      "/home/jupyter/plant_microRNA_prediction/secondary_structure/mfold/AMWY020001621|6865-7284|201-220|-90,100,\n",
      "190,200,\n",
      "Structure plots generated.\n",
      "All done.\n",
      "4\t110,120,11\t280,290,110,120,23\t130,mfold version 3.6\n",
      "300,310,9\t8\t130,25\t210,140,\n",
      "Structure plots generated.\n",
      "All done.\n",
      "140,150,REUSE= NO\n",
      "220,320,330,13\t12\t160,7\t150,160,SEQ.pnt created.\n",
      "11\t340,350,230,/home/jupyter/plant_microRNA_prediction6\t170,13\t\n",
      "/home/jupyter/plant_microRNA_prediction/secondary_structure/mfold/AMWY020002411|8615-9032|201-218|+Sequence length is 420\n",
      "RNA free energy files (version 2.3) at 20 degrees created.\n",
      "170,360,370,380,390,240,\n",
      "16\t2\t180,190,180,10,20,19\t400,410, \n",
      "End of Fill\n",
      "Save file created using nafold.\n",
      "Suboptimal foldings created.\n",
      "Energy dot plot created.\n",
      "10\t250,5\t6\tmfold version 3.6\n",
      "/home/jupyter/plant_microRNA_prediction200,\n",
      "30,40,50,60,260,270,12\t\n",
      "Structure plots generated.\n",
      "All done.\n",
      "Minimum folding energy is -112.60 kcal/mol.\n",
      "Energy increment is 5.63 kcal/mol.\n",
      "/home/jupyter/plant_microRNA_prediction/secondary_structure/mfold/AMWY020002741|1953-2370|201-218|+\n",
      "24\t190,200,26\t70,80,90,210,REUSE= NO\n",
      "280,9\tH-num file created from plot file.\n",
      "210,100,220,SEQ.pnt created.\n",
      "mfold version 3.6\n",
      "290,300,1,2,3,4,220,13\t110,120,8\tSequence length is 418\n",
      "230,REUSE= NO\n",
      "14\t310,12\t5,6,7,230,RNA free energy files (version 2.3) at 20 degrees created.\n",
      "10,130,140,7\t240,/home/jupyter/plant_microRNA_predictionSEQ.pnt created.\n",
      "Sequence length is 418\n",
      "\n",
      "320,330,8,9,10,14\t3\t/home/jupyter/plant_microRNA_prediction/secondary_structure/mfold/AMWY020002741|1538-1956|201-219|-\n",
      "17\t240,340,350,1\tRNA free energy files (version 2.3) at 20 degrees created.\n",
      "11,12,250,260,7\t20,30,40,50,60,150,250,360,370,11\t10,20,30,40,13,14,15,16,17,18,6\t70,80,90,100,270,280,25\tmfold version 3.6\n",
      "260,380,390,400,160,170,13\t19,20,21,22,27\t50,60,70,\n",
      "Structure plots generated.\n",
      "290,REUSE= NO\n",
      "All done.\n",
      "110,120,130,180,190,270,410, \n",
      "End of Fill\n",
      "Save file created using nafold.\n",
      "23,24,25,26,27,80,90,100,300,310,320,280,290,14\t140,150,9\t200,28,29,30, \n",
      "Minimum folding energy is -120.10 kcal/mol.\n",
      "Energy increment is 6.00 kcal/mol.\n",
      "15\tSEQ.pnt created.\n",
      "Sequence length is 419\n",
      "110,120,330,160,170,300,8\t13\tH-num file created from plot file.\n",
      "15\t4\tRNA free energy files (version 2.3) at 20 degrees created.\n",
      "310,320,130,210,20\t340,350,360,370,380,18\t1,2,3,/home/jupyter/plant_microRNA_prediction180,\n",
      "7\t330,140,150,220,10,20,30,40,/home/jupyter/plant_microRNA_prediction/secondary_structure/mfold/AMWY020002791|11386-11804|201-219|-190,200,390,400,410, \n",
      "End of Fill\n",
      "Save file created using nafold.\n",
      "\n",
      "8\t26\t4,5,6,7,8,12\t230,240,160,50,60,70,80,14\t340,350,360,Minimum folding energy is -171.50 kcal/mol.\n",
      "Energy increment is 8.57 kcal/mol.\n",
      "28\tmfold version 3.6\n",
      "9,10,11,12,13,14,15,250,90,100,110,370,380,390,170,210,16,17,18,15\tREUSE= NO\n",
      "120,130,\n",
      "Structure plots generated.\n",
      "All done.\n",
      "260,400,410, \n",
      "End of Fill\n",
      "Save file created using nafold.\n",
      "H-num file created from plot file.\n",
      "220,230,180,190,16\t19,20,SEQ.pnt created.\n",
      "Sequence length is 419\n",
      "1\t140,150,270,280,9\t240,1,2,3,4,5,Minimum folding energy is -141.10 kcal/mol.\n",
      "Energy increment is 7.05 kcal/mol.\n",
      "200,21,22,23,16\t5\t290,160,170,6,7,8,250,260,210,19\tRNA free energy files (version 2.3) at 20 degrees created.\n",
      "8\t\n",
      "Structure plots generated.\n",
      "All done.\n",
      "300,24,25,26,2\t180,190,H-num file created from plot file.\n",
      "9,10,11,12,10,20,30,40,50,220,270,280,290,310,27,28,29,30,27\t200,15\t13\t1,2,3,4,5,60,70,13,14,300,310,/home/jupyter/plant_microRNA_prediction230,Suboptimal foldings created.\n",
      "Energy dot plot created.\n",
      "\n",
      "/home/jupyter/plant_microRNA_prediction/secondary_structure/mfold/AMWY020002591|394-814|201-221|-320,31,32,33,29\t210,\n",
      "80,90,15,16,10\t320,330,6,7,8,9,10, \n",
      "16\t240,220,330,340,100,110,17, \n",
      "mfold version 3.6\n",
      "340,350,11,12,13,14,230,250,260,350,2\t/home/jupyter/plant_microRNA_prediction\n",
      "/home/jupyter/plant_microRNA_prediction/secondary_structure/mfold/AMWY020002001|10002-10422|201-221|+120,17\t\n",
      "6\t\n",
      "Structure plots generated.\n",
      "All done.\n",
      "360,370,380,240,250,15,16,17,270,REUSE= NO\n",
      "117\t360,370,380,390,400,410, \n",
      "End of Fill\n",
      "Save file created using nafold.\n",
      "3\tSEQ.pnt created.\n",
      "260,270,18, \n",
      "280,130,140,150,9\t\tmfold version 3.6\n",
      "Minimum folding energy is -135.40 kcal/mol.\n",
      "Energy increment is 6.77 kcal/mol.\n",
      "390,400,410, \n",
      "End of Fill\n",
      "Save file created using nafold.\n",
      "14\tSequence length is 421\n",
      "160,290,300,280,Minimum folding energy is -137.10 kcal/mol.\n",
      "Energy increment is 6.86 kcal/mol.\n",
      "REUSE= NO\n",
      "170,28\tRNA free energy files (version 2.3) at 20 degrees created.\n",
      "10,2\tH-num file created from plot file.\n",
      "11\t290,300,SEQ.pnt created.\n",
      "Sequence length is 421\n",
      "1\t310,320,\n",
      "/home/jupyter/plant_microRNA_prediction180,1,20,30,40,/home/jupyter/plant_microRNA_prediction/secondary_structure/mfold/AMWY020002101|5539-5957|201-219|-17\t330,H-num file created from plot file.\n",
      "RNA free energy files (version 2.3) at 20 degrees created.\n",
      "310,\n",
      "2,3,4,5,6,50,60,70,80,340,350,190,1,2,3,4,5,6,Suboptimal foldings created.\n",
      "Energy dot plot created.\n",
      "18\t3\t320,330,340,10,20,30,40,7,8,9,10,11,12,90,100,110,18\t200,360,370,10\t4\t7,8,9,mfold version 3.6\n",
      "20\t7\t210,220,350,360,370,380,All done.\n",
      "\n",
      "Structure plots generated.\n",
      "13,14,15,16,120,130,140,Suboptimal foldings created.\n",
      "Energy dot plot created.\n",
      "50,60,70,380,390,400,410, \n",
      "REUSE= NO\n",
      "10,11,12,13,230,17,18,19,20,21,80,90,100,110,15\t390,400,410, \n",
      "End of Fill\n",
      "Save file created using nafold.\n",
      "End of Fill\n",
      "Save file created using nafold.\n",
      "Minimum folding energy is -124.10 kcal/mol.\n",
      "Energy increment is 6.21 kcal/mol.\n",
      "29\t150,160,10\t14, \n",
      "SEQ.pnt created.\n",
      "Sequence length is 419\n",
      "Minimum folding energy is -115.90 kcal/mol.\n",
      "Energy increment is 5.79 kcal/mol.\n",
      "240,22, \n",
      "170,180,RNA free energy files (version 2.3) at 20 degrees created.\n",
      "Suboptimal foldings created.\n",
      "2\t30\t120,130,250,10,12\tH-num file created from plot file.\n",
      "1,190,140,150,160,Energy dot plot created.\n",
      "10\t/home/jupyter/plant_microRNA_prediction\n",
      "/home/jupyter/plant_microRNA_prediction/secondary_structure/mfold/AMWY020002481|3478-3899|201-222|+2,3,4,5,6,7,260,270,20,30,40,50,H-num file created from plot file.\n",
      "18\t\n",
      "19\t170,200,210,11\t19\t280,1,2,3,8,9,10,11,12,5\t4\t8\tmfold version 3.6\n",
      "Suboptimal foldings created.\n",
      "Energy dot plot created.\n",
      "180,60,70,80,2\t220,230,13,14,15,16,17,18,19,20,290,300,4,5,6,7,8,9,10,11,11\t10\tREUSE= NO\n",
      "90,100,110,120,2\t21,22,23,240,250,260,190,200,1\t310,12,13,14,130,140,270,24,25,26,3\t210,320,330,15,16,17,11\t27,28,29,30,150,SEQ.pnt created.\n",
      "Sequence length is 422\n",
      "31\t280,290,300,Suboptimal foldings created.\n",
      "Energy dot plot created.\n",
      "18,19,20,21,340, \n",
      "RNA free energy files (version 2.3) at 20 degrees created.\n",
      "13\t310,160,220,230,1\t22,23,24,10\t12\t1\t19\t10,20,30,40,320,330,340,170,180,190,5\t6\t350,360,370,25,26,240,350,360,50,60,70,3\t200,27, \n",
      "9\t250,260,380,390,400,410, \n",
      "End of Fill\n",
      "Save file created using nafold.\n",
      "11\t80,90,100,30\t370,380,390,400,410,12\t270,110,120,2\t210,220,Minimum folding energy is -142.90 kcal/mol.\n",
      "Energy increment is 7.14 kcal/mol.\n",
      "4\t420, \n",
      "End of Fill\n",
      "Save file created using nafold.\n",
      "Minimum folding energy is -119.50 kcal/mol.\n",
      "130,140,12\t230,240,280,Energy increment is 5.97 kcal/mol.\n",
      "150,11\tH-num file created from plot file.\n",
      "290,300,3\t250,14\tH-num file created from plot file.\n",
      "20\t13\t310,1\t160,260,270,1,2,3,4,20\t1,2,3,7\t6\t280,170,180,320,5,6,7,4,5,6,7,8,9,\n",
      "Structure plots generated.\n",
      "All done.\n",
      "190,200,Suboptimal foldings created.\n",
      "330,340,350,4\t31\t8,9,10,11,12,13,12\t290,300,310,10,11,12,13\t360,370,380,390,Energy dot plot created.\n",
      "10\t3\t14,15,16,17,5\t13,14,15,16,17,320,330,210,Suboptimal foldings created.\n",
      "Energy dot plot created.\n",
      "400,410,420, \n",
      "End of Fill\n",
      "13\t18,19,20,21,18,19,20,21,22,340,350,360,370,220,230,Save file created using nafold.\n",
      "Minimum folding energy is -106.80 kcal/mol.\n",
      "10\t22,23,23,24,25,12\t240,4\t380,390,400,410, \n",
      "End of Fill\n",
      "Save file created using nafold.\n",
      "Energy increment is 5.34 kcal/mol.\n",
      "2\t15\t/home/jupyter/plant_microRNA_prediction \n",
      "\n",
      "Minimum folding energy is -107.10 kcal/mol.\n",
      "Energy increment is 5.36 kcal/mol.\n",
      "14\t \n",
      "\n",
      "/home/jupyter/plant_microRNA_prediction/secondary_structure/mfold/AMWY020001951|3001-3263|44-63|-250,20\tH-num file created from plot file.8\t21\t7\t260,270,\n",
      "1,2,3,4,5,6,mfold version 3.6\n",
      "H-num file created from plot file.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "13\t3\t5\t14\t280,11\t1,7,8,9,4\tREUSE= NO\n",
      "290,300,2,3,4,5,6,6\t10, \n",
      "14\t310,SEQ.pnt created.\n",
      "Sequence length is 263\n",
      "5\t7,8,11\t13\t320,330,340,350,3\t16\tRNA free energy files (version 2.3) at 20 degrees created.\n",
      "9,10,11,12,13,15\t360,21\t10,20,30,8\t14,15, \n",
      "Suboptimal foldings created.\n",
      "Energy dot plot created.\n",
      "10\t22\tSuboptimal foldings created.\n",
      "Energy dot plot created.\n",
      "9\t370,380,390,15\t6\t10\t40,50,60,400,410,420, \n",
      "End of Fill\n",
      "Save file created using nafold.\n",
      "Minimum folding energy is -134.20 kcal/mol.\n",
      "Energy increment is 6.71 kcal/mol.\n",
      "14\t4\t70,80,Suboptimal foldings created.\n",
      "Energy dot plot created.\n",
      "5\t12\t15\t7\tH-num file created from plot file.\n",
      "90,100,14\t10\t17\t6\t110,120,12\t1,2,3,4,5,4\tSuboptimal foldings created.\n",
      "9\t6,7,8,9,10,130,140,Energy dot plot created.\n",
      "10\t16\t1\t11\t\n",
      "Structure plots generated.\n",
      "All done.16\t150,160,23\t\n",
      "22\t11, \n",
      "7\t170,180,1\t5\t190,200,16\t6\t8\t11\t15\t210,220,13\t18\t7\t230,240,250,260, \n",
      "End of Fill\n",
      "Save file created using nafold.\n",
      "13\t5\t\n",
      "Structure plots generated.\n",
      "All done.\n",
      "/home/jupyter/plant_microRNA_prediction\n",
      "/home/jupyter/plant_microRNA_prediction/secondary_structure/mfold/AMWY020001921|3070-3487|201-218|-\n",
      "Minimum folding energy is -88.50 kcal/mol.\n",
      "Energy increment is 4.42 kcal/mol.\n",
      "Suboptimal foldings created.\n",
      "17\t11\t12\t2\tH-num file created from plot file.\n",
      "Energy dot plot created.\n",
      "10\t17\tmfold version 3.6\n",
      "23\t24\t1,2,6\t8\tREUSE= NO\n",
      "3,4,5,6,7,8,/home/jupyter/plant_microRNA_prediction\n",
      "2\t/home/jupyter/plant_microRNA_prediction/secondary_structure/mfold/AMWY020002371|533-804|201-219|+9,10,11,\n",
      "9\t17\tSEQ.pnt created.\n",
      "Sequence length is 418\n",
      "16\t7\t12,13,14,15,12\t14\tmfold version 3.6\n",
      "19\t \n",
      "RNA free energy files (version 2.3) at 20 degrees created.\n",
      "14\t8\tREUSE= NO\n",
      "12\t1\t13\t11\t10,20,30,40,6\t18\t3\t25\t50,60,70,24\tSEQ.pnt created.\n",
      "Sequence length is 272\n",
      "9\t80,90,100,7\tRNA free energy files (version 2.3) at 20 degrees created.\n",
      "Suboptimal foldings created.\n",
      "Energy dot plot created.\n",
      "\n",
      "Structure plots generated.\n",
      "All done.\n",
      "15\t3\t110,120,8\t17\t13\t10\t18\t130,140,10,20,30,40,9\t1\t50,60,70,150,13\t80,90,160,15\t7\t14\t/home/jupyter/plant_microRNA_prediction\n",
      "1\t/home/jupyter/plant_microRNA_prediction/secondary_structure/mfold/AMWY020002591|1183-1601|201-219|+\n",
      "100,1\t170,2\t110,120,mfold version 3.6\n",
      "26\t180,\n",
      "Structure plots generated.\n",
      "All done.\n",
      "4\t8\t25\t130,140,190,16\tREUSE= NO\n",
      "11\t150,9\t4\t19\t160,18\t14\t200,SEQ.pnt created.\n",
      "Sequence length is 419\n",
      "170,180,14\t210,20\t190,200,\n",
      "Structure plots generated.\n",
      "All done.\n",
      "/home/jupyter/plant_microRNA_prediction\n",
      "220,/home/jupyter/plant_microRNA_prediction/secondary_structure/mfold/AMWY020002791|717-1133|201-217|-16\t\n",
      "8\t2\t3\t15\t210,220,230,RNA free energy files (version 2.3) at 20 degrees created.\n",
      "230,240,250,10,20,30,40,240,2\t260,270, \n",
      "End of Fill\n",
      "Save file created using nafold.\n",
      "50,60,70,80,mfold version 3.6\n",
      "250,9\t27\t2\tMinimum folding energy is -67.00 kcal/mol.\n",
      "Energy increment is 3.35 kcal/mol.\n",
      "REUSE= NO\n",
      "90,260,5\t12\t/home/jupyter/plant_microRNA_prediction\n",
      "100,110,5\t17\t/home/jupyter/plant_microRNA_prediction/secondary_structure/mfold/AMWY020001851|6155-6574|201-220|+\n",
      "H-num file created from plot file.\n",
      "\n",
      "Structure plots generated.\n",
      "All done.\n",
      "1\t270,120,19\t1,2,3,SEQ.pnt created.\n",
      "Sequence length is 417\n",
      "280,290,mfold version 3.6\n",
      "130,15\t15\tRNA free energy files (version 2.3) at 20 degrees created.\n",
      "9\t300,4,5,6,7,8,REUSE= NO\n",
      "4\t17\t140,150,310,320,21\t9, \n",
      "160,10,20,30,330,340,/home/jupyter/plant_microRNA_prediction3\t16\t170,\n",
      "3\t/home/jupyter/plant_microRNA_prediction/secondary_structure/mfold/AMWY020002481|3480-3899|201-220|+SEQ.pnt created.\n",
      "Sequence length is 420\n",
      "40,50,60,\n",
      "28\t350,360,\n",
      "Structure plots generated.\n",
      "All done.\n",
      "3\t180,70,Suboptimal foldings created.\n",
      "Energy dot plot created.\n",
      "370,380,390,mfold version 3.6\n",
      "13\t190,80,90,RNA free energy files (version 2.3) at 20 degrees created.\n",
      "10,6\t400,410, \n",
      "End of Fill\n",
      "Save file created using nafold.\n",
      "1\t20\t200,1\tREUSE= NO\n",
      "6\t20,30,40,50,60,100,18\tMinimum folding energy is -132.80 kcal/mol.\n",
      "Energy increment is 6.64 kcal/mol.\n",
      "1\t210,220,\n",
      "Structure plots generated.\n",
      "All done.\n",
      "16\t70,80,90,110,120,SEQ.pnt created.\n",
      "Sequence length is 420\n",
      "230,H-num file created from plot file.\n",
      "5\t/home/jupyter/plant_microRNA_prediction100,110,\n",
      "/home/jupyter/plant_microRNA_prediction/secondary_structure/mfold/AMWY020002691|3764-4181|201-218|-130,18\t\n",
      "22\t240,1,2,3,4,RNA free energy files (version 2.3) at 20 degrees created.\n",
      "120,130,140,150,250,mfold version 3.6\n",
      "5,10,20,30,40,4\t140,150,160,29\t260,4\t6,7,8,2\t50,60,70,REUSE= NO\n",
      "/home/jupyter/plant_microRNA_prediction160,\n",
      "17\t14\t170,4\t270,80,90,/home/jupyter/plant_microRNA_prediction/secondary_structure/mfold/AMWY020002851|8518-8936|201-219|-9,10,11,\n",
      "20\t170,180,7\t21\t100,110,280,290, \n",
      "2\tSEQ.pnt created.\n",
      "Sequence length is 418\n",
      "19\t180,190,17\t7\tmfold version 3.6\n",
      "120,130,300,6\tRNA free energy files (version 2.3) at 20 degrees created.\n",
      "190,19\tREUSE= NO\n",
      "200,140,310,320,200,210,210,10,20,30,23\t150,160,330,340,220,220,SEQ.pnt created.\n",
      "Sequence length is 419\n",
      "40,50,60,70,80,2\t170,350,360,370,230,5\t5\t18\tSuboptimal foldings created.\n",
      "Energy dot plot created.\n",
      "10\t230,15\t90,100,5\tRNA free energy files (version 2.3) at 20 degrees created.\n",
      "10,20,180,190,22\t3\t240,250,380,390,400,21\t240,200,110,120,130,30,40,50,8\t1\t260,270,3\t250,8\t140,60,70,80,210,410, \n",
      "End of Fill\n",
      "Save file created using nafold.\n",
      "Minimum folding energy is -152.10 kcal/mol.\n",
      "Energy increment is 7.61 kcal/mol.\n",
      "7\t18\t280,150,160,260,1\t220,230,24\t90,100,290,300,H-num file created from plot file.\n",
      "270,240,170,310,110,120,130,1,2,3,280,30\t6\t250,320,4,5,6,7,140,150,180,11290,6\t1\t190,200,4\t260,330,19\t8,9,10,11,22\t6\t160,300,20\t\t23\t9\t270,280,340,350,360,210,310,12,13,14,170,4\t9\t290,300,370,380,220,230,180,190,19\t8\t15, \n",
      "320,330,340,25\t310,320,200,240,390,400,410, \n",
      "End of Fill\n",
      "Save file created using nafold.\n",
      "350,360,330,340,20\t210,370,380,390,400,410,Minimum folding energy is -114.60 kcal/mol.\n",
      "Energy increment is 5.73 kcal/mol.\n",
      "250,260,270,350,360,370,380,7\t220,230, \n",
      "End of Fill\n",
      "Save file created using nafold.\n",
      "1\t390,400,410, \n",
      "End of Fill\n",
      "280,290,3\t2\t21\t7\tH-num file created from plot file.\n",
      "2\t7\t300,310,Save file created using nafold.\n",
      "Minimum folding energy is -134.00 kcal/mol.\n",
      "Energy increment is 6.70 kcal/mol.\n",
      "5\tMinimum folding energy is -122.30 kcal/mol.\n",
      "Energy increment is 6.12 kcal/mol.\n",
      "240,\n",
      "Structure plots generated.\n",
      "All done.\n",
      "24\t1,2,3,4,1\t5\tSuboptimal foldings created.\n",
      "Energy dot plot created.\n",
      "10\t\n",
      "Structure plots generated.\n",
      "All done.\n",
      "320,330,250,260,H-num file created from plot file.\n",
      "5,6,7,8,9,9\tH-num file created from plot file.\n",
      "26\t340,350,360,370,270,1,2,3,4,10,11,12,13,1\t1,2,3,4,280,290,380,390,400,410, \n",
      "End of Fill\n",
      "5,6,7,8,9,14,15,21\t5,6,7,8,Save file created using nafold.\n",
      "Minimum folding energy is -97.10 kcal/mol.\n",
      "Energy increment is 4.86 kcal/mol.\n",
      "300,310,10,11,12,13,14,16,17,18,19,/home/jupyter/plant_microRNA_prediction9,10,\n",
      "8\t320,330,/home/jupyter/plant_microRNA_prediction15,16,17,\n",
      "20,21,22,/home/jupyter/plant_microRNA_prediction/secondary_structure/mfold/AMWY020002891|2-420|201-219|-2\t\n",
      "4\t3\t/home/jupyter/plant_microRNA_prediction/secondary_structure/mfold/AMWY020002901|421-840|201-220|+22\t11, \n",
      "\n",
      "340,350,360,3\t8\t18,19,20,21, \n",
      "H-num file created from plot file.\n",
      "8\t6\t370,380,390,mfold version 3.6\n",
      "22,23,24,25,26, \n",
      "mfold version 3.6\n",
      "11\t1,2,3,4,5,6,20\t25\t400,410, \n",
      "End of Fill\n",
      "Save file created using nafold.\n",
      "Minimum folding energy is -129.90 kcal/mol.\n",
      "Energy increment is 6.50 kcal/mol.\n",
      "Structure plots generated.\n",
      "All done.\n",
      "6\t27\tREUSE= NO\n",
      "7,8,9,\n",
      "REUSE= NO\n",
      "SEQ.pnt created.\n",
      "10,11,12,13,20\tSEQ.pnt created.\n",
      "22\tH-num file created from plot file.\n",
      "Sequence length is 420\n",
      "14,15, \n",
      "Suboptimal foldings created.\n",
      "Energy dot plot created.\n",
      "Sequence length is 419\n",
      "1,2,3,4,3\t10\tRNA free energy files (version 2.3) at 20 degrees created.\n",
      "9\t4\tRNA free energy files (version 2.3) at 20 degrees created.\n",
      "5,6,7,8,9\t/home/jupyter/plant_microRNA_prediction10,20,30,40,50,5\t\n",
      "4\t23\t10,20,30,40,9,10,7\t9\t/home/jupyter/plant_microRNA_prediction/secondary_structure/mfold/AMWY020002911|1-281|201-219|-60,70,\n",
      "50,60,70,21\t11,12,13,14,15,26\t80,90,7\t \n",
      "80,90,28\t12\tmfold version 3.6\n",
      "100,110,Suboptimal foldings created.\n",
      "Energy dot plot created.\n",
      "10\t21\t100,110,120,Suboptimal foldings created.\n",
      "Energy dot plot created.\n",
      "REUSE= NO\n",
      "120,130,140,130,140,23\t\n",
      "Structure plots generated.\n",
      "All done.\n",
      "Suboptimal foldings created.\n",
      "Energy dot plot created.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "SEQ.pnt created.\n",
      "All done.\n",
      "\n",
      "Structure plots generated.\n",
      "5\t150,160,4\t10\t150,160,6\t10\t11\tSequence length is 281\n",
      "170,5\t8\t170,24\tRNA free energy files (version 2.3) at 20 degrees created.\n",
      "10,180,190,\n",
      "Structure plots generated.\n",
      "All done.\n",
      "180,200,Suboptimal foldings created.\n",
      "Energy dot plot created.\n",
      "10\t20,30,40,50,13\t8\t27\t190,22\t210,/home/jupyter/plant_microRNA_prediction60,70,29\t\n",
      "/home/jupyter/plant_microRNA_prediction/secondary_structure/mfold/AMWY020003031|4770-5187|201-218|+200,\n",
      "/home/jupyter/plant_microRNA_prediction11\t\n",
      "/home/jupyter/plant_microRNA_prediction/secondary_structure/mfold/AMWY020002411|5665-6079|201-215|+220,80,\n",
      "11\t210,220,230,90,100,/home/jupyter/plant_microRNA_prediction\n",
      "mfold version 3.6\n",
      "22\t/home/jupyter/plant_microRNA_prediction/secondary_structure/mfold/AMWY020003061|8575-8993|201-219|-\n",
      "mfold version 3.6\n",
      "24\t11\t230,5\t240,6\t110,REUSE= NO\n",
      "REUSE= NO\n",
      "1\tmfold version 3.6\n",
      "240,6\t250,9\tSEQ.pnt created.\n",
      "Sequence length is 4187\t120,25\t\n",
      "250,260,28\t130,23\tREUSE= NO\n",
      "SEQ.pnt created.\n",
      "Sequence length is 415\n",
      "2\t11\tRNA free energy files (version 2.3) at 20 degrees created.\n",
      "9\t140,150,SEQ.pnt created.\n",
      "260,270,14\t270,280,12\tSequence length is 419\n",
      "280,160,10,20,30,40,50,60,RNA free energy files (version 2.3) at 20 degrees created.\n",
      "10,25\t290,300,170,180,12\t7\tRNA free energy files (version 2.3) at 20 degrees created.\n",
      "70,80,90,100,23\t290,300,310,6\t310,320,330,340,190,2\t20,30,40,50,60,70,80,110,10,20,30,40,12\t320,330,8\t200,\n",
      "Structure plots generated.\n",
      "All done.\n",
      "350,360,7\t340,350,50,60,90,100,110,120,130,26\t370,380,390,400,29\t210,220,360,370,140,150,120,130,70,80,2\t12\t230,240,30\t15\t380,390,160,410, \n",
      "End of Fill\n",
      "Save file created using nafold.\n",
      "Minimum folding energy is -135.80 kcal/mol.\n",
      "140,90,100,\n",
      "Structure plots generated.\n",
      "All done.\n",
      "170,Energy increment is 6.79 kcal/mol.\n",
      "250,260,270,280, \n",
      "13\t400,410, \n",
      "End of Fill\n",
      "Save file created using nafold.\n",
      "110,120,150,160,26\t180,H-num file created from plot file.\n",
      "/home/jupyter/plant_microRNA_prediction13\tEnd of Fill\n",
      "Save file created using nafold.\n",
      "Minimum folding energy is -84.90 kcal/mol.\n",
      "Energy increment is 4.25 kcal/mol.\n",
      "Minimum folding energy is -133.80 kcal/mol.\n",
      "Energy increment is 6.69 kcal/mol.\n",
      "\n",
      "/home/jupyter/plant_microRNA_prediction/secondary_structure/mfold/AMWY020003211|678-1096|201-219|+\n",
      "170,130,140,190,8\tH-num file created from plot file.24\tH-num file created from plot file.\n",
      "1,2,3,7\t3\t180,150,160,200,mfold version 3.6\n",
      "8\t13\t1,2,4,5,6,7,9\t\n",
      "1,2,3,190,/home/jupyter/plant_microRNA_prediction\n",
      "8,9,170,3,4,5,6,7,/home/jupyter/plant_microRNA_prediction/secondary_structure/mfold/AMWY020003261|12626-13043|201-218|+REUSE= NO\n",
      "200,210,4,5,6,7, \n",
      "\n",
      "10,11,12,180,SEQ.pnt created.\n",
      "Sequence length is 419\n",
      "2\t220,13\t210,27\t8,9,10,11,3\t3\t190, \n",
      "1\t230,12,13,14,15,200,mfold version 3.6\n",
      "220,Suboptimal foldings created.\n",
      "Energy dot plot created.\n",
      "1\t16,17,RNA free energy files (version 2.3) at 20 degrees created.\n",
      "10,14\t230,9\t240,27\t \n",
      "14\t20,30,40,50,4\t8\tREUSE= NO\n",
      "210,250,260,240,25\t\n",
      "Structure plots generated.\n",
      "All done.\n",
      "60,70,270,220,230,SEQ.pnt created.\n",
      "250,9\t14\t80,90,100,280,290,Suboptimal foldings created.\n",
      "Energy dot plot created.\n",
      "240,Sequence length is 418\n",
      "260,270,30\t14\t110,120,28\t300,4\t250,10\tRNA free energy files (version 2.3) at 20 degrees created.\n",
      "10,20,30,2\t280,4\t130,140,310,320,260,2\t40,50,60,70,\n",
      "Structure plots generated.\n",
      "All done.\n",
      "290,300,310,/home/jupyter/plant_microRNA_prediction150,330,15\t\n",
      "/home/jupyter/plant_microRNA_prediction/secondary_structure/mfold/AMWY020002441|2151-2568|201-218|+\n",
      "270,80,90,100,2\t340,350,Suboptimal foldings created.\n",
      "Energy dot plot created.\n",
      "10\t160,320,330,5\t280,290,300,9\t110,15\tmfold version 3.6\n",
      "340,350,360,370,170,180,2\t310,15\t120,130,REUSE= NO\n",
      "380,390,190,\n",
      "Structure plots generated.\n",
      "All done.\n",
      "360,370,380,320,140,SEQ.pnt created.\n",
      "Sequence length is 418\n",
      "200,/home/jupyter/plant_microRNA_prediction400,410, \n",
      "End of Fill\n",
      "Save file created using nafold.\n",
      "\n",
      "/home/jupyter/plant_microRNA_prediction/secondary_structure/mfold/AMWY020003271|905-1322|201-218|+390,400,410, \n",
      "End of Fill\n",
      "Save file created using nafold.\n",
      "330,340,\n",
      "31\t11\t150,160,15\t5\tMinimum folding energy is -124.20 kcal/mol.\n",
      "Energy increment is 6.21 kcal/mol.\n",
      "210,Minimum folding energy is -137.10 kcal/mol.\n",
      "Energy increment is 6.86 kcal/mol.\n",
      "350,360,370,380,170,RNA free energy files (version 2.3) at 20 degrees created.\n",
      "mfold version 3.6\n",
      "29\t3\t220,3\t390,400,410, \n",
      "End of Fill\n",
      "5\tH-num file created from plot file.\n",
      "10,20,30,40,50,60,REUSE= NO\n",
      "180,H-num file created from plot file.\n",
      "Save file created using nafold.\n",
      "Minimum folding energy is -123.20 kcal/mol.\n",
      "Energy increment is 6.16 kcal/mol.\n",
      "3\t/home/jupyter/plant_microRNA_prediction\n",
      "16\t230,240,70,80,90,/home/jupyter/plant_microRNA_prediction/secondary_structure/mfold/AMWY020003281|6110-6527|201-218|-1,2,3,4,5,6,190,6\t1,2,3,4,5,\n",
      "250,1\t100,11\tSEQ.pnt created.\n",
      "Sequence length is 418\n",
      "3\t7,8,9,10,11,12,6,7,8,\n",
      "Structure plots generated.\n",
      "All done.\n",
      "200,260,110,120,16\tRNA free energy files (version 2.3) at 20 degrees created.\n",
      "210, \n",
      "H-num file created from plot file.\n",
      "13,14,15,16,mfold version 3.6\n",
      "32\t270,280,130,140,150,10,20,30,40,50,220,1,2,3,4,17,18,19,20,21,22,23, \n",
      "REUSE= NO\n",
      "12\t6\t290,160,5,6,7,8,230,60,70,SEQ.pnt created.\n",
      "1\t300,310,4\t2\t170,180,4\t80,90,9,10,240,/home/jupyter/plant_microRNA_predictionSuboptimal foldings created.\n",
      "\n",
      "/home/jupyter/plant_microRNA_prediction/secondary_structure/mfold/AMWY020003291|3059-3479|201-221|-\n",
      "190,320,6\tSequence length is 418\n",
      "4\t11, \n",
      "250,100,110,2\t12\tEnergy dot plot created.\n",
      "1\t17\t330,340,260,200,mfold version 3.6\n",
      "RNA free energy files (version 2.3) at 20 degrees created.\n",
      "10,20,30,120,130,7\t4\t40,50,60,70,210,220,230,350,360,370,380,270,140,33\t17\t390,400,410, \n",
      "End of Fill\n",
      "240,80,90,100,150,160,7\t280,290,REUSE= NO\n",
      "1\t110,120,250,260,Save file created using nafold.\n",
      "Minimum folding energy is -145.00 kcal/mol.\n",
      "Energy increment is 7.25 kcal/mol.\n",
      "170,Suboptimal foldings created.\n",
      "Energy dot plot created.\n",
      "2\t300,310,5\tSEQ.pnt created.\n",
      "Sequence length is 421\n",
      "30\t5\t130,270,180,320,10\tSuboptimal foldings created.\n",
      "Energy dot plot created.H-num file created from plot file.\n",
      "18\tRNA free energy files (version 2.3) at 20 degrees created.\n",
      "280,290,7\t2\t140,150,160,13\t330,340,350,3\t5\t190,200,\n",
      "10\t10,20,30,40,1,2,3,4,300,310,170,180,360,370,380,390,18\t50,60,70,8\t190,5,6,7,8,9,210,5\t320,330,400,410, \n",
      "End of Fill\n",
      "Save file created using nafold.\n",
      "80,90,100,3\t200,10,11,12,13,14,8\t220,230,Minimum folding energy is -135.30 kcal/mol.\n",
      "Energy increment is 6.76 kcal/mol.\n",
      "340,350,360,370,2\t110,210,15,16,6\t3\t240,11\t220,380,390,400,410, \n",
      "End of Fill\n",
      "120,130,6\t17,18,19,3\t230,240,H-num file created from plot file.\n",
      "1,250,Save file created using nafold.\n",
      "Minimum folding energy is -155.70 kcal/mol.\n",
      "Energy increment is 7.79 kcal/mol.\n",
      "3\t140,150,160,14\t4\t19\t20,21,22,23,24,8\t11\t250,260,270,6\t2,3,4,5,170,25,26,27, \n",
      "19\t9\t6,7,8,9,10,180,190,H-num file created from plot file.\n",
      "280,290,260,270,280,6\t1,2,3,4,3\t11,12,200,300,310,9\t4\t290,300,5,6,7\t210,1\t13,14,15,16,17,18, \n",
      "4\t7\t320,310,320,7,8,9,10,4\t220,5\t12\t15\t330,340,350,1\t7\t9\t11,12,330,340,350,360,230,240,4\t360,370,380,370,380,390,13,14,250,\n",
      "Structure plots generated.\n",
      "390,400,410, \n",
      "End of Fill\n",
      "All done.\n",
      "400,410, \n",
      "End of Fill\n",
      " \n",
      "260,1\tSave file created using nafold.\n",
      "Minimum folding energy is -143.60 kcal/mol.\n",
      "Energy increment is 7.18 kcal/mol.\n",
      "Save file created using nafold.\n",
      "Minimum folding energy is -115.40 kcal/mol.\n",
      "Energy increment is 5.77 kcal/mol.\n",
      "270,\n",
      "Structure plots generated.\n",
      "All done.\n",
      "7\t5\t4\t\n",
      "Structure plots generated.\n",
      "All done.\n",
      "280,290,H-num file created from plot file.\n",
      "H-num file created from plot file.\n",
      "300,5\t5\t8\tSuboptimal foldings created.\n",
      "Energy dot plot created.\n",
      "10\tSuboptimal foldings created.\n",
      "Energy dot plot created.\n",
      "1,2,3,2\t1,2,3,310,320,/home/jupyter/plant_microRNA_prediction10\t13\t\n",
      "16\t8\t4,5,6,7,/home/jupyter/plant_microRNA_prediction/secondary_structure/mfold/AMWY020003341|4703-5044|201-219|+6\t330,340,4,5,6,7,8,\n",
      "5\t\n",
      "Structure plots generated.\n",
      "All done.\n",
      "20\t/home/jupyter/plant_microRNA_prediction\n",
      "8,9,10,11,350,360,370,/home/jupyter/plant_microRNA_prediction/secondary_structure/mfold/AMWY020002921|2640-3061|201-222|-9,10,11,12,13,14,\n",
      "Suboptimal foldings created.\n",
      "Energy dot plot created.\n",
      "10\t/home/jupyter/plant_microRNA_prediction\n",
      "mfold version 3.6\n",
      "/home/jupyter/plant_microRNA_prediction/secondary_structure/mfold/AMWY020002591|8479-8900|201-222|-380,390,400,410,420,\n",
      "15,16,17,12, \n",
      "20\tmfold version 3.6\n",
      " \n",
      "End of Fill\n",
      "Save file created using nafold.\n",
      "REUSE= NO\n",
      "8\t18, \n",
      "mfold version 3.6\n",
      "REUSE= NO\n",
      "6\tMinimum folding energy is -129.90 kcal/mol.\n",
      "Energy increment is 6.50 kcal/mol.\n",
      "5\tREUSE= NO\n",
      "SEQ.pnt created.\n",
      "Sequence length is 342\n",
      "/home/jupyter/plant_microRNA_prediction\n",
      "6\tSEQ.pnt created.\n",
      "Sequence length is 422\n",
      "RNA free energy files (version 2.3) at 20 degrees created.\n",
      "H-num file created from plot file.\n",
      "11\t3\t11\t6\t7\t9\tSEQ.pnt created.\n",
      "Sequence length is 422\n",
      "14\t10,20,30,1,2,3,RNA free energy files (version 2.3) at 20 degrees created.\n",
      "Suboptimal foldings created.\n",
      "Energy dot plot created.\n",
      "10\t9\t6\t21\t17\t40,50,60,RNA free energy files (version 2.3) at 20 degrees created.\n",
      "10,20,30,40,4,5,6,7,8,70,80,90,9,10,11,12,13,14,50,60,70,10,20,30,40,21\t11\t100,110,15,16,17,18,19,80,90,50,60,70,Suboptimal foldings created.\n",
      "9\t7\t120,130, \n",
      "80,90,Energy dot plot created.\n",
      "10\t6\t100,110,140,7\t4\t100,110,7\t12\t120,12\t\n",
      "Structure plots generated.\n",
      "All done.\n",
      "15\t150,8\t120,22\t130,140,1\t7\t\n",
      "Structure plots generated.\n",
      "11\t160,170,130,All done.\n",
      "150,180,140,150,160,170,12\t190,\n",
      "Structure plots generated.\n",
      "160,All done.\n",
      "/home/jupyter/plant_microRNA_prediction\n",
      "11\t200,22\t170,8\t180,7\t210,8\t180,190,/home/jupyter/plant_microRNA_predictionSuboptimal foldings created.\n",
      "Energy dot plot created.\n",
      "190,\n",
      "/home/jupyter/plant_microRNA_prediction/secondary_structure/mfold/AMWY020002741|1881-2301|201-221|+5\t\n",
      "220,230,13\t200,10\t200,13\t9\t8\t2\t240,250,210,210,12\tmfold version 3.6\n",
      "16\t/home/jupyter/plant_microRNA_prediction260,270,2\t\n",
      "220,/home/jupyter/plant_microRNA_prediction/secondary_structure/mfold/AMWY020002791|1687-2105|201-219|-\n",
      "220,8\tREUSE= NO\n",
      "280,290,300,230,230,SEQ.pnt created.\n",
      "Sequence length is 421\n",
      "310,320,330,340, \n",
      "mfold version 3.6\n",
      "240,250,240,RNA free energy files (version 2.3) at 20 degrees created.\n",
      "13\tEnd of Fill\n",
      "Save file created using nafold.\n",
      "Minimum folding energy is -128.20 kcal/mol.\n",
      "Energy increment is 6.41 kcal/mol.\n",
      "23\tREUSE= NO\n",
      "9\t260,250,10,20,30,8\t12\tSEQ.pnt created.\n",
      "Sequence length is 419\n",
      "270,\n",
      "Structure plots generated.\n",
      "40,50,60,70,260,H-num file created from plot file.\n",
      "All done.\n",
      "9\t280,11\t14\t80,90,9\t270,6\t1,2,3,RNA free energy files (version 2.3) at 20 degrees created.\n",
      "10,290,1\t14\t100,280,290,4,5,6,7,8,20,30,40,50,60,17\t3\t3\t300,310,\n",
      "Structure plots generated.\n",
      "All done.\n",
      "110,120,70,80,300,9,10,11,12,13,14,320,330,130,90,100,310,15,16,17,18,19, \n",
      "/home/jupyter/plant_microRNA_prediction340,350,\n",
      "140,/home/jupyter/plant_microRNA_prediction/secondary_structure/mfold/AMWY020002691|3764-4182|201-219|-\n",
      "14\t110,120,320,330,24\t13\t360,370,380,390,150,\n",
      "Structure plots generated.\n",
      "All done.\n",
      "130,340,350,360,mfold version 3.6\n",
      "REUSE= NO\n",
      "400,410,420, \n",
      "End of Fill\n",
      "Save file created using nafold.\n",
      "160,170,\n",
      "Structure plots generated.\n",
      "All done.\n",
      "9\t/home/jupyter/plant_microRNA_prediction\n",
      "140,150,370,380,390,SEQ.pnt created.\n",
      "Minimum folding energy is -144.40 kcal/mol.\n",
      "Energy increment is 7.22 kcal/mol.\n",
      "12\t\n",
      "Structure plots generated.\n",
      "All done.\n",
      "7\t15\t15\t160,400,410,420, \n",
      "End of Fill\n",
      "Save file created using nafold.\n",
      "Sequence length is 419\n",
      "180,H-num file created from plot file.\n",
      "4\t2\t170,RNA free energy files (version 2.3) at 20 degrees created.\n",
      "10,Minimum folding energy is -115.40 kcal/mol.\n",
      "Energy increment is 5.77 kcal/mol.\n",
      "190,4\t18\t1,2,3,4,180,20,30,40,50,60,H-num file created from plot file.\n",
      "200,/home/jupyter/plant_microRNA_prediction\n",
      "Suboptimal foldings created.\n",
      "Energy dot plot created.\n",
      "5,6,7,8,9,190,/home/jupyter/plant_microRNA_prediction70,80,90,\n",
      "1,2,3,4,/home/jupyter/plant_microRNA_prediction/secondary_structure/mfold/AMWY020002881|1667-2085|201-219|+\n",
      "210,10\t10,11,12,13,14,200,100,110,1\t5,6,7,8,/home/jupyter/plant_microRNA_prediction220,\n",
      "/home/jupyter/plant_microRNA_prediction/secondary_structure/mfold/AMWY020002741|1954-2372|201-219|+\n",
      "25\t15,16,17,18,19,14\tmfold version 3.6\n",
      "210,\n",
      "Structure plots generated.\n",
      "All done.\n",
      "120,130,9,10,11,12,13,230,20, \n",
      "REUSE= NO\n",
      "SEQ.pnt created.\n",
      "Sequence length is 419\n",
      "220,mfold version 3.6\n",
      "REUSE= NO\n",
      "140,150,14,15,16,17,240,13\tSEQ.pnt created.\n",
      "230,8\t16\t160,16\t18, \n",
      "250,5\tRNA free energy files (version 2.3) at 20 degrees created.\n",
      "Sequence length is 419\n",
      "5\t3\t240,170,260,19\t10,20,30,40,50,60,RNA free energy files (version 2.3) at 20 degrees created.\n",
      "180,250,/home/jupyter/plant_microRNA_prediction270,\n",
      "/home/jupyter/plant_microRNA_prediction/secondary_structure/mfold/AMWY020002901|430-847|201-218|-11\t70,80,\n",
      "190,200,10,20,30,40,50,260,270,280,2\t90,100,210,60,70,280,mfold version 3.6\n",
      "290,110,220,26\t80,90,15\t290,REUSE= NO\n",
      "SEQ.pnt created.\n",
      "300,310,230,120,100,110,300,Sequence length is 418\n",
      "320,17\t240,Suboptimal foldings created.\n",
      "Energy dot plot created.\n",
      "10\t120,130,140,310,320,14\tRNA free energy files (version 2.3) at 20 degrees created.\n",
      "10,20,30,330,340,9\tSuboptimal foldings created.\n",
      "Energy dot plot created.\n",
      "17\t6\t4\t250,130,150,6\t330,340,40,50,60,70,350,360,10\t140,150,260,270,1\t160,350,360,80,90,370,380,390,160,280,170,180,12\t370,380,390,100,400,410,420, \n",
      "End of Fill\n",
      "Save file created using nafold.\n",
      "170,290,300,3\t190,400,410, \n",
      "End of Fill\n",
      "Save file created using nafold.\n",
      "Minimum folding energy is -136.80 kcal/mol.\n",
      "Minimum folding energy is -126.40 kcal/mol.\n",
      "Energy increment is 6.32 kcal/mol.\n",
      "110,120,310,320,180,200,2\tEnergy increment is 6.84 kcal/mol.\n",
      "16\t130,330,340,350,190,210,H-num file created from plot file.\n",
      "1,H-num file created from plot file.\n",
      "140,150,15\t\n",
      "Structure plots generated.\n",
      "All done.\n",
      "200,18\t360,370,380,390,400,410,220,11\t5\t1,2,3,4,5,2,3,4,5,6,7,160,7\t7\t18\t \n",
      "End of Fill\n",
      "Save file created using nafold.\n",
      "Minimum folding energy is -97.10 kcal/mol.\n",
      "Energy increment is 4.86 kcal/mol.\n",
      "230,6,7,8,8,9,10,11,12,170,11\t210,9,10,11,13,14,15,16,17,20\t180,13\t220,240,H-num file created from plot file.\n",
      "12,13,14, \n",
      "18,19, \n",
      "190,200,230,/home/jupyter/plant_microRNA_prediction250,260,1,2,3,4,5,\n",
      "/home/jupyter/plant_microRNA_prediction/secondary_structure/mfold/AMWY020003071|2184-2602|201-219|-\n",
      "4\t240,6,7,8,9,270,210,3\t17\tmfold version 3.6\n",
      "REUSE= NO\n",
      "280,10,11,12,13,220,250,16\t290,14,15, \n",
      "230,1\t6\t12\t260,SEQ.pnt created.\n",
      "Sequence length is 419\n",
      "8\t300,310,8\t240,19\t270,Suboptimal foldings created.\n",
      "RNA free energy files (version 2.3) at 20 degrees created.\n",
      "320,12\t250,280,Energy dot plot created.\n",
      "10\t10,20,30,40,21\t14\t330,340,260,290,50,60,70,350,360,270,280,300,310,Suboptimal foldings created.\n",
      "Energy dot plot created.\n",
      "80,90,5\t370,380,390,400,290,300,320,10\t4\t100,410, \n",
      "End of Fill\n",
      "Save file created using nafold.\n",
      "Minimum folding energy is -106.90 kcal/mol.\n",
      "18\t17\t310,330,340,2\t110,120,Energy increment is 5.34 kcal/mol.\n",
      "7\tSuboptimal foldings created.\n",
      "Energy dot plot created.\n",
      "10\t350,360,320,130,1\tH-num file created from plot file.\n",
      "13\t9\t370,380,390,330,340,9\t1,2,3,4,140,15\t13\t400,410, \n",
      "End of Fill\n",
      "Save file created using nafold.\n",
      "350,360,370,11\t22\t150,5,6,380,390,400,410, \n",
      "End of Fill\n",
      "Minimum folding energy is -124.10 kcal/mol.\n",
      "Energy increment is 6.21 kcal/mol.\n",
      "160,7, \n",
      "Save file created using nafold.\n",
      "Minimum folding energy is -152.80 kcal/mol.\n",
      "Energy increment is 7.64 kcal/mol.\n",
      "170,6\t11\tH-num file created from plot file.\n",
      "1,180,1\t5\tH-num file created from plot file.\n",
      "2,3,4,5,6,190,18\t3\t8\t7,8,9,10,11,12,1,2,3,4,5,6,200,\n",
      "Structure plots generated.\n",
      "All done.\n",
      "Suboptimal foldings created.\n",
      "Energy dot plot created.\n",
      "20\t11\t13,14,15,16,17,7,8,9,10,11,\n",
      "Structure plots generated.\n",
      "All done.\n",
      "14\t1\t14\t210,18,19,20,21,22,12,13,14,15,16\t23,24,25,26,16,17,18,220,23\t12\t27,28,29,30,230,19,20,21,22,23,24,25,12\t \n",
      " \n",
      "240,/home/jupyter/plant_microRNA_prediction\n",
      "7\t/home/jupyter/plant_microRNA_prediction\n",
      "/home/jupyter/plant_microRNA_prediction/secondary_structure/mfold/AMWY020002901|2830-3246|201-217|+\n",
      "250,6\t2\t19\t260,9\tmfold version 3.6\n",
      "REUSE= NO\n",
      "12\t21\t4\t15\tSEQ.pnt created.\n",
      "Sequence length is 417\n",
      "270,280,2\t17\tRNA free energy files (version 2.3) at 20 degrees created.\n",
      "15\t290,13\t2\t10,20,30,300,13\t40,50,60,310,70,80,90,8\t320,330,7\t3\t1\t\n",
      "Structure plots generated.\n",
      "All done.\n",
      "100,110,340,13\t120,130,350,360,370,Suboptimal foldings created.\n",
      "Energy dot plot created.\n",
      "10\t140,150,22\t3\t5\t16\t380,390,400,410, \n",
      "End of Fill\n",
      "16\tSuboptimal foldings created.\n",
      "Energy dot plot created.\n",
      "18\t160,170,Save file created using nafold.\n",
      "Minimum folding energy is -119.00 kcal/mol.\n",
      "Energy increment is 5.95 kcal/mol.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "10\t14\t3\t180,/home/jupyter/plant_microRNA_prediction\n",
      "/home/jupyter/plant_microRNA_prediction/secondary_structure/mfold/AMWY020003291|25-445|201-221|-\n",
      "14\t190,H-num file created from plot file.\n",
      "200,1,2,3,4,5,mfold version 3.6\n",
      "REUSE= NO\n",
      "9\t8\t2\t210,6,7,8,9,10,11,12,SEQ.pnt created.\n",
      "Sequence length is 421\n",
      "4\t14\t220,13,14,15,16,17,RNA free energy files (version 2.3) at 20 degrees created.\n",
      "6\t17\t230,18,19,20,21,10,20,30,11\t23\t17\t240,22,23,24, \n",
      "40,50,60,70,4\t19\t4\t250,260,11\t80,90,100,1\t270,280,15\t110,290,\n",
      "Structure plots generated.\n",
      "All done.\n",
      "120,130,300,9\t3\t140,150,5\t310,320,330,15\t160,340,350,360,7\t18\t170,12\t370,380,390,400,410, \n",
      "End of Fill\n",
      "1\t24\t180,5\t18\t/home/jupyter/plant_microRNA_prediction\n",
      "Save file created using nafold.\n",
      "Minimum folding energy is -114.20 kcal/mol.\n",
      "Energy increment is 5.71 kcal/mol.\n",
      "5\t190,12\t2\t16\t200,210,Suboptimal foldings created.\n",
      "H-num file created from plot file.\n",
      "Energy dot plot created.\n",
      "10\t6\t220,1\t1,2,3,4,\n",
      "Structure plots generated.\n",
      "All done.\n",
      "4\t230,240,5,6,7,8,9,10,11,12,250,13,14,15,16,17,8\t2\t19\t25\t13\t260,18,19,20,13\t1\t6\t270,280, \n",
      "6\t17\t/home/jupyter/plant_microRNA_prediction\n",
      "/home/jupyter/plant_microRNA_prediction/secondary_structure/mfold/AMWY020002791|8185-8603|201-219|-\n",
      "290,3\t300,310,11\tmfold version 3.6\n",
      "REUSE= NO\n",
      "320,330,340,2\tSEQ.pnt created.\n",
      "Sequence length is 419\n",
      "5\t350,360,7\tRNA free energy files (version 2.3) at 20 degrees created.\n",
      "370,380,390,3\t1\t10,20,30,40,50,400,410,420, \n",
      "End of Fill\n",
      "Save file created using nafold.\n",
      "7\t26\t14\t9\t60,70,80,90,Minimum folding energy is -117.70 kcal/mol.\n",
      "Energy increment is 5.88 kcal/mol.\n",
      "14\t2\t7\t18\tSuboptimal foldings created.\n",
      "Energy dot plot created.\n",
      "100,110,H-num file created from plot file.\n",
      "10\t120,130,1,2,3,4\t140,150,4,5,6,7,8,9,12\t3\t160,170,10,11,12,13,8\t6\t180,14,15,16,17,20\t190,27\t4\t15\t\n",
      "Structure plots generated.\n",
      "All done.\n",
      "18,19,20,21,22,\n",
      "Structure plots generated.\n",
      "All done.\n",
      "200,3\t \n",
      "8\t210,220,15\t11\t19\t230,5\t240,13\t4\t7\t250,260,/home/jupyter/plant_microRNA_prediction/home/jupyter/plant_microRNA_prediction\n",
      "\n",
      "/home/jupyter/plant_microRNA_prediction/secondary_structure/mfold/AMWY020002881|1667-2086|201-220|+/home/jupyter/plant_microRNA_prediction/secondary_structure/mfold/AMWY020003261|2778-3194|201-217|-\n",
      "\n",
      "270,280,9\tmfold version 3.6\n",
      "REUSE= NO\n",
      "mfold version 3.6\n",
      "REUSE= NO\n",
      "2\t5\t290,SEQ.pnt created.\n",
      "Sequence length is 417\n",
      "SEQ.pnt created.\n",
      "Sequence length is 420\n",
      "16\t300,310,2\tRNA free energy files (version 2.3) at 20 degrees created.\n",
      "RNA free energy files (version 2.3) at 20 degrees created.\n",
      "16\t4\t320,330,Suboptimal foldings created.\n",
      "1\t9\t12\t10,20,30,40,10,20,30,40,6\t340,350,360,Energy dot plot created.\n",
      "10\t50,60,70,80,50,60,70,370,380,390,400,5\t90,100,14\t80,90,410, \n",
      "End of Fill\n",
      "Save file created using nafold.\n",
      "Minimum folding energy is -119.20 kcal/mol.\n",
      "8\t\n",
      "Structure plots generated.\n",
      "All done.\n",
      "110,100,110,Energy increment is 5.96 kcal/mol.\n",
      "120,H-num file created from plot file.\n",
      "120,6\t3\t17\t130,140,130,140,1,2,3,4,17\t3\t150,150,5,6,7,8,9,10,5\t\n",
      "Structure plots generated.\n",
      "All done.\n",
      "13\t160,/home/jupyter/plant_microRNA_prediction\n",
      "160,11,12,13,14,15,/home/jupyter/plant_microRNA_prediction/secondary_structure/mfold/AMWY020003281|2967-3385|201-219|+\n",
      "7\t11\t2\t170,170,180,16,17,18,19, \n",
      "6\t180,mfold version 3.6\n",
      "REUSE= NO\n",
      "190,15\t9\tSEQ.pnt created.\n",
      "Sequence length is 419\n",
      "190,200,RNA free energy files (version 2.3) at 20 degrees created.\n",
      "/home/jupyter/plant_microRNA_prediction200,\n",
      "7\t/home/jupyter/plant_microRNA_prediction/secondary_structure/mfold/AMWY020003061|3061-3478|201-218|+210,\n",
      "10,20,30,40,210,220,18\t4\t18\t4\tmfold version 3.6\n",
      "REUSE= NO\n",
      "50,60,70,220,230,SEQ.pnt created.\n",
      "Sequence length is 418\n",
      "80,90,230,12\t6\t8\t14\tRNA free energy files (version 2.3) at 20 degrees created.\n",
      "100,110,240,240,3\t7\tSuboptimal foldings created.\n",
      "Energy dot plot created.\n",
      "10,20,30,40,120,250,260,250,10\t50,60,70,8\t\n",
      "Structure plots generated.\n",
      "All done.\n",
      "130,270,260,16\t80,90,100,140,280,290,270,280,110,120,150,5\t300,5\t290,300,19\t19\t130,140,310,160,310,7\t150,9\t320,330,340,170,320,330,/home/jupyter/plant_microRNA_prediction13\t\n",
      "/home/jupyter/plant_microRNA_prediction/secondary_structure/mfold/AMWY020003291|3585-4004|201-220|+\n",
      "4\t15\t350,360,180,340,350,160,370,380,390,400,410,360,370,380,390,400,190,8\t9\tmfold version 3.6\n",
      "REUSE= NO\n",
      "170,180,17\t \n",
      "End of Fill\n",
      "Save file created using nafold.\n",
      "Minimum folding energy is -133.70 kcal/mol.\n",
      "Energy increment is 6.68 kcal/mol.\n",
      "410, \n",
      "End of Fill\n",
      "Save file created using nafold.\n",
      "Minimum folding energy is -106.90 kcal/mol.\n",
      "Energy increment is 5.34 kcal/mol.\n",
      "200,11\tSEQ.pnt created.\n",
      "Sequence length is 420\n",
      "190,210,RNA free energy files (version 2.3) at 20 degrees created.\n",
      "6\t200,H-num file created from plot file.\n",
      "H-num file created from plot file.\n",
      "220,1\t10,20,30,40,6\t210,1,2,3,4,5,6,1,2,3,4,50,60,70,220,7,8,9,10,11,5,6, \n",
      "230,8\t\n",
      "Structure plots generated.\n",
      "All done.\n",
      "1\t14\t80,90,100,16\t230,12, \n",
      "240,5\t110,240,9\t\n",
      "Structure plots generated.\n",
      "All done.\n",
      "250,120,130,250,12\t260,18\t140,260,Suboptimal foldings created.\n",
      "Energy dot plot created.\n",
      "1\t270,7\t150,160,270,/home/jupyter/plant_microRNA_prediction\n",
      "/home/jupyter/plant_microRNA_prediction/secondary_structure/mfold/AMWY020002791|4721-5139|201-219|+\n",
      "280,7\t170,280,290,Suboptimal foldings created.\n",
      "Energy dot plot created.\n",
      "20\t290,180,mfold version 3.6\n",
      "REUSE= NO\n",
      "300,310,/home/jupyter/plant_microRNA_prediction10\t\n",
      "/home/jupyter/plant_microRNA_prediction/secondary_structure/mfold/AMWY020003341|4703-5044|201-220|+\n",
      "20\t15\t9\t300,310,17\t190,200,SEQ.pnt created.\n",
      "Sequence length is 419\n",
      "320,330,6\t320,330,mfold version 3.6\n",
      "REUSE= NO\n",
      "RNA free energy files (version 2.3) at 20 degrees created.\n",
      "340,350,\n",
      "Structure plots generated.\n",
      "All done.\n",
      "340,350,SEQ.pnt created.\n",
      "Sequence length is 342\n",
      "210,10,20,30,360,370,380,19\t360,370,220,RNA free energy files (version 2.3) at 20 degrees created.\n",
      "13\t40,50,60,70,2\t390,400,410, \n",
      "End of Fill\n",
      "Save file created using nafold.\n",
      "8\t380,390,400,410, \n",
      "End of Fill\n",
      "10,20,30,40,230,80,90,100,Minimum folding energy is -135.70 kcal/mol.\n",
      "Energy increment is 6.79 kcal/mol.\n",
      "Save file created using nafold.\n",
      "Minimum folding energy is -138.70 kcal/mol.\n",
      "Energy increment is 6.93 kcal/mol.\n",
      "50,60,70,240,250,110,H-num file created from plot file.\n",
      "8\t/home/jupyter/plant_microRNA_prediction\n",
      "11\t80,90,100,120,130,260,21\t1,2,3,4,21\t16\t\n",
      "Structure plots generated.\n",
      "All done.\n",
      "H-num file created from plot file.\n",
      "140,110,120,270,7\t18\t5,6,7,8,1,2,3,4,5,130,140,280,150,9,10,11,12, \n",
      "6,7,8,9,10,150,290,160,170,3\t11,12,13,14,9\t1\t160,300,310,180,15,16,17,18,/home/jupyter/plant_microRNA_prediction170,180,190,\n",
      "320,330,/home/jupyter/plant_microRNA_prediction/secondary_structure/mfold/AMWY020002961|1232-1655|201-224|+14\t\n",
      "19,20, \n",
      "9\t190,200,340,350,200,mfold version 3.6\n",
      "REUSE= NO\n",
      "17\t22\t210,360,370,380,210,12\t22\t220,230,220,390,400,410, \n",
      "End of Fill\n",
      "Save file created using nafold.\n",
      "Minimum folding energy is -126.10 kcal/mol.\n",
      "19\tSuboptimal foldings created.\n",
      "Energy dot plot created.\n",
      "10\t8\tSEQ.pnt created.\n",
      "Sequence length is 424\n",
      "240,250,230,Energy increment is 6.30 kcal/mol.\n",
      "RNA free energy files (version 2.3) at 20 degrees created.\n",
      "4\t260,270,280,240,250,H-num file created from plot file.\n",
      "10,20,30,40,50,290,300,310,320,330,260,1,2,3,4,5,\n",
      "Structure plots generated.\n",
      "All done.\n",
      "20\t60,70,80,15\t340, \n",
      "End of Fill\n",
      "Save file created using nafold.\n",
      "Minimum folding energy is -128.20 kcal/mol.\n",
      "270,6,7,8,9,\n",
      "Structure plots generated.\n",
      "All done.\n",
      "90,100,Energy increment is 6.41 kcal/mol.\n",
      "280,10,11,12,13,23\t18\t110,120,Suboptimal foldings created.\n",
      "Energy dot plot created.\n",
      "10\t1\tH-num file created from plot file.\n",
      "290,300,14,15,16, \n",
      "23\t130,140,1,2,3,4,5,6,7,8,310,320,9\t150,/home/jupyter/plant_microRNA_prediction9,10,11,12,13,14,1\t\n",
      "11\t/home/jupyter/plant_microRNA_prediction/secondary_structure/mfold/AMWY020003221|714-1132|201-219|-330,340,350,\n",
      "5\t160,/home/jupyter/plant_microRNA_prediction15,16,17,18,19, \n",
      "\n",
      "360,370,380,170,mfold version 3.6\n",
      "REUSE= NO\n",
      "390,400,410, \n",
      "End of Fill\n",
      "Save file created using nafold.\n",
      "21\t16\t180,SEQ.pnt created.\n",
      "Sequence length is 419\n",
      "Minimum folding energy is -125.80 kcal/mol.\n",
      "Energy increment is 6.29 kcal/mol.\n",
      "190,200,RNA free energy files (version 2.3) at 20 degrees created.\n",
      "H-num file created from plot file.\n",
      "19\t11\t210,2\t10,20,30,40,50,1,2,3,4,24\t220,60,70,80,Suboptimal foldings created.\n",
      "Energy dot plot created.\n",
      "10\t24\t\n",
      "Structure plots generated.\n",
      "All done.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "5,6,7,8,9,10,230,20\tSuboptimal foldings created.\n",
      "Energy dot plot created.\n",
      "90,100,110,12\t11,12,13,14,15,16,6\t240,10\t120,17,18,19, \n",
      "250,130,140,22\t260,150,17\t/home/jupyter/plant_microRNA_prediction270,280,\n",
      "160,3\t12\t170,25\t1\t11\t290,300,25\t180,2\t310,190,\n",
      "Structure plots generated.\n",
      "All done.\n",
      "11\t1\t320,330,340,200,23\t350,360,370,210,380,390,400,410,420, \n",
      "End of Fill\n",
      "Save file created using nafold.\n",
      "Suboptimal foldings created.\n",
      "Energy dot plot created.\n",
      "10\t220,13\tMinimum folding energy is -155.80 kcal/mol.\n",
      "Energy increment is 7.79 kcal/mol.\n",
      "18\t230,4\t12\tH-num file created from plot file.\n",
      "26\t20\t240,/home/jupyter/plant_microRNA_prediction\n",
      "2\t1,2,3,4,5,250,260,12\t3\t2\t6,7,8,9,10,270,24\t11,12,13, \n",
      "280,290,300,310,11\t320,19\t27\t21\t330,340,13\t14\t350,360,5\t370,380,390,400,410, \n",
      "End of Fill\n",
      "13\t3\tSave file created using nafold.\n",
      "Minimum folding energy is -159.00 kcal/mol.\n",
      "Energy increment is 7.95 kcal/mol.\n",
      "4\t3\tSuboptimal foldings created.\n",
      "Energy dot plot created.\n",
      "10\t2\tH-num file created from plot file.\n",
      "12\t22\t1\t1,2,3,4,5,6,14\t28\t7,8,9,10,11,12,15\t6\t13,14,15,16,17,14\t5\t4\t18,19,20,21,4\t22,23,3\t11\t24, \n",
      "13\t2\t15\t2\t16\t7\t29\t15\t5\t14\t5\t6\t12\t3\t16\t4\t3\t2\t17\t16\t8\tSuboptimal foldings created.\n",
      "6\tEnergy dot plot created.\n",
      "10\t15\t6\t7\t5\t4\t1\t13\t4\t9\t18\t30\t7\t17\t7\t11\t16\t5\t8\t6\t2\t1\t5\t19\t\n",
      "Structure plots generated.\n",
      "All done.\n",
      "3\t18\t8\t7\t12\t6\t8\t17\t2\t6\t9\t3\t/home/jupyter/plant_microRNA_prediction\n",
      "/home/jupyter/plant_microRNA_prediction/secondary_structure/mfold/AMWY020003261|6422-6841|201-220|-\n",
      "4\tmfold version 3.6\n",
      "REUSE= NO\n",
      "1\tSEQ.pnt created.\n",
      "Sequence length is 420\n",
      "9\t9\t13\tRNA free energy files (version 2.3) at 20 degrees created.\n",
      "19\t8\t7\t7\t10,20,30,40,50,60,3\t70,80,4\t\n",
      "Structure plots generated.\n",
      "All done.\n",
      "18\t90,100,110,120,\n",
      "Structure plots generated.\n",
      "All done.\n",
      "5\t20\t130,\n",
      "Structure plots generated.\n",
      "All done.\n",
      "9\t1\t14\t140,150,8\t/home/jupyter/plant_microRNA_prediction\n",
      "/home/jupyter/plant_microRNA_prediction/secondary_structure/mfold/AMWY020002901|3469-3886|201-218|-\n",
      "160,8\t170,4\tmfold version 3.6\n",
      "REUSE= NO\n",
      "5\t/home/jupyter/plant_microRNA_prediction\n",
      "/home/jupyter/plant_microRNA_prediction/secondary_structure/mfold/AMWY020002901|6817-7233|201-217|-\n",
      "180,19\tSEQ.pnt created.\n",
      "Sequence length is 418\n",
      "2\t/home/jupyter/plant_microRNA_prediction\n",
      "/home/jupyter/plant_microRNA_prediction/secondary_structure/mfold/AMWY020003061|6941-7355|201-215|+\n",
      "190,200,6\tmfold version 3.6\n",
      "REUSE= NO\n",
      "RNA free energy files (version 2.3) at 20 degrees created.\n",
      "210,SEQ.pnt created.\n",
      "Sequence length is 417\n",
      "mfold version 3.6\n",
      "REUSE= NO\n",
      "10,20,30,40,50,60,2\t15\t9\t220,RNA free energy files (version 2.3) at 20 degrees created.\n",
      "SEQ.pnt created.\n",
      "Sequence length is 415\n",
      "70,80,\n",
      "Structure plots generated.\n",
      "All done.\n",
      "230,240,10,20,30,40,50,60,RNA free energy files (version 2.3) at 20 degrees created.\n",
      "90,100,9\t250,70,80,10,20,30,40,50,60,110,120,1\t5\t260,90,100,70,80,90,130,6\t270,110,120,100,110,140,3\t7\t/home/jupyter/plant_microRNA_prediction\n",
      "280,290,/home/jupyter/plant_microRNA_prediction/secondary_structure/mfold/AMWY020003071|892-1309|201-218|-\n",
      "130,120,130,150,300,140,150,140,150,3\t160,16\tmfold version 3.6\n",
      "REUSE= NO\n",
      "\n",
      "Structure plots generated.\n",
      "All done.\n",
      "310,320,160,160,170,170,SEQ.pnt created.\n",
      "Sequence length is 418\n",
      "\n",
      "Structure plots generated.\n",
      "All done.\n",
      "330,170,180,190,180,RNA free energy files (version 2.3) at 20 degrees created.\n",
      "6\t340,350,360,7\t180,190,10,20,30,40,50,60,2\t370,380,390,400,410, \n",
      "End of Fill\n",
      "70,200,210,190,200,200,/home/jupyter/plant_microRNA_prediction\n",
      "/home/jupyter/plant_microRNA_prediction/secondary_structure/mfold/AMWY020003291|25-446|201-222|+\n",
      "Save file created using nafold.\n",
      "Minimum folding energy is -134.40 kcal/mol.\n",
      "Energy increment is 6.72 kcal/mol.\n",
      "8\t4\t220,80,90,100,mfold version 3.6\n",
      "REUSE= NO\n",
      "210,220,230,240,210,110,/home/jupyter/plant_microRNA_prediction\n",
      "/home/jupyter/plant_microRNA_prediction/secondary_structure/mfold/AMWY020002811|1788-2204|201-217|+\n",
      "SEQ.pnt created.\n",
      "Sequence length is 422\n",
      "4\tH-num file created from plot file.\n",
      "1,2,3,230,250,260,120,130,220,RNA free energy files (version 2.3) at 20 degrees created.\n",
      "10,17\t4,5,6,7,8,9,240,270,230,140,mfold version 3.6\n",
      "REUSE= NO\n",
      "20,30,40,50,60,3\t10,11,12,13,14,15,16,250,280,150,SEQ.pnt created.\n",
      "Sequence length is 417\n",
      "240,70,80,90,7\t8\t17,18,19,20,21,290,260,160,170,RNA free energy files (version 2.3) at 20 degrees created.\n",
      "10,20,250,260,100,110,22,23,24,300,310,320,270,180,270,30,40,50,60,5\t120,130,25, \n",
      "330,340,350,190,200,280,70,80,90,5\t9\t140,150,360,370,380,390,400,280,100,110,290,160,170,410, \n",
      "End of Fill\n",
      "Save file created using nafold.\n",
      "Minimum folding energy is -118.00 kcal/mol.\n",
      "Energy increment is 5.90 kcal/mol.\n",
      "290,300,210,120,130,300,310,18\t180,310,220,140,320,330,4\tH-num file created from plot file.\n",
      "190,230,320,330,150,160,340,350,9\t8\t1,2,3,4,5,6,7,200,240,340,350,360,170,360,370,380,390,400,8,9,10,\n",
      "Structure plots generated.\n",
      "All done.\n",
      "210,250,260,370,380,390,400,410,180,410, \n",
      "End of Fill\n",
      "Save file created using nafold.\n",
      "Minimum folding energy is -114.90 kcal/mol.\n",
      "Energy increment is 5.75 kcal/mol.\n",
      "6\t11,12, \n",
      "270,220, \n",
      "End of Fill\n",
      "Save file created using nafold.\n",
      "Minimum folding energy is -118.80 kcal/mol.\n",
      "Energy increment is 5.94 kcal/mol.\n",
      "190,6\t280,290,230,200,H-num file created from plot file.\n",
      "1,H-num file created from plot file.\n",
      "240,300,2,3,4,5,6,7,5\t19\tSuboptimal foldings created.\n",
      "Energy dot plot created.\n",
      "10\t1,2,3,4,5,6,250,260,210,310,320,8,9,10,11,12,/home/jupyter/plant_microRNA_prediction\n",
      "/home/jupyter/plant_microRNA_prediction/secondary_structure/mfold/AMWY020002741|2930-3348|201-219|-\n",
      "7,8,9,10,11,12,270,220,330,340,350,\n",
      "Structure plots generated.\n",
      "13, \n",
      "All done.\n",
      "Suboptimal foldings created.\n",
      "Energy dot plot created.\n",
      "10\t230,280,290,13,14,15,360,370,380,390,400,9\tmfold version 3.6\n",
      "REUSE= NO\n",
      "300,240,16, \n",
      "7\t410, \n",
      "End of Fill\n",
      "Save file created using nafold.\n",
      "Minimum folding energy is -136.10 kcal/mol.\n",
      "Energy increment is 6.80 kcal/mol.\n",
      "SEQ.pnt created.\n",
      "Sequence length is 419\n",
      "7\t310,320,250,RNA free energy files (version 2.3) at 20 degrees created.\n",
      "1\t330,340,350,260,H-num file created from plot file.\n",
      "10,20,30,40,50,/home/jupyter/plant_microRNA_prediction\n",
      "/home/jupyter/plant_microRNA_prediction/secondary_structure/mfold/AMWY020003321|2820-3237|201-218|-\n",
      "Suboptimal foldings created.\n",
      "Energy dot plot created.\n",
      "10\t360,370,380,270,1,2,3,4,5,6,7,60,70,80,6\t280,290,390,400,410,420, \n",
      "End of Fill\n",
      "Save file created using nafold.\n",
      "8,9,10,11,12,13,90,100,11\tmfold version 3.6\n",
      "REUSE= NO\n",
      "Minimum folding energy is -119.60 kcal/mol.\n",
      "Energy increment is 5.98 kcal/mol.\n",
      "300,310,14,15,16,17,18,19,110,\n",
      "Structure plots generated.\n",
      "All done.\n",
      "SEQ.pnt created.\n",
      "Sequence length is 418\n",
      "Suboptimal foldings created.\n",
      "Energy dot plot created.\n",
      "10\t320,330,20,21, \n",
      "120,130,11\tRNA free energy files (version 2.3) at 20 degrees created.\n",
      "10,20,8\t8\tH-num file created from plot file.\n",
      "1,2,3,340,350,360,140,30,40,50,60,4,5,6,7,8,9,370,380,390,400,410, \n",
      "150,70,80,90,20\t10,11,12,13,14,15,End of Fill\n",
      "Save file created using nafold.\n",
      "Minimum folding energy is -148.80 kcal/mol.\n",
      "Energy increment is 7.44 kcal/mol.\n",
      "160,11\t100,110,120,/home/jupyter/plant_microRNA_prediction\n",
      "16,17,18,19,20,21,170,180,130,140,12\tH-num file created from plot file.\n",
      "22,23,24,25,26,27,190,7\t150,160,28,29,30,31,1,2,3,4,5,6,200,11\t9\t170,32,33,34,35,7,8,9,10,11,12,13,12\t9\t180, \n",
      "14,15,16,17,210,21\t190,200,Suboptimal foldings created.\n",
      "Energy dot plot created.\n",
      "10\t18,19,20,220,21, \n",
      "230,12\t210,13\t240,8\t220,250,12\t230,240,\n",
      "Structure plots generated.\n",
      "All done.\n",
      "260,\n",
      "Structure plots generated.\n",
      "All done.\n",
      "250,260,1\t270,270,280,22\t11\t280,290,290,300,13\t14\t300,310,Suboptimal foldings created.\n",
      "9\t310,/home/jupyter/plant_microRNA_prediction\n",
      "/home/jupyter/plant_microRNA_prediction/secondary_structure/mfold/AMWY020003281|5599-6016|201-218|+\n",
      "320,Energy dot plot created.\n",
      "10\t320,330,/home/jupyter/plant_microRNA_prediction\n",
      "Suboptimal foldings created.\n",
      "Energy dot plot created.\n",
      "330,340,mfold version 3.6\n",
      "REUSE= NO\n",
      "13\t340,350,10\t350,360,370,SEQ.pnt created.\n",
      "Sequence length is 418\n",
      "360,370,380,390,380,390,400,410, \n",
      "End of Fill\n",
      "RNA free energy files (version 2.3) at 20 degrees created.\n",
      "400,410, \n",
      "End of Fill\n",
      "Save file created using nafold.\n",
      "12\tSave file created using nafold.\n",
      "Minimum folding energy is -144.30 kcal/mol.\n",
      "Energy increment is 7.21 kcal/mol.\n",
      "2\t10,20,30,40,50,23\t1\tMinimum folding energy is -147.60 kcal/mol.\n",
      "Energy increment is 7.38 kcal/mol.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "15\tH-num file created from plot file.\n",
      "\n",
      "Structure plots generated.\n",
      "All done.\n",
      "60,70,80,H-num file created from plot file.\n",
      "1,2,11\t1,2,3,4,90,100,3,4,5,6,7,8,9,5,6,7,8,9,10,14\t110,120,10,11, \n",
      "11\t11,12,13,14,15,16,130,140,3\t17,18,19,20,150,160,/home/jupyter/plant_microRNA_prediction\n",
      "21, \n",
      "170,13\t16\t2\t24\t180,12\tSuboptimal foldings created.\n",
      "Energy dot plot created.\n",
      "10\t190,15\t200,210,220,12\t4\t230,3\t14\t17\t240,2\t11\t250,Suboptimal foldings created.\n",
      "Energy dot plot created.\n",
      "10\t260,270,13\t13\t280,16\t290,300,18\t310,320,5\t15\t330,340,4\t3\t350,360,370,1\t14\t380,390,400,410, \n",
      "End of Fill\n",
      "Save file created using nafold.\n",
      "Minimum folding energy is -143.70 kcal/mol.\n",
      "Energy increment is 7.18 kcal/mol.\n",
      "11\t1\t14\tH-num file created from plot file.\n",
      "16\t19\t1,2,3,4,5,6\t6,7,8,9,10,5\t4\t11,12,13, \n",
      "2\t15\t2\t15\t12\t1\t17\t6\t7\t5\tSuboptimal foldings created.\n",
      "Energy dot plot created.\n",
      "10\t3\t16\t3\t18\t16\t13\t20\t7\t8\t6\t11\t4\t17\t17\t4\t21\t19\t8\t9\t14\t5\t12\t7\t18\t18\t22\t9\t5\t1\t15\t\n",
      "Structure plots generated.\n",
      "All done.\n",
      "8\t13\t6\t23\t19\t19\t6\t\n",
      "Structure plots generated.\n",
      "All done.\n",
      "/home/jupyter/plant_microRNA_prediction\n",
      "16\t9\t20\t1\t24\t1\t1\t7\t/home/jupyter/plant_microRNA_prediction\n",
      "7\t\n",
      "Structure plots generated.\n",
      "All done.\n",
      "17\t21\t2\t8\t20\t20\t8\t25\t/home/jupyter/plant_microRNA_prediction\n",
      "/home/jupyter/plant_microRNA_prediction/secondary_structure/mfold/AMWY020003231|1-360|186-205|-\n",
      "18\t2\tmfold version 3.6\n",
      "REUSE= NO\n",
      "SEQ.pnt created.\n",
      "Sequence length is 360\n",
      "2\t21\t21\t3\t9\tRNA free energy files (version 2.3) at 20 degrees created.\n",
      "10,20,9\t30,40,50,60,70,80,90,100,110,19\t120,130,3\t140,150,22\t2\t160,170,4\t\n",
      "Structure plots generated.\n",
      "All done.\n",
      "\n",
      "Structure plots generated.\n",
      "All done.\n",
      "3\t180,190,200,210,1\t220,230,4\t240,250,260,23\t270,280,290,3\t/home/jupyter/plant_microRNA_prediction/home/jupyter/plant_microRNA_prediction\n",
      "\n",
      "5\t300,310,320,330,340,350, \n",
      "End of Fill\n",
      "4\tSave file created using nafold.\n",
      "Minimum folding energy is -139.30 kcal/mol.\n",
      "Energy increment is 6.96 kcal/mol.\n",
      "H-num file created from plot file.\n",
      "20\t5\t1,2,3,4,5,6,7,8,9,24\t10,11,12,13,14,15,16,17,18,19,20,4\t21, \n",
      "6\t5\t6\t21\t25\t5\tSuboptimal foldings created.\n",
      "7\tEnergy dot plot created.\n",
      "10\t6\t7\t2\t6\t8\t11\t26\t7\t8\t3\t9\t12\t27\t7\t9\t\n",
      "Structure plots generated.\n",
      "All done.\n",
      "8\t4\t13\t28\t8\t\n",
      "Structure plots generated.\n",
      "All done.\n",
      "/home/jupyter/plant_microRNA_prediction\n",
      "9\t14\t5\t/home/jupyter/plant_microRNA_prediction29\t\n",
      "9\t\n",
      "Structure plots generated.\n",
      "All done.\n",
      "15\t6\t2\t\n",
      "Structure plots generated.\n",
      "All done.\n",
      "/home/jupyter/plant_microRNA_prediction\n",
      "16\t7\t/home/jupyter/plant_microRNA_prediction\n",
      "30\t17\t8\t31\t18\t9\t32\t19\t\n",
      "Structure plots generated.\n",
      "All done.\n",
      "33\t1\t/home/jupyter/plant_microRNA_prediction\n",
      "34\t20\t35\t21\t3\t2\t4\t3\t5\t4\t6\t5\t7\t6\t8\t7\t9\t8\t\n",
      "Structure plots generated.\n",
      "All done.\n",
      "9\t/home/jupyter/plant_microRNA_prediction\n",
      "\n",
      "Structure plots generated.\n",
      "All done.\n",
      "/home/jupyter/plant_microRNA_prediction\n"
     ]
    }
   ],
   "source": [
    "remove_lock = False\n",
    "def run_mfold(tag):\n",
    "    tag = reformat(tag)\n",
    "    %cd {base + tag}\n",
    "    !mfold  SEQ=\"SEQ.FASTA\" T=20     \n",
    "    if(not remove_lock):\n",
    "        !find . -not -name \"*.ct\" -not -name \"*.pdf\" -not -name \"*SEQ.FASTA\" -not -type d -delete\n",
    "    %cd ../../..\n",
    "\n",
    "if __name__ == '__main__':        \n",
    "    pool = mp.Pool(mp.cpu_count())  \n",
    "    pool.map(run_mfold, df['tag'].iloc[:200])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-11-06T15:12:52.776002Z",
     "start_time": "2021-11-06T15:10:32.633727Z"
    },
    "hidden": true
   },
   "outputs": [],
   "source": [
    "'''\n",
    "base = \"secondary_structure/mfold/\"\n",
    "for directory in glob.glob(f\"{base}*\"):    \n",
    "    tag = directory[len(base):]\n",
    "    ct_files = glob.glob(f'{directory}/*.ct')        \n",
    "    try:\n",
    "        ct_files.remove(f'{base}{tag}/SEQ.ct')\n",
    "    except:\n",
    "        print(directory)\n",
    "        print(ct_files)\n",
    "        print(\"*****************\")\n",
    "    for file in ct_files:        \n",
    "        shutil.copy(file, './1.ct')\n",
    "        #dot = ct2dot_bracket('./1.ct')\n",
    "        #dot = dot.split('\\n')\n",
    "        #with open('./2.ct', 'w') as stream:\n",
    "            #stream.write(bracket_to_ct(tag, dot[0] , dot[1] , \"(0)\"))        \n",
    "        #ct1 = '\\n'.join(reformatCT('./1.ct').split('\\n')[1:])\n",
    "        #ct2 = '\\n'.join(reformatCT('./2.ct').split('\\n')[1:])\n",
    "        #if(ct1 != ct2):\n",
    "            #print(file)\n",
    "        ct = reformatCT('./1.ct')\n",
    "        [nucleotide, index, values] = get_ct_data(ct)        \n",
    "        #print(is_nested( index,  values))\n",
    "        if(not is_nested( index,  values)):\n",
    "            print(\"************\")             \n",
    "'''"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-11-06T15:10:27.456082Z",
     "start_time": "2021-11-06T15:10:27.450725Z"
    },
    "hidden": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "heading_collapsed": true,
    "hidden": true
   },
   "source": [
    "## Mxfold2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-11-06T16:20:40.729087Z",
     "start_time": "2021-11-06T16:20:40.725251Z"
    },
    "hidden": true
   },
   "outputs": [],
   "source": [
    "#!wget https://github.com/keio-bioinformatics/mxfold2/releases/download/v0.1.1/mxfold2-0.1.1.tar.gz\n",
    "#!pip3 install mxfold2-0.1.1.tar.gz\n",
    "#!rm mxfold2-0.1.1.tar.gz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-11-06T16:22:35.682900Z",
     "start_time": "2021-11-06T16:20:48.694439Z"
    },
    "hidden": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      ">AMWY02059828.1:2832-3256(+)\n",
      "AAAGAATCAGCAATGGAAAAATAACCGGTTCTTAATTCAGcataacaaattattcaattataatatagcTGTAAAAGAAATCTAAGTCTATTTGATATAGATCGGAATTTacgcaaattaaaaatttccaaataaGCAGTTCCGACCTGAGATCTGAACCGAAAACGCAAGATCCATCTAAACTCTCACCTCGGTCTCCGATTCAGTTGATGCAAGGCGGGATCCAATTCGCCTTTTCATTCAATTACATTCACCAATAACAGCTCGCCATCtggcttttaataaaaagttgcCAATCGGTTCCCGACCTGCACCAAGCGAATTAGAGACCGCCGGTAACTGAATCATTCTACATTAATCCCCGACTCCTCCTTTTACACATAGCAACTTCGCCCAAGAagactaaaaagaaaaggaagctAAC\n",
      "........................((((...........((..........................))..........................................................................................................................(((...(((...............((((..........)))).................................(((....))).....................)))...)))..........................))))........................................................................................ (35.0)\n",
      ">AMWY02004761.1:1853-2277(+)\n",
      "actaataatgCATGGCCATATATATCAAATCTACCATATgccatttaataattttccttttttcttcttctttctttttctctctaaataCCATATATAtgcccttcttcttcttctcataCATGAACAGTTGAGGGGAATGTTGTCTGGTTCGAGATCATTCTTCGGCGATCTCAAGTTTAATTCTCTTCTTCAGAATGATTTCGGACCAGGCTTCATTCCCCCCAACTtggtttaattgttaatatatattctcgaTCGATTCGATGATGTTATATTTATGCATGCAACGGCCTATAGCATTATGTGTTTTTTGGTTAGTTCTTCCTAATAATCTGCTGCTTGATCGAGTATTCAGGTTGATTTTCTCGACGTTTGTTGAATCATTTGATCCAAAAGGGGGAGGCAGAAAATGAGCCGCCGG\n",
      ".....................................................................................................................................................................(((((....)..............................................(((((.............................((.......))..........................((.....)).................................................................................................))))).(((.........))))))). (40.1)\n",
      ">AMWY02033394.1:469-893(-)\n",
      "CTCCCCTTGTCTACCATCCCCAACTAGCGAGAGAGACATTACCTACCTGAATAGAAGATCTCTCTCGAGCTCTCGagctctctctttttctataTCTCTGTCTCTTTGTGTCTCTGGAGCTTGTACTAACATTAATATCGTGCACCAGCAGCAGTTGAAGCTGCCAGCATGATCTAAACTTCCTTCTCTGTAAAGGATAGATCGGATCATGTGGTAGCTTCACCTGTTGATGGGATCACGAAAGCGCCCCTCTTACTACTCTACATTAATTCTTTCTCGTTATACAACCTCCCAGTAAGCATGCTTTCAAAACCAACTTGAGtaagttaatttgtttagcttttgtttttggctcttcctttactttaaattttctcatctgggtttttgttatatatatgtactgttttatatatgtattcct\n",
      "..................(((......(....)..................................................................................................................((.((.......)).))...................................................................(((....(....)...)))...................................................................................................................................)))........................................ (36.4)\n",
      ">AMWY02098706.1:0-407(-)\n",
      "TGGGGGCCAACGGCATTGGCGGCGGGTTTTTGGGCCGTCGTGGTTCGATTTGAGTGGAGGGAGGTCaagttaaaattgatatttgtgGTTAAAAGGGGTCAATTTATAGACGACTTGTGCAAGTATAACAACTCCACATGTAACTTGTAATATgccctttttttattctcttaccAAACTCGCTTTAAATATCGGCCGAGTGGAGTGGAGTGGAGTGGAGTGGAGCAtcttttctgatttttctATCTCATCTAATCTCATCACCTCCTTTTAAAAGACATCGTATTAaacacttcttttttttttttttgtttttaaatttttagaaaggCCAACTAAACACTTCTTTACCTTTACGTTATGTGACCAGCAAGCAAGCACATAATCCATACACTTCAAAATACTCT\n",
      ".(..((((..((((.....((((..........)))).((.....))...................................................................................................................................................))))..............................(......).............................................................................................))))..)....................................................................... (33.7)\n",
      ">AMWY02003241.1:1997-2420(+)\n",
      "AAAACAGATCCAACACAGGAACaagacaaaattatcaagtaaACAAGAGAAATGTTTTCCAAAACGGTGCGTACCGTGGGATCGAAAGGAGAATGCGAGGGAGTTCCGTTGTACGGATGAAACGAGCGCCTGCCCGCGTTTTGGACGCGAATTAGTATAGgctgagagagagagagagaaagaacgTGCGTTGGTCAAAGTGGAGTGGAGTGGAGTGGAGTGGAGTCGAGTTGATTTCGTTATTGGTATTATTTCAGTTTTCTTTACGAtttggagagagagagagagagagagatcgtTAATATAATTTCGGTTCAGTTCAGAAACGAAGGGCTGGCTCGCGGAAACTGGAAGGCAACTTCAGCCAGCAAATTACAACGCTGACGTTAACATATTGGTATTTTGCTTGCATTTTTCGcttct\n",
      ".........((......)).............................................((..(((...(....)..(....)......((..(......(((.....)))......)..))(((.(((((((.......)))..................................................................................................................................................................((..............))..)))).)))....(....).....(((.......)))............)))...))..................................... (42.1)\n",
      ">AMWY02103996.1:37-460(-)\n",
      "tattattcttatCTTCTTTTNCATATTTTTNGAANTCAGTATTggatttttcttcatataaatCTTTGTGTTCGATCAGTTAAAGCTTAATTCGTTGCATGGATAGATTGACAGTCGTTGTTTGCGGGTGTAGCATCATCAAGATTCACATGCAAATGCACGGTAGGGATGCGATAATCTCCGCCAAGGTAGTTTGAAAGTGAGAATCTTGATGATGCTGCATCAGCAATAAACGACTATAATAAGagtatatattcaaaatcaagCCTCGCCAGTGCCACTTTGCACTTAAGCATTCAGCAGTACTCTAAACAAGAAATGTGGTTCTTTTACTTNCCTTTCTGTTGCTGTCAGGTGGGTatatttccttattttcttgGCNAGTtctataaactatatatatatatatattattggaatTGG\n",
      "...........................................................................................................................(((((....((..(.....)........((....))...........))......).))))...................................((....))...........................................(((...((......))......((.....))........................................................((.......))..........))).......................................... (29.3)\n",
      ">AMWY02089009.1:1704-2127(+)\n",
      "TTCCTCAGTTCTATTTGTTCCTTTTTGAAGGTTTTGTCTCGGGAACAATACGTCATATTATGTTTGTTATGGAGGACGCAATATTTTAACAGTCGTTGTTCGCTGATGCAGTGTCATCAAGATTCACATTCAAGCTAAAGGCAGAATTAAGCAGATGCAGtgatatctatatatttctgCTTGCCTTTGGTTTTTGGAAGTGAGAATCTTGATGATGCTGCATCGGCGATAAACGACTATAACAAGGTATCGCAGCATAATCTCCCGGAACAGACGGATCTTGATTTTTCTGCCGTCGAATCTGTTAATTAAAGGTCNAAAGTGAGTGNATCCCagatcaatattatttttttaagaaaaaaaaaaaNNGNAATTATTTCAGCCTGTGCTTTTCatgtttttggtttctttttgaaaatatct\n",
      "........................................(((..(.....).................................................((....))........................((.....((........))....)).........................((..........))..........................((((......(...........)..................(((.......)))..............))))....................................)))......................................................................................... (32.9)\n",
      ">AMWY02083127.1:3090-3455(+)\n",
      "TTCCTCAGTTCTATTTGTTCCTTTTTGAAGGTTTTGTCTCGGGAACAATACGTCATATTATGTTTGTTATGGAGGACGCAATATTTTAACAGTCGTTGTTCGCTGATGCAGTGTCATCAAGATTCACATTCAAGCTAAAGGCAGAATTAAGCAGATGCAGtgatatctatatatttctgCTTGCCTTTGGTTTTTGGAAGTGAGAATCTTGATGATGCTGCATCGGCGATAAACGACTATAACAAGGTATCGCAGCATAATCTCCCGGAACAGACGGATCTTGATTTTTCTGCCGTCGAATCTGTTAATTAAAGGTCAAAGTGAGTGCATCCcagatcaatatatttttttaagaaaaaaaaaaa\n",
      "........................................(((..(.....).................................................((....))........................((.....((........))....)).........................((..........))..........................((((......(...........)..................(((.......)))..............))))...................................)))................................ (27.2)\n",
      ">AMWY02043062.1:414-837(-)\n",
      "tcttattattcttatcttcttttcatatttttgaatcAGTATTggatttttcttcatataaatCTTTGTGTTCGATCAGTTAAAGCTTAATTCGTTGCATGGATAGATTGACAGTCGTTGTTTGCGGGTGTAGCATCATCAAGATTCACATGCAAATGCACGGTAGGGATGCGATAATCTCCGCCAAGGTAGTTTGAAAGTGAGAATCTTGATGATGCTGCATCAGCAATAAACGACTATAATAAGagtatatattcaaaatcaagCCTCGCCAGTGCCACTTTGCACTTAAGCATTCAGCAGTACTCTAAACAAGAAATGTGGttcttttactttcctttCTGTTGCTGTCAGGTGGGTatatttccttattttcttgGCAGTtctataaactatatatatatatatattattggaatGGAT\n",
      "...........................................................................................................................(((((....((..(.....)........((....))...........))......).))))...................................((....))...........................................(((...((......))......((.....))........................................................((.......))..........))).......................................... (28.7)\n",
      ">AMWY02038760.1:1824-2247(+)\n",
      "tcttccttcaccTTCTAACCAGTTTCAACCTCcaagtaaataaaagttattgaaatgatttctttaatcttatttttgtaaatatatatattgattgattCCTGCAGAATCGAATCGATGAACAACAGTCCATCAATCAGTCATTGTTTGCTGATGCAGCATCATTAAGATTCACATGCGAATGGTTAGAAGCAAAGCAGTGAGAATCTTGATGATGCTGCATCGGCCATAAATGACTATATAACATCATCAGGATTTACGGATCTAGATCAATTTTGCTACCTTTGCTTCTGATACaggtgaaattatttttaatttgtcccCATCttgttctcttttatttgttagcCGACAGTAATTGAAAGTAGTATCTGGGgtctcttttattttccatgTTCATACTATTTCTCTTGCAGGTAGAGA\n",
      "....................................................................................................((.((.....((...))................................((....))....................((............))..........................((....))..................................................((.......))..............................(.((((.................................................)))).).................................)).))...... (28.5)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      ">AMWY02002487.1:2516-2939(+)\n",
      "tctctcttctttttcttccttaacTTTTCATCCAGCTTCAACCTCCATTTAGATCaaagttattgaatttttttttcatcttatttatgtaaatatatattgtttccTGCGGAAACGAATCCATGAACAACAGTCAATCAGTCATTGTTTGCTGATGCAGCGTCATCAAGATTCGCATGCTGATGGGTCGAGCAAAGCAGTGAGAATCTTGATGATGCTGCATCGGCCATAATTGACTATAtctcgtcatcatcatcatcatcatccagtTTCAACCTCCATGTAAATCaagttattgaattatttggtAAATAGATACTGATTCCCGCAGAATTGAATCAATGAACAACAGTCAATCAGTCATTGTTTGCTGATGCAGCATCATCAAGATTCACATGCGAATGGGTCGAACCAAAGCAGTGA\n",
      "............................................................................................................((((....(....)............................((....)).(((...........))).........((.((.((...((..................)).))..)).)).................................................................................................))))........................................((....))....................((....((......))...))..... (30.2)\n",
      ">AMWY02089812.1:0-419(-)\n",
      "TCtgtctatatttattttcttctcattcACTGTAGTAATTTAAGCCTATACAGTTCTGAGTTGACCNatttctttatataaagtTNATTTCGAGAAGTCTACNAAGGGAAAAAGTGAAGCTGCCAGCATGATCTAGCTTCGGTTAGGAAATAAGTTAAAGCTAAGGAAGAATGAATGAACCGGTGGACTAACCCTAACTAGGTCATGCTCTGACAGCCTCACTCCTTCCTATTTGGAGACCCAACAAATTCCAGATTATCCCTCAAATCTTTAAcaagataaagcaaaaaacAAAGGTATAAGATCTCTCTGTAagtatattcttttactttgtttttgaaatctCACGCATGCGTTCTCTTCTGATCTgatcacttttttcttttcccctttCTCTTCTATGTTGCTCGTTGATTTCC\n",
      ".........................................................................................................(((..............((.((........))...)).......................................((.(......))).................................((.....)).......................)))............................................................................................................................................................. (31.7)\n",
      ">AMWY02033396.1:16-439(-)\n",
      "TTCATCTgtctatattattttcttctcattcACTGTAGTAATTTAAGCCTATACAGTTCTGAGTTGAccatttctttatataaagttatttcGAGAAGTCTACAAGGGAAAAAGTGAAGCTGCCAGCATGATCTAGCTTCGGTTAGGAAATAAGTTAAAGCTAAGGAAGAATGAATGAACCGGTGGACTAACCCTAACTAGGTCATGCTCTGACAGCCTCACTCCTTCCTATTTGGAGACCCAACAAATTCCAGATTATCCCTCAAATCTTTAAcaagataaagcaaaaaacAAAGGTATAAGATCTCTCTGTAagtatattcttttactttgtttttgaaatctCACGCATGCGTTCTCTTCTGATCTgatcacttttttcttttcccctttCTCTTCTATGTTGCTCGTTGATTTCCTACT\n",
      ".........................................................................................................(((..............((.((........))...)).......................................((.(......))).................................((.....)).......................)))................................................................................................................................................................. (30.5)\n",
      "[W NNPACK.cpp:79] Could not initialize NNPACK! Reason: Unsupported hardware.\n",
      ">AMWY02048808.1:0-283(-)\n",
      "tatatatatatatatacacatatattattcatatgtctttgcttgtatatatataaatttggaaGCATTGCTAGCTATACGACGTTCTCAAGTGTaagagtaaataaatattaatgcattGTCTTGTATGGAAGAGAAAGAGGGTGGGTTTGAGCCAAGGATGACTTGCCGGTGTGTATAtgtacataatattaatatacatgtttTATTTGTGATCTGGGAAGTCTATCCTTGGCTCTTGTTCGTCCTCTTTCATCCCACATAAGACCTTGGAGATAAATTC\n",
      ".............................................................................................................................................(((.((.......))............................................................(..((........))..)......................)))........................ (15.1)\n",
      ">AMWY02067626.1:7483-7906(+)\n",
      "TTTATTTCCTATGCTTTAATGCGTCTGCTAACTTTCTTGCTTCAGGTTCTTTAAGGTTACCAGCATCAACCAATTGTATGATGTTTCCCTTTTGTGGGTTAAAAGCGTTAGGAAAATATATACGGACGTTTTGATACACAAGTGATCATTtgcatatgtttatttataaaaacgtGTATGCATGGAAGTATTAGATAGTATGCCTGGCTCCCTGTATGCCATTTGCTAAGCCCATCAGAAGATTGGTGGTCTTCGTGGATAGCGTGCGAGGAGCCATGCATGCCAATTTACTGTCATTTGCAGGTGtgtttttgattcttttgttttcattccTGGAAAATAAACTTAAGCATTTTCATCAAACTGCTGTGGACGAGGCCTGAGTTATCCCTCTCAACTTCAAAGAGTTCCTTGGTGATGAGG\n",
      ".......((...((......((....))..........................................................(((......)))...........................................................................................................(((.........)))....((...((((..(....).........(.....)............)).))...))......................................................................................))...))................................................... (36.1)\n",
      ">AMWY02064143.1:2949-3372(+)\n",
      "CATGAAGAATTTTGTTctggataaaaatttatctaagtTACTCATTCATTTTCAGGCATTTCTGGTTCACATATATCCATTTGTATACATTCAGTATCTTGTTCAATGACGCCGAACTTCATATATAAACGCTTGGATATATGCTATTAGATGTATTTACATGTGTTTATTACAAGATcatgcatatttatataagtatgTGCCTGGCTCCCTGTATGCCATTTGCAGAGCACATCGAATATCGATGGCCTCCGTGAATGGCGTATGAGGAGCCATGCATATGTCATGTATTATATCCATGCTTTTTGCAGGGGGTGTTTTCCTGTTTTTTCCAAAGACTAAGAACACAAAGGGCAGCATTGTAATTTTGTAAGATGCAGGTTCAGCCTTGTATTTAAGGTACAGTTAATCTCATCTTCGTCT\n",
      "......................................................(((......((...........))....................................................((..........)).........................................................(((.(((.........))).......................................(((.........)))......................((..((.....))..))((......)).............................)))..............................)))................................... (35.5)\n",
      ">AMWY02022832.1:1734-2156(-)\n",
      "tgcatgtaatattattatgcagTAAGTATAATTTGTTAGCGTTTGGTTTGTTAGATCTGTGTTGTATGAAATGTTTACGCCGATCTGTGTGGTGTAGAGTTGAAGATTTTGCTGCTAGAAGAATCGTAGATCTTGATTaacttgttttttgtttttggttttttctatttgaaaaagtaaaacgAGAAGGGTGAAGTCACTGGAGGCAGCGGTTCATCGATCTCTTCCTGGCAATTTTGTTGtgaaaataacacaaatacaCGAATCGATCGATAAACCTCTGCATCCAGCGCTTACTCTCCCtttcttctcaaatatttcgcctcttttgcttattttcttttattttttattttttatgttgtgttTGTGTGCTAAGAAATATTAGATTTGAAGTTTCTGTAATTAAGAGGAGaaatatt\n",
      "..........................................................................................................................................................................................................................................................................................((.....))................................................................................................................................... (26.7)\n",
      ">AMWY02099822.1:1487-1909(-)\n",
      "tttttttttttttcttctcttggtAAATTTTTGCGTGGTGTGNTTTTTTCCTATATATGGAAAGCATTATCTTCTGCTTTTTATNcattattcattatattttcttcttctttcttctctcattCTTGGAACATAGAAATTGACAGAAGAGAGTGAGCACACAAAGGGCACTTGTATGACTATACTTTTNGCCTTTGCGTGCTCACTTCTCTTTCTGTCAGCTTCCAGTGCCGGAATCTGGTCTTTCTCCGACCAGTTCTCCGCCTGTGTTTTGTTCCAAATTCATCTCTCAATACTTTCCCTTTTCCCTCCCTCTCTTTATCGCCTTCGTCGGAGTTGAAGAAGGAAGACAACACAACACACAGTTGACAACTACTCTAAGTTTGTGTGTAttctattttcctttctttctgtCTTTACtc\n",
      ".................................................((.......))..........................................................................................................(((.(......)............((....))......................((.......))(((...(.((.(.......).)).)....)))............................................................)))................................................................................................ (34.6)\n",
      ">AMWY02082313.1:5813-6235(+)\n",
      "GggtgtttttttttagtttcagaTCTGGGTAATTAGCTAGTCTTAATGACAGGAAATTAGTGTGAGTTAATGGCTAATAATAACATGTAATCatcaaagaaatttagaaaaaatatgaagaggAAAAACAAGAGAGCTACTGACAGAAGAGAGTGAGCACACGCAGGCAATTGTATGAGGAAATATCTTTGCGGGTGCGTGCTCACTTCTCTTTCTGTCAGCTTCATTTTCGTAAGTGTTTCCTCACaacagcaaaaacaaaaatacttcCAGAAAAAAGCTTACAGTAATTTCCTCTgatatctctctttctctagtACTTTCTTCCTTCCAGATGTCTCTTTACATCTCCCAGAAACAGCAGGCAGTACATAACATGATCATCAAAATCTCTTGgttattctcttttcttccaaaaccCT\n",
      "(((.......................(((............(.....)............................................................................................................((....))..........................((....))........................................................................................................................................................))).....................................(....)......................))). (27.2)\n",
      ">AMWY02072689.1:0-378(+)\n",
      "GTTTTTTCCTATATATGGAAAGCATTATCTTCTgctttttatcattattcattatattttcttcttctttcttctctcattCTTGGAACATAGAAATTGACAGAAGAGAGTGAGCACACAAAGGGCACTTGTATGACTATACTTTTGCCTTTGCGTGCTCACTTCTCTTTCTGTCAGCTTCCAGTGCCGGAATCTGGTCTTTCTCCGACCAGTTCTCCGCCTGTGTTTTGTTCCAAATTCATCTCTCAATACTTTCCCTTTTCCCTCCCTCTCTTTATCGCCTTCGTCGGAGTTGAAGAAGGAAGACAACACAACACACAGTTGACAACTACTCTAAGTTTGTGTGTAttctattttcctttctttctgtCTTTACtc\n",
      ".......((.......)).........................................................................................................(((.(......)...........((....))......................((.......))(((...(.((.(.......).)).)....)))............................................................)))................................................................................................ (30.3)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      ">AMWY02001968.1:1243-1665(+)\n",
      "CTACCTAAACTCCATGCATGGCTCGTGCTAGCTTTctggtttcttcttttttctttaagggcttattataaatttgcaGCAAGCCTAAACCCTTCTTAATTTCAAGATCTCTCTTcatttgattctttctttctttttctagggattcttcttcttcttcttcttcttgtttgctGCTGGTGTATGTTGGTTTGAGAGATTGAAGCTGCCAGCATGATCTGGTAATATGGAAcctaatattatacatatacatatctatatctatatatatagatagatagttttagatttactctttaattatattaattcctCCACCTATAGTTTTAGAGttactctttaattatattaattcttccACCTCTTTGTATAGATAGATAGAAACATATAGAAGGTCTTAGATTTCCTTTGCTTTTAGATCC\n",
      "...((....(.((..((...((....))...............................(((....................)))..........................................................................................)).)).)......))........................................................................................................................................................................................................................................ (27.0)\n",
      ">AMWY02039981.1:1214-1636(-)\n",
      "atttatgttttctattttataattaaaaaataaaaaaaaaataataaggtaatctctctctctttttcttttatattcatatgaGGTCTCCCTTAAACCACCCACTTTCCTCTATAATAGAGAGAAAacgaagagagaaagaaagagagaagaaaaagtctCTTTGGGTTTAATTGTTTCTACTGCTAGTAGAGGAAGCATCCAAAGGGATCGCATTGATCCTGAGCAAGCTTAATTTccatacttaaattttttttggatcaTGCTATCCCTTTGGATTCCTCCTTTGCTGGctcataattatgtataattattccAACATATTCATGTCATGATATGCCTGCatactcatcatcatcatcatcatcttcttcttcttcttcataacTCGGCAACAAGTTTCCGATGAGTATTTGCAGCCA\n",
      ".....................................................................................................................................................................................................((..((...(((...((......((...................................)).....))...)))...))...........((.(((............................................))).)).............................................(((..........)))..........))..... (27.5)\n",
      ">AMWY02071394.1:544-966(+)\n",
      "CCGGggaatcaaatgaaaaatgtacCAACAGTACTTCTGTAGTATACCGGCAAGTCATCCTTGGCTGCGTCCAACTCTCTACCTCTTCAGGCAAGACTCTTTTGACTAAGCTGCAGGCAGCCGTTTAATATCGATCGTATCTAGCTAGGTACTGAAAATTTTGCCATTCCAGTCTGACACGAGAAAGAGAGAGCCCCATGTAGCCAAGGACAAACTTGCCGGCAATTAAATAAGGAAACATTCAATGATTCCGGCCGGCAAGTCATCCTCGGCTGCACCCCACTCTCTACCTCATCGTGCAAGGCTCCAATGGCAGTACTACTTCTTTTGTATATATGGATGTTTTCTTGTTATGCTGTTGTAACTTGTATAAGCTACTACATATTATTGTAGTGcgaatcatatatatatatatatgtgtg\n",
      "..((((...(....)................................((((...........((..(.(.......).)..))..........................((.....)).))))...............................................................................(((..((.....(..((((((..............................))))))..)....))..)))....))))............................................................................................................................................. (42.3)\n",
      ">AMWY02097334.1:771-1193(-)\n",
      "CCCTCCGCTTTCTTTCTTGTCTTCTATAAATCAAACCCCCCACCCTCAACCCCCAACCCNAACTCCTCTTCCTGTATTTTCACATTCTCCCTCCATACCCaaacctagctagctagcttgcttgctcttcttcttctttcttcttcatattttaaCTTCCAAGCCGGTCCAAGTTTTCTTGATTAGTGCAGGTGGAGAGTTCCAAAGGGATCGCATTGATCTGATggcataaaattatacatacatatatgctTCTTATTCGATTGATTCAGTTCGATTTTATTGAtcgattttgtttcttattagaTCATGCGATCCCTTAGGAATTTCCCATCAATAGCTCTAAACTTACTCCTGCCGCTGCTGgtaagtaaatataattatatactattattaCAGATATATATTGACtaattattctt\n",
      ".......................................................................................................((..((..((..((......))............................................................................((...(((..(((......(....)................................................((...........))......................)))..)))...))...................................)).)).))....................................................... (35.1)\n"
     ]
    }
   ],
   "source": [
    "!mxfold2 predict ./extended.txt > secondary_structure/mxfold2_result.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 143,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-10-31T17:46:46.009981Z",
     "start_time": "2021-10-31T17:46:45.980600Z"
    },
    "hidden": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>tag</th>\n",
       "      <th>data</th>\n",
       "      <th>bracket</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AMWY02059828.1:2832-3256(+)</td>\n",
       "      <td>AAAGAATCAGCAATGGAAAAATAACCGGTTCTTAATTCAGcataac...</td>\n",
       "      <td>........................((((...........((........</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AMWY02004761.1:1853-2277(+)</td>\n",
       "      <td>actaataatgCATGGCCATATATATCAAATCTACCATATgccattt...</td>\n",
       "      <td>.................................................</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                           tag  \\\n",
       "0  AMWY02059828.1:2832-3256(+)   \n",
       "1  AMWY02004761.1:1853-2277(+)   \n",
       "\n",
       "                                                data  \\\n",
       "0  AAAGAATCAGCAATGGAAAAATAACCGGTTCTTAATTCAGcataac...   \n",
       "1  actaataatgCATGGCCATATATATCAAATCTACCATATgccattt...   \n",
       "\n",
       "                                             bracket  \n",
       "0  ........................((((...........((........  \n",
       "1  .................................................  "
      ]
     },
     "execution_count": 143,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = fasta_to_df('secondary_structure/mxfold2_result.txt')\n",
    "df = df.apply(lambda row: bracket_row(row) , axis=1)\n",
    "df.head(2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 145,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-10-31T17:48:37.583164Z",
     "start_time": "2021-10-31T17:48:37.308743Z"
    },
    "hidden": true
   },
   "outputs": [],
   "source": [
    "base = \"./secondary_structure/mxfold2/\"\n",
    "!rm -r {base}\n",
    "!mkdir -p {base}\n",
    "for index, row in df.iterrows():    \n",
    "    if(not os.path.exists(base + reformat(row['tag']))):\n",
    "        os.makedirs(base + reformat(row['tag']))        \n",
    "    tag = reformat(row['tag'])\n",
    "    with open(base + f\"{tag}/{tag}.ct\",'w') as file:\n",
    "        bracket = row['bracket'].split(' ')[0]\n",
    "        deltaG = row['bracket'].split(' ')[1]\n",
    "        ct = bracket_to_ct(row['tag'], row['data'], bracket, deltaG)\n",
    "        file.write(ct)    "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "heading_collapsed": true,
    "hidden": true
   },
   "source": [
    "## SPOT-RNA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-11-02T17:07:56.140509Z",
     "start_time": "2021-11-02T17:07:56.137052Z"
    },
    "hidden": true,
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "#!git clone https://github.com/jaswindersingh2/SPOT-RNA.git\n",
    "#%cd SPOT-RNA\n",
    "#!wget 'https://www.dropbox.com/s/dsrcf460nbjqpxa/SPOT-RNA-models.tar.gz' || wget -O SPOT-RNA-models.tar.gz 'https://app.nihaocloud.com/f/fbf3315a91d542c0bdc2/?dl=1'\n",
    "#!tar -xvzf SPOT-RNA-models.tar.gz && rm SPOT-RNA-models.tar.gz\n",
    "#!sudo apt-get install python3.6\n",
    "#!python3.6 -m pip install tensorflow==1.14.0 # or for gpu: tensorflow-gpu==1.14.0\n",
    "#! python3.6 -m pip install -r requirements.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-11-04T15:54:27.343924Z",
     "start_time": "2021-11-04T15:54:27.088380Z"
    },
    "hidden": true
   },
   "outputs": [],
   "source": [
    "base = \"./secondary_structure/spot_rna/\"\n",
    "!rm -r {base}\n",
    "!mkdir -p {base}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-11-04T16:06:02.696464Z",
     "start_time": "2021-11-04T16:02:34.125112Z"
    },
    "hidden": true,
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      ">> Opening FASTA file...\n",
      ">> Converting FASTA file from multiline to single line and writing to file.\n",
      ">> Done!\n",
      "\n",
      "Preparing tfr records file for SPOT-RNA:\n",
      "100%|█████████████████████████████████████████████| 1/1 [00:04<00:00,  4.21s/it]\n",
      "\n",
      "Predicting for SPOT-RNA model 0\n",
      "100%|█████████████████████████████████████████████| 1/1 [00:05<00:00,  5.23s/it]\n",
      "\n",
      "Predicting for SPOT-RNA model 1\n",
      "100%|█████████████████████████████████████████████| 1/1 [00:09<00:00,  9.51s/it]\n",
      "\n",
      "Predicting for SPOT-RNA model 2\n",
      "100%|█████████████████████████████████████████████| 1/1 [00:13<00:00, 13.10s/it]\n",
      "\n",
      "Predicting for SPOT-RNA model 3\n",
      "100%|█████████████████████████████████████████████| 1/1 [00:16<00:00, 16.44s/it]\n",
      "\n",
      "Predicting for SPOT-RNA model 4\n",
      "100%|█████████████████████████████████████████████| 1/1 [00:18<00:00, 18.18s/it]\n",
      "\n",
      "Post Processing and Saving Output\n",
      "\n",
      "Finished!\n",
      "\n",
      "Processsing Time 203.32813096046448 seconds\n"
     ]
    }
   ],
   "source": [
    "!python3.6 ./SPOT-RNA/SPOT-RNA.py  --inputs ./extended.txt  --outputs '{base}'  --cpu 32 --plots True"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-11-04T16:28:01.500658Z",
     "start_time": "2021-11-04T16:28:01.122310Z"
    },
    "hidden": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "rm: cannot remove './RNA_secondary_structure/spot_rna//*.bpseq': No such file or directory\n",
      "rm: cannot remove './RNA_secondary_structure/spot_rna//*.prob': No such file or directory\n",
      "\u001b[1;35mWARNING: \u001b[0m\u001b[1mbases 193 and 343 (GA) can't pair!\u001b[0m\n",
      "\u001b[1;35mWARNING: \u001b[0m\u001b[1mbases 194 and 342 (GG) can't pair!\u001b[0m\n",
      "\u001b[1;35mWARNING: \u001b[0m\u001b[1mbases 195 and 341 (UU) can't pair!\u001b[0m\n",
      "\u001b[1;35mWARNING: \u001b[0m\u001b[1mbases 196 and 340 (CC) can't pair!\u001b[0m\n",
      "\u001b[1;35mWARNING: \u001b[0m\u001b[1mbases 198 and 334 (CC) can't pair!\u001b[0m\n",
      "\u001b[1;35mWARNING: \u001b[0m\u001b[1mbases 201 and 333 (AC) can't pair!\u001b[0m\n",
      "\u001b[1;35mWARNING: \u001b[0m\u001b[1mbases 203 and 331 (UC) can't pair!\u001b[0m\n",
      "\u001b[1;35mWARNING: \u001b[0m\u001b[1mbases 204 and 330 (CC) can't pair!\u001b[0m\n",
      "\u001b[1;35mWARNING: \u001b[0m\u001b[1mbases 205 and 329 (AA) can't pair!\u001b[0m\n",
      "\u001b[1;35mWARNING: \u001b[0m\u001b[1mbases 206 and 328 (GG) can't pair!\u001b[0m\n",
      "\u001b[1;35mWARNING: \u001b[0m\u001b[1mbases 212 and 316 (GA) can't pair!\u001b[0m\n",
      "\u001b[1;35mWARNING: \u001b[0m\u001b[1mbases 213 and 315 (CC) can't pair!\u001b[0m\n",
      "\u001b[1;35mWARNING: \u001b[0m\u001b[1mbases 214 and 314 (AC) can't pair!\u001b[0m\n",
      "\u001b[1;35mWARNING: \u001b[0m\u001b[1mbases 215 and 313 (AA) can't pair!\u001b[0m\n",
      "\u001b[1;35mWARNING: \u001b[0m\u001b[1mbases 217 and 311 (GG) can't pair!\u001b[0m\n",
      "\u001b[1;35mWARNING: \u001b[0m\u001b[1mbases 218 and 305 (CC) can't pair!\u001b[0m\n",
      "\u001b[36mExternal loop\u001b[0m                           : \u001b[32m -371\u001b[0m\n",
      "\u001b[36mInterior loop\u001b[0m (  3, 32) \u001b[1mAU\u001b[0m; (  4, 31) \u001b[1mGC\u001b[0m: \u001b[32m -256\u001b[0m\n",
      "\u001b[36mInterior loop\u001b[0m (  4, 31) \u001b[1mGC\u001b[0m; (  5, 30) \u001b[1mAU\u001b[0m: \u001b[32m -294\u001b[0m\n",
      "\u001b[36mInterior loop\u001b[0m (  5, 30) \u001b[1mAU\u001b[0m; (  6, 29) \u001b[1mAU\u001b[0m: \u001b[32m -122\u001b[0m\n",
      "\u001b[36mHairpin  loop\u001b[0m (  6, 29) \u001b[1mAU\u001b[0m              : \u001b[32m  716\u001b[0m\n",
      "\u001b[36mInterior loop\u001b[0m ( 38, 71) \u001b[1mCG\u001b[0m; ( 39, 70) \u001b[1mAU\u001b[0m: \u001b[32m -255\u001b[0m\n",
      "\u001b[36mInterior loop\u001b[0m ( 39, 70) \u001b[1mAU\u001b[0m; ( 40, 69) \u001b[1mGC\u001b[0m: \u001b[32m -256\u001b[0m\n",
      "\u001b[36mInterior loop\u001b[0m ( 40, 69) \u001b[1mGC\u001b[0m; ( 41, 68) \u001b[1mCG\u001b[0m: \u001b[32m -403\u001b[0m\n",
      "\u001b[36mInterior loop\u001b[0m ( 41, 68) \u001b[1mCG\u001b[0m; ( 49, 64) \u001b[1mAU\u001b[0m: \u001b[32m  618\u001b[0m\n",
      "\u001b[36mInterior loop\u001b[0m ( 49, 64) \u001b[1mAU\u001b[0m; ( 50, 63) \u001b[1mUA\u001b[0m: \u001b[32m -155\u001b[0m\n",
      "\u001b[36mInterior loop\u001b[0m ( 50, 63) \u001b[1mUA\u001b[0m; ( 51, 62) \u001b[1mUA\u001b[0m: \u001b[32m -122\u001b[0m\n",
      "\u001b[36mInterior loop\u001b[0m ( 51, 62) \u001b[1mUA\u001b[0m; ( 52, 61) \u001b[1mAU\u001b[0m: \u001b[32m -165\u001b[0m\n",
      "\u001b[36mHairpin  loop\u001b[0m ( 52, 61) \u001b[1mAU\u001b[0m              : \u001b[32m  502\u001b[0m\n",
      "\u001b[36mInterior loop\u001b[0m ( 87,101) \u001b[1mUA\u001b[0m; ( 88,100) \u001b[1mCG\u001b[0m: \u001b[32m -294\u001b[0m\n",
      "\u001b[36mInterior loop\u001b[0m ( 88,100) \u001b[1mCG\u001b[0m; ( 89, 99) \u001b[1mUA\u001b[0m: \u001b[32m -256\u001b[0m\n",
      "\u001b[36mInterior loop\u001b[0m ( 89, 99) \u001b[1mUA\u001b[0m; ( 90, 98) \u001b[1mAU\u001b[0m: \u001b[32m -165\u001b[0m\n",
      "\u001b[36mHairpin  loop\u001b[0m ( 90, 98) \u001b[1mAU\u001b[0m              : \u001b[32m  573\u001b[0m\n",
      "\u001b[36mInterior loop\u001b[0m (151,174) \u001b[1mGC\u001b[0m; (152,173) \u001b[1mAU\u001b[0m: \u001b[32m -294\u001b[0m\n",
      "\u001b[36mInterior loop\u001b[0m (152,173) \u001b[1mAU\u001b[0m; (153,172) \u001b[1mUA\u001b[0m: \u001b[32m -155\u001b[0m\n",
      "\u001b[36mInterior loop\u001b[0m (153,172) \u001b[1mUA\u001b[0m; (154,171) \u001b[1mCG\u001b[0m: \u001b[32m -294\u001b[0m\n",
      "\u001b[36mInterior loop\u001b[0m (154,171) \u001b[1mCG\u001b[0m; (155,170) \u001b[1mUA\u001b[0m: \u001b[32m -256\u001b[0m\n",
      "\u001b[36mHairpin  loop\u001b[0m (155,170) \u001b[1mUA\u001b[0m              : \u001b[32m  505\u001b[0m\n",
      "\u001b[36mInterior loop\u001b[0m (193,343) \u001b[1mGA\u001b[0m; (194,342) \u001b[1mGG\u001b[0m: \u001b[32m  105\u001b[0m\n",
      "\u001b[36mInterior loop\u001b[0m (194,342) \u001b[1mGG\u001b[0m; (195,341) \u001b[1mUU\u001b[0m: \u001b[32m  105\u001b[0m\n",
      "\u001b[36mInterior loop\u001b[0m (195,341) \u001b[1mUU\u001b[0m; (196,340) \u001b[1mCC\u001b[0m: \u001b[32m  105\u001b[0m\n",
      "\u001b[36mInterior loop\u001b[0m (196,340) \u001b[1mCC\u001b[0m; (197,339) \u001b[1mUA\u001b[0m: \u001b[32m  -74\u001b[0m\n",
      "\u001b[36mInterior loop\u001b[0m (197,339) \u001b[1mUA\u001b[0m; (198,334) \u001b[1mCC\u001b[0m: \u001b[32m  513\u001b[0m\n",
      "\u001b[36mInterior loop\u001b[0m (198,334) \u001b[1mCC\u001b[0m; (201,333) \u001b[1mAC\u001b[0m: \u001b[32m  437\u001b[0m\n",
      "\u001b[36mInterior loop\u001b[0m (201,333) \u001b[1mAC\u001b[0m; (202,332) \u001b[1mUG\u001b[0m: \u001b[32m   74\u001b[0m\n",
      "\u001b[36mInterior loop\u001b[0m (202,332) \u001b[1mUG\u001b[0m; (203,331) \u001b[1mUC\u001b[0m: \u001b[32m   10\u001b[0m\n",
      "\u001b[36mInterior loop\u001b[0m (203,331) \u001b[1mUC\u001b[0m; (204,330) \u001b[1mCC\u001b[0m: \u001b[32m  105\u001b[0m\n",
      "\u001b[36mInterior loop\u001b[0m (204,330) \u001b[1mCC\u001b[0m; (205,329) \u001b[1mAA\u001b[0m: \u001b[32m  105\u001b[0m\n",
      "\u001b[36mInterior loop\u001b[0m (205,329) \u001b[1mAA\u001b[0m; (206,328) \u001b[1mGG\u001b[0m: \u001b[32m  105\u001b[0m\n",
      "\u001b[36mInterior loop\u001b[0m (206,328) \u001b[1mGG\u001b[0m; (207,327) \u001b[1mUA\u001b[0m: \u001b[32m  -74\u001b[0m\n",
      "\u001b[36mInterior loop\u001b[0m (207,327) \u001b[1mUA\u001b[0m; (208,326) \u001b[1mUG\u001b[0m: \u001b[32m -193\u001b[0m\n",
      "\u001b[36mInterior loop\u001b[0m (208,326) \u001b[1mUG\u001b[0m; (209,324) \u001b[1mGU\u001b[0m: \u001b[32m  395\u001b[0m\n",
      "\u001b[36mInterior loop\u001b[0m (209,324) \u001b[1mGU\u001b[0m; (210,323) \u001b[1mAU\u001b[0m: \u001b[32m -193\u001b[0m\n",
      "\u001b[36mInterior loop\u001b[0m (210,323) \u001b[1mAU\u001b[0m; (211,322) \u001b[1mUA\u001b[0m: \u001b[32m -155\u001b[0m\n",
      "\u001b[36mInterior loop\u001b[0m (211,322) \u001b[1mUA\u001b[0m; (212,316) \u001b[1mGA\u001b[0m: \u001b[32m  550\u001b[0m\n",
      "\u001b[36mInterior loop\u001b[0m (212,316) \u001b[1mGA\u001b[0m; (213,315) \u001b[1mCC\u001b[0m: \u001b[32m  105\u001b[0m\n",
      "\u001b[36mInterior loop\u001b[0m (213,315) \u001b[1mCC\u001b[0m; (214,314) \u001b[1mAC\u001b[0m: \u001b[32m  105\u001b[0m\n",
      "\u001b[36mInterior loop\u001b[0m (214,314) \u001b[1mAC\u001b[0m; (215,313) \u001b[1mAA\u001b[0m: \u001b[32m  105\u001b[0m\n",
      "\u001b[36mInterior loop\u001b[0m (215,313) \u001b[1mAA\u001b[0m; (216,312) \u001b[1mGC\u001b[0m: \u001b[32m -163\u001b[0m\n",
      "\u001b[36mInterior loop\u001b[0m (216,312) \u001b[1mGC\u001b[0m; (217,311) \u001b[1mGG\u001b[0m: \u001b[32m -187\u001b[0m\n",
      "\u001b[36mInterior loop\u001b[0m (217,311) \u001b[1mGG\u001b[0m; (218,305) \u001b[1mCC\u001b[0m: \u001b[32m  550\u001b[0m\n",
      "\u001b[36mInterior loop\u001b[0m (218,305) \u001b[1mCC\u001b[0m; (219,303) \u001b[1mGC\u001b[0m: \u001b[32m  254\u001b[0m\n",
      "\u001b[36mInterior loop\u001b[0m (219,303) \u001b[1mGC\u001b[0m; (221,302) \u001b[1mGU\u001b[0m: \u001b[32m  230\u001b[0m\n",
      "\u001b[36mInterior loop\u001b[0m (222,235) \u001b[1mAU\u001b[0m; (226,234) \u001b[1mAU\u001b[0m: \u001b[32m  475\u001b[0m\n",
      "\u001b[36mHairpin  loop\u001b[0m (226,234) \u001b[1mAU\u001b[0m              : \u001b[32m  558\u001b[0m\n",
      "\u001b[36mInterior loop\u001b[0m (268,275) \u001b[1mCG\u001b[0m; (269,274) \u001b[1mCG\u001b[0m: \u001b[32m -385\u001b[0m\n",
      "\u001b[36mHairpin  loop\u001b[0m (269,274) \u001b[1mCG\u001b[0m              : \u001b[32m  383\u001b[0m\n",
      "\u001b[36mInterior loop\u001b[0m (276,289) \u001b[1mCG\u001b[0m; (277,288) \u001b[1mUA\u001b[0m: \u001b[32m -256\u001b[0m\n",
      "\u001b[36mHairpin  loop\u001b[0m (277,288) \u001b[1mUA\u001b[0m              : \u001b[32m  619\u001b[0m\n",
      "\u001b[36mMulti    loop\u001b[0m (221,302) \u001b[1mGU\u001b[0m              : \u001b[32m  331\u001b[0m\n",
      "\u001b[36mInterior loop\u001b[0m (388,401) \u001b[1mCG\u001b[0m; (389,400) \u001b[1mUA\u001b[0m: \u001b[32m -256\u001b[0m\n",
      "\u001b[36mInterior loop\u001b[0m (389,400) \u001b[1mUA\u001b[0m; (390,399) \u001b[1mUA\u001b[0m: \u001b[32m -122\u001b[0m\n",
      "\u001b[36mInterior loop\u001b[0m (390,399) \u001b[1mUA\u001b[0m; (391,398) \u001b[1mCG\u001b[0m: \u001b[32m -294\u001b[0m\n",
      "\u001b[36mHairpin  loop\u001b[0m (391,398) \u001b[1mCG\u001b[0m              : \u001b[32m  200\u001b[0m\n",
      "AAAGAAUCAGCAAUGGAAAAAUAACCGGUUCUUAAUUCAGCAUAACAAAUUAUUCAAUUAUAAUAUAGCUGUAAAAGAAAUCUAAGUCUAUUUGAUAUAGAUCGGAAUUUACGCAAAUUAAAAAUUUCCAAAUAAGCAGUUCCGACCUGAGAUCUGAACCGAAAACGCAAGAUCCAUCUAAACUCUCACCUCGGUCUCCGAUUCAGUUGAUGCAAGGCGGGAUCCAAUUCGCCUUUUCAUUCAAUUACAUUCACCAAUAACAGCUCGCCAUCUGGCUUUUAAUAAAAAGUUGCCAAUCGGUUCCCGACCUGCACCAAGCGAAUUAGAGACCGCCGGUAACUGAAUCAUUCUACAUUAAUCCCCGACUCCUCCUUUUACACAUAGCAACUUCGCCCAAGAAGACUAAAAAGAAAAGGAAGCUAAC\n",
      "..((((......................)))).....((((.......((((........))))...))))...............((((.......)))).................................................(((((..............)))))..................((((((..(((((((((((((((((((.((...(.......))................................((....))((..........))............)).).....)))))).....))).)))))))))....)))))............................................((((......)))).......................\u001b[32m ( 29.73)\u001b[0m\n"
     ]
    }
   ],
   "source": [
    "!rm {base}/*.bpseq\n",
    "!rm {base}/*.prob\n",
    "for file in glob.glob(f\"{base}*.ct\"):    \n",
    "    f = file[len(base):-3] # .ct        \n",
    "    f = reformat(f)        \n",
    "    if(not os.path.exists(base + f)):\n",
    "        os.makedirs(base + f)  \n",
    "    header = reformatCT(file).split(\"\\n\")[0]    \n",
    "    with open(f\"{base}{f}.dot\", 'w') as stream:        \n",
    "        stream.write(ct2dot_bracket(file))\n",
    "    !RNAeval \"{base}{f}.dot\" -T 20 -v \n",
    "    #shutil.move(file, f\"{base}{f}/{f}.ct\")    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-11-04T17:28:02.481342Z",
     "start_time": "2021-11-04T17:28:02.464531Z"
    },
    "hidden": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "   424 dG =       0.0 AMWY02059828.1:2832-3256(+)\n",
      "     1 A      0      2      0       1\n",
      "     2 A      1      3      0       2\n",
      "     3 A      2      4     32       3\n",
      "     4 G      3      5     31       4\n",
      "     5 A      4      6     30       5\n",
      "     6 A      5      7     29       6\n",
      "     7 U      6      8      0       7\n",
      "     8 C      7      9      0       8\n",
      "     9 A      8     10      0       9\n",
      "    10 G      9     11      0      10\n",
      "    11 C     10     12      0      11\n",
      "    12 A     11     13      0      12\n",
      "    13 A     12     14      0      13\n",
      "    14 U     13     15      0      14\n",
      "    15 G     14     16      0      15\n",
      "    16 G     15     17      0      16\n",
      "    17 A     16     18      0      17\n",
      "    18 A     17     19      0      18\n",
      "    19 A     18     20      0      19\n",
      "    20 A     19     21      0      20\n",
      "    21 A     20     22      0      21\n",
      "    22 U     21     23      0      22\n",
      "    23 A     22     24      0      23\n",
      "    24 A     23     25      0      24\n",
      "    25 C     24     26      0      25\n",
      "    26 C     25     27      0      26\n",
      "    27 G     26     28      0      27\n",
      "    28 G     27     29      0      28\n",
      "    29 U     28     30      6      29\n",
      "    30 U     29     31      5      30\n",
      "    31 C     30     32      4      31\n",
      "    32 U     31     33      3      32\n",
      "    33 U     32     34      0      33\n",
      "    34 A     33     35      0      34\n",
      "    35 A     34     36      0      35\n",
      "    36 U     35     37      0      36\n",
      "    37 U     36     38      0      37\n",
      "    38 C     37     39     71      38\n",
      "    39 A     38     40     70      39\n",
      "    40 G     39     41     69      40\n",
      "    41 C     40     42     68      41\n",
      "    42 A     41     43      0      42\n",
      "    43 U     42     44      0      43\n",
      "    44 A     43     45      0      44\n",
      "    45 A     44     46      0      45\n",
      "    46 C     45     47      0      46\n",
      "    47 A     46     48      0      47\n",
      "    48 A     47     49      0      48\n",
      "    49 A     48     50     64      49\n",
      "    50 U     49     51     63      50\n",
      "    51 U     50     52     62      51\n",
      "    52 A     51     53     61      52\n",
      "    53 U     52     54      0      53\n",
      "    54 U     53     55      0      54\n",
      "    55 C     54     56      0      55\n",
      "    56 A     55     57      0      56\n",
      "    57 A     56     58      0      57\n",
      "    58 U     57     59      0      58\n",
      "    59 U     58     60      0      59\n",
      "    60 A     59     61      0      60\n",
      "    61 U     60     62     52      61\n",
      "    62 A     61     63     51      62\n",
      "    63 A     62     64     50      63\n",
      "    64 U     63     65     49      64\n",
      "    65 A     64     66      0      65\n",
      "    66 U     65     67      0      66\n",
      "    67 A     66     68      0      67\n",
      "    68 G     67     69     41      68\n",
      "    69 C     68     70     40      69\n",
      "    70 U     69     71     39      70\n",
      "    71 G     70     72     38      71\n",
      "    72 U     71     73      0      72\n",
      "    73 A     72     74      0      73\n",
      "    74 A     73     75      0      74\n",
      "    75 A     74     76      0      75\n",
      "    76 A     75     77      0      76\n",
      "    77 G     76     78      0      77\n",
      "    78 A     77     79      0      78\n",
      "    79 A     78     80      0      79\n",
      "    80 A     79     81      0      80\n",
      "    81 U     80     82      0      81\n",
      "    82 C     81     83      0      82\n",
      "    83 U     82     84      0      83\n",
      "    84 A     83     85      0      84\n",
      "    85 A     84     86      0      85\n",
      "    86 G     85     87      0      86\n",
      "    87 U     86     88    101      87\n",
      "    88 C     87     89    100      88\n",
      "    89 U     88     90     99      89\n",
      "    90 A     89     91     98      90\n",
      "    91 U     90     92      0      91\n",
      "    92 U     91     93      0      92\n",
      "    93 U     92     94      0      93\n",
      "    94 G     93     95      0      94\n",
      "    95 A     94     96      0      95\n",
      "    96 U     95     97      0      96\n",
      "    97 A     96     98      0      97\n",
      "    98 U     97     99     90      98\n",
      "    99 A     98    100     89      99\n",
      "   100 G     99    101     88     100\n",
      "   101 A    100    102     87     101\n",
      "   102 U    101    103      0     102\n",
      "   103 C    102    104      0     103\n",
      "   104 G    103    105      0     104\n",
      "   105 G    104    106      0     105\n",
      "   106 A    105    107      0     106\n",
      "   107 A    106    108      0     107\n",
      "   108 U    107    109      0     108\n",
      "   109 U    108    110      0     109\n",
      "   110 U    109    111      0     110\n",
      "   111 A    110    112      0     111\n",
      "   112 C    111    113      0     112\n",
      "   113 G    112    114      0     113\n",
      "   114 C    113    115      0     114\n",
      "   115 A    114    116      0     115\n",
      "   116 A    115    117      0     116\n",
      "   117 A    116    118      0     117\n",
      "   118 U    117    119      0     118\n",
      "   119 U    118    120      0     119\n",
      "   120 A    119    121      0     120\n",
      "   121 A    120    122      0     121\n",
      "   122 A    121    123      0     122\n",
      "   123 A    122    124      0     123\n",
      "   124 A    123    125      0     124\n",
      "   125 U    124    126      0     125\n",
      "   126 U    125    127      0     126\n",
      "   127 U    126    128      0     127\n",
      "   128 C    127    129      0     128\n",
      "   129 C    128    130      0     129\n",
      "   130 A    129    131      0     130\n",
      "   131 A    130    132      0     131\n",
      "   132 A    131    133      0     132\n",
      "   133 U    132    134      0     133\n",
      "   134 A    133    135      0     134\n",
      "   135 A    134    136      0     135\n",
      "   136 G    135    137      0     136\n",
      "   137 C    136    138      0     137\n",
      "   138 A    137    139      0     138\n",
      "   139 G    138    140      0     139\n",
      "   140 U    139    141      0     140\n",
      "   141 U    140    142      0     141\n",
      "   142 C    141    143      0     142\n",
      "   143 C    142    144      0     143\n",
      "   144 G    143    145      0     144\n",
      "   145 A    144    146      0     145\n",
      "   146 C    145    147      0     146\n",
      "   147 C    146    148      0     147\n",
      "   148 U    147    149      0     148\n",
      "   149 G    148    150      0     149\n",
      "   150 A    149    151      0     150\n",
      "   151 G    150    152    174     151\n",
      "   152 A    151    153    173     152\n",
      "   153 U    152    154    172     153\n",
      "   154 C    153    155    171     154\n",
      "   155 U    154    156    170     155\n",
      "   156 G    155    157      0     156\n",
      "   157 A    156    158      0     157\n",
      "   158 A    157    159      0     158\n",
      "   159 C    158    160      0     159\n",
      "   160 C    159    161      0     160\n",
      "   161 G    160    162      0     161\n",
      "   162 A    161    163      0     162\n",
      "   163 A    162    164      0     163\n",
      "   164 A    163    165      0     164\n",
      "   165 A    164    166      0     165\n",
      "   166 C    165    167      0     166\n",
      "   167 G    166    168      0     167\n",
      "   168 C    167    169      0     168\n",
      "   169 A    168    170      0     169\n",
      "   170 A    169    171    155     170\n",
      "   171 G    170    172    154     171\n",
      "   172 A    171    173    153     172\n",
      "   173 U    172    174    152     173\n",
      "   174 C    173    175    151     174\n",
      "   175 C    174    176      0     175\n",
      "   176 A    175    177      0     176\n",
      "   177 U    176    178      0     177\n",
      "   178 C    177    179      0     178\n",
      "   179 U    178    180      0     179\n",
      "   180 A    179    181      0     180\n",
      "   181 A    180    182      0     181\n",
      "   182 A    181    183      0     182\n",
      "   183 C    182    184      0     183\n",
      "   184 U    183    185      0     184\n",
      "   185 C    184    186      0     185\n",
      "   186 U    185    187      0     186\n",
      "   187 C    186    188      0     187\n",
      "   188 A    187    189      0     188\n",
      "   189 C    188    190      0     189\n",
      "   190 C    189    191      0     190\n",
      "   191 U    190    192      0     191\n",
      "   192 C    191    193      0     192\n",
      "   193 G    192    194    343     193\n",
      "   194 G    193    195    342     194\n",
      "   195 U    194    196    341     195\n",
      "   196 C    195    197    340     196\n",
      "   197 U    196    198    339     197\n",
      "   198 C    197    199    334     198\n",
      "   199 C    198    200      0     199\n",
      "   200 G    199    201      0     200\n",
      "   201 A    200    202    333     201\n",
      "   202 U    201    203    332     202\n",
      "   203 U    202    204    331     203\n",
      "   204 C    203    205    330     204\n",
      "   205 A    204    206    329     205\n",
      "   206 G    205    207    328     206\n",
      "   207 U    206    208    327     207\n",
      "   208 U    207    209    326     208\n",
      "   209 G    208    210    324     209\n",
      "   210 A    209    211    323     210\n",
      "   211 U    210    212    322     211\n",
      "   212 G    211    213    316     212\n",
      "   213 C    212    214    315     213\n",
      "   214 A    213    215    314     214\n",
      "   215 A    214    216    313     215\n",
      "   216 G    215    217    312     216\n",
      "   217 G    216    218    311     217\n",
      "   218 C    217    219    305     218\n",
      "   219 G    218    220    303     219\n",
      "   220 G    219    221      0     220\n",
      "   221 G    220    222    302     221\n",
      "   222 A    221    223    235     222\n",
      "   223 U    222    224      0     223\n",
      "   224 C    223    225      0     224\n",
      "   225 C    224    226      0     225\n",
      "   226 A    225    227    234     226\n",
      "   227 A    226    228      0     227\n",
      "   228 U    227    229      0     228\n",
      "   229 U    228    230      0     229\n",
      "   230 C    229    231      0     230\n",
      "   231 G    230    232      0     231\n",
      "   232 C    231    233      0     232\n",
      "   233 C    232    234      0     233\n",
      "   234 U    233    235    226     234\n",
      "   235 U    234    236    222     235\n",
      "   236 U    235    237      0     236\n",
      "   237 U    236    238      0     237\n",
      "   238 C    237    239      0     238\n",
      "   239 A    238    240      0     239\n",
      "   240 U    239    241      0     240\n",
      "   241 U    240    242      0     241\n",
      "   242 C    241    243      0     242\n",
      "   243 A    242    244      0     243\n",
      "   244 A    243    245      0     244\n",
      "   245 U    244    246      0     245\n",
      "   246 U    245    247      0     246\n",
      "   247 A    246    248      0     247\n",
      "   248 C    247    249      0     248\n",
      "   249 A    248    250      0     249\n",
      "   250 U    249    251      0     250\n",
      "   251 U    250    252      0     251\n",
      "   252 C    251    253      0     252\n",
      "   253 A    252    254      0     253\n",
      "   254 C    253    255      0     254\n",
      "   255 C    254    256      0     255\n",
      "   256 A    255    257      0     256\n",
      "   257 A    256    258      0     257\n",
      "   258 U    257    259      0     258\n",
      "   259 A    258    260      0     259\n",
      "   260 A    259    261      0     260\n",
      "   261 C    260    262      0     261\n",
      "   262 A    261    263      0     262\n",
      "   263 G    262    264      0     263\n",
      "   264 C    263    265      0     264\n",
      "   265 U    264    266      0     265\n",
      "   266 C    265    267      0     266\n",
      "   267 G    266    268      0     267\n",
      "   268 C    267    269    275     268\n",
      "   269 C    268    270    274     269\n",
      "   270 A    269    271      0     270\n",
      "   271 U    270    272      0     271\n",
      "   272 C    271    273      0     272\n",
      "   273 U    272    274      0     273\n",
      "   274 G    273    275    269     274\n",
      "   275 G    274    276    268     275\n",
      "   276 C    275    277    289     276\n",
      "   277 U    276    278    288     277\n",
      "   278 U    277    279      0     278\n",
      "   279 U    278    280      0     279\n",
      "   280 U    279    281      0     280\n",
      "   281 A    280    282      0     281\n",
      "   282 A    281    283      0     282\n",
      "   283 U    282    284      0     283\n",
      "   284 A    283    285      0     284\n",
      "   285 A    284    286      0     285\n",
      "   286 A    285    287      0     286\n",
      "   287 A    286    288      0     287\n",
      "   288 A    287    289    277     288\n",
      "   289 G    288    290    276     289\n",
      "   290 U    289    291      0     290\n",
      "   291 U    290    292      0     291\n",
      "   292 G    291    293      0     292\n",
      "   293 C    292    294      0     293\n",
      "   294 C    293    295      0     294\n",
      "   295 A    294    296      0     295\n",
      "   296 A    295    297      0     296\n",
      "   297 U    296    298      0     297\n",
      "   298 C    297    299      0     298\n",
      "   299 G    298    300      0     299\n",
      "   300 G    299    301      0     300\n",
      "   301 U    300    302      0     301\n",
      "   302 U    301    303    221     302\n",
      "   303 C    302    304    219     303\n",
      "   304 C    303    305      0     304\n",
      "   305 C    304    306    218     305\n",
      "   306 G    305    307      0     306\n",
      "   307 A    306    308      0     307\n",
      "   308 C    307    309      0     308\n",
      "   309 C    308    310      0     309\n",
      "   310 U    309    311      0     310\n",
      "   311 G    310    312    217     311\n",
      "   312 C    311    313    216     312\n",
      "   313 A    312    314    215     313\n",
      "   314 C    313    315    214     314\n",
      "   315 C    314    316    213     315\n",
      "   316 A    315    317    212     316\n",
      "   317 A    316    318      0     317\n",
      "   318 G    317    319      0     318\n",
      "   319 C    318    320      0     319\n",
      "   320 G    319    321      0     320\n",
      "   321 A    320    322      0     321\n",
      "   322 A    321    323    211     322\n",
      "   323 U    322    324    210     323\n",
      "   324 U    323    325    209     324\n",
      "   325 A    324    326      0     325\n",
      "   326 G    325    327    208     326\n",
      "   327 A    326    328    207     327\n",
      "   328 G    327    329    206     328\n",
      "   329 A    328    330    205     329\n",
      "   330 C    329    331    204     330\n",
      "   331 C    330    332    203     331\n",
      "   332 G    331    333    202     332\n",
      "   333 C    332    334    201     333\n",
      "   334 C    333    335    198     334\n",
      "   335 G    334    336      0     335\n",
      "   336 G    335    337      0     336\n",
      "   337 U    336    338      0     337\n",
      "   338 A    337    339      0     338\n",
      "   339 A    338    340    197     339\n",
      "   340 C    339    341    196     340\n",
      "   341 U    340    342    195     341\n",
      "   342 G    341    343    194     342\n",
      "   343 A    342    344    193     343\n",
      "   344 A    343    345      0     344\n",
      "   345 U    344    346      0     345\n",
      "   346 C    345    347      0     346\n",
      "   347 A    346    348      0     347\n",
      "   348 U    347    349      0     348\n",
      "   349 U    348    350      0     349\n",
      "   350 C    349    351      0     350\n",
      "   351 U    350    352      0     351\n",
      "   352 A    351    353      0     352\n",
      "   353 C    352    354      0     353\n",
      "   354 A    353    355      0     354\n",
      "   355 U    354    356      0     355\n",
      "   356 U    355    357      0     356\n",
      "   357 A    356    358      0     357\n",
      "   358 A    357    359      0     358\n",
      "   359 U    358    360      0     359\n",
      "   360 C    359    361      0     360\n",
      "   361 C    360    362      0     361\n",
      "   362 C    361    363      0     362\n",
      "   363 C    362    364      0     363\n",
      "   364 G    363    365      0     364\n",
      "   365 A    364    366      0     365\n",
      "   366 C    365    367      0     366\n",
      "   367 U    366    368      0     367\n",
      "   368 C    367    369      0     368\n",
      "   369 C    368    370      0     369\n",
      "   370 U    369    371      0     370\n",
      "   371 C    370    372      0     371\n",
      "   372 C    371    373      0     372\n",
      "   373 U    372    374      0     373\n",
      "   374 U    373    375      0     374\n",
      "   375 U    374    376      0     375\n",
      "   376 U    375    377      0     376\n",
      "   377 A    376    378      0     377\n",
      "   378 C    377    379      0     378\n",
      "   379 A    378    380      0     379\n",
      "   380 C    379    381      0     380\n",
      "   381 A    380    382      0     381\n",
      "   382 U    381    383      0     382\n",
      "   383 A    382    384      0     383\n",
      "   384 G    383    385      0     384\n",
      "   385 C    384    386      0     385\n",
      "   386 A    385    387      0     386\n",
      "   387 A    386    388      0     387\n",
      "   388 C    387    389    401     388\n",
      "   389 U    388    390    400     389\n",
      "   390 U    389    391    399     390\n",
      "   391 C    390    392    398     391\n",
      "   392 G    391    393      0     392\n",
      "   393 C    392    394      0     393\n",
      "   394 C    393    395      0     394\n",
      "   395 C    394    396      0     395\n",
      "   396 A    395    397      0     396\n",
      "   397 A    396    398      0     397\n",
      "   398 G    397    399    391     398\n",
      "   399 A    398    400    390     399\n",
      "   400 A    399    401    389     400\n",
      "   401 G    400    402    388     401\n",
      "   402 A    401    403      0     402\n",
      "   403 C    402    404      0     403\n",
      "   404 U    403    405      0     404\n",
      "   405 A    404    406      0     405\n",
      "   406 A    405    407      0     406\n",
      "   407 A    406    408      0     407\n",
      "   408 A    407    409      0     408\n",
      "   409 A    408    410      0     409\n",
      "   410 G    409    411      0     410\n",
      "   411 A    410    412      0     411\n",
      "   412 A    411    413      0     412\n",
      "   413 A    412    414      0     413\n",
      "   414 A    413    415      0     414\n",
      "   415 G    414    416      0     415\n",
      "   416 G    415    417      0     416\n",
      "   417 A    416    418      0     417\n",
      "   418 A    417    419      0     418\n",
      "   419 G    418    420      0     419\n",
      "   420 C    419    421      0     420\n",
      "   421 U    420    422      0     421\n",
      "   422 A    421    423      0     422\n",
      "   423 A    422    424      0     423\n",
      "   424 C    423      0      0     424\n",
      "\n"
     ]
    }
   ],
   "source": [
    "df = fasta_to_df('./secondary_structure/spot_rna/AMWY02059828.1:2832-3256(+).dot')\n",
    "df = df.apply(lambda row: bracket_row(row) , axis=1)\n",
    "bracket = df['bracket'][0].split(' ')[0]\n",
    "ct = bracket_to_ct(df['tag'][0], df['data'][0], bracket, \"(0)\")\n",
    "print(ct)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "heading_collapsed": true,
    "hidden": true
   },
   "source": [
    "## Vienna package"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 92,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-11-14T16:28:14.874416Z",
     "start_time": "2021-11-14T16:28:14.870733Z"
    },
    "hidden": true
   },
   "outputs": [],
   "source": [
    "#!wget https://www.tbi.univie.ac.at/RNA/download/ubuntu/ubuntu_20_04/viennarna_2.4.18-1_amd64.deb -O viennarna.deb\n",
    "#!sudo dpkg -i ./viennarna.deb\n",
    "#!sudo apt-get -f install\n",
    "#!rm viennarna.deb"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 131,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-11-14T19:35:57.889695Z",
     "start_time": "2021-11-14T19:35:57.480149Z"
    },
    "hidden": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "rm: cannot remove './secondary_structure/viennarna_result.txt': No such file or directory\r\n"
     ]
    }
   ],
   "source": [
    "base = \"./secondary_structure/viennarna/\"\n",
    "!rm -r {base}\n",
    "!rm ./secondary_structure/viennarna_result.txt\n",
    "!mkdir -p {base}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 132,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-11-14T19:52:09.128413Z",
     "start_time": "2021-11-14T19:36:03.739898Z"
    },
    "hidden": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/home/jupyter/plant_microRNA_prediction/secondary_structure/viennarna\n",
      "/home/jupyter/plant_microRNA_prediction\n"
     ]
    }
   ],
   "source": [
    "%cd {base}\n",
    "!RNAfold --jobs=0 --infile ../../Temp/extended_modified.txt  --noPS -T 20 > ../viennarna_result.txt\n",
    "%cd ../../"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 95,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-11-14T17:36:50.680457Z",
     "start_time": "2021-11-14T17:36:28.498386Z"
    },
    "hidden": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(21225, 3)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>tag</th>\n",
       "      <th>data</th>\n",
       "      <th>bracket</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AMWY02000003.1:4084-4506|200-222(+)</td>\n",
       "      <td>auacaauugucacauaguuuacauuauuaauuuccgcuuaauuuau...</td>\n",
       "      <td>....((((((...))))))................((((((...((...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AMWY02000003.1:4391-4810|200-219(+)</td>\n",
       "      <td>uucuaauugcaaauuuauguuauauuuuuaaauagaaagggagauu...</td>\n",
       "      <td>.....((.(((......))).))......(((((((.(((((((.....</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                   tag  \\\n",
       "0  AMWY02000003.1:4084-4506|200-222(+)   \n",
       "1  AMWY02000003.1:4391-4810|200-219(+)   \n",
       "\n",
       "                                                data  \\\n",
       "0  auacaauugucacauaguuuacauuauuaauuuccgcuuaauuuau...   \n",
       "1  uucuaauugcaaauuuauguuauauuuuuaaauagaaagggagauu...   \n",
       "\n",
       "                                             bracket  \n",
       "0  ....((((((...))))))................((((((...((...  \n",
       "1  .....((.(((......))).))......(((((((.(((((((.....  "
      ]
     },
     "execution_count": 95,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = fasta_to_df('secondary_structure/viennarna_result.txt')\n",
    "df = df.apply(lambda row: bracket_row(row) , axis=1)\n",
    "print(df.shape)\n",
    "df.head(2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 125,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-11-04T13:12:19.517498Z",
     "start_time": "2021-11-04T13:12:19.487872Z"
    },
    "hidden": true
   },
   "outputs": [],
   "source": [
    "for index, row in df.iterrows():    \n",
    "    tag = reformat(row['tag'])\n",
    "    if(not os.path.exists(base + tag)):\n",
    "        os.makedirs(base + tag)      \n",
    "    with open(base + f\"{tag}/{tag}.ct\",'w') as file:\n",
    "        bracket = row['bracket'].split(' ')[0]\n",
    "        deltaG = row['bracket'].split(' ')[1]\n",
    "        ct = bracket_to_ct(row['tag'], row['data'], bracket, deltaG, False)\n",
    "        file.write(ct)    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 126,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-11-04T13:12:20.455326Z",
     "start_time": "2021-11-04T13:12:20.449583Z"
    },
    "hidden": true
   },
   "outputs": [],
   "source": [
    "import glob\n",
    "for file in glob.glob(f\"{base}*.ps\"):    \n",
    "    f = file[len(base):-6] # _ss.ps \n",
    "    f = reformat(f)        \n",
    "    shutil.move(file, f\"{base}{f}/{f}.ps\")    "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "heading_collapsed": true,
    "hidden": true
   },
   "source": [
    "## ContraFold"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-11-06T15:47:08.282696Z",
     "start_time": "2021-11-06T15:47:08.278027Z"
    },
    "hidden": true
   },
   "outputs": [],
   "source": [
    "#!wget http://contra.stanford.edu/contrafold/contrafold_v2_02.tar.gz\n",
    "#!tar -xvzf contrafold_v2_02.tar.gz && rm contrafold_v2_02.tar.gz\n",
    "#%cd contrafold/src\n",
    "#!make clean\n",
    "#!make \n",
    "# to file must changed to be complieable # utility.hpp and optimization.c++ files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-11-06T15:52:08.807315Z",
     "start_time": "2021-11-06T15:52:08.274966Z"
    },
    "hidden": true
   },
   "outputs": [],
   "source": [
    "counter = 0\n",
    "base = \"./secondary_structure/contrafold/\"\n",
    "!rm -r {base}\n",
    "!mkdir -p {base}\n",
    "df = fasta_to_df('./Temp/extended.txt')\n",
    "\n",
    "for index, row in df.iterrows():    \n",
    "    tag = reformat(row['tag'])\n",
    "    if(not os.path.exists(base + tag)):\n",
    "        os.makedirs(base + tag)            \n",
    "    with open(base + f\"{tag}/{tag}.FASTA\",'w') as file:\n",
    "        file.write(f\">{row['tag']}\\n{row['data']}\")\n",
    "    counter += 1    \n",
    "    if(counter >= 10):\n",
    "        break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-11-06T15:52:20.318477Z",
     "start_time": "2021-11-06T15:52:18.362143Z"
    },
    "hidden": true
   },
   "outputs": [],
   "source": [
    "def run_contrafold(tag):\n",
    "    tag = reformat(tag)    \n",
    "    %cd contrafold/src\n",
    "    !./contrafold predict ../..{base[1:]}{tag}/{tag}.FASTA > ../..{base[1:]}{tag}/{tag}.dot\n",
    "    with open(f\"../..{base[1:]}{tag}/{tag}.dot\", 'r') as file:\n",
    "        text = file.read()\n",
    "    text = [l for l in text.split(\"\\n\") if l[:len(\">structure\")] != \">structure\"]    \n",
    "    header = text[0]\n",
    "    with open(f\"../..{base[1:]}{tag}/{tag}.dot\", 'w') as file:\n",
    "        file.write('\\n'.join(text[1:]))    \n",
    "    !RNAeval  ../..{base[1:]}{tag}/{tag}.dot -T 20 > ../..{base[1:]}{tag}/{tag}.dotdg    \n",
    "    with open(f\"../..{base[1:]}{tag}/{tag}.dotdg\", 'r') as file:\n",
    "        text = file.read()\n",
    "    with open(f\"../..{base[1:]}{tag}/{tag}.dot\", 'w') as file:\n",
    "        file.write(header + \"\\n\" + text)    \n",
    "    \n",
    "    df = fasta_to_df(f'../..{base[1:]}{tag}/{tag}.dot')\n",
    "    df = df.apply(lambda row: bracket_row(row) , axis=1)        \n",
    "    tag = reformat(df['tag'][0])\n",
    "    with open(f'../..{base[1:]}{tag}/{tag}.ct','w') as file:\n",
    "        bracket = df['bracket'][0].split(' ')[0]        \n",
    "        deltaG = df['bracket'][0].split(' ')[1]\n",
    "        ct = bracket_to_ct(df['tag'][0], df['data'][0], bracket, deltaG, False)\n",
    "        file.write(ct)    \n",
    "    #!rm ../..{base[1:]}{tag}/{tag}.dot\n",
    "    #!rm ../..{base[1:]}{tag}/{tag}.dotdg\n",
    "    !rm ../..{base[1:]}{tag}/{tag}.FASTA\n",
    "    %cd ../../        \n",
    "\n",
    "if __name__ == '__main__':        \n",
    "    pool = mp.Pool(mp.cpu_count() - 1)  \n",
    "    pool.map(run_contrafold, df['tag'].iloc[:10])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-11-06T16:04:49.091652Z",
     "start_time": "2021-11-06T16:04:49.084840Z"
    },
    "hidden": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "A U\n",
      "( )\n"
     ]
    }
   ],
   "source": [
    "s = 'CUCCCCUUGUCUACCAUCCCCAACUAGCGAGAGAGACAUUACCUACCUGAAUAGAAGAUCUCUCUCGAGCUCUCGagcucucucuuuuucuauaUCUCUGUCUCUUUGUGUCUCUGGAGCUUGUACUAACAUUAAUAUCGUGCACCAGCAGCAGUUGAAGCUGCCAGCAUGAUCUAAACUUCCUUCUCUGUAAAGGAUAGAUCGGAUCAUGUGGUAGCUUCACCUGUUGAUGGGAUCACGAAAGCGCCCCUCUUACUACUCUACAUUAAUUCUUUCUCGUUAUACAACCUCCCAGUAAGCAUGCUUUCAAAACCAACUUGAGuaaguuaauuuguuuagcuuuuguuuuuggcucuuccuuuacuuuaaauuuucucaucuggguuuuuguuauauauauguacuguuuuauauauguauuccu'\n",
    "d = '............................((((((((..(...(((......))).)..))))))))(((((....)))))...................................((((.((((...(((.......(((((..(((.((((((.((((((((((.(((((((((.(.(((((((.......))))).)).).))))))))))))))))))).)))))).)))...))))).....................................)))..)))).))))....((((()))))..((((((....((((.(((((((.....)))(((.........)))................)))).))))....))))))....(((((((((((......)))))))))))....'\n",
    "print(s[300],s[301])\n",
    "print(d[300],d[301])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-11-06T16:04:27.288085Z",
     "start_time": "2021-11-06T16:04:27.151101Z"
    },
    "hidden": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[36mExternal loop\u001b[0m                           : \u001b[32m -364\u001b[0m\r\n",
      "\u001b[36mInterior loop\u001b[0m ( 29, 66) \u001b[1mGC\u001b[0m; ( 30, 65) \u001b[1mAU\u001b[0m: \u001b[32m -294\u001b[0m\r\n",
      "\u001b[36mInterior loop\u001b[0m ( 30, 65) \u001b[1mAU\u001b[0m; ( 31, 64) \u001b[1mGC\u001b[0m: \u001b[32m -256\u001b[0m\r\n",
      "\u001b[36mInterior loop\u001b[0m ( 31, 64) \u001b[1mGC\u001b[0m; ( 32, 63) \u001b[1mAU\u001b[0m: \u001b[32m -294\u001b[0m\r\n",
      "\u001b[36mInterior loop\u001b[0m ( 32, 63) \u001b[1mAU\u001b[0m; ( 33, 62) \u001b[1mGC\u001b[0m: \u001b[32m -256\u001b[0m\r\n",
      "\u001b[36mInterior loop\u001b[0m ( 33, 62) \u001b[1mGC\u001b[0m; ( 34, 61) \u001b[1mAU\u001b[0m: \u001b[32m -294\u001b[0m\r\n",
      "\u001b[36mInterior loop\u001b[0m ( 34, 61) \u001b[1mAU\u001b[0m; ( 35, 60) \u001b[1mGC\u001b[0m: \u001b[32m -256\u001b[0m\r\n",
      "\u001b[36mInterior loop\u001b[0m ( 35, 60) \u001b[1mGC\u001b[0m; ( 36, 59) \u001b[1mAU\u001b[0m: \u001b[32m -294\u001b[0m\r\n",
      "\u001b[36mInterior loop\u001b[0m ( 36, 59) \u001b[1mAU\u001b[0m; ( 39, 56) \u001b[1mUA\u001b[0m: \u001b[32m  103\u001b[0m\r\n",
      "\u001b[36mInterior loop\u001b[0m ( 39, 56) \u001b[1mUA\u001b[0m; ( 43, 54) \u001b[1mCG\u001b[0m: \u001b[32m  305\u001b[0m\r\n",
      "\u001b[36mInterior loop\u001b[0m ( 43, 54) \u001b[1mCG\u001b[0m; ( 44, 53) \u001b[1mUA\u001b[0m: \u001b[32m -256\u001b[0m\r\n",
      "\u001b[36mInterior loop\u001b[0m ( 44, 53) \u001b[1mUA\u001b[0m; ( 45, 52) \u001b[1mAU\u001b[0m: \u001b[32m -165\u001b[0m\r\n",
      "\u001b[36mHairpin  loop\u001b[0m ( 45, 52) \u001b[1mAU\u001b[0m              : \u001b[32m  493\u001b[0m\r\n",
      "\u001b[36mInterior loop\u001b[0m ( 67, 80) \u001b[1mGC\u001b[0m; ( 68, 79) \u001b[1mAU\u001b[0m: \u001b[32m -294\u001b[0m\r\n",
      "\u001b[36mInterior loop\u001b[0m ( 68, 79) \u001b[1mAU\u001b[0m; ( 69, 78) \u001b[1mGC\u001b[0m: \u001b[32m -256\u001b[0m\r\n",
      "\u001b[36mInterior loop\u001b[0m ( 69, 78) \u001b[1mGC\u001b[0m; ( 70, 77) \u001b[1mCG\u001b[0m: \u001b[32m -403\u001b[0m\r\n",
      "\u001b[36mInterior loop\u001b[0m ( 70, 77) \u001b[1mCG\u001b[0m; ( 71, 76) \u001b[1mUA\u001b[0m: \u001b[32m -256\u001b[0m\r\n",
      "\u001b[36mHairpin  loop\u001b[0m ( 71, 76) \u001b[1mUA\u001b[0m              : \u001b[32m  533\u001b[0m\r\n",
      "\u001b[36mInterior loop\u001b[0m (116,292) \u001b[1mGC\u001b[0m; (117,291) \u001b[1mGC\u001b[0m: \u001b[32m -385\u001b[0m\r\n",
      "\u001b[36mInterior loop\u001b[0m (117,291) \u001b[1mGC\u001b[0m; (118,290) \u001b[1mAU\u001b[0m: \u001b[32m -294\u001b[0m\r\n",
      "\u001b[36mInterior loop\u001b[0m (118,290) \u001b[1mAU\u001b[0m; (119,289) \u001b[1mGC\u001b[0m: \u001b[32m -256\u001b[0m\r\n",
      "\u001b[36mInterior loop\u001b[0m (119,289) \u001b[1mGC\u001b[0m; (121,287) \u001b[1mUA\u001b[0m: \u001b[32m   83\u001b[0m\r\n",
      "\u001b[36mInterior loop\u001b[0m (121,287) \u001b[1mUA\u001b[0m; (122,286) \u001b[1mUA\u001b[0m: \u001b[32m -122\u001b[0m\r\n",
      "\u001b[36mInterior loop\u001b[0m (122,286) \u001b[1mUA\u001b[0m; (123,285) \u001b[1mGC\u001b[0m: \u001b[32m -255\u001b[0m\r\n",
      "\u001b[36mInterior loop\u001b[0m (123,285) \u001b[1mGC\u001b[0m; (124,284) \u001b[1mUA\u001b[0m: \u001b[32m -270\u001b[0m\r\n",
      "\u001b[36mInterior loop\u001b[0m (124,284) \u001b[1mUA\u001b[0m; (128,281) \u001b[1mAU\u001b[0m: \u001b[32m  411\u001b[0m\r\n",
      "\u001b[36mInterior loop\u001b[0m (128,281) \u001b[1mAU\u001b[0m; (129,280) \u001b[1mAU\u001b[0m: \u001b[32m -122\u001b[0m\r\n",
      "\u001b[36mInterior loop\u001b[0m (129,280) \u001b[1mAU\u001b[0m; (130,279) \u001b[1mCG\u001b[0m: \u001b[32m -270\u001b[0m\r\n",
      "\u001b[36mInterior loop\u001b[0m (130,279) \u001b[1mCG\u001b[0m; (138,241) \u001b[1mUA\u001b[0m: \u001b[32m  774\u001b[0m\r\n",
      "\u001b[36mInterior loop\u001b[0m (138,241) \u001b[1mUA\u001b[0m; (139,240) \u001b[1mCG\u001b[0m: \u001b[32m -294\u001b[0m\r\n",
      "\u001b[36mInterior loop\u001b[0m (139,240) \u001b[1mCG\u001b[0m; (140,239) \u001b[1mGC\u001b[0m: \u001b[32m -284\u001b[0m\r\n",
      "\u001b[36mInterior loop\u001b[0m (140,239) \u001b[1mGC\u001b[0m; (141,238) \u001b[1mUA\u001b[0m: \u001b[32m -270\u001b[0m\r\n",
      "\u001b[36mInterior loop\u001b[0m (141,238) \u001b[1mUA\u001b[0m; (142,237) \u001b[1mGC\u001b[0m: \u001b[32m -255\u001b[0m\r\n",
      "\u001b[36mInterior loop\u001b[0m (142,237) \u001b[1mGC\u001b[0m; (145,233) \u001b[1mCG\u001b[0m: \u001b[32m   52\u001b[0m\r\n",
      "\u001b[36mInterior loop\u001b[0m (145,233) \u001b[1mCG\u001b[0m; (146,232) \u001b[1mCG\u001b[0m: \u001b[32m -385\u001b[0m\r\n",
      "\u001b[36mInterior loop\u001b[0m (146,232) \u001b[1mCG\u001b[0m; (147,231) \u001b[1mAU\u001b[0m: \u001b[32m -255\u001b[0m\r\n",
      "\u001b[36mInterior loop\u001b[0m (147,231) \u001b[1mAU\u001b[0m; (149,229) \u001b[1mCG\u001b[0m: \u001b[32m   83\u001b[0m\r\n",
      "\u001b[36mInterior loop\u001b[0m (149,229) \u001b[1mCG\u001b[0m; (150,228) \u001b[1mAU\u001b[0m: \u001b[32m -255\u001b[0m\r\n",
      "\u001b[36mInterior loop\u001b[0m (150,228) \u001b[1mAU\u001b[0m; (151,227) \u001b[1mGU\u001b[0m: \u001b[32m  -74\u001b[0m\r\n",
      "\u001b[36mInterior loop\u001b[0m (151,227) \u001b[1mGU\u001b[0m; (152,226) \u001b[1mCG\u001b[0m: \u001b[32m -305\u001b[0m\r\n",
      "\u001b[36mInterior loop\u001b[0m (152,226) \u001b[1mCG\u001b[0m; (153,225) \u001b[1mAU\u001b[0m: \u001b[32m -255\u001b[0m\r\n",
      "\u001b[36mInterior loop\u001b[0m (153,225) \u001b[1mAU\u001b[0m; (154,224) \u001b[1mGC\u001b[0m: \u001b[32m -256\u001b[0m\r\n",
      "\u001b[36mInterior loop\u001b[0m (154,224) \u001b[1mGC\u001b[0m; (156,222) \u001b[1mUA\u001b[0m: \u001b[32m   83\u001b[0m\r\n",
      "\u001b[36mInterior loop\u001b[0m (156,222) \u001b[1mUA\u001b[0m; (157,221) \u001b[1mGC\u001b[0m: \u001b[32m -255\u001b[0m\r\n",
      "\u001b[36mInterior loop\u001b[0m (157,221) \u001b[1mGC\u001b[0m; (158,220) \u001b[1mAU\u001b[0m: \u001b[32m -294\u001b[0m\r\n",
      "\u001b[36mInterior loop\u001b[0m (158,220) \u001b[1mAU\u001b[0m; (159,219) \u001b[1mAU\u001b[0m: \u001b[32m -122\u001b[0m\r\n",
      "\u001b[36mInterior loop\u001b[0m (159,219) \u001b[1mAU\u001b[0m; (160,218) \u001b[1mGC\u001b[0m: \u001b[32m -256\u001b[0m\r\n",
      "\u001b[36mInterior loop\u001b[0m (160,218) \u001b[1mGC\u001b[0m; (161,217) \u001b[1mCG\u001b[0m: \u001b[32m -403\u001b[0m\r\n",
      "\u001b[36mInterior loop\u001b[0m (161,217) \u001b[1mCG\u001b[0m; (162,216) \u001b[1mUA\u001b[0m: \u001b[32m -256\u001b[0m\r\n",
      "\u001b[36mInterior loop\u001b[0m (162,216) \u001b[1mUA\u001b[0m; (163,215) \u001b[1mGU\u001b[0m: \u001b[32m -132\u001b[0m\r\n",
      "\u001b[36mInterior loop\u001b[0m (163,215) \u001b[1mGU\u001b[0m; (164,214) \u001b[1mCG\u001b[0m: \u001b[32m -305\u001b[0m\r\n",
      "\u001b[36mInterior loop\u001b[0m (164,214) \u001b[1mCG\u001b[0m; (165,213) \u001b[1mCG\u001b[0m: \u001b[32m -385\u001b[0m\r\n",
      "\u001b[36mInterior loop\u001b[0m (165,213) \u001b[1mCG\u001b[0m; (167,212) \u001b[1mGU\u001b[0m: \u001b[32m  254\u001b[0m\r\n",
      "\u001b[36mInterior loop\u001b[0m (167,212) \u001b[1mGU\u001b[0m; (168,211) \u001b[1mCG\u001b[0m: \u001b[32m -305\u001b[0m\r\n",
      "\u001b[36mInterior loop\u001b[0m (168,211) \u001b[1mCG\u001b[0m; (169,210) \u001b[1mAU\u001b[0m: \u001b[32m -255\u001b[0m\r\n",
      "\u001b[36mInterior loop\u001b[0m (169,210) \u001b[1mAU\u001b[0m; (170,209) \u001b[1mUA\u001b[0m: \u001b[32m -155\u001b[0m\r\n",
      "\u001b[36mInterior loop\u001b[0m (170,209) \u001b[1mUA\u001b[0m; (171,208) \u001b[1mGC\u001b[0m: \u001b[32m -255\u001b[0m\r\n",
      "\u001b[36mInterior loop\u001b[0m (171,208) \u001b[1mGC\u001b[0m; (172,207) \u001b[1mAU\u001b[0m: \u001b[32m -294\u001b[0m\r\n",
      "\u001b[36mInterior loop\u001b[0m (172,207) \u001b[1mAU\u001b[0m; (173,206) \u001b[1mUA\u001b[0m: \u001b[32m -155\u001b[0m\r\n",
      "\u001b[36mInterior loop\u001b[0m (173,206) \u001b[1mUA\u001b[0m; (174,205) \u001b[1mCG\u001b[0m: \u001b[32m -294\u001b[0m\r\n",
      "\u001b[36mInterior loop\u001b[0m (174,205) \u001b[1mCG\u001b[0m; (175,204) \u001b[1mUG\u001b[0m: \u001b[32m -264\u001b[0m\r\n",
      "\u001b[36mInterior loop\u001b[0m (175,204) \u001b[1mUG\u001b[0m; (177,202) \u001b[1mAU\u001b[0m: \u001b[32m  176\u001b[0m\r\n",
      "\u001b[36mInterior loop\u001b[0m (177,202) \u001b[1mAU\u001b[0m; (179,200) \u001b[1mCG\u001b[0m: \u001b[32m   83\u001b[0m\r\n",
      "\u001b[36mInterior loop\u001b[0m (179,200) \u001b[1mCG\u001b[0m; (180,199) \u001b[1mUA\u001b[0m: \u001b[32m -256\u001b[0m\r\n",
      "\u001b[36mInterior loop\u001b[0m (180,199) \u001b[1mUA\u001b[0m; (181,197) \u001b[1mUA\u001b[0m: \u001b[32m  295\u001b[0m\r\n",
      "\u001b[36mInterior loop\u001b[0m (181,197) \u001b[1mUA\u001b[0m; (182,196) \u001b[1mCG\u001b[0m: \u001b[32m -294\u001b[0m\r\n",
      "\u001b[36mInterior loop\u001b[0m (182,196) \u001b[1mCG\u001b[0m; (183,195) \u001b[1mCG\u001b[0m: \u001b[32m -385\u001b[0m\r\n",
      "\u001b[36mInterior loop\u001b[0m (183,195) \u001b[1mCG\u001b[0m; (184,194) \u001b[1mUA\u001b[0m: \u001b[32m -256\u001b[0m\r\n",
      "\u001b[36mInterior loop\u001b[0m (184,194) \u001b[1mUA\u001b[0m; (185,193) \u001b[1mUA\u001b[0m: \u001b[32m -122\u001b[0m\r\n",
      "\u001b[36mHairpin  loop\u001b[0m (185,193) \u001b[1mUA\u001b[0m              : \u001b[32m  552\u001b[0m\r\n",
      "\u001b[36mInterior loop\u001b[0m (297,306) \u001b[1mAU\u001b[0m; (298,305) \u001b[1mAU\u001b[0m: \u001b[32m -122\u001b[0m\r\n",
      "\u001b[36mInterior loop\u001b[0m (298,305) \u001b[1mAU\u001b[0m; (299,304) \u001b[1mGC\u001b[0m: \u001b[32m -256\u001b[0m\r\n",
      "\u001b[36mInterior loop\u001b[0m (299,304) \u001b[1mGC\u001b[0m; (300,303) \u001b[1mCG\u001b[0m: \u001b[32m -403\u001b[0m\r\n",
      "\u001b[36mInterior loop\u001b[0m (300,303) \u001b[1mCG\u001b[0m; (301,302) \u001b[1mAU\u001b[0m: \u001b[32m -255\u001b[0m\r\n",
      "\u001b[36mHairpin  loop\u001b[0m (301,302) \u001b[1mAU\u001b[0m              : \u001b[32m10000000\u001b[0m\r\n",
      "\u001b[36mInterior loop\u001b[0m (309,388) \u001b[1mAU\u001b[0m; (310,387) \u001b[1mAU\u001b[0m: \u001b[32m -122\u001b[0m\r\n",
      "\u001b[36mInterior loop\u001b[0m (310,387) \u001b[1mAU\u001b[0m; (311,386) \u001b[1mAU\u001b[0m: \u001b[32m -122\u001b[0m\r\n",
      "\u001b[36mInterior loop\u001b[0m (311,386) \u001b[1mAU\u001b[0m; (312,385) \u001b[1mAU\u001b[0m: \u001b[32m -122\u001b[0m\r\n",
      "\u001b[36mInterior loop\u001b[0m (312,385) \u001b[1mAU\u001b[0m; (313,384) \u001b[1mCG\u001b[0m: \u001b[32m -270\u001b[0m\r\n",
      "\u001b[36mInterior loop\u001b[0m (313,384) \u001b[1mCG\u001b[0m; (314,383) \u001b[1mCG\u001b[0m: \u001b[32m -385\u001b[0m\r\n",
      "\u001b[36mInterior loop\u001b[0m (314,383) \u001b[1mCG\u001b[0m; (319,378) \u001b[1mUA\u001b[0m: \u001b[32m  121\u001b[0m\r\n",
      "\u001b[36mInterior loop\u001b[0m (319,378) \u001b[1mUA\u001b[0m; (320,377) \u001b[1mGC\u001b[0m: \u001b[32m -255\u001b[0m\r\n",
      "\u001b[36mInterior loop\u001b[0m (320,377) \u001b[1mGC\u001b[0m; (321,376) \u001b[1mAU\u001b[0m: \u001b[32m -294\u001b[0m\r\n",
      "\u001b[36mInterior loop\u001b[0m (321,376) \u001b[1mAU\u001b[0m; (322,375) \u001b[1mGC\u001b[0m: \u001b[32m -256\u001b[0m\r\n",
      "\u001b[36mInterior loop\u001b[0m (322,375) \u001b[1mGC\u001b[0m; (324,373) \u001b[1mAU\u001b[0m: \u001b[32m   64\u001b[0m\r\n",
      "\u001b[36mInterior loop\u001b[0m (324,373) \u001b[1mAU\u001b[0m; (325,372) \u001b[1mAU\u001b[0m: \u001b[32m -122\u001b[0m\r\n",
      "\u001b[36mInterior loop\u001b[0m (325,372) \u001b[1mAU\u001b[0m; (326,371) \u001b[1mGU\u001b[0m: \u001b[32m  -74\u001b[0m\r\n",
      "\u001b[36mInterior loop\u001b[0m (326,371) \u001b[1mGU\u001b[0m; (327,370) \u001b[1mUA\u001b[0m: \u001b[32m -180\u001b[0m\r\n",
      "\u001b[36mInterior loop\u001b[0m (328,338) \u001b[1mUA\u001b[0m; (329,337) \u001b[1mAU\u001b[0m: \u001b[32m -165\u001b[0m\r\n",
      "\u001b[36mInterior loop\u001b[0m (329,337) \u001b[1mAU\u001b[0m; (330,336) \u001b[1mAU\u001b[0m: \u001b[32m -122\u001b[0m\r\n",
      "\u001b[36mHairpin  loop\u001b[0m (330,336) \u001b[1mAU\u001b[0m              : \u001b[32m  424\u001b[0m\r\n",
      "\u001b[36mInterior loop\u001b[0m (339,353) \u001b[1mGC\u001b[0m; (340,352) \u001b[1mCG\u001b[0m: \u001b[32m -403\u001b[0m\r\n",
      "\u001b[36mInterior loop\u001b[0m (340,352) \u001b[1mCG\u001b[0m; (341,351) \u001b[1mUG\u001b[0m: \u001b[32m -264\u001b[0m\r\n",
      "\u001b[36mHairpin  loop\u001b[0m (341,351) \u001b[1mUG\u001b[0m              : \u001b[32m  544\u001b[0m\r\n",
      "\u001b[36mMulti    loop\u001b[0m (327,370) \u001b[1mUA\u001b[0m              : \u001b[32m  589\u001b[0m\r\n",
      "\u001b[36mInterior loop\u001b[0m (393,420) \u001b[1mAU\u001b[0m; (394,419) \u001b[1mUA\u001b[0m: \u001b[32m -155\u001b[0m\r\n",
      "\u001b[36mInterior loop\u001b[0m (394,419) \u001b[1mUA\u001b[0m; (395,418) \u001b[1mAU\u001b[0m: \u001b[32m -165\u001b[0m\r\n",
      "\u001b[36mInterior loop\u001b[0m (395,418) \u001b[1mAU\u001b[0m; (396,417) \u001b[1mUG\u001b[0m: \u001b[32m -180\u001b[0m\r\n",
      "\u001b[36mInterior loop\u001b[0m (396,417) \u001b[1mUG\u001b[0m; (397,416) \u001b[1mAU\u001b[0m: \u001b[32m -132\u001b[0m\r\n",
      "\u001b[36mInterior loop\u001b[0m (397,416) \u001b[1mAU\u001b[0m; (398,415) \u001b[1mUA\u001b[0m: \u001b[32m -155\u001b[0m\r\n",
      "\u001b[36mInterior loop\u001b[0m (398,415) \u001b[1mUA\u001b[0m; (399,414) \u001b[1mAU\u001b[0m: \u001b[32m -165\u001b[0m\r\n",
      "\u001b[36mInterior loop\u001b[0m (399,414) \u001b[1mAU\u001b[0m; (400,413) \u001b[1mUA\u001b[0m: \u001b[32m -155\u001b[0m\r\n",
      "\u001b[36mInterior loop\u001b[0m (400,413) \u001b[1mUA\u001b[0m; (401,412) \u001b[1mGU\u001b[0m: \u001b[32m -132\u001b[0m\r\n",
      "\u001b[36mInterior loop\u001b[0m (401,412) \u001b[1mGU\u001b[0m; (402,411) \u001b[1mUA\u001b[0m: \u001b[32m -180\u001b[0m\r\n",
      "\u001b[36mInterior loop\u001b[0m (402,411) \u001b[1mUA\u001b[0m; (403,410) \u001b[1mAU\u001b[0m: \u001b[32m -165\u001b[0m\r\n",
      "\u001b[36mHairpin  loop\u001b[0m (403,410) \u001b[1mAU\u001b[0m              : \u001b[32m  494\u001b[0m\r\n",
      "CUCCCCUUGUCUACCAUCCCCAACUAGCGAGAGAGACAUUACCUACCUGAAUAGAAGAUCUCUCUCGAGCUCUCGagcucucucuuuuucuauaUCUCUGUCUCUUUGUGUCUCUGGAGCUUGUACUAACAUUAAUAUCGUGCACCAGCAGCAGUUGAAGCUGCCAGCAUGAUCUAAACUUCCUUCUCUGUAAAGGAUAGAUCGGAUCAUGUGGUAGCUUCACCUGUUGAUGGGAUCACGAAAGCGCCCCUCUUACUACUCUACAUUAAUUCUUUCUCGUUAUACAACCUCCCAGUAAGCAUGCUUUCAAAACCAACUUGAGuaaguuaauuuguuuagcuuuuguuuuuggcucuuccuuuacuuuaaauuuucucaucuggguuuuuguuauauauauguacuguuuuauauauguauuccu\r\n",
      "............................((((((((..(...(((......))).)..))))))))(((((....)))))...................................((((.((((...(((.......(((((..(((.((((((.((((((((((.(((((((((.(.(((((((.......))))).)).).))))))))))))))))))).)))))).)))...))))).....................................)))..)))).))))....((((()))))..((((((....((((.(((((((.....)))(((.........)))................)))).))))....))))))....(((((((((((......)))))))))))....\u001b[32m (99862.42)\u001b[0m\r\n"
     ]
    }
   ],
   "source": [
    "'''path = 'secondary_structure/contrafold/AMWY020333941_469-893_-_/AMWY020333941_469-893_-_.dot'\n",
    "!RNAeval  {path} -T 20 -v'''; "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "heading_collapsed": true
   },
   "source": [
    "# Visualization"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-11-04T15:58:52.008179Z",
     "start_time": "2021-11-04T15:58:52.003841Z"
    },
    "hidden": true
   },
   "outputs": [],
   "source": [
    "#https://github.com/ViennaRNA/forna\n",
    "#http://varna.lri.fr/"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# CT Analizer"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 125,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-11-18T17:28:44.564952Z",
     "start_time": "2021-11-18T17:28:33.694894Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(200, 2)\n"
     ]
    }
   ],
   "source": [
    "# only select those not ran before\n",
    "base = \"./secondary_structure/mfold/\"\n",
    "df = fasta_to_df('./Temp/extended_modified.txt')\n",
    "\n",
    "index_list =[]\n",
    "for index, row in df.iterrows():    \n",
    "    tag = reformat(row['tag'])    \n",
    "    if(len(glob.glob(f'{base + tag}/*.ct')) != 0):\n",
    "        index_list.append(index)\n",
    "df = df.iloc[index_list,:]\n",
    "print(df.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 256,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-11-18T17:53:20.653821Z",
     "start_time": "2021-11-18T17:53:20.648566Z"
    }
   },
   "outputs": [],
   "source": [
    "def get_tag_info(tag):\n",
    "    data = tag.split('|')\n",
    "    hit_start = int(data[2].split('-')[0]) - 1 \n",
    "    hit_end = int(data[2].split('-')[1])\n",
    "    sign = data[3]\n",
    "    return [hit_start, hit_end, sign]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 257,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-11-18T17:53:21.012878Z",
     "start_time": "2021-11-18T17:53:21.007008Z"
    }
   },
   "outputs": [],
   "source": [
    "def get_deltaG(ct):\n",
    "    ct_head = ct.split('\\n')[0]\n",
    "    if(\"dG = \" in ct_head):    \n",
    "        dG_patter = \"dG = \" \n",
    "    elif(\"dG= \" in ct_head):    \n",
    "        dG_patter = \"dG= \"\n",
    "    elif(\"dG=\" in ct_head):    \n",
    "        dG_patter = \"dG=\"\n",
    "    elif(\"dG =\" in ct_head):    \n",
    "        dG_patter = \"dG =\"\n",
    "    else:\n",
    "        print('there is no dG')\n",
    "    return float(ct_head.split(dG_patter)[-1].split(' ')[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 258,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-11-18T17:53:21.327146Z",
     "start_time": "2021-11-18T17:53:21.322248Z"
    }
   },
   "outputs": [],
   "source": [
    "def get_complementarity_in_hit_region(inc_srange, hit_len):    \n",
    "    if(sum(inc_srange == 0) == hit_len):\n",
    "        return \"no\"\n",
    "    elif(sum(inc_srange != 0) == hit_len):\n",
    "        return \"fully_connected\"\n",
    "    else:    \n",
    "        return \"yes\"    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 259,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-11-18T17:53:21.593215Z",
     "start_time": "2021-11-18T17:53:21.589071Z"
    }
   },
   "outputs": [],
   "source": [
    "def get_hit_self_complementarity(hit_start, hit_end, inc_srange):    \n",
    "    if(((inc_srange <= hit_start) | (inc_srange > hit_end)).all()):\n",
    "        return \"no\"\n",
    "    return \"yes\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 260,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-11-18T17:53:21.956194Z",
     "start_time": "2021-11-18T17:53:21.951557Z"
    }
   },
   "outputs": [],
   "source": [
    "def get_istar_min_max(inc_srange, hit_self_complementarity):\n",
    "    nonzero_data_srange  = inc_srange[inc_srange!=0]\n",
    "    if(hit_self_complementarity == 'yes'):\n",
    "        return [np.nan, np.nan]\n",
    "    return [nonzero_data_srange.min(), nonzero_data_srange.max()]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 261,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-11-18T17:53:22.223786Z",
     "start_time": "2021-11-18T17:53:22.219095Z"
    }
   },
   "outputs": [],
   "source": [
    "def get_continuous_pairing(hit_start, hit_end, istar_min, istar_max, hit_self_complementarity):    \n",
    "    if(hit_self_complementarity == 'yes'):\n",
    "        return \"undifined\"\n",
    "    if(hit_end < istar_max and (hit_start+1) > istar_min):\n",
    "        return \"no\"    \n",
    "    return  \"yes\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 299,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-11-18T18:09:08.053839Z",
     "start_time": "2021-11-18T18:09:08.046407Z"
    }
   },
   "outputs": [],
   "source": [
    "def get_mir_type(hit_start, hit_end, istar_min, istar_max, continuous_pairing, complementarity_in_hit_region, hit_self_complementarity):        \n",
    "    if(continuous_pairing == \"yes\" and complementarity_in_hit_region != \"no\" and hit_self_complementarity == \"no\"):\n",
    "        if( hit_end < istar_min):\n",
    "            return \"5p\" \n",
    "        if( (hit_start+1) > istar_max):\n",
    "            return \"3p\"     \n",
    "    else:\n",
    "        if(continuous_pairing == \"no\" and hit_self_complementarity == \"yes\"):\n",
    "            return \"discontinuous star strand and hit self complementarity\"\n",
    "        elif(continuous_pairing == \"no\"):\n",
    "            return \"discontinuous star strand\"\n",
    "        elif(hit_self_complementarity == \"yes\"):\n",
    "            return \"hit self complementarity\"\n",
    "\n",
    "    if(complementarity_in_hit_region == \"no\"):\n",
    "        return \"no complementarity in hit region\"  \n",
    "    print(hit_start, hit_end, istar_min, istar_max, continuous_pairing, complementarity_in_hit_region, hit_self_complementarity)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 300,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-11-18T18:09:08.517364Z",
     "start_time": "2021-11-18T18:09:08.510899Z"
    }
   },
   "outputs": [],
   "source": [
    "def get_star_start(hit_start, hit_end, values):\n",
    "    c = 0\n",
    "    i = hit_end - 3 - c\n",
    "    while(values[i] == 0 and i >= 0):\n",
    "        c += 1\n",
    "        i = hit_end - 3 - c        \n",
    "    if(values[i] - c < 1):                    \n",
    "        return [max(values[i] - c,1), \"negative value\"]\n",
    "    if(i < hit_start):\n",
    "        return [values[i] - c, 'less than hit start']\n",
    "    return [values[i] - c, '']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 301,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-11-18T18:09:08.723604Z",
     "start_time": "2021-11-18T18:09:08.717815Z"
    }
   },
   "outputs": [],
   "source": [
    "def get_star_end(hit_start, hit_end, values):\n",
    "    if(hit_start - 2 >= 0 ):\n",
    "        a = 0    \n",
    "    else:\n",
    "        a = abs(hit_start - 2)\n",
    "    \n",
    "    i = hit_start - 2 + a\n",
    "    while(values[i] == 0 and i <= hit_end):\n",
    "        a += 1\n",
    "        i = hit_start - 2 + a\n",
    "    \n",
    "    if(i <= hit_end):\n",
    "        return values[i] + a        \n",
    "    return np.nan"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 302,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-11-18T18:09:08.907641Z",
     "start_time": "2021-11-18T18:09:08.902488Z"
    }
   },
   "outputs": [],
   "source": [
    "def get_num_of_linking_residues(hit_start,hit_end, star_start, star_end, mir_type):\n",
    "    if(mir_type == '5p'):\n",
    "        return str(star_start - hit_end - 1)\n",
    "    elif(mir_type == '3p'):\n",
    "        return str(hit_start - star_end)\n",
    "    elif(mir_type == \"discontinuous star strand\"):\n",
    "        return \"discontinuous star strand\"\n",
    "    elif(mir_type == \"no complementarity in hit region\"):\n",
    "        print('error')    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 303,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-11-18T18:09:09.144815Z",
     "start_time": "2021-11-18T18:09:09.140138Z"
    }
   },
   "outputs": [],
   "source": [
    "def get_star_branching(star_start, star_end, star_range, values):    \n",
    "    return not ((values[star_range-1] < star_start) | (values[star_range-1] > star_end)).all()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 304,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-11-18T18:09:09.445272Z",
     "start_time": "2021-11-18T18:09:09.427011Z"
    }
   },
   "outputs": [],
   "source": [
    "def getBOI_5p(hit_start, hit_end, values):\n",
    "    # first calc latest non zero value\n",
    "    for i in range(hit_end-1, 0, -1):\n",
    "        if(values[i] != 0):\n",
    "            last_v = values[i]\n",
    "            last_i = i\n",
    "            place = i\n",
    "            break            \n",
    "            \n",
    "    for i in range(place-1, 0, -1):        \n",
    "        v = values[i]\n",
    "        if(v == 0):\n",
    "            continue\n",
    "        if(v < last_v):\n",
    "            return [last_i + 1, last_v]                                                \n",
    "        \n",
    "        if((v - last_v) >= 3):\n",
    "            s1 = set(range(last_v+1, v))\n",
    "            s2 = set([values[i-1] for i in range(last_v+1, v)])\n",
    "            if(len(s1.intersection(s2)) > 0):\n",
    "                return [last_i + 1, last_v]    \n",
    "        last_v = v            \n",
    "        last_i = i            \n",
    "    for i in range(0,hit_end):\n",
    "        if(values[i] != 0 ):\n",
    "            return [i + 1, values[i]]    \n",
    "    print(\"Error\")\n",
    "                \n",
    "def getBOI_3p(hit_start, hit_end, values):\n",
    "    # first calc latest non zero value\n",
    "    for i in range(hit_start, len(values)):    \n",
    "        if(values[i] != 0):\n",
    "            last_v = values[i]\n",
    "            last_i = i\n",
    "            place = i\n",
    "            break            \n",
    "            \n",
    "    for i in range(place + 1, len(values)):\n",
    "        v = values[i]\n",
    "        if(v == 0):\n",
    "            continue\n",
    "        if(v > last_v):\n",
    "            return [last_i + 1, last_v]                                                \n",
    "        \n",
    "        if((last_v - v) >= 3):\n",
    "            s1 = set(range(v+1, last_v))\n",
    "            s2 = set([values[i-1] for i in range(v+1, last_v)])\n",
    "            if(len(s1.intersection(s2)) > 0):\n",
    "                return [last_i + 1, last_v]    \n",
    "        last_v = v            \n",
    "        last_i = i            \n",
    "    for i in range(hit_start, len(values)):\n",
    "        if(values[i] != 0 ):\n",
    "            return [i + 1, values[i]]    \n",
    "    print(\"Error\")            \n",
    "    \n",
    "    \n",
    "def get_boi(hit_start, hit_end, values, mir_type):\n",
    "    if(mir_type not in ['3p','5p']):\n",
    "        return [np.nan, np.nan]\n",
    "    if(mir_type == '5p'):\n",
    "        return getBOI_5p(hit_start, hit_end, values)\n",
    "    if(mir_type == '3p'):\n",
    "        return getBOI_3p(hit_start, hit_end, values)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 323,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-11-18T18:19:26.396304Z",
     "start_time": "2021-11-18T18:19:26.374194Z"
    }
   },
   "outputs": [],
   "source": [
    "server_url = \"http://jupyter.sysmanager.ir/tree/plant_microRNA_prediction\"\n",
    "def get_row(tag, path):\n",
    "    #print(path)\n",
    "    result = {}    \n",
    "    ct = reformatCT(path)\n",
    "    result['seq name'] = tag\n",
    "    result['ct name'] = \"\"\n",
    "    result['ct'] = f'=HYPERLINK(\"{server_url + path[1:]}\")'\n",
    "    result['pdf'] = f'=HYPERLINK(\"{server_url + path[1:-3] + \".pdf\"}\")'        \n",
    "    [hit_start, hit_end, sign] = get_tag_info(tag)\n",
    "    result['hit_start'] = hit_start + 1\n",
    "    result['hit_end'] =  hit_end\n",
    "    result['sign'] = sign\n",
    "    dg = get_deltaG(ct)\n",
    "    result['delta G'] = dg\n",
    "    [nucleotide, index, values] = get_ct_data(ct)\n",
    "    hit_seq = ''.join(nucleotide[hit_start:hit_end])\n",
    "    result['hit_seq'] = hit_seq\n",
    "    hit_range = index[hit_start:hit_end]\n",
    "    hit_len = len(hit_range)\n",
    "    result['hit_len'] = hit_len\n",
    "    inc_srange = values[hit_start:hit_end] # Incomplete_Star_range\n",
    "\n",
    "    complementarity_in_hit_region = get_complementarity_in_hit_region(inc_srange, hit_len)\n",
    "    result['complementarity_in_hit_region'] = complementarity_in_hit_region    \n",
    "    if(complementarity_in_hit_region == \"no\"):\n",
    "        msg = \"no complementarity in hit region\"\n",
    "        result['hit_self_complementarity'] = msg\n",
    "        result['continuous_pairing'] = msg\n",
    "        result['mir_type'] = msg\n",
    "        result['star_start'] = msg\n",
    "        result['star_end'] = msg\n",
    "        result['star_seq'] = msg\n",
    "        result['num_of_linking_residues'] = msg\n",
    "        result['star_branching'] = msg\n",
    "        result['boi_start']  = msg\n",
    "        result['boi_end'] = msg\n",
    "        return pd.Series(result) \n",
    "    \n",
    "    hit_self_complementarity = get_hit_self_complementarity(hit_start, hit_end, inc_srange)\n",
    "    result['hit_self_complementarity'] = hit_self_complementarity\n",
    "    \n",
    "    [istar_min, istar_max] = get_istar_min_max(inc_srange, hit_self_complementarity)  \n",
    "    result['istar_min']  = istar_min\n",
    "    result['istar_max']  = istar_max\n",
    "    continuous_pairing = get_continuous_pairing(hit_start, hit_end, istar_min, istar_max, hit_self_complementarity)\n",
    "    result['continuous_pairing'] = continuous_pairing    \n",
    "    if(continuous_pairing == \"no\"):\n",
    "        msg = \"discontinuous star strand\"        \n",
    "        result['mir_type'] = msg\n",
    "        result['star_start'] = msg\n",
    "        result['star_end'] = msg\n",
    "        result['star_seq'] = msg\n",
    "        result['num_of_linking_residues'] = msg\n",
    "        result['star_branching'] = msg\n",
    "        result['boi_start']  = msg\n",
    "        result['boi_end'] = msg\n",
    "        return pd.Series(result) \n",
    "    \n",
    "    mir_type = get_mir_type(hit_start, hit_end, istar_min, istar_max, continuous_pairing, complementarity_in_hit_region, hit_self_complementarity)\n",
    "    result['mir_type'] = mir_type    \n",
    "    if(mir_type not in ['3p', '5p']):\n",
    "        msg = mir_type        \n",
    "        result['star_start'] = msg\n",
    "        result['star_end'] = msg\n",
    "        result['star_seq'] = msg\n",
    "        result['num_of_linking_residues'] = msg\n",
    "        result['star_branching'] = msg\n",
    "        result['boi_start']  = msg\n",
    "        result['boi_end'] = msg\n",
    "        return pd.Series(result) \n",
    "    \n",
    "    [star_start, star_start_msg] = get_star_start(hit_start, hit_end, values)\n",
    "    star_end = get_star_end(hit_start, hit_end, values)\n",
    "    result['star_start'] = star_start \n",
    "    result['star_start_msg'] = star_start_msg     \n",
    "    result['star_end'] = star_end    \n",
    "    \n",
    "    star_range = index[star_start - 1:star_end]\n",
    "    star_seq = ''.join(nucleotide[star_start - 1:star_end])\n",
    "    result['star_seq'] = star_seq\n",
    "    num_of_linking_residues = get_num_of_linking_residues(hit_start,hit_end, star_start, star_end, mir_type)\n",
    "    result['num_of_linking_residues'] = num_of_linking_residues\n",
    "    star_branching = get_star_branching(star_start, star_end, star_range, values)\n",
    "    result['star_branching'] = star_branching\n",
    "    [boi_start, boi_end] = get_boi(hit_start, hit_end, values, mir_type)\n",
    "    result['boi_start'] = boi_start\n",
    "    result['boi_end'] =  boi_end\n",
    "    return pd.Series(result)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 324,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-11-18T18:19:26.403184Z",
     "start_time": "2021-11-18T18:19:26.399836Z"
    }
   },
   "outputs": [],
   "source": [
    "'''def get_df_by_tag(tag):\n",
    "    ct_files = glob.glob(f'{base}{reformat(tag)}/SEQ_*.ct')\n",
    "    return pd.Series(ct_files).apply(lambda path: get_row(tag, path))\n",
    "df['tag'].apply(lambda t : get_df_by_tag(t))''';"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 325,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-11-18T18:19:31.837980Z",
     "start_time": "2021-11-18T18:19:27.995832Z"
    }
   },
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "a5ba9e6164cc45af9032abd07214c00b",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "  0%|          | 0/200 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "updating: ct_analizer_result.csv (deflated 95%)\r\n"
     ]
    }
   ],
   "source": [
    "def get_df_by_tag(tag):\n",
    "    ct_files = glob.glob(f'{base}{reformat(tag)}/SEQ_*.ct')\n",
    "    return pd.Series(ct_files).apply(lambda path: get_row(tag, path))\n",
    "dfs = []\n",
    "for d in process_map(get_df_by_tag , df['tag'], tqdm_class=tqdm, max_workers=mp.cpu_count()- 1, chunksize=5):\n",
    "    dfs.append(d)\n",
    "df_result = pd.concat(dfs,axis=0)\n",
    "df_result.to_csv(\"ct_analizer_result.csv\", index=False)\n",
    "!zip -r ct_analizer_result.csv.zip ./ct_analizer_result.csv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 326,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-11-18T18:19:32.926854Z",
     "start_time": "2021-11-18T18:19:32.872994Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seq name</th>\n",
       "      <th>ct name</th>\n",
       "      <th>ct</th>\n",
       "      <th>pdf</th>\n",
       "      <th>hit_start</th>\n",
       "      <th>hit_end</th>\n",
       "      <th>sign</th>\n",
       "      <th>delta G</th>\n",
       "      <th>hit_seq</th>\n",
       "      <th>hit_len</th>\n",
       "      <th>...</th>\n",
       "      <th>continuous_pairing</th>\n",
       "      <th>mir_type</th>\n",
       "      <th>star_start</th>\n",
       "      <th>star_end</th>\n",
       "      <th>star_seq</th>\n",
       "      <th>num_of_linking_residues</th>\n",
       "      <th>star_branching</th>\n",
       "      <th>boi_start</th>\n",
       "      <th>boi_end</th>\n",
       "      <th>star_start_msg</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AMWY02000003.1|4085-4506|201-222|+</td>\n",
       "      <td></td>\n",
       "      <td>=HYPERLINK(\"http://jupyter.sysmanager.ir/tree/...</td>\n",
       "      <td>=HYPERLINK(\"http://jupyter.sysmanager.ir/tree/...</td>\n",
       "      <td>201</td>\n",
       "      <td>222</td>\n",
       "      <td>+</td>\n",
       "      <td>-100.01</td>\n",
       "      <td>ATAAAAACTTTTGAAAAATTCA</td>\n",
       "      <td>22</td>\n",
       "      <td>...</td>\n",
       "      <td>no</td>\n",
       "      <td>discontinuous star strand</td>\n",
       "      <td>discontinuous star strand</td>\n",
       "      <td>discontinuous star strand</td>\n",
       "      <td>discontinuous star strand</td>\n",
       "      <td>discontinuous star strand</td>\n",
       "      <td>discontinuous star strand</td>\n",
       "      <td>discontinuous star strand</td>\n",
       "      <td>discontinuous star strand</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AMWY02000003.1|4085-4506|201-222|+</td>\n",
       "      <td></td>\n",
       "      <td>=HYPERLINK(\"http://jupyter.sysmanager.ir/tree/...</td>\n",
       "      <td>=HYPERLINK(\"http://jupyter.sysmanager.ir/tree/...</td>\n",
       "      <td>201</td>\n",
       "      <td>222</td>\n",
       "      <td>+</td>\n",
       "      <td>-98.95</td>\n",
       "      <td>ATAAAAACTTTTGAAAAATTCA</td>\n",
       "      <td>22</td>\n",
       "      <td>...</td>\n",
       "      <td>yes</td>\n",
       "      <td>5p</td>\n",
       "      <td>355</td>\n",
       "      <td>376</td>\n",
       "      <td>cgttcttcgacctcacgcattt</td>\n",
       "      <td>132</td>\n",
       "      <td>False</td>\n",
       "      <td>196</td>\n",
       "      <td>379</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AMWY02000003.1|4085-4506|201-222|+</td>\n",
       "      <td></td>\n",
       "      <td>=HYPERLINK(\"http://jupyter.sysmanager.ir/tree/...</td>\n",
       "      <td>=HYPERLINK(\"http://jupyter.sysmanager.ir/tree/...</td>\n",
       "      <td>201</td>\n",
       "      <td>222</td>\n",
       "      <td>+</td>\n",
       "      <td>-99.99</td>\n",
       "      <td>ATAAAAACTTTTGAAAAATTCA</td>\n",
       "      <td>22</td>\n",
       "      <td>...</td>\n",
       "      <td>yes</td>\n",
       "      <td>5p</td>\n",
       "      <td>221</td>\n",
       "      <td>418</td>\n",
       "      <td>CAtttttctctgtatgtctctttcattctattattgtccttccgat...</td>\n",
       "      <td>-2</td>\n",
       "      <td>True</td>\n",
       "      <td>213</td>\n",
       "      <td>228</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AMWY02000003.1|4085-4506|201-222|+</td>\n",
       "      <td></td>\n",
       "      <td>=HYPERLINK(\"http://jupyter.sysmanager.ir/tree/...</td>\n",
       "      <td>=HYPERLINK(\"http://jupyter.sysmanager.ir/tree/...</td>\n",
       "      <td>201</td>\n",
       "      <td>222</td>\n",
       "      <td>+</td>\n",
       "      <td>-96.88</td>\n",
       "      <td>ATAAAAACTTTTGAAAAATTCA</td>\n",
       "      <td>22</td>\n",
       "      <td>...</td>\n",
       "      <td>no</td>\n",
       "      <td>discontinuous star strand</td>\n",
       "      <td>discontinuous star strand</td>\n",
       "      <td>discontinuous star strand</td>\n",
       "      <td>discontinuous star strand</td>\n",
       "      <td>discontinuous star strand</td>\n",
       "      <td>discontinuous star strand</td>\n",
       "      <td>discontinuous star strand</td>\n",
       "      <td>discontinuous star strand</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AMWY02000003.1|4085-4506|201-222|+</td>\n",
       "      <td></td>\n",
       "      <td>=HYPERLINK(\"http://jupyter.sysmanager.ir/tree/...</td>\n",
       "      <td>=HYPERLINK(\"http://jupyter.sysmanager.ir/tree/...</td>\n",
       "      <td>201</td>\n",
       "      <td>222</td>\n",
       "      <td>+</td>\n",
       "      <td>-101.61</td>\n",
       "      <td>ATAAAAACTTTTGAAAAATTCA</td>\n",
       "      <td>22</td>\n",
       "      <td>...</td>\n",
       "      <td>no</td>\n",
       "      <td>discontinuous star strand</td>\n",
       "      <td>discontinuous star strand</td>\n",
       "      <td>discontinuous star strand</td>\n",
       "      <td>discontinuous star strand</td>\n",
       "      <td>discontinuous star strand</td>\n",
       "      <td>discontinuous star strand</td>\n",
       "      <td>discontinuous star strand</td>\n",
       "      <td>discontinuous star strand</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>AMWY02000334.1|4703-5044|201-220|+</td>\n",
       "      <td></td>\n",
       "      <td>=HYPERLINK(\"http://jupyter.sysmanager.ir/tree/...</td>\n",
       "      <td>=HYPERLINK(\"http://jupyter.sysmanager.ir/tree/...</td>\n",
       "      <td>201</td>\n",
       "      <td>220</td>\n",
       "      <td>+</td>\n",
       "      <td>-125.23</td>\n",
       "      <td>TCGTTGAGGGCAGCGTTGAT</td>\n",
       "      <td>20</td>\n",
       "      <td>...</td>\n",
       "      <td>undifined</td>\n",
       "      <td>hit self complementarity</td>\n",
       "      <td>hit self complementarity</td>\n",
       "      <td>hit self complementarity</td>\n",
       "      <td>hit self complementarity</td>\n",
       "      <td>hit self complementarity</td>\n",
       "      <td>hit self complementarity</td>\n",
       "      <td>hit self complementarity</td>\n",
       "      <td>hit self complementarity</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>AMWY02000334.1|4703-5044|201-220|+</td>\n",
       "      <td></td>\n",
       "      <td>=HYPERLINK(\"http://jupyter.sysmanager.ir/tree/...</td>\n",
       "      <td>=HYPERLINK(\"http://jupyter.sysmanager.ir/tree/...</td>\n",
       "      <td>201</td>\n",
       "      <td>220</td>\n",
       "      <td>+</td>\n",
       "      <td>-121.85</td>\n",
       "      <td>TCGTTGAGGGCAGCGTTGAT</td>\n",
       "      <td>20</td>\n",
       "      <td>...</td>\n",
       "      <td>no</td>\n",
       "      <td>discontinuous star strand</td>\n",
       "      <td>discontinuous star strand</td>\n",
       "      <td>discontinuous star strand</td>\n",
       "      <td>discontinuous star strand</td>\n",
       "      <td>discontinuous star strand</td>\n",
       "      <td>discontinuous star strand</td>\n",
       "      <td>discontinuous star strand</td>\n",
       "      <td>discontinuous star strand</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>AMWY02000334.1|4703-5044|201-220|+</td>\n",
       "      <td></td>\n",
       "      <td>=HYPERLINK(\"http://jupyter.sysmanager.ir/tree/...</td>\n",
       "      <td>=HYPERLINK(\"http://jupyter.sysmanager.ir/tree/...</td>\n",
       "      <td>201</td>\n",
       "      <td>220</td>\n",
       "      <td>+</td>\n",
       "      <td>-127.69</td>\n",
       "      <td>TCGTTGAGGGCAGCGTTGAT</td>\n",
       "      <td>20</td>\n",
       "      <td>...</td>\n",
       "      <td>yes</td>\n",
       "      <td>3p</td>\n",
       "      <td>167</td>\n",
       "      <td>206</td>\n",
       "      <td>caagctgtggccagctatggcaaagaaaagctcgTCGTTG</td>\n",
       "      <td>-6</td>\n",
       "      <td>True</td>\n",
       "      <td>208</td>\n",
       "      <td>197</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>AMWY02000334.1|4703-5044|201-220|+</td>\n",
       "      <td></td>\n",
       "      <td>=HYPERLINK(\"http://jupyter.sysmanager.ir/tree/...</td>\n",
       "      <td>=HYPERLINK(\"http://jupyter.sysmanager.ir/tree/...</td>\n",
       "      <td>201</td>\n",
       "      <td>220</td>\n",
       "      <td>+</td>\n",
       "      <td>-122.25</td>\n",
       "      <td>TCGTTGAGGGCAGCGTTGAT</td>\n",
       "      <td>20</td>\n",
       "      <td>...</td>\n",
       "      <td>undifined</td>\n",
       "      <td>hit self complementarity</td>\n",
       "      <td>hit self complementarity</td>\n",
       "      <td>hit self complementarity</td>\n",
       "      <td>hit self complementarity</td>\n",
       "      <td>hit self complementarity</td>\n",
       "      <td>hit self complementarity</td>\n",
       "      <td>hit self complementarity</td>\n",
       "      <td>hit self complementarity</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>AMWY02000334.1|4703-5044|201-220|+</td>\n",
       "      <td></td>\n",
       "      <td>=HYPERLINK(\"http://jupyter.sysmanager.ir/tree/...</td>\n",
       "      <td>=HYPERLINK(\"http://jupyter.sysmanager.ir/tree/...</td>\n",
       "      <td>201</td>\n",
       "      <td>220</td>\n",
       "      <td>+</td>\n",
       "      <td>-122.17</td>\n",
       "      <td>TCGTTGAGGGCAGCGTTGAT</td>\n",
       "      <td>20</td>\n",
       "      <td>...</td>\n",
       "      <td>yes</td>\n",
       "      <td>3p</td>\n",
       "      <td>25</td>\n",
       "      <td>47</td>\n",
       "      <td>ccatgaaaggcctagcagccatg</td>\n",
       "      <td>153</td>\n",
       "      <td>False</td>\n",
       "      <td>225</td>\n",
       "      <td>18</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3395 rows × 24 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                              seq name ct name  \\\n",
       "0   AMWY02000003.1|4085-4506|201-222|+           \n",
       "1   AMWY02000003.1|4085-4506|201-222|+           \n",
       "2   AMWY02000003.1|4085-4506|201-222|+           \n",
       "3   AMWY02000003.1|4085-4506|201-222|+           \n",
       "4   AMWY02000003.1|4085-4506|201-222|+           \n",
       "..                                 ...     ...   \n",
       "14  AMWY02000334.1|4703-5044|201-220|+           \n",
       "15  AMWY02000334.1|4703-5044|201-220|+           \n",
       "16  AMWY02000334.1|4703-5044|201-220|+           \n",
       "17  AMWY02000334.1|4703-5044|201-220|+           \n",
       "18  AMWY02000334.1|4703-5044|201-220|+           \n",
       "\n",
       "                                                   ct  \\\n",
       "0   =HYPERLINK(\"http://jupyter.sysmanager.ir/tree/...   \n",
       "1   =HYPERLINK(\"http://jupyter.sysmanager.ir/tree/...   \n",
       "2   =HYPERLINK(\"http://jupyter.sysmanager.ir/tree/...   \n",
       "3   =HYPERLINK(\"http://jupyter.sysmanager.ir/tree/...   \n",
       "4   =HYPERLINK(\"http://jupyter.sysmanager.ir/tree/...   \n",
       "..                                                ...   \n",
       "14  =HYPERLINK(\"http://jupyter.sysmanager.ir/tree/...   \n",
       "15  =HYPERLINK(\"http://jupyter.sysmanager.ir/tree/...   \n",
       "16  =HYPERLINK(\"http://jupyter.sysmanager.ir/tree/...   \n",
       "17  =HYPERLINK(\"http://jupyter.sysmanager.ir/tree/...   \n",
       "18  =HYPERLINK(\"http://jupyter.sysmanager.ir/tree/...   \n",
       "\n",
       "                                                  pdf  hit_start  hit_end  \\\n",
       "0   =HYPERLINK(\"http://jupyter.sysmanager.ir/tree/...        201      222   \n",
       "1   =HYPERLINK(\"http://jupyter.sysmanager.ir/tree/...        201      222   \n",
       "2   =HYPERLINK(\"http://jupyter.sysmanager.ir/tree/...        201      222   \n",
       "3   =HYPERLINK(\"http://jupyter.sysmanager.ir/tree/...        201      222   \n",
       "4   =HYPERLINK(\"http://jupyter.sysmanager.ir/tree/...        201      222   \n",
       "..                                                ...        ...      ...   \n",
       "14  =HYPERLINK(\"http://jupyter.sysmanager.ir/tree/...        201      220   \n",
       "15  =HYPERLINK(\"http://jupyter.sysmanager.ir/tree/...        201      220   \n",
       "16  =HYPERLINK(\"http://jupyter.sysmanager.ir/tree/...        201      220   \n",
       "17  =HYPERLINK(\"http://jupyter.sysmanager.ir/tree/...        201      220   \n",
       "18  =HYPERLINK(\"http://jupyter.sysmanager.ir/tree/...        201      220   \n",
       "\n",
       "   sign  delta G                 hit_seq  hit_len  ... continuous_pairing  \\\n",
       "0     +  -100.01  ATAAAAACTTTTGAAAAATTCA       22  ...                 no   \n",
       "1     +   -98.95  ATAAAAACTTTTGAAAAATTCA       22  ...                yes   \n",
       "2     +   -99.99  ATAAAAACTTTTGAAAAATTCA       22  ...                yes   \n",
       "3     +   -96.88  ATAAAAACTTTTGAAAAATTCA       22  ...                 no   \n",
       "4     +  -101.61  ATAAAAACTTTTGAAAAATTCA       22  ...                 no   \n",
       "..  ...      ...                     ...      ...  ...                ...   \n",
       "14    +  -125.23    TCGTTGAGGGCAGCGTTGAT       20  ...          undifined   \n",
       "15    +  -121.85    TCGTTGAGGGCAGCGTTGAT       20  ...                 no   \n",
       "16    +  -127.69    TCGTTGAGGGCAGCGTTGAT       20  ...                yes   \n",
       "17    +  -122.25    TCGTTGAGGGCAGCGTTGAT       20  ...          undifined   \n",
       "18    +  -122.17    TCGTTGAGGGCAGCGTTGAT       20  ...                yes   \n",
       "\n",
       "                     mir_type                 star_start  \\\n",
       "0   discontinuous star strand  discontinuous star strand   \n",
       "1                          5p                        355   \n",
       "2                          5p                        221   \n",
       "3   discontinuous star strand  discontinuous star strand   \n",
       "4   discontinuous star strand  discontinuous star strand   \n",
       "..                        ...                        ...   \n",
       "14   hit self complementarity   hit self complementarity   \n",
       "15  discontinuous star strand  discontinuous star strand   \n",
       "16                         3p                        167   \n",
       "17   hit self complementarity   hit self complementarity   \n",
       "18                         3p                         25   \n",
       "\n",
       "                     star_end  \\\n",
       "0   discontinuous star strand   \n",
       "1                         376   \n",
       "2                         418   \n",
       "3   discontinuous star strand   \n",
       "4   discontinuous star strand   \n",
       "..                        ...   \n",
       "14   hit self complementarity   \n",
       "15  discontinuous star strand   \n",
       "16                        206   \n",
       "17   hit self complementarity   \n",
       "18                         47   \n",
       "\n",
       "                                             star_seq  \\\n",
       "0                           discontinuous star strand   \n",
       "1                              cgttcttcgacctcacgcattt   \n",
       "2   CAtttttctctgtatgtctctttcattctattattgtccttccgat...   \n",
       "3                           discontinuous star strand   \n",
       "4                           discontinuous star strand   \n",
       "..                                                ...   \n",
       "14                           hit self complementarity   \n",
       "15                          discontinuous star strand   \n",
       "16           caagctgtggccagctatggcaaagaaaagctcgTCGTTG   \n",
       "17                           hit self complementarity   \n",
       "18                            ccatgaaaggcctagcagccatg   \n",
       "\n",
       "      num_of_linking_residues             star_branching  \\\n",
       "0   discontinuous star strand  discontinuous star strand   \n",
       "1                         132                      False   \n",
       "2                          -2                       True   \n",
       "3   discontinuous star strand  discontinuous star strand   \n",
       "4   discontinuous star strand  discontinuous star strand   \n",
       "..                        ...                        ...   \n",
       "14   hit self complementarity   hit self complementarity   \n",
       "15  discontinuous star strand  discontinuous star strand   \n",
       "16                         -6                       True   \n",
       "17   hit self complementarity   hit self complementarity   \n",
       "18                        153                      False   \n",
       "\n",
       "                    boi_start                    boi_end star_start_msg  \n",
       "0   discontinuous star strand  discontinuous star strand            NaN  \n",
       "1                         196                        379                 \n",
       "2                         213                        228                 \n",
       "3   discontinuous star strand  discontinuous star strand            NaN  \n",
       "4   discontinuous star strand  discontinuous star strand            NaN  \n",
       "..                        ...                        ...            ...  \n",
       "14   hit self complementarity   hit self complementarity            NaN  \n",
       "15  discontinuous star strand  discontinuous star strand            NaN  \n",
       "16                        208                        197                 \n",
       "17   hit self complementarity   hit self complementarity            NaN  \n",
       "18                        225                         18                 \n",
       "\n",
       "[3395 rows x 24 columns]"
      ]
     },
     "execution_count": 326,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_result"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "heading_collapsed": true
   },
   "source": [
    "# BLASTX or DIMOND"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "heading_collapsed": true
   },
   "source": [
    "# DIAMOND"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-10-17T20:40:11.194148Z",
     "start_time": "2021-10-17T20:40:11.187523Z"
    },
    "hidden": true
   },
   "source": [
    "https://github.com/bbuchfink/diamond"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "hidden": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "hidden": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "hidden": true
   },
   "outputs": [],
   "source": [
    "'''\n",
    "import hashlib\n",
    "import os\n",
    "parallel = [os.path.join(dp, f) for dp, dn, filenames in os.walk(\"./PRNA_secondary_structure\") for f in filenames ]\n",
    "series = [os.path.join(dp, f) for dp, dn, filenames in os.walk(\"./SRNA_secondary_structure\") for f in filenames ]\n",
    "for i in range(len(parallel)):        \n",
    "    md5_hash = hashlib.md5()\n",
    "    with open(parallel[i],\"rb\") as file:        \n",
    "        md5_hash.update(file.read())\n",
    "        digest1 = md5_hash.hexdigest()                \n",
    "    md5_hash = hashlib.md5()\n",
    "    with open(series[i],\"rb\") as file:        \n",
    "        md5_hash.update(file.read())\n",
    "        digest2 = md5_hash.hexdigest()                \n",
    "    if(digest1 != digest2):\n",
    "        print(parallel[i])\n",
    "        print(series[i])\n",
    "        print(\"***********\")     \n",
    "'''"
   ]
  }
 ],
 "metadata": {
  "hide_input": false,
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.10"
  },
  "varInspector": {
   "cols": {
    "lenName": 16,
    "lenType": 16,
    "lenVar": 40
   },
   "kernels_config": {
    "python": {
     "delete_cmd_postfix": "",
     "delete_cmd_prefix": "del ",
     "library": "var_list.py",
     "varRefreshCmd": "print(var_dic_list())"
    },
    "r": {
     "delete_cmd_postfix": ") ",
     "delete_cmd_prefix": "rm(",
     "library": "var_list.r",
     "varRefreshCmd": "cat(var_dic_list()) "
    }
   },
   "types_to_exclude": [
    "module",
    "function",
    "builtin_function_or_method",
    "instance",
    "_Feature"
   ],
   "window_display": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
